---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r global options, include = FALSE}
knitr::opts_chunk$set(echo=T, include = T, warning=FALSE, message=FALSE, tidy = TRUE, tidy.opts = list(comment = FALSE))
```

# ```{r cachecontrol, cache = TRUE, cache.extra = tools::md5sum(expr_data_norm)}
# knitr::opts_chunk$set(cache.rebuild = TRUE)
# ```

######################################################
#### LOAD LIBRARIES
######################################################
```{r}
library(DESeq2)
library(preprocessCore)
library(reshape)
library(ggplot2)
library(doBy)
library(ggpubr)
library(gridExtra)
library(Hmisc)
library(plyr)
library(pheatmap)
library(Gviz)
library(rlist)
library(GenomicRanges)
library(GenomicFeatures)
library(TxDb.Mmusculus.UCSC.mm9.knownGene)
library(rtracklayer)
library(trackViewer)
library(plyr)
library(ComplexHeatmap)
theme_set(theme_classic2())
source("~/software/myScripts/R-scripts/FUNCTIONS.R")
```

######################################################
#### CUSTOM FUNCTIONS
######################################################
```{r}
plot_profile <- function(x, y, plot) {
  ## plot: promoter_class or promoter_expr
  test.sub <- x
  test.sub$class <- as.character(test.sub$class)

  if(missing(plot)) { plot<-"promoter_class"; }
  if(missing(y)) { y<-"NA"; }
  
  #col=c("#006837", "#1a9850", "#66bd63", "#a6d96a", "#fdae61", "#f46d43", "#d73027", "#a50026")
  col <- c(rep("#e31a1c", 3), rep("#1f78b4",2), rep("#33a02c",3))
  shp <- c(rep(17,3), rep(16,2), rep(18,3))

  if(plot=="promoter_class") {
    test.sub$promoter <- "narrow"
    test.sub[which(test.sub$distance > 500 & test.sub$distance <= 1000),]$promoter <- "medium"
    test.sub[which(test.sub$distance > 1000),]$promoter <- "broad"
    
    test.melt <- melt(test.sub[,c(1:8,9,16)])
    
    ggplot(test.melt, aes(variable, group=1)) +
      stat_smooth(data = test.melt[which(test.melt$promoter=="narrow"),], aes(y = log(value), group=1, colour="narrow"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      stat_smooth(data = test.melt[which(test.melt$promoter=="medium"),], aes(y = log(value), group=1, colour="medium"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      stat_smooth(data = test.melt[which(test.melt$promoter=="broad"),], aes(y = log(value), group=1, colour="broad"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      ## colors are arranged alphabetically, meaning broad, medium and narrow
      scale_colour_manual("promoter", breaks = c("narrow", "medium", "broad"), values = c("#4faf4a","#ef7e1b", "#e31e1e")) +
      theme_bw() +
      facet_grid(.~class) +
      ggtitle(y) +
      scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
      theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
      theme(legend.position="none")
  } else if(plot=="promoter_expr") {
    test.sub$promoter <- "low"
    # test.sub[which(log(test.sub$expr_esc) > as.vector(summary(log(test.sub$expr_esc))[2]) & log(test.sub$expr_esc) < as.vector(summary(log(test.sub$expr_esc))[5])),]$promoter <- "average"
    #test.sub[which(log(test.sub$expr_esc) > as.vector(summary(log(test.sub$expr_esc))[5])),]$promoter <- "high"
    test.sub[which(log(test.sub$expr_esc+1) > 2 & log(test.sub$expr_esc) <= 4),]$promoter <- "average"
    test.sub[which(log(test.sub$expr_esc+1) > 4),]$promoter <- "high"
    
    test.melt <- melt(test.sub[,c(1:8,9,16)])
    
    ggplot(test.melt, aes(variable, group=1)) +
      stat_smooth(data = test.melt[which(test.melt$promoter=="low"),], aes(y = log(value), group=1, colour="low"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      stat_smooth(data = test.melt[which(test.melt$promoter=="average"),], aes(y = log(value), group=1, colour="average"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      stat_smooth(data = test.melt[which(test.melt$promoter=="high"),], aes(y = log(value), group=1, colour="high"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      ## colors are arranged alphabetically, meaning broad, medium and narrow
      scale_colour_manual("promoter", breaks = c("low", "average", "high"), values = c("#4faf4a","#ef7e1b", "#e31e1e")) +
      theme_bw() +
      ggtitle(y) +
      scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
      theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
      theme(legend.position="none")
  } else if(plot=="promoter_directionality") {
      test.melt <- melt(test.sub[,c(1:8,9,ncol(test.sub))])
      colnames(test.melt) <- c("class", "promoter", "variable", "value")
    
      ggplot(test.melt, aes(variable, group=1)) +
        stat_smooth(data = test.melt[which(test.melt$promoter=="1"),], aes(y = log(value), group=1, colour="1"), method=lm, formula = y ~ poly(x,5), level=0.95) +
        stat_smooth(data = test.melt[which(test.melt$promoter=="0"),], aes(y = log(value), group=1, colour="0"), method=lm, formula = y ~ poly(x,5), level=0.95) +
        stat_smooth(data = test.melt[which(test.melt$promoter=="-1"),], aes(y = log(value), group=1, colour="-1"), method=lm, formula = y ~ poly(x,5), level=0.95) +
        ## colors are arranged alphabetically, meaning broad, medium and narrow
        scale_colour_manual("promoter", breaks = c("1", "0", "-1"), values = c("#ef7e1b", "#e31e1e", "#4faf4a")) +
        theme_bw() +
        ggtitle(y) +
        scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
        theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
        theme(legend.position="none")
  }
}

plot_profile_by_geneFC <- function(x, y, plot) {
  ## plot: what to plot (frequency: frequency plot;
  ##                     promoter: class+promoter class, 
  ##                     class: only class)
  test.sub <- x
  test.sub$geneClass <- as.character(test.sub$geneClass)
  
  #col=c("#006837", "#1a9850", "#66bd63", "#a6d96a", "#fdae61", "#f46d43", "#d73027", "#a50026")
  col <- c(rep("#e31a1c", 3), rep("#1f78b4",2), rep("#33a02c",3))
  shp <- c(rep(17,3), rep(16,2), rep(18,3))
  
  if(missing(plot)) { plot="class"; }

  if(plot=="frequency") {
    df.melt <- as.data.frame(table(test.sub[,c(12,10,13)]))
    colnames(df.melt) <- c("xlabels", "class", "cluster", "freq")
    total <- as.vector(unlist(ddply(df.melt, .(class, cluster), summarise, X2=sum(freq))[[3]]))
    df.melt$total <- c(rep(total[1],3), rep(total[4],3), rep(total[2],3), rep(total[5],3), rep(total[3],3), rep(total[6],3))
    df.melt$density <- df.melt$freq/df.melt$total
    #c(table(gene_expr[which(gene_expr$chd4_class=="down" & gene_expr$cluster=="Y2"),]$promoter_class))/table(gene_expr[which(gene_expr$cluster=="Y2"),]$promoter_class)
    
    p <- ggplot(df.melt, aes(xlabels, density, fill=factor(cluster))) + 
      geom_bar(stat="identity") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
      facet_grid(.~class+cluster) +
      geom_text(data = df.melt, aes(label = freq)) +
      theme(legend.position="none") +
      xlab("")
    return(p)
  } else if(plot=="promoter") {
    test.melt <- melt(test.sub[,c(2:9,10,12,13)])
    p <- ggplot(test.melt, aes(variable, group=1)) +
            stat_smooth(data = test.melt[which(test.melt$geneClass=="down"),], aes(y = log(value), group=1, colour="down"), method=lm, formula = y ~ poly(x,5), level=0.95) +
            stat_smooth(data = test.melt[which(test.melt$geneClass=="neutral"),], aes(y = log(value), group=1, colour="neutral"), method=lm, formula = y ~ poly(x,5), level=0.95, linetype="dashed") +
            stat_smooth(data = test.melt[which(test.melt$geneClass=="up"),], aes(y = log(value), group=1, colour="up"), method=lm, formula = y ~ poly(x,5), level=0.95) +
            scale_colour_manual("geneClass", breaks = c("down", "neutral", "up"), values = c("#4393c3","#000000", "#d6604d")) +
            theme_bw() +
            facet_grid(.~class+promoter_class) +
            ggtitle(y) +
            scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
            theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
            theme(legend.position="none")
    return(p)
  } else if(plot=="CpG") {
    test.melt <- melt(test.sub)
    p <- ggplot(test.melt, aes(variable, group=1)) +
      stat_smooth(data = test.melt[which(test.melt$CpG_length=="small"),], aes(y = log(value), group=1, colour="down"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      stat_smooth(data = test.melt[which(test.melt$CpG_length=="moderate"),], aes(y = log(value), group=1, colour="neutral"), method=lm, formula = y ~ poly(x,5), level=0.95, linetype="dashed") +
      stat_smooth(data = test.melt[which(test.melt$CpG_length=="big"),], aes(y = log(value), group=1, colour="up"), method=lm, formula = y ~ poly(x,5), level=0.95) +
      scale_colour_manual("CpG_length", breaks = c("small", "moderate", "big"), values = c("#4393c3","#000000", "#d6604d")) +
      theme_bw() +
      facet_grid(.~geneClass+promoter_class) +
      ggtitle(y) +
      scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
      theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
      theme(legend.position="none")
    return(p)
  } else if(plot=="ep400") {
    test.melt <- melt(test.sub)
    p <- ggplot(test.melt, aes(variable, group=1)) +
          stat_smooth(data = test.melt[which(test.melt$ep400_cluster=="low"),], aes(y = log(value), group=1, colour="down"), method=lm, formula = y ~ poly(x,5), level=0.95) +
          stat_smooth(data = test.melt[which(test.melt$ep400_cluster=="average"),], aes(y = log(value), group=1, colour="neutral"), method=lm, formula = y ~ poly(x,5), level=0.95, linetype="dashed") +
          stat_smooth(data = test.melt[which(test.melt$ep400_cluster=="high"),], aes(y = log(value), group=1, colour="up"), method=lm, formula = y ~ poly(x,5), level=0.95) +
          scale_colour_manual("ep400_cluster", breaks = c("low", "average", "high"), values = c("#4393c3","#000000", "#d6604d")) +
          theme_bw() +
          facet_grid(.~geneClass) +
          ggtitle(y) +
          scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
          theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
          theme(legend.position="none")
    return(p)
  } else {
    test.melt <- melt(test.sub[,c(2:9,10,12)])
    p <- ggplot(test.melt, aes(variable, group=1)) +
          stat_smooth(data = test.melt[which(test.melt$geneClass=="down"),], aes(y = log(value), group=1, colour="down"), method=lm, formula = y ~ poly(x,5), level=0.95) +
          stat_smooth(data = test.melt[which(test.melt$geneClass=="neutral"),], aes(y = log(value), group=1, colour="neutral"), method=lm, formula = y ~ poly(x,5), level=0.95, linetype="dashed") +
          stat_smooth(data = test.melt[which(test.melt$geneClass=="up"),], aes(y = log(value), group=1, colour="up"), method=lm, formula = y ~ poly(x,5), level=0.95) +
          scale_colour_manual("geneClass", breaks = c("down", "neutral", "up"), values = c("#4393c3","#000000", "#d6604d")) +
          theme_bw() +
          facet_grid(.~class) +
          ggtitle(y) +
          scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
          theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
          theme(legend.position="none")
    return(p)
  }
}
```

######################################################
#### CUSTOM ORDER
######################################################
```{r}
factors_order <- c("jarid2", "ezh2", "ring1b", "h3k27me3", "suz12", "h3k9me3", "cbx7",
                   "paf1", "leo1", "cdc73",
                   "ctcf", "brg1", "h3v3", "h3Only", "h3k4me1", "chd6", "chd4", "max", "mll34", "smc1a", "mll2", "kdm5b", 
                   "h2az", "kdm2a", "h2azAC", "h3k4me2",
                   "chd2", "chd9", "brd4",  "chd1", "ep400", "h3k27ac", "p300",
                   "cdk8", "yy1", "nipbl", "med1", "med12", "myc", "tbp",
                   "chd8", "set1a", "cxxc1", "ash2l", "hdac1", "lsd1", "nelfa",
                   "h3k9ac", "h3k4me3", "h3k36me3", "rad21", "mbd3")

enzyme_order <- c("chd2","chd9", "chd1", "chd6", "brd4", "chd8", "chd4", "brg1", "ep400", "ring1", "ezh2", "suz12", "mll34", "mll2", "h2az")

promoter_class_order <- c("narrow", "medium", "broad")

profile_order_y2 <- c("h3k4me1", "h3k36me3", "h3k4me2", "chd9", "chd1", "chd2",
                      "h2az", "h2azAC", "kdm2a", "h3k4me3", "h3k9ac",
                      "set1a", "yy1", "mll34", "cdk8", "nelfa", "lsd1", "tbp", "smc1a", "med12", "med1", "nipbl", "myc", "hdac1", "max", 
                      "ctcf", "kdm5b", "h3v3", "h3Only",
                      "pol2", "chd8", "ep400", "ash2l", "cxxc1", "mll2", "p300", "h3k27ac", "brd4", "brg1", "chd4", "chd6",   
                      "jarid2", "ezh2", "ring1b", "h3k27me3", "suz12", "h3k9me3", "cbx7",
                      "rad21", "mbd3", "esrrb", "kdm5c", "cdk9", "cdk7")

profile_order_y1 <- c("h3k4me1", "h3k36me3", "h3k4me2", "cdc73", "paf1", "leo1", "chd9", "chd1", "chd2",
                      "h3k27ac", "brd4", "chd8", "ep400", "brg1", "chd4", "chd6",
                      "mll2", "esrrb", "p300", "hdac1", "cdk7", "yy1", "set1a", "nelfa", 
                      "cdk8", "cdk9", "mll34", "smc1a", "med12", "lsd1", "tbp", "med1", "nipbl",
                      "myc", "kdm2a", "max", "kdm5c", "h2az", "h2azAC", "h3k4me3", "h3k9ac",
                      "kdm5b", "h3v3", "h3Only",
                      "ctcf", "h3k9me3", "ring1b", "ezh2", "h3k27me3", "jarid2", "cbx7", "suz12",
                      "pol2", "pol2_ser2p", "pol2_ser5p", "pol2_ser7p", "ash2l", "cxxc1", "rad21", "mbd3", "mnase", "atac")
```

######################################################
#### LOAD OR CREATE FILES
######################################################
```{r}
##----------------------------------------------------##
## Load gene promoter file (esc)
##----------------------------------------------------##

df_esc <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.GENE")
colnames(df_esc) <- c("chr", "start", "end", "gene", "score", "strand", "promoter_defined", 
                  "promoter", "dist_tss_up", "dist_tss_down",
                  "h3k27ac_promoter", "h3k27me3_promoter", "h3k4me1_promoter", "h3k4me3_promoter",
                  "h3k9me3_promoter", "h3k36me3_promoter", "atac_promoter", 
                  "h3k27ac_distal", "h3k27me3_distal", "h3k4me1_distal", "h3k4me3_distal",
                  "h3k9me3_distal", "h3k36me3_distal", "atac_distal", 
                  "h3k27ac_distal_ws", "h3k27me3_distal_ws", "h3k4me1_distal_ws", "h3k4me3_distal_ws",
                  "h3k9me3_distal_ws", "h3k36me3_distal_ws", "atac_distal_ws", 
                  "enhancer_active", "enhancer_poised", "enhancer_primed", "promoter_active", 
                  "promoter_bivalent", "repressed_heterochromatin", "repressed_polycomb", "transcribed",
                  "cluster", "expr_mef", "expr_mef48", "expr_ipsc", "expr_esc", "closest_tss", "CpG_length", "perCpG")

df_esc[which(df_esc$closest_tss=="."),]$closest_tss <- NA
df_esc$closest_tss <- as.numeric(df_esc$closest_tss)
hk_genes <- read.table("~/project/genome_annotations/mouse_house_keeping_genes.txt")
df_esc$house_keeping <- ifelse(df_esc$gene %in% hk_genes$V1, 1, 0)

df_esc$promoter_class <- "narrow"
df_esc[which(df_esc$dist_tss_down >500 & df_esc$dist_tss_down <=1000),]$promoter_class <- "medium"
df_esc[which(df_esc$dist_tss_down >1000),]$promoter_class <- "broad"
df_esc$promoter_class <- factor(df_esc$promoter_class, levels=promoter_class_order, ordered = T)

##----------------------------------------------------##
## Load file containing normalized signal (RPKM) for various chromatin modifers (CM) at gene promoter in ESC
##----------------------------------------------------##

enzyme_tpm <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.TPM")
colnames(enzyme_tpm) <- c("chr", "start", "end", "name", "score", "strand", "promoter_defined", "description", 
                          "dist_tss_up", "dist_tss_down", "class", "ash2l", "brg1", "cbx7", "chd1",
                          "chd2", "chd4", "chd6", "chd8", "chd9", "cxxc1", "ep400", "ezh2", "hdac1", "kdm5b",
                          "mll2", "mll34", "p300", "rad21", "ring1b", "set1a", "suz12", "h3k27ac", "h3k27me3", "h3k4me1",
                          "h3k4me3", "h3k9me3", "h3k36me3", "h2az", "h3k4me2", "h3k9ac", "brd4", "lsd1", "cdk8", "ctcf", 
                          "jarid2", "kdm2a", "med12", "med1", "nipbl", "smc1a", "tbp", "yy1", "h2azAC", "h3v3", "h3Only",
                          "atac_brg1_kd", "atac_chd4_kd", "atac_ep400_kd", "atac_wt", 
                          "pol2_brg1_kd", "pol2_chd4_kd", "pol2_ep400_kd", "pol2_wt",
                          "myc", "max", "pol2_ser2p", "pol2_ser5p", "nelfa", "pol2_ser7p",
                          "paf1", "leo1", "cdc73")

df <- read.table(pipe("cut -f 4,12 /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.TPM.additional"))
colnames(df) <- c("gene", "h3v3_1")
enzyme_tpm$h3v3 <- df$h3v3_1

##----------------------------------------------------##
## Load file containing gene expression in esc (wt), and upon KD of:
## brg1, chd1, chd2, chd4, chd6, chd8, chd9, ep400, ezh2, ring1, mll34, mll2 and h2az
##----------------------------------------------------##

gene_expr <- read.table("/home/pundhir/project/chip-seq-analysis/data/gene/esc/GENE_EXPR.MATRIX")

colnames(gene_expr) <- c("gene", "brd4_fc", "brd4_pVal", "brd4_qVal", "brd4_class",
                         "brg1_fc", "brg1_pVal", "brg1_qVal", "brg1_class",
                         "chd1_fc", "chd1_pVal", "chd1_qVal", "chd1_class",
                         "chd2_fc", "chd2_pVal", "chd2_qVal", "chd2_class",
                         "chd4_fc", "chd4_pVal", "chd4_qVal", "chd4_class",
                         "chd6_fc", "chd6_pVal", "chd6_qVal", "chd6_class",
                         "chd8_fc", "chd8_pVal", "chd8_qVal", "chd8_class",
                         "chd9_fc", "chd9_pVal", "chd9_qVal", "chd9_class",
                         "ep400_fc", "ep400_pVal", "ep400_qVal", "ep400_class",
                         "ezh2_fc", "ezh2_pVal", "ezh2_qVal", "ezh2_class",
                         "ring1_fc", "ring1_pVal", "ring1_qVal", "ring1_class",
                         "suz12_fc", "suz12_pVal", "suz12_qVal", "suz12_class",
                         "mll34_fc", "mll34_pVal", "mll34_qVal", "mll34_class",
                         "mll2_fc", "mll2_pVal", "mll2_qVal", "mll2_class",
                         "h2az_fc", "h2az_pVal", "h2az_qVal", "h2az_class")

gene_expr <- merge(gene_expr, df_esc[,c(4,10,40,49)], by.x="gene", by.y="gene", all=F)

write.table(gene_expr, "/home/pundhir/project/chip-seq-analysis/data/gene/esc/GENE_EXPR.MATRIX.CLUSTER", quote = F, col.names=T, row.names = F, sep="\t")

##----------------------------------------------------##
## Load file containing spatial profile of various chromatin modifers (CM) at gene promoters
##----------------------------------------------------##

profile <- read.table(pipe("less /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.profileTPM | grep -vw N1 | grep -vw N2 | grep -vw N3"), comment.char = "")
colnames(profile) <- c("chr", "start", "end", "name", "score", "strand", "description", "ash2l", "brg1", "cbx7", "chd1",
                       "chd2", "chd4", "chd6", "chd8", "chd9", "cxxc1", "ep400", "ezh2", "hdac1", "kdm5b",
                       "mll2", "mll34", "p300", "rad21", "ring1b", "set1a", "suz12", "h3k27ac", "h3k27me3", "h3k4me1",
                       "h3k4me3", "h3k9me3", "h3k36me3", "h2az", "h3k4me2", "h3k9ac", "brd4", "lsd1", "cdk8", "ctcf", 
                       "jarid2", "kdm2a", "med12", "med1", "nipbl", "smc1a", "tbp", "yy1", "h2azAC", "h3v3", "h3Only", 
                       "mbd3", "pol2", "myc", "max", "nelfa", "esrrb", "kdm5c", "cdk9", "cdk7", "paf1", "leo1", "cdc73",
                       "pol2_ser2p", "pol2_ser5p", "pol2_ser7p", "mnase", "atac")

class <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$description))
distance <- as.numeric(apply(profile[grep("HFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[7]))
name <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$name))
gene <- apply(profile[grep("HFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[4])
profile.lst <- lapply(profile[,c(8:ncol(profile))], function(x) as.data.frame(matrix(x, nrow=length(x)/8, byrow = T)))
profile.lst <- lapply(profile.lst, setNames, nm <- c("2up", "1up", "up", "hfr", "nfr", "down", "1down", "2down"))
profile.lst <- lapply(profile.lst, cbind, class)
profile.lst <- lapply(profile.lst, cbind, distance)
profile.lst <- lapply(profile.lst, cbind, gene)

lst <- lapply(names(profile.lst), function(x) { y <- cbind(profile.lst[[x]], 
                                                  (rowSums(profile.lst[[x]][,c(5:8)])-rowSums(profile.lst[[x]][,c(1:4)]))/
                                                    (rowSums(profile.lst[[x]][,c(5:8)])+rowSums(profile.lst[[x]][,c(1:4)])));
                                              colnames(y)[ncol(y)] <- sprintf("profile_directionality_%s", x); return(y); })
names(lst) <- names(profile.lst)
profile.lst <- lst

##----------------------------------------------------##
## Create file by merging enzyme signal and gene expression information
##----------------------------------------------------##

gene_enzyme <- merge(enzyme_tpm[,c(4,12:56,60,64:73)], gene_expr, by.x="name", by.y="gene")
gene_enzyme <- merge(df_esc[,c(1:6,41:48)], gene_enzyme, by.x="gene", by.y="name")
gene_enzyme <- gene_enzyme[,c(2:4,1,5:ncol(gene_enzyme))]
gene_enzyme <- gene_enzyme[which(gene_enzyme$cluster=="Y1" | gene_enzyme$cluster=="Y2"),]
gene_enzyme <- dplyr::arrange(gene_enzyme, cluster, promoter_class, desc(ep400))
gene_enzyme$cluster <- as.factor(as.character(gene_enzyme$cluster))
gene_enzyme$ep400_cluster <- "low"
gene_enzyme[which(gene_enzyme$ep400 > summary(gene_enzyme$pol2_wt)[2] & gene_enzyme$ep400 <= summary(gene_enzyme$pol2_wt)[5]),]$ep400_cluster <- "average"
gene_enzyme[which(gene_enzyme$ep400 > summary(gene_enzyme$pol2_wt)[5]),]$ep400_cluster <- "high"
# gene_enzyme$ep400_cluster <- ifelse(gene_enzyme$ep400 > summary(gene_enzyme$ep400)[3], "yes", "no")
# hist(gene_enzyme$ep400)
table(gene_enzyme[,c(134,133,132)])
gene_enzyme$CpG_cluster <- "small"
gene_enzyme[which(gene_enzyme$CpG_length > summary(gene_enzyme$CpG_length)[3] & gene_enzyme$CpG_length <= summary(gene_enzyme$CpG_length)[5]),]$CpG_cluster <- "moderate"
gene_enzyme[which(gene_enzyme$CpG_length > summary(gene_enzyme$CpG_length)[5]),]$CpG_cluster <- "big"

table(gene_enzyme[,c(135,133,132)])

gene_enzyme[which(gene_enzyme$gene=="Ep400" | gene_enzyme$gene=="Chd4" | gene_enzyme$gene=="Smarca4" | gene_enzyme$gene=="Chd8"),c(4,75,78,87,90,103,106,131:134)]
write.table(gene_enzyme, "/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/gene_enzyme.bed", quote = F, col.names=T, row.names = F, sep="\t")

##----------------------------------------------------##
## Create file by merging enzyme signal, gene expression, enhancer and pol2 pausing information together (gene_enzyme_enhancer)
##----------------------------------------------------##

df <- read.table((pipe("cat /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/brg1_ko/genes.bed.activeRP.tpm")))
colnames(df) <- c("chr", "start", "end", "name", "score", "strand", "ep400", "dist_down", "cluster", "promoter_class", "ep400_cluster", 
                  "h3k27ac_wt_E", "h3k27ac_wt_ws_E", "h3k27ac_ko_E", "h3k27ac_ko_ws_E",
                  "h3k27me3_wt_E", "h3k27me3_wt_ws_E", "h3k27me3_ko_E", "h3k27me3_ko_ws_E",
                  "h3k27ac_E", "h3k27ac_ws_E", "h3k27me3_E", "h3k27me3_ws_E", "brg1_wt_E", "brg1_wt_ws_E", "yy1_wt_E", "yy1_wt_ws_E",
                  "enhancer_count", "fire_rep1", "fire_rep2", "fire_rep3",
                  "h3k27ac_wt_rep1", "h3k27ac_wt_rep2", "h3k27ac_ko_rep1", "h3k27ac_ko_rep2",
                  "h3k27me3_wt_rep1", "h3k27me3_wt_rep2", "h3k27me3_ko_rep1", "h3k27me3_ko_rep2")
df <- merge(gene_enzyme, df[,c(4,12:39)], by.x="gene", by.y="name")
df <- df[,c(2:4,1,5:ncol(df))]

df$promoter_class_merged <- "bivalent"
df[which(df$cluster=="Y2"),]$promoter_class_merged <- as.character(df[which(df$cluster=="Y2"),]$promoter_class)
df$promoter_class_merged <- as.factor(df$promoter_class_merged)

df$ep400_cluster_merged <- "bivalent"
df[which(df$cluster=="Y2"),]$ep400_cluster_merged <- as.character(df[which(df$cluster=="Y2"),]$ep400_cluster)
df$ep400_cluster_merged <- factor(df$ep400_cluster_merged, levels=c("bivalent", "high", "average", "low"))

df <- merge(df, profile.lst$brg1[,c(11,12)], by.x="gene", by.y="gene")
df <- merge(df, profile.lst$chd4[,c(11,12)], by.x="gene", by.y="gene")
df <- merge(df, profile.lst$ep400[,c(11,12)], by.x="gene", by.y="gene")
df <- merge(df, profile.lst$ezh2[,c(11,12)], by.x="gene", by.y="gene")
gene_enzyme_enhancer <- df[,c(2:4,1,5:169)]
gene_enzyme_enhancer <- dplyr::arrange(gene_enzyme_enhancer, cluster, promoter_class, desc(ep400))
gene_enzyme_enhancer$directionality_class <- "0"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$profile_directionality_chd4 > 0.3),]$directionality_class <- "1"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$profile_directionality_chd4 < -0.3),]$directionality_class <- "-1"
gene_enzyme_enhancer$directionality_class_merged <- "bivalent_1"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y2"),]$directionality_class_merged <- "active_0"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$profile_directionality_chd4 > 0.3 & gene_enzyme_enhancer$cluster=="Y2"),]$directionality_class_merged <- "active_1"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$profile_directionality_chd4 < -0.3 & gene_enzyme_enhancer$cluster=="Y2"),]$directionality_class_merged <- "active_-1"
gene_enzyme_enhancer$directionality_class_merged <- factor(gene_enzyme_enhancer$directionality_class_merged, levels=c("bivalent_1", "active_-1", "active_0", "active_1"), ordered = T)
gene_enzyme_enhancer$directionality_class <- factor(gene_enzyme_enhancer$directionality_class, levels=c("-1", "0", "1"), ordered=T)

gene_enzyme_enhancer$ratio_chd4_brg1 <-  ((gene_enzyme_enhancer$chd4+1)/(gene_enzyme_enhancer$brg1+1))
gene_enzyme_enhancer$ratio_chd4_brg1_class <- "brg1"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$ratio_chd4_brg1 > summary(gene_enzyme_enhancer$ratio_chd4_brg1)[2] & gene_enzyme_enhancer$ratio_chd4_brg1 <= summary(gene_enzyme_enhancer$ratio_chd4_brg1)[5]),]$ratio_chd4_brg1_class <- "equal"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$ratio_chd4_brg1 > summary(gene_enzyme_enhancer$ratio_chd4_brg1)[5]),]$ratio_chd4_brg1_class <- "chd4"
table(gene_enzyme_enhancer$ratio_chd4_brg1_class)
table(gene_enzyme_enhancer$ep400_cluster_merged)
table(gene_enzyme_enhancer$directionality_class_merged)

## pol2 pausing data
df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_pol2pausing.bed")
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss", 
                  "pausing_brg1_wt_rep1", "pausing_brg1_wt_rep2",
                  "pausing_brg1_ko_rep1", "pausing_brg1_ko_rep2",
                  "pausing_mbd3_24h_rep1", "pausing_mbd3_24h_rep2", "pausing_mbd3_24h_rep3",
                  "pausing_mbd3_0h_rep1", "pausing_mbd3_0h_rep2", "pausing_mbd3_0h_rep3",
                  "pausing_wt_rep1", "pausing_wt_rep2", "pausing_brg1_kd_rep1", "pausing_brg1_kd_rep2",
                  "pausing_chd4_kd_rep1", "pausing_chd4_kd_rep2", "pausing_ep400_kd_rep1", "pausing_ep400_kd_rep2",
                  "pausing_wt_rep3", "pausing_brg1_kd_rep3", "pausing_chd4_kd_rep3")

df$gene_coor <- sprintf("%s:%d-%d", df$chr, df$start, df$end)
df$gene_length <- df$end - df$start
gene_enzyme_enhancer <- merge(gene_enzyme_enhancer[,c(1:173)], df[,c(4,8:ncol(df))], by.x="gene", by.y="gene", all.x=T)
gene_enzyme_enhancer <- gene_enzyme_enhancer[,c(2:4,1,5:ncol(gene_enzyme_enhancer))]

## pol2 pausing data (counts)
df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_pol2signal.bed")
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss", 
                  "pausing_brg1_wt_rep1", "pausing_brg1_wt_rep2",
                  "pausing_brg1_ko_rep1", "pausing_brg1_ko_rep2",
                  "pausing_mbd3_24h_rep1", "pausing_mbd3_24h_rep2", "pausing_mbd3_24h_rep3",
                  "pausing_mbd3_0h_rep1", "pausing_mbd3_0h_rep2", "pausing_mbd3_0h_rep3",
                  "pausing_wt_rep1", "pausing_wt_rep2", "pausing_brg1_kd_rep1", "pausing_brg1_kd_rep2",
                  "pausing_chd4_kd_rep1", "pausing_chd4_kd_rep2", "pausing_ep400_kd_rep1", "pausing_ep400_kd_rep2",
                  "pausing_wt_rep3", "pausing_brg1_kd_rep3", "pausing_chd4_kd_rep3")

## REPLICATE1
mat <- t(apply(df[,c(11:ncol(df))], 1, function(x) as.numeric(gsub("^.*/", "", x))))
res.brg1 <- matrix2deseq2(mat[,c(1:2,3:4)], c("wt,wt,ko,ko"))
res.chd4 <- matrix2deseq2(mat[,c(6:7,9:10)], c("wt,wt,ko,ko"))
df$pol2_fc_gb_Brg1_rep1 <- res.brg1$log2FoldChange
df$pol2_fc_gb_Chd4_rep1 <- res.chd4$log2FoldChange
mat <- t(apply(df[,c(11:ncol(df))], 1, function(x) as.numeric(gsub("/.*$", "", x))))
res.brg1 <- matrix2deseq2(mat[,c(1:2,3:4)], c("wt,wt,ko,ko"))
res.chd4 <- matrix2deseq2(mat[,c(6:7,9:10)], c("wt,wt,ko,ko"))
df$pol2_fc_tssr_Brg1_rep1 <- res.brg1$log2FoldChange
df$pol2_fc_tssr_Chd4_rep1 <- res.chd4$log2FoldChange

## REPLICATE2
mat <- t(apply(df[,c(11:ncol(df))], 1, function(x) as.numeric(gsub("^.*/", "", x))))
res.brg1 <- matrix2deseq2(mat[,c(11:12,13:14)], c("wt,wt,ko,ko"))
res.chd4 <- matrix2deseq2(mat[,c(11:12,15:16)], c("wt,wt,ko,ko"))
df$pol2_fc_gb_Brg1_rep2 <- res.brg1$log2FoldChange
df$pol2_fc_gb_Chd4_rep2 <- res.chd4$log2FoldChange
mat <- t(apply(df[,c(11:ncol(df))], 1, function(x) as.numeric(gsub("/.*$", "", x))))
res.brg1 <- matrix2deseq2(mat[,c(11:12,13:14)], c("wt,wt,ko,ko"))
res.chd4 <- matrix2deseq2(mat[,c(11:12,15:16)], c("wt,wt,ko,ko"))
df$pol2_fc_tssr_Brg1_rep2 <- res.brg1$log2FoldChange
df$pol2_fc_tssr_Chd4_rep2 <- res.chd4$log2FoldChange

gene_enzyme_enhancer <- merge(gene_enzyme_enhancer[,c(1:ncol(gene_enzyme_enhancer))], df[,c(4,32:39)], by.x="gene", by.y="gene", all.x=T)
gene_enzyme_enhancer <- gene_enzyme_enhancer[,c(2:4,1,5:ncol(gene_enzyme_enhancer))]

## GRO-seq data (directionality)
df <- read.table(pipe("cat /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.GENE.groseqDirectionality | cut -f 1-12"))
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "dist_tss_down", "cluster", 
                  "groseq_wt_down_rep1", "groseq_wt_up_rep1", "groseq_wt_down_rep2", "groseq_wt_up_rep2")
NORM_METHOD="qu"
df$groseq_directionality_rep1 <- apply(scale_df(df[,c(9:10)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
df$groseq_directionality_rep2 <- apply(scale_df(df[,c(11:12)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
gene_enzyme_enhancer <- merge(gene_enzyme_enhancer[,c(1:ncol(gene_enzyme_enhancer))], df[,c(4,13,14)], by.x="gene", by.y="gene")
gene_enzyme_enhancer <- gene_enzyme_enhancer[,c(2:4,1,5:ncol(gene_enzyme_enhancer))]
gene_enzyme_enhancer$groseq_class <- "0"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$groseq_directionality_rep2 < -0.35),]$groseq_class <- "-1"
gene_enzyme_enhancer[which(gene_enzyme_enhancer$groseq_directionality_rep2 > 0.35),]$groseq_class <- "1"
gene_enzyme_enhancer$groseq_class <- factor(gene_enzyme_enhancer$groseq_class, levels=c("-1", "0", "1"), ordered=T)
table(gene_enzyme_enhancer$groseq_class)

write.table(gene_enzyme_enhancer, "/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/gene_enzyme_enhancer.bed", quote = F, col.names=T, row.names = F, sep="\t")

## GRO-seq data (pausing)

# df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_divergentTranscription.bed")
# colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss", 
#                   "groseq_wt_rep1", "groseq_wt_rep2", 
#                   "cage_wt_rep1", "cage_wt_rep2", "cage_rrp40_rep1", "cage_rrp40_rep2", "cage_rbm7_rep1", "cage_rbm7_rep2",
#                   "proseq_wt_rep1", "proseq_wt_rep2", "proseq_rrp40_rep1", "proseq_rrp40_rep2", "proseq_rbm7_rep1", "proseq_rbm7_rep2",
#                   "nasseq_wt_rep1", "nasseq_wt_rep2", "nasseq_wt_rep3", "nasseq_kd_rep1", "nasseq_kd_rep2", "nasseq_kd_rep3")
# mat <- t(apply(df[,c(25:30)], 1, function(x) as.numeric(gsub("^.*/", "", x))))
# res.chd4 <- matrix2deseq2(mat, c("wt,wt,wt,ko,ko,ko"))
# df$groseq_fc_gb_Chd4 <- res.chd4$log2FoldChange
# hist(res.chd4[which(res.chd4$padj<0.05),]$log2FoldChange)
# EnhancedVolcano::EnhancedVolcano(res.chd4, lab="", x="log2FoldChange", y="padj", ylim = c(0,1))
# 
# mat <- t(apply(df[,c(25:30)], 1, function(y) unlist(lapply(y, function(x) sum(as.numeric(unlist(strsplit(as.character(x), "/"))[1:2]))))))
# res.chd4 <- matrix2deseq2(mat, c("wt,wt,wt,ko,ko,ko"))
# df$groseq_fc_tssr_Chd4 <- res.chd4$log2FoldChange
# 
# EnhancedVolcano::EnhancedVolcano(res.chd4, lab="", x="log2FoldChange", y="padj", ylim = c(0,1))

##----------------------------------------------------##
## Create file by merging enzyme signal, gene expression and divergent transcript information together (gene_enzyme_divergent)
##----------------------------------------------------##

df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_divergentTranscription.bed")
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss", 
                  "groseq_wt_rep1", "groseq_wt_rep2", 
                  "cage_wt_rep1", "cage_wt_rep2", "cage_rrp40_rep1", "cage_rrp40_rep2", "cage_rbm7_rep1", "cage_rbm7_rep2",
                  "proseq_wt_rep1", "proseq_wt_rep2", "proseq_rrp40_rep1", "proseq_rrp40_rep2", "proseq_rbm7_rep1", "proseq_rbm7_rep2",
                  "nasseq_wt_rep1", "nasseq_wt_rep2", "nasseq_wt_rep3", "nasseq_kd_rep1", "nasseq_kd_rep2", "nasseq_kd_rep3")
revTrans <- apply(df[,c(11:30)], 2, function(x) groseq2revTrans(x))
colnames(revTrans) <- sprintf("%s_revTrans", colnames(revTrans))
pausing <- apply(df[,c(11:30)], 2, function(x) groseq2pausing(x))
colnames(pausing) <- sprintf("%s_pausing", colnames(pausing))
df <- cbind(df[,c(1:10),], revTrans, pausing)

                  
# 
# df <- read.table("promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.GENE.groseqDirectionality")
# colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "dist_tss_down", "cluster", 
#                   "groseq_wt_down_rep1", "groseq_wt_up_rep1", "groseq_wt_down_rep2", "groseq_wt_up_rep2",
#                   "cage_wt_down_rep1", "cage_wt_up_rep1", "cage_wt_down_rep2", "cage_wt_up_rep2",
#                   "cage_rrp40_down_rep1", "cage_rrp40_up_rep1", "cage_rrp40_down_rep2", "cage_rrp40_up_rep2",
#                   "cage_rbm7_down_rep1", "cage_rbm7_up_rep1", "cage_rbm7_down_rep2", "cage_rbm7_up_rep2",
#                   "proseq_wt_down_rep1", "proseq_wt_up_rep1", "proseq_wt_down_rep2", "proseq_wt_up_rep2",
#                   "proseq_rrp40_down_rep1", "proseq_rrp40_up_rep1", "proseq_rrp40_down_rep2", "proseq_rrp40_up_rep2",
#                   "proseq_rbm7_down_rep1", "proseq_rbm7_up_rep1", "proseq_rbm7_down_rep2", "proseq_rbm7_up_rep2")
# NORM_METHOD="qu"
# df$groseq_directionality_rep1 <- apply(scale_df(df[,c(9:10)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$groseq_directionality_rep2 <- apply(scale_df(df[,c(11:12)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$cage_directionality_rep1 <- apply(scale_df(df[,c(13:14)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$cage_directionality_rep2 <- apply(scale_df(df[,c(15:16)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$cage_directionality_rrp40_rep1 <- apply(scale_df(df[,c(17:18)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$cage_directionality_rrp40_rep2 <- apply(scale_df(df[,c(19:20)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$cage_directionality_rbm7_rep1 <- apply(scale_df(df[,c(21:22)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$cage_directionality_rbm7_rep2 <- apply(scale_df(df[,c(23:24)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$proseq_directionality_rep1 <- apply(scale_df(df[,c(25:26)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$proseq_directionality_rep2 <- apply(scale_df(df[,c(27:28)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$proseq_directionality_rrp40_rep1 <- apply(scale_df(df[,c(29:30)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$proseq_directionality_rrp40_rep2 <- apply(scale_df(df[,c(31:32)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$proseq_directionality_rbm7_rep1 <- apply(scale_df(df[,c(33:34)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))
# df$proseq_directionality_rbm7_rep2 <- apply(scale_df(df[,c(35:36)], NORM_METHOD), 1, function(x) (x[1]-x[2])/(x[1]+x[2]))

# gene_enzyme_divergent <- merge(gene_enzyme_enhancer[,c(1:207)], df[,c(4,9:ncol(df))], by.x="gene", by.y="gene")
gene_enzyme_divergent <- merge(gene_enzyme_enhancer[,c(1:207)], df[,c(4,11:ncol(df))], by.x="gene", by.y="gene")
gene_enzyme_divergent <- gene_enzyme_divergent[,c(2:4,1,5:ncol(gene_enzyme_divergent))]
gene_enzyme_divergent$groseq_class <- "0"
gene_enzyme_divergent[which(gene_enzyme_divergent$groseq_wt_rep1_revTrans > 0.7),]$groseq_class <- "-1"
gene_enzyme_divergent[which(gene_enzyme_divergent$groseq_wt_rep1_revTrans < 0.3),]$groseq_class <- "1"
gene_enzyme_divergent$groseq_class <- factor(gene_enzyme_divergent$groseq_class, levels=c("-1", "0", "1"), ordered=T)
table(gene_enzyme_divergent$groseq_class)

## control for similar expression across the three chd4 classes (-1, 0 and 1)
test <- gene_enzyme_divergent[which(gene_enzyme_divergent$cluster=="Y2"),]
test.gene <- rbind(subSample_median(test[which(test$directionality_class=="-1"),], test[which(test$directionality_class=="1"),10], 500, 10),
                   subSample_median(test[which(test$directionality_class=="0"),], test[which(test$directionality_class=="1"),10], 500, 10),
                   test[which(test$directionality_class=="1"),])
test.pol2 <- rbind(subSample_median(test[which(test$directionality_class=="-1"),], test[which(test$directionality_class=="1"),61], 500, 61),
                   subSample_median(test[which(test$directionality_class=="0"),], test[which(test$directionality_class=="1"),61], 500, 61),
                   test[which(test$directionality_class=="1"),])
test.h3k27ac <- rbind(subSample_median(test[which(test$directionality_class=="-1"),], test[which(test$directionality_class=="1"),36], 500, 36),
                    subSample_median(test[which(test$directionality_class=="0"),], test[which(test$directionality_class=="1"),36], 500, 36),
                    test[which(test$directionality_class=="1"),])
test.h3k4me1 <- rbind(subSample_median(test[which(test$directionality_class=="-1"),], test[which(test$directionality_class=="1"),39], 500, 39),
                    subSample_median(test[which(test$directionality_class=="0"),], test[which(test$directionality_class=="1"),39], 500, 39),
                    test[which(test$directionality_class=="1"),])
#test.sub <- test.gene[test.gene$gene %in% test.pol2$gene,]
test.sub <- test.h3k27ac[test.h3k27ac$gene %in% test.pol2$gene,]

# test.sub <- test.sub[which(test.sub$dist_closest_tss>1000),]

## control by randomly selecting identical number of promoters from the three chd4 classes (-1, 0 and 1)
# test.sub <- rbind(test.sub[which(test.sub$directionality_class=="-1"),][sample(nrow(test.sub[which(test.sub$directionality_class=="-1"),]),479),],
#                   test.sub[which(test.sub$directionality_class=="0"),][sample(nrow(test.sub[which(test.sub$directionality_class=="0"),]),479),],
#                   test.sub[which(test.sub$directionality_class=="1"),])

# test.sub <- rbind(test.sub[which(test.sub$directionality_class_merged=="active_-1"),], 
#                   sample_df(test.sub[which(test.sub$directionality_class_merged=="active_0"),], 258),
#                   sample_df(test.sub[which(test.sub$directionality_class_merged=="active_1"),], 258))
table(test.sub$directionality_class_merged)

gene_enzyme_divergent$equalExpr <- "NotNorm"
gene_enzyme_divergent[which(gene_enzyme_divergent$gene %in% test.sub$gene),]$equalExpr <- "ExprNorm"

write.table(gene_enzyme_divergent, "/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/gene_enzyme_divergent.bed", quote = F, col.names=T, row.names = F, sep="\t")

##----------------------------------------------------##
## Create file by merging chd4/brg1 directionality changes with gro-seq directionality in esc and mef
##----------------------------------------------------##

## Step-A ##
esc_mef <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/brg1_esc_mef.bed")
colnames(esc_mef) <- c("chr", "start", "end", "gene", "score", "strand", "expr_esc", "expr_mef", "promoter_class", "directionality_esc", "directionality_esc_class", "directionality_mef", "directionality_mef_class", "diff")

esc_mef$fc <- log2((esc_mef$expr_mef+1)/(esc_mef$expr_esc+1))

#-- approach-I --#
esc_mef$directionality_class <- "0_0"
esc_mef[which(esc_mef$diff < -0.3),]$directionality_class <- "-1_1"
esc_mef[which(esc_mef$diff > 0.3),]$directionality_class <- "1_-1"

#-- approach-II --#
# esc_mef$diff <- rank(esc_mef$directionality_esc)-rank(esc_mef$directionality_mef)
# esc_mef$directionality_class <- "0_0"
# esc_mef[which(esc_mef$diff < -7000),]$directionality_class <- "-1_1"
# esc_mef[which(esc_mef$diff > 7000),]$directionality_class <- "1_-1"

#-- approach-III --#
esc_mef$diff <- rank(esc_mef$directionality_esc)-rank(esc_mef$directionality_mef)
esc_mef$rank_esc <- rank(esc_mef$directionality_esc)
esc_mef$rank_mef <- rank(esc_mef$directionality_mef)
summary(esc_mef[,c(17,18)])
esc_mef$directionality_class <- "0_0"
# esc_mef[which(esc_mef$rank_esc < 6100 & esc_mef$rank_mef > 18300),]$directionality_class <- "-1_1"
# esc_mef[which(esc_mef$rank_esc > 18300 & esc_mef$rank_mef < 6100),]$directionality_class <- "1_-1"
esc_mef[which(esc_mef$rank_esc < 9000 & esc_mef$rank_mef > 15446),]$directionality_class <- "-1_1"
esc_mef[which(esc_mef$rank_esc > 15446 & esc_mef$rank_mef < 9000),]$directionality_class <- "1_-1"

#-- approach-IV --#
# esc_mef$directionality_class <- "0_0"
# esc_mef[which(esc_mef$directionality_esc_class == -1 & esc_mef$directionality_mef_class == 1),]$directionality_class <- "-1_1"
# esc_mef[which(esc_mef$directionality_esc_class == 1 & esc_mef$directionality_mef_class == -1),]$directionality_class <- "1_-1"

esc_mef <- esc_mef[,c(1:16)]
esc_mef$directionality_class <- factor(esc_mef$directionality_class, levels=c("-1_1", "0_0", "1_-1"), ordered=T)
table(esc_mef$directionality_class)

## Step-B ##
groseq_esc <- read.table(pipe("cat /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.GENE.groseqDirectionality | cut -f 1-12"))
colnames(groseq_esc) <- c("chr", "start", "end", "gene", "score", "strand", "dist_tss_down", "cluster_esc", 
                  "groseq_wt_down_rep1_esc", "groseq_wt_up_rep1_esc", "groseq_wt_down_rep2_esc", "groseq_wt_up_rep2_esc")

groseq_esc$groseq_directionality_rep1_esc <- 0
THRESHOLD <- summary(groseq_esc$groseq_wt_down_rep1_esc)[3]
groseq_esc[which(groseq_esc$groseq_wt_down_rep1_esc > THRESHOLD),]$groseq_directionality_rep1_esc <- groseq_esc[which(groseq_esc$groseq_wt_down_rep1_esc > THRESHOLD),]$groseq_wt_up_rep1_esc/(groseq_esc[which(groseq_esc$groseq_wt_down_rep1_esc > THRESHOLD),]$groseq_wt_up_rep1_esc + groseq_esc[which(groseq_esc$groseq_wt_down_rep1_esc > THRESHOLD),]$groseq_wt_down_rep1_esc)

groseq_esc$groseq_directionality_rep2_esc <- 0
THRESHOLD <- summary(groseq_esc$groseq_wt_down_rep2_esc)[3]
groseq_esc[which(groseq_esc$groseq_wt_down_rep2_esc > THRESHOLD),]$groseq_directionality_rep2_esc <- groseq_esc[which(groseq_esc$groseq_wt_down_rep2_esc > THRESHOLD),]$groseq_wt_up_rep2_esc/(groseq_esc[which(groseq_esc$groseq_wt_down_rep2_esc > THRESHOLD),]$groseq_wt_up_rep2_esc + groseq_esc[which(groseq_esc$groseq_wt_down_rep2_esc > THRESHOLD),]$groseq_wt_down_rep2_esc)

groseq_esc$groseq_class_esc <- "bidirectional"
groseq_esc[which(groseq_esc$groseq_directionality_rep2_esc <= 0.1),]$groseq_class_esc <- "unidirectional"

## Step-C ##
groseq_mef <- read.table(pipe("cat /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/mef/GENES_PROMOTER.BED.RC.CLUSTERS.GENE.groseqDirectionality | cut -f 1-14"))
colnames(groseq_mef) <- c("chr", "start", "end", "gene", "score", "strand", "dist_tss_down", "cluster_mef", 
                          "groseq_wt_down_rep1_mef", "groseq_wt_up_rep1_mef", "groseq_wt_down_rep2_mef", "groseq_wt_up_rep2_mef", 
                          "groseq_wt_down_rep3_mef", "groseq_wt_up_rep3_mef")

groseq_mef$groseq_directionality_rep1_mef <- 0
THRESHOLD <- summary(groseq_mef$groseq_wt_down_rep1_mef)[3]
groseq_mef[which(groseq_mef$groseq_wt_down_rep1_mef > THRESHOLD),]$groseq_directionality_rep1_mef <- groseq_mef[which(groseq_mef$groseq_wt_down_rep1_mef > THRESHOLD),]$groseq_wt_up_rep1_mef/(groseq_mef[which(groseq_mef$groseq_wt_down_rep1_mef > THRESHOLD),]$groseq_wt_up_rep1_mef + groseq_mef[which(groseq_mef$groseq_wt_down_rep1_mef > THRESHOLD),]$groseq_wt_down_rep1_mef)

groseq_mef$groseq_directionality_rep2_mef <- 0
THRESHOLD <- summary(groseq_mef$groseq_wt_down_rep2_mef)[3]
groseq_mef[which(groseq_mef$groseq_wt_down_rep2_mef > THRESHOLD),]$groseq_directionality_rep2_mef <- groseq_mef[which(groseq_mef$groseq_wt_down_rep2_mef > THRESHOLD),]$groseq_wt_up_rep2_mef/(groseq_mef[which(groseq_mef$groseq_wt_down_rep2_mef > THRESHOLD),]$groseq_wt_up_rep2_mef + groseq_mef[which(groseq_mef$groseq_wt_down_rep2_mef > THRESHOLD),]$groseq_wt_down_rep2_mef)

groseq_mef$groseq_directionality_rep3_mef <- 0
THRESHOLD <- summary(groseq_mef$groseq_wt_down_rep3_mef)[3]
groseq_mef[which(groseq_mef$groseq_wt_down_rep3_mef > THRESHOLD),]$groseq_directionality_rep3_mef <- groseq_mef[which(groseq_mef$groseq_wt_down_rep3_mef > THRESHOLD),]$groseq_wt_up_rep3_mef/(groseq_mef[which(groseq_mef$groseq_wt_down_rep3_mef > THRESHOLD),]$groseq_wt_up_rep3_mef + groseq_mef[which(groseq_mef$groseq_wt_down_rep3_mef > THRESHOLD),]$groseq_wt_down_rep3_mef)

groseq_mef$groseq_class_mef <- "bidirectional"
groseq_mef[which(groseq_mef$groseq_directionality_rep3_mef <= 0.1),]$groseq_class_mef <- "unidirectional"

## Step-D ##
esc_mef <- merge(merge(esc_mef, groseq_esc[,c(4,8:15)], by.x="gene", by.y="gene"), groseq_mef[,c(4,8:18)])
esc_mef <- esc_mef[,c(2:4,1,5:35)]

## Step-E ##
gene_esc <- read.table("/home/pundhir/project/chip-seq-analysis/data/gene/esc/mll34/genes_esc_specific.txt")
gene_mef <- read.table("/home/pundhir/project/chip-seq-analysis/data/gene/esc/mll34/genes_mef_specific.txt")

esc_mef$gene_specificity <- "none"
esc_mef[esc_mef$gene %in% gene_esc$V1,]$gene_specificity <- "esc"
esc_mef[esc_mef$gene %in% gene_mef$V1,]$gene_specificity <- "mef"

esc_mef$promoter_class <- as.character("none")
#esc_mef[which((grepl("Y", esc_mef$cluster_esc) | grepl("Y", esc_mef$cluster_mef)) & !is.na(esc_mef$expr_esc) & !is.na(esc_mef$expr_mef)),]$promoter_class <- "sig"
esc_mef[which((esc_mef$expr_esc > 1 | esc_mef$expr_mef > 1) & !is.na(esc_mef$expr_esc) & !is.na(esc_mef$expr_mef) & abs(esc_mef$fc) > 1),]$promoter_class <- "sig"
# esc_mef[which((esc_mef$expr_esc > 1 | esc_mef$expr_mef > 1) & !is.na(esc_mef$expr_esc) & !is.na(esc_mef$expr_mef)),]$promoter_class <- "sig"
table(esc_mef$promoter_class)
table(esc_mef[which(esc_mef$promoter_class=="sig"),]$gene_specificity)

esc_mef$directionality_esc_class <- as.factor(esc_mef$directionality_esc_class)
esc_mef$directionality_mef_class <- as.factor(esc_mef$directionality_mef_class)

write.table(esc_mef, "/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/brg1_esc_mef.bed.sig", quote = F, col.names=T, row.names = F, sep="\t")

##----------------------------------------------------##
## ANALYSIS IN MEF CELLS
##----------------------------------------------------##

##----------------------------------------------------##
## Load gene promoter file (mef)
##----------------------------------------------------##

df_mef <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/mef/GENES_PROMOTER.BED.RC.CLUSTERS.GENE")
colnames(df_mef) <- c("chr", "start", "end", "gene", "score", "strand", "promoter_defined", 
                  "promoter", "dist_tss_up", "dist_tss_down",
                  "h3k27ac_promoter", "h3k27me3_promoter", "h3k4me1_promoter", "h3k4me3_promoter",
                  "h3k9me3_promoter", "h3k36me3_promoter", "atac_promoter", 
                  "h3k27ac_distal", "h3k27me3_distal", "h3k4me1_distal", "h3k4me3_distal",
                  "h3k9me3_distal", "h3k36me3_distal", "atac_distal", 
                  "h3k27ac_distal_ws", "h3k27me3_distal_ws", "h3k4me1_distal_ws", "h3k4me3_distal_ws",
                  "h3k9me3_distal_ws", "h3k36me3_distal_ws", "atac_distal_ws",
                  "h3k27ac_NA", "h3k27me3_NA", "h3k4me1_NA", "h3k4me3_NA",
                  "h3k9me3_NA", "h3k36me3_NA", "atac_NA",
                  "enhancer_active", "enhancer_poised", "enhancer_primed", "promoter_active", 
                  "promoter_bivalent", "repressed_heterochromatin", "repressed_polycomb", "transcribed",
                  "cluster", "expr_mef", "expr_mef48", "expr_ipsc", "expr_esc", "closest_tss", "CpG_length", "perCpG")

df_mef[which(df_mef$closest_tss=="."),]$closest_tss <- NA
df_mef$closest_tss <- as.numeric(df_mef$closest_tss)
hk_genes <- read.table("~/project/genome_annotations/mouse_house_keeping_genes.txt")
df_mef$house_keeping <- ifelse(df_mef$gene %in% hk_genes$V1, 1, 0)

df_mef$promoter_class <- "narrow"
df_mef[which(df_mef$dist_tss_down >500 & df_mef$dist_tss_down <=1000),]$promoter_class <- "medium"
df_mef[which(df_mef$dist_tss_down >1000),]$promoter_class <- "broad"
df_mef$promoter_class <- factor(df_mef$promoter_class, levels=promoter_class_order, ordered = T)

##----------------------------------------------------##
## Load file containing normalized signal (RPKM) for various chromatin modifers (CM) at gene promoter
##----------------------------------------------------##

enzyme_tpm_mef <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/mef/GENES_PROMOTER.BED.RC.CLUSTERS.TPM")
colnames(enzyme_tpm_mef) <- c("chr_mef", "start_mef", "end_mef", "name_mef", "score_mef", "strand_mef", "promoter_defined_mef", "description_mef", 
                          "dist_tss_up_mef", "dist_tss_down_mef", "class_mef", "brg1_mef", "chd4_mef", "mbd3_mef", "h3k27ac_mef", "h3k27me3_mef",
                          "h3k4me1_mef", "h3k4me3_mef", "h3k9me3_mef", "h3k36me3_mef", "h2az_mef", "h3k4me2_mef", "h3k9ac_mef", "atac_mef", 
                          "pol2_wt_mef", "pol2_ser5p_mef", "pol2_ser2p_mef", "h3v3_mef", "ezh2_mef")
gene_enzyme_mef <- merge(enzyme_tpm_mef, df_esc[,c(4,41:44,45:48)], by.x="name_mef", by.y="gene")
gene_enzyme_mef <- gene_enzyme_mef[,c(2:4,1,5:ncol(gene_enzyme_mef))]

## pol2 pausing data
df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/mef/allGenes_pol2pausing.bed")
colnames(df) <- c("chr_mef", "start_mef", "end_mef", "gene_mef", "score_mef", "strand_mef", "id_mef", "TSSR_mef", "pol2_TSSR_mef", 
                  "dist_closest_tss_mef", "pausing_wt_rep1_mef")

df$gene_coor <- sprintf("%s:%d-%d", df$chr, df$start, df$end)
df$gene_length <- df$end - df$start
gene_enzyme_mef <- merge(gene_enzyme_mef[,c(1:36)], df[,c(4,11)], by.x="name_mef", by.y="gene_mef", all.x=T)

gene_enzyme_mef <- gene_enzyme_mef[which(!is.na(gene_enzyme_mef$expr_mef)),]
table(gene_enzyme_mef$class_mef)
# gene_enzyme_mef <- gene_enzyme_mef[which(gene_enzyme_mef$class_mef=="Y1"),]

## gene promoter class
gene_enzyme_mef$class_mef_broad <- "other"
gene_enzyme_mef[which(gene_enzyme_mef$class_mef=="N3"),]$class_mef_broad <- "repressed"
gene_enzyme_mef[which(gene_enzyme_mef$class_mef=="N4" | 
                        gene_enzyme_mef$class_mef=="Y1"),]$class_mef_broad <- "active"
table(gene_enzyme_mef$class_mef_broad)

## gene expr class
gene_enzyme_mef$expr_cluster_mef <- "NA"
gene_enzyme_mef[which(gene_enzyme_mef$pol2_wt_mef <= summary(gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad!="other"),]$pol2_wt_mef)[2]
                      ),]$expr_cluster_mef <- "low"
gene_enzyme_mef[which(gene_enzyme_mef$pol2_wt_mef > summary(gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad!="other"),]$pol2_wt_mef)[2] &
                        gene_enzyme_mef$pol2_wt_mef <= summary(gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad!="other"),]$pol2_wt_mef)[5]
                      ),]$expr_cluster_mef <- "average"
gene_enzyme_mef[which(gene_enzyme_mef$pol2_wt_mef > summary(gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad!="other"),]$pol2_wt_mef)[5]
                      ),]$expr_cluster_mef <- "high"
gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad=="other"),]$expr_cluster_mef <- "other"
gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad=="repressed"),]$expr_cluster_mef <- "repressed"
table(gene_enzyme_mef$expr_cluster_mef)

# gene_enzyme_mef$expr_cluster_mef <- "low"
# gene_enzyme_mef[which(gene_enzyme_mef$expr_mef > summary(gene_enzyme_mef$expr_mef)[2] & gene_enzyme_mef$expr_mef <= summary(gene_enzyme_mef$expr_mef)[5]),]$expr_cluster_mef <- "average"
# gene_enzyme_mef[which(gene_enzyme_mef$expr_mef > summary(gene_enzyme_mef$expr_mef)[5]),]$expr_cluster_mef <- "high"
# table(gene_enzyme_mef$expr_cluster_mef)
# gene_enzyme_mef[which(gene_enzyme_mef$class_mef=="N3"),]$expr_cluster_mef <- "repressed"
# 
# gene_enzyme_mef$expr_cluster_esc <- "low"
# gene_enzyme_mef[which(gene_enzyme_mef$expr_esc > summary(gene_enzyme_mef$expr_esc)[2] & gene_enzyme_mef$expr_esc <= summary(gene_enzyme_mef$expr_esc)[5]),]$expr_cluster_esc <- "average"
# gene_enzyme_mef[which(gene_enzyme_mef$expr_esc > summary(gene_enzyme_mef$expr_esc)[5]),]$expr_cluster_esc <- "high"
# table(gene_enzyme_mef$expr_cluster_esc)

gene_enzyme_mef$fc <- log2((gene_enzyme_mef$expr_mef+1)/(gene_enzyme_mef$expr_esc+1))

gene_enzyme_mef <- merge(gene_enzyme_mef, enzyme_tpm, by.x="name_mef", by.y="name")
gene_enzyme_mef <- gene_enzyme_mef[,c(2:4,1,5:ncol(gene_enzyme_mef))]

# gene_enzyme_mef$ratio_chd4_brg1_mef <-  ((gene_enzyme_mef$chd4_mef+1)/(gene_enzyme_mef$brg1_mef+1))
# gene_enzyme_mef$ratio_chd4_brg1 <-  ((gene_enzyme_mef$chd4+1)/(gene_enzyme_mef$brg1+1))

# gene_enzyme_mef[grep("2410004B18Rik", gene_enzyme_mef$name_mef),c("name_mef", "chd4", "brg1", "pol2_ser5p")]
# gene_enzyme_enhancer[grep("2410004B18Rik", gene_enzyme_enhancer$gene), c("gene", "chd4", "brg1", "pol2_ser5p")]
# enzyme_tpm[grep("2410004B18Rik", enzyme_tpm$name), c("name", "chd4", "brg1", "pol2_ser5p")]
```

######################################################
#### PICK RANDOM POL2 REGIONS FOR CHIP-QPCR
######################################################
```{r}
corrplot(cor(gene_enzyme_enhancer[,c("pol2_wt", "pol2_ser5p", "pol2_ser2p")]), method="pie", type="upper")
test <- gene_enzyme_enhancer[order(-gene_enzyme_enhancer$pol2_wt),c("chr", "start", "end", "gene", "pol2_wt", "pol2_ser5p", "pol2_ser2p")]
high <- test[which(test$pol2_wt > summary(test$pol2_wt)[5]),]
set.seed(6); high[sample(nrow(high),5),]
med <- test[which(test$pol2_wt > summary(test$pol2_wt)[3] & test$pol2_wt <= summary(test$pol2_wt)[5]),]
set.seed(6); med[sample(nrow(med),5),]
low <- test[which(test$pol2_wt <= summary(test$pol2_wt)[3]),]
set.seed(6); low[sample(nrow(low),5),]
neg <- df_esc[which(df_esc$promoter_defined=="N" & df_esc$expr_esc<=0),c("chr", "start", "end", "gene")]
set.seed(15); neg[sample(nrow(neg),10),]
#dim(test[which(test$pol2_wt <= summary(test$pol2_wt)[2]),])

#-----------------------------------------#
## specificity of ESC and MEF-specific genes (Plath data) with our gene clusters (bivalent, high, average, low)
table(gene_enzyme_enhancer[gene_enzyme_enhancer$gene %in% gene_esc$V1,]$ep400_cluster_merged)
 barplot(table(gene_enzyme_enhancer[gene_enzyme_enhancer$gene %in% gene_esc$V1,]$ep400_cluster_merged))
```

######################################################
#### PRIMARY ANALYSIS AT PROMOTERS (esc)
######################################################
```{r}
##----------------------------------------------------##
## gene promoters: expression, house keeping genes and closest gene analysis (esc)
##----------------------------------------------------##

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_esc_stats.pdf")
par(mfrow=c(4,1))
color_esc = c(rep("red", as.data.frame(table(df_esc$cluster))[1,2]), rep("green", as.data.frame(table(df_esc$cluster))[2,2]), rep("red", as.data.frame(table(df_esc$cluster))[3,2]), rep("green", as.data.frame(table(df_esc$cluster))[4,2]), rep("red", as.data.frame(table(df_esc$cluster))[5,2]))
barplot(df_esc$expr_esc, xlab="genes", ylab="gene expression", border = color_esc, col=color_esc)
image(as.matrix(df_esc$expr_esc), col=rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.001)(256)))

barplot(df_esc$house_keeping, xlab="genes", ylab="house keeping", border = NA)
# df.melt <- melt(as.data.frame(split(df_esc[which(df_esc$cluster=="Y2"),]$house_keeping, ceiling(seq_along(df_esc[which(df_esc$cluster=="Y2"),]$house_keeping)/2))))
# remainder(length(df_esc[which(df_esc$cluster=="Y2"),]$house_keeping),1:1000)
# ggplot(df.melt, aes(variable, value))  + geom_boxplot()

barplot(df_esc$closest_tss, border = color_esc, col=color_esc, xlab="genes", ylab="dist to closest TSS")
df_esc.melt <- melt(df_esc[,c(40,45)])
ggplot(df_esc.melt, aes(cluster, log(value))) + geom_boxplot() + theme_bw() + ylab("dist to closest TSS (log)")


plot(log(df_esc[which(df_esc$cluster=="Y2" & !is.na(df_esc$expr_esc)),]$expr_esc+1), df_esc[which(df_esc$cluster=="Y2" & !is.na(df_esc$expr_esc)),]$dist_tss_down)

test <- melt(df_esc[,c(40,49,44)])
ggplot(test, aes(log(value+1), color=promoter_class)) + stat_ecdf() + facet_grid(.~cluster)

par(mfrow=c(2,1))
barplot(df_esc$CpG_length, xlab="genes", ylab="CpG island length", border = color_esc, col=color_esc)
image(as.matrix(df_esc$CpG_length), col=rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.2)(256)))
dev.off()

##----------------------------------------------------##
## gene promoters: normalized signal (RPKM) of variuos chromatin modifiers (CM)
##----------------------------------------------------##

set.seed(1)
n1 <- enzyme_tpm[which(enzyme_tpm$class=="N1"),]
y2 <- enzyme_tpm[which(enzyme_tpm$class=="Y2"),]
test <- rbind(enzyme_tpm[which(enzyme_tpm$class=="Y1"),], 
              n1[sample(nrow(n1), length(enzyme_tpm[which(enzyme_tpm$class=="Y1"),]$class)),],
              y2[sample(nrow(y2), length(enzyme_tpm[which(enzyme_tpm$class=="Y1"),]$class)),])
# enzyme_tpm.melt <- melt(enzyme_tpm[sample(nrow(enzyme_tpm), length(enzyme_tpm[which(enzyme_tpm$class=="Y2"),]$class)),c(11:56)])
enzyme_tpm.melt <- melt(test[,c(11:56,64:66,69,71:73)])
enzyme_tpm.melt$value <- log2(enzyme_tpm.melt$value)
med <- summaryBy(value ~ variable, data = enzyme_tpm.melt, FUN = list(median, mean, sd))
colnames(med) <- c("variable", "median", "mean", "sd")
med$variable <- factor(med$variable, levels=factors_order, ordered = T)

enzyme_tpm$class <- as.character(enzyme_tpm$class)
enzyme_tpm.melt <- melt(enzyme_tpm[which(enzyme_tpm$class!="N1" & enzyme_tpm$class!="N2" & enzyme_tpm$class!="N3"),c(11:56,64:66,69,71:73)])
enzyme_tpm.melt$variable <- factor(enzyme_tpm.melt$variable, levels=factors_order, ordered = T)

p <- ggplot(enzyme_tpm.melt[which(enzyme_tpm.melt$variable!="h3k36me3" & enzyme_tpm.melt$variable!="rad21"),], 
            aes(class, log2(value), fill=class)) + 
  geom_boxplot(outlier.size = 0.5) +
  scale_fill_manual(values=c("#b2df8a", "#fb9a99")) +
  facet_grid(.~variable) +
  geom_hline(data=med[which(med$variable!="h3k36me3" & med$variable!="rad21"),], 
             aes(yintercept = median), color="#1f78b4") +
  ylab("Signal (TPKM, log2)") + xlab("")
## boxplot median below the dashed line are not significant
wilcox.test(enzyme_tpm[which(enzyme_tpm$class=="Y2"),]$ezh2, enzyme_tpm$ezh2, alternative = "g")$p.value
## boxplot median above the dashed line are significant
wilcox.test(enzyme_tpm[which(enzyme_tpm$class=="Y2"),]$ring1b, enzyme_tpm$ring1b, alternative = "g")$p.value
ggsave(p, dpi=300, height=10, width=25, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chromatin_remodellers_esc_promoters_barplot.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## correlation between Chd4 and Brg1 binding at promoters
##----------------------------------------------------##

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/correlation_chd4_brg1.pdf")
par(mfrow=c(1,2))
smoothScatter(log(enzyme_tpm$chd4), log(enzyme_tpm$brg1), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (TPM; log)", ylab = "brg1 (TPM; log)", main="all genes")
lines(lowess(log(enzyme_tpm$chd4), log(enzyme_tpm$brg1)),col='red', lty="dashed")
cor(log(enzyme_tpm$chd4), log(enzyme_tpm$brg1))
smoothScatter(gene_enzyme_enhancer$chd4, gene_enzyme_enhancer$brg1, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (TPM)", ylab = "brg1 (TPM)", main="bivalent and active genes")
lines(lowess(gene_enzyme_enhancer$chd4, gene_enzyme_enhancer$brg1),col='red', lty="dashed")
cor(gene_enzyme_enhancer$chd4, gene_enzyme_enhancer$brg1)
dev.off()

##----------------------------------------------------##
## plot predicted nucleosome positions at gene promoters
##----------------------------------------------------##

class <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$description))
distance <- as.numeric(apply(profile[grep("HFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[7]))
distance_1DN <- ifelse(profile[grep("NFR", profile$name),]$strand=="+", profile[grep("NFR", profile$name),]$end-as.numeric(apply(profile[grep("NFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2])), as.numeric(apply(profile[grep("NFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2]))-profile[grep("NFR", profile$name),]$start)
distance_2DN <- ifelse(profile[grepl("^DOWN", profile$name),]$strand=="+", profile[grepl("^DOWN", profile$name),]$end-as.numeric(apply(profile[grepl("^DOWN", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2])), as.numeric(apply(profile[grepl("^DOWN", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2]))-profile[grepl("^DOWN", profile$name),]$start)
distance_3DN <- ifelse(profile[grepl("^EXTR_DOWN", profile$name),]$strand=="+", profile[grepl("^EXTR_DOWN", profile$name),]$end-as.numeric(apply(profile[grepl("^EXTR_DOWN", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2])), as.numeric(apply(profile[grepl("^EXTR_DOWN", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2]))-profile[grepl("^EXTR_DOWN", profile$name),]$start)

# distance_1UP <- ifelse(profile[grep("NFR", profile$name),]$strand=="+", profile[grep("NFR", profile$name),]$start-as.numeric(apply(profile[grep("NFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2])), as.numeric(apply(profile[grep("NFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[2]))-profile[grep("NFR", profile$name),]$end)

## line plot for position of +1, +1'  and +1" nucleosomes
pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/predicted_nucleosome_positions.pdf")
par(mfrow=c(2,1))
plot(distance_1DN[which(class == "Y1")], type="l", col="black", lty=2, ylim=c(-2500,2500), ylab="")
lines(distance_2DN[which(class == "Y1")], type="l", col="black", lty=3, ylim=c(-2500,2500))
lines(distance_3DN[which(class == "Y1")], type="l", col="black", lty=4, ylim=c(-2500,2500))
plot(distance_1DN[which(class == "Y2")], type="l", col="black", lty=2, ylim=c(-2500,2500), ylab="")
lines(distance_2DN[which(class == "Y2")], type="l", col="black", lty=3, ylim=c(-2500,2500))
lines(distance_3DN[which(class == "Y2")], type="l", col="black", lty=4, ylim=c(-2500,2500))
dev.off()

```

######################################################
#### PRIMARY ANALYSIS AT PROMOTERS (mef)
######################################################
```{r}
##----------------------------------------------------##
## gene promoters: expression, house keeping genes and closest gene analysis (esc)
##----------------------------------------------------##

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_mef_stats.pdf")
par(mfrow=c(4,1))
color_mef = c(rep("red", as.data.frame(table(df_mef$cluster))[1,2]), rep("green", as.data.frame(table(df_mef$cluster))[2,2]), rep("red", as.data.frame(table(df_mef$cluster))[3,2]), rep("green", as.data.frame(table(df_mef$cluster))[4,2]), rep("red", as.data.frame(table(df_mef$cluster))[5,2]))
barplot(df_mef$expr_mef, xlab="genes", ylab="gene expression", border = color_mef, col=color_mef)
image(as.matrix(df_mef$expr_mef), col=rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.001)(256)))

barplot(df_mef$house_keeping, xlab="genes", ylab="house keeping", border = NA)
# df.melt <- melt(as.data.frame(split(df_mef[which(df_mef$cluster=="Y2"),]$house_keeping, ceiling(seq_along(df_mef[which(df_mef$cluster=="Y2"),]$house_keeping)/2))))
# remainder(length(df_mef[which(df_mef$cluster=="Y2"),]$house_keeping),1:1000)
# ggplot(df.melt, aes(variable, value))  + geom_boxplot()

barplot(df_mef$closest_tss, border = color_mef, col=color_mef, xlab="genes", ylab="dist to closest TSS")
df_mef.melt <- melt(df_mef[,c("cluster","closest_tss")])
ggplot(df_mef.melt, aes(cluster, log(value))) + geom_boxplot() + theme_bw() + ylab("dist to closest TSS (log)")


plot(log(df_mef[which(df_mef$cluster=="Y1" & !is.na(df_mef$expr_mef)),]$expr_mef+1), df_mef[which(df_mef$cluster=="Y1" & !is.na(df_mef$expr_mef)),]$dist_tss_down)

test <- melt(df_mef[,c("cluster","promoter_class","expr_mef")])
ggplot(test, aes(log(value+1), color=promoter_class)) + stat_ecdf() + facet_grid(.~cluster)

par(mfrow=c(2,1))
barplot(df_mef$CpG_length, xlab="genes", ylab="CpG island length", border = color_mef, col=color_mef)
image(as.matrix(df_mef$CpG_length), col=rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.2)(256)))
dev.off()
```

##################################################
#### ANALYSIS SHOWING THE ROLE OF POL2 AND PRC2 IN EXPLAINING THE ANATAGONISTIC FUNCTION OF CHD4 AND BRG1
##################################################
```{r}
##----------------------------------------------------##
## volcano plot showing that Chd4 and Brg1 directionality do not explain anatagonistic function of Chd4 and Brg1
##----------------------------------------------------##

dev.off()
pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_expression_volcano.pdf", height=15, width=18)
p1 <- lapply(c(75,87,91,95,103,123), function(x) {
  test <- gene_enzyme_enhancer[,c(x,x+1,x+3,132,171)]
  colnames(test) <- c("fc", "pVal", "class", "cluster", "promoter_class")
  ggplot(test, aes(x=fc, y=-log10(pVal))) +
    geom_point(aes(color = class), size=1.2) +
    scale_color_manual(values = c("up"= "#a50026", "down"="#313695",  "neutral"= "gray")) +
    facet_grid(.~cluster+promoter_class) +
    ggtitle(colnames(gene_enzyme[x])) +
    theme(legend.position="none")
})

p2 <- lapply(c(75,87,91,95,103,123), function(x) {
      test <- gene_enzyme[,c(x,x+1,x+3,132,133)]
      colnames(test) <- c("fc", "pVal", "class", "cluster", "promoter_class")
      ggplot(test, aes(x=fc, y=-log10(pVal))) +
        geom_point(aes(color = class), size=1.2) +
        scale_color_manual(values = c("up"= "#a50026", "down"="#313695",  "neutral"= "gray")) +
        facet_grid(.~cluster+promoter_class) +
        ggtitle(colnames(gene_enzyme[x])) +
        theme(legend.position="none")
    })

g <- do.call(grid.arrange, c(p1, p2))
g

p1 <- lapply(c(75,87,91,95,103,123), function(x) {
  test <- gene_enzyme_enhancer[,c(x,x+1,x+3,165,61)]
  colnames(test) <- c("fc", "pVal", "class", "ep400_cluster_merged")
  ggplot(test, aes(x=fc, y=-log10(pVal))) +
    geom_point(aes(color = class), size=1.2) +
    scale_color_manual(values = c("up"= "#a50026", "down"="#313695",  "neutral"= "gray")) +
    facet_grid(.~ep400_cluster_merged) +
    ggtitle(colnames(gene_enzyme[x])) +
    theme(legend.position="none")
})

g <- do.call(grid.arrange, p1)
g
dev.off()

# ggsave(g, dpi=300, height=15, width=18, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_expression_volcano.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## expression of genes from the four classes during differentiation
##----------------------------------------------------##

df.melt <- melt(gene_enzyme_enhancer[,c("expr_mef", "expr_mef48", "expr_ipsc", "expr_esc", "ep400_cluster_merged")])
df.melt$variableNumeric <- as.numeric(df.melt$variable)
df.melt$value <- df.melt$value
df.melt <- summarySE(df.melt[!is.na(df.melt$value),], measurevar = "value", groupvars = c("ep400_cluster_merged", "variable"))

p1 <- ggplot(df.melt, aes(variable, value, color=ep400_cluster_merged)) +  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(ep400_cluster_merged), color=factor(ep400_cluster_merged)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  #scale_color_manual(values=c("#313695", "#000000", "#a50026")) +
  theme_bw() +  theme(legend.position="top") +
  geom_smooth(method = "lm", formula = y ~ poly(x, 4), se = F)
  #facet_wrap(~ep400_cluster_merged, scales="free")
ggsave(p1, dpi=300, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_expression_differentiation.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## volcano plot analysis
##----------------------------------------------------##

p1 <- ggplot(gene_enzyme_enhancer, aes(x=chd4_fc, y=-log10(chd4_pVal))) +
  geom_point(aes(colour = chd4_fc, fill = chd4_fc), shape = 21, size = 3, stroke = 1) +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "PuOr"), bias=0.7)(256))) + 
  scale_color_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "PuOr"), bias=0.7)(256))) +
  theme(legend.position="none") + theme_classic() + scale_x_reverse() +
  geom_vline(xintercept=c(-0.41, 0.41), linetype = 2) +
  scale_y_continuous(position = "right")

p2 <- ggplot(gene_enzyme_enhancer, aes(x=brg1_fc, y=-log10(brg1_pVal))) +
  geom_point(aes(colour = brg1_fc, fill = brg1_fc), shape = 21, size = 3, stroke = 1) +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "PuOr"), bias=1.1)(256))) + 
  scale_color_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "PuOr"), bias=1.1)(256))) +
  theme(legend.position="none") + theme_classic() + scale_x_reverse() +
  geom_vline(xintercept=c(-0.41, 0.41), linetype = 2) +
  scale_y_continuous(position = "right")

test <- gene_enzyme_enhancer
test$chd4_class <- as.character(cut2(test$chd4_fc, m=300))
test[which(gene_enzyme_enhancer$chd4_class=="neutral"),]$chd4_class <- "0"
test$chd4_class <- as.factor(as.numeric(gsub("]", "", gsub("[)]+", "", gsub("^.*,", "", test$chd4_class)))))
test.melt <- melt(summaryBy(pol2_wt ~ chd4_class, data = test, FUN = list(median)))
p3 <- ggplot(test.melt, aes(chd4_class, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=1.1)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()

test.melt <- melt(summaryBy(ezh2 ~ chd4_class, data = test, FUN = list(median)))
p4 <- ggplot(test.melt, aes(chd4_class, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=0.7)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()

test$brg1_class <- as.character(cut2(test$brg1_fc, m=500))
test[which(gene_enzyme_enhancer$brg1_class=="neutral"),]$brg1_class <- "0"
test$brg1_class <- as.factor(as.numeric(gsub("]", "", gsub("[)]+", "", gsub("^.*,", "", test$brg1_class)))))
test.melt <- melt(summaryBy(pol2_wt ~ brg1_class, data = test, FUN = list(median)))
p5 <- ggplot(test.melt, aes(brg1_class, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=1.1)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()

test.melt <- melt(summaryBy(ezh2 ~ brg1_class, data = test, FUN = list(median)))
p6 <- ggplot(test.melt, aes(brg1_class, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=0.7)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()

test.melt <- melt(summaryBy(chd4_fc ~ chd4_class, data = test, FUN = list(median)))
p7 <- ggplot(test.melt, aes(chd4_class, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "PuOr"), bias=0.7)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()

test.melt <- melt(summaryBy(brg1_fc ~ brg1_class, data = test, FUN = list(median)))
p8 <- ggplot(test.melt, aes(brg1_class, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "PuOr"), bias=1.2)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()

freqPlot(melt(table(gene_enzyme_enhancer[,c(165,78)])), "density") + theme_bw()
freqPlot(melt(table(gene_enzyme_enhancer[,c(165,90)])), "density") + theme_bw()

g <- ggarrange(p1, p2, p3, p4, p5, p6, p7, p8, nrow=4, ncol=2)
g
ggsave(g, dpi=300, height=15, width=18, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_expression_volcano1.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## heatmap showing that mutually exclusive ep400 and prc2 binding explain anatagonistic function of Chd4 and Brg1
##----------------------------------------------------##

## All (promoter_class)
test <- gene_enzyme_enhancer
test$diff_fc <- log((test$expr_esc+1)/(test$expr_mef+1))
test$ep400_cluster <- cut2(test$pol2_wt, m=25)
test$gene_length <- log(test$gene_length)

test.melt <- melt(summaryBy(chd4_fc ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
#barplot(test.melt[order(test.melt$promoter_class_merged, test.melt$ep400_cluster),]$value)
p1 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=0.3)(56))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_text(size=10, angle=45))

# test.melt <- melt(summaryBy(chd4_pVal ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
# ggplot(test.melt, aes(ep400_cluster, variable, fill=-log10(value))) + geom_tile() + 
#   scale_fill_gradientn(colours = (colorRampPalette(brewer.pal(9, "PRGn"), bias=3)(256))) +
#   facet_grid(.~cluster + promoter_class_merged) +
#   theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(summaryBy(brg1_fc ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
#barplot(test.melt[order(test.melt$promoter_class_merged, test.melt$ep400_cluster),]$value)
p2 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=1.1)(56))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_text(size=10, angle=45))

# test.melt <- melt(summaryBy(brg1_pVal ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
# ggplot(test.melt, aes(ep400_cluster, variable, fill=-log10(value))) + geom_tile() + 
#   scale_fill_gradientn(colours = (colorRampPalette(brewer.pal(9, "PRGn"), bias=3)(256))) +
#   #scale_fill_manual(values = ifelse(test.melt$value < 0.05, "red", "green"))
#   facet_grid(.~cluster + promoter_class_merged) +
#   theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(summaryBy(ring1_fc ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=0.85)(56))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(CpG_length ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
p4 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.8)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(expr_esc + expr_ipsc + expr_mef48 + expr_mef ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
p5 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=log(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=1)(10))) +
  facet_grid(.~cluster+promoter_class_merged) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(diff_fc ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
p6 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=1)(10))) +
  facet_grid(.~cluster+promoter_class_merged) +
  theme(axis.text.x = element_blank())

# comparison between gene expression changes upon chd4 and mbd3 kd
mbd3 <- read.table(pipe("grep -v Gene.symbol /home/pundhir/project/chip-seq-analysis/data/gene/esc/other/mbd3.fc.uniqueGeneId"))
colnames(mbd3) <- c("gene", "mbd3_fc", "mbd3_pVal", "mbd3_qVal", "mbd3_class")
test <- merge(test, mbd3, by.x="gene", by.y="gene")

test.melt <- melt(summaryBy(mbd3_fc ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
p7 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=1)(56))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(summaryBy(gene_length ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
p8 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.1)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_text(size=10, angle=45))

test <- merge(test, enzyme_tpm[,c("name", "atac_brg1_kd", "atac_chd4_kd", "atac_ep400_kd")], by.x="gene", by.y="name")
test$atac_fc_brg1KD <- log(test$atac_brg1_kd/test$atac_wt)
test$atac_fc_chd4KD <- log(test$atac_chd4_kd/test$atac_wt)

# test.melt <- melt(summaryBy(atac_fc_brg1KD ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
# ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
#   facet_grid(.~cluster + promoter_class_merged) +
#   theme(axis.text.x = element_text(size=10, angle=45))
# 
# test.melt <- melt(summaryBy(atac_fc_chd4KD ~ ep400_cluster + cluster + promoter_class_merged, data = test, FUN = list(median)))
# ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
#   facet_grid(.~cluster + promoter_class_merged) +
#   theme(axis.text.x = element_text(size=10, angle=45))

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/cor_chd4_mbd3_expr_fc.pdf")
test.melt <- summaryBy(chd4_fc +  mbd3_fc ~ ep400_cluster, data = test, FUN = list(median))
smoothScatter(test.melt$chd4_fc.median, test.melt$mbd3_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (logFC)", ylab = "mbd3 (logFC)")
lines(lowess(test.melt$chd4_fc.median, test.melt$mbd3_fc.median),col='red', lty="dashed")
# # reg<-lm(chd4_fc.median ~ mbd3_fc.median, data = test.melt)
# # abline(reg, col="red", lty="dashed")
cor(test.melt$chd4_fc.median, test.melt$mbd3_fc.median, method="spearman")

test.melt <- summaryBy(atac_fc_brg1KD +  atac_fc_chd4KD ~ ep400_cluster, data = test, FUN = list(median))
smoothScatter(test.melt$atac_fc_chd4KD.median, test.melt$atac_fc_brg1KD.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "ATAC (Chd4 KD logFC)", ylab = "ATAC (Brg1 KD logFC)")
lines(lowess(test.melt$atac_fc_chd4KD.median, test.melt$atac_fc_brg1KD.median),col='red', lty="dashed")
cor(test.melt$atac_fc_chd4KD.median, test.melt$atac_fc_brg1KD.median, method="spearman")
dev.off()

g <- arrangeGrob(p1, p2, p3, p4, p6, p7, p8, nrow = 8, ncol=1)
ggsave(g, dpi=300, height=25, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/cluster_4_analysis_ALL.pdf", useDingbats=FALSE)

## All (ep400_class)
test <- gene_enzyme
test$ep400_cluster <- cut2(test$ep400, m=20)

test.melt <- melt(summaryBy(ep400_fc + chd4_fc + chd6_fc + chd8_fc + brg1_fc ~ ep400_cluster + cluster, data = test, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=0.9)(56))) +
  # scale_fill_gradientn(colours=c("blue", "blue", "white", "red", "red"), breaks=c(-2.5,-0.1,-0.05, 0.05,0.1,2.5)) +
  facet_grid(.~cluster) +
  theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(summaryBy(ring1_fc ~ ep400_cluster + cluster, data = test, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=0.27)(56))) +
  facet_grid(.~cluster) +
  theme(axis.text.x = element_text(size=10, angle=45))

#summary(df_esc[which(df_esc$cluster=="N1"),]$CpG_length)

test.melt <- melt(summaryBy(CpG_length ~ ep400_cluster + cluster, data = test, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.95)(256))) +
  facet_grid(.~cluster) +
  theme(axis.text.x = element_text(size=10, angle=45))

g <- arrangeGrob(p1, p2, p3, nrow = 3, ncol=1)
ggsave(g, dpi=300, height=25, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/cluster_4_analysis_ALL_ep400_sort.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## analysis to explain relationship with enhancers
##----------------------------------------------------##

# df.melt <- melt(gene_enzyme_enhancer[,c(155,139)])
# ggplot(df.melt, aes(promoter_class_merged, log(value))) + geom_boxplot() + facet_grid(.~variable) + ylab("Brg1 at enhancers (log)")
# df.melt <- melt(gene_enzyme_enhancer[,c(155,127,129)])
# ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~promoter_class_merged) + ylab("H3K27ac at enhancers (log)") + theme(axis.text.x = element_text(size=10, angle=45))
# df.melt <- melt(gene_enzyme_enhancer[,c(155,133,135)])
# ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~promoter_class_merged) + ylab("H3K27me3 at enhancers (log)") + theme(axis.text.x = element_text(size=10, angle=45))
# 
# df.melt <- melt(gene_enzyme_enhancer[,c(155,16)])
# ggplot(df.melt, aes(promoter_class_merged, log(value))) + geom_boxplot() + ylab("Brg1 at promoters (log)")
# df.melt <- melt(gene_enzyme_enhancer[,c(155,147:150)])
# ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~promoter_class_merged) + ylab("H3K27ac at promoters (log)") + theme(axis.text.x = element_text(size=10, angle=45))
# df.melt <- melt(gene_enzyme_enhancer[,c(155,151:154)])
# ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~promoter_class_merged) + ylab("H3K27me3 at promoters (log)") + theme(axis.text.x = element_text(size=10, angle=45))

## ALSO CONSIDER ANALYZING FOR SUPER-ENHANCERS
```

##################################################
#### CHD4 AND BRG1 BINDING FOLD CHANGE AND POL2 PAUSING ANALYSIS (ESC)
##################################################
```{r}
test <- gene_enzyme_enhancer
# test[which(test$pol2_TSSR < 1 | test$dist_closest_tss < 3000),c(171:186)] <- NA
test$pausing_brg1_wt <- rowMeans(test[,c(177:178)])
test$pausing_brg1_ko <- rowMeans(test[,c(179:180)])
test$pausing_mbd3_wt <- rowMeans(test[,c(182:183)])
test$pausing_mbd3_ko <- rowMeans(test[,c(185:186)])

test$pausing_wt <- rowMeans(test[,c(187:188)])
test$pausing_brg1_kd <- rowMeans(test[,c(189:190)])
test$pausing_chd4_kd <- rowMeans(test[,c(191:192)])

#test$ratio_chd4_brg1 <- cut2(log2((test$chd4)/(test$brg1)), m=25)
test$ratio_chd4_brg1 <- cut2((test$chd4+1)/(test$brg1+1), m=25)
test$ep400_cluster_merged_col <- "red"
test[which(test$ep400_cluster_merged=="low"),]$ep400_cluster_merged_col <- "green"
test[which(test$ep400_cluster_merged=="average"),]$ep400_cluster_merged_col <- "blue"
test[which(test$ep400_cluster_merged=="high"),]$ep400_cluster_merged_col <- "yellow"

# test[,c("pol2_ser2p", "pol2_ser5p")] <- normalize.quantiles(as.matrix(test[,c("pol2_ser2p", "pol2_ser5p")]))
# test$ratio_ser2p_ser5p <- test$pol2_ser2p/test$pol2_ser5p
test$ratio_ser2p_ser5p <- as.numeric(normalize.quantiles(as.matrix(test$pol2_ser2p))/normalize.quantiles(as.matrix(test$pol2_ser5p)))
# test$ratio_ser5p_wt <- as.numeric(normalize.quantiles(as.matrix(test$pol2_ser5p))/normalize.quantiles(as.matrix(test$pol2_wt)))
# test$ratio_ser2p_wt <- as.numeric(normalize.quantiles(as.matrix(test$pol2_ser2p))/normalize.quantiles(as.matrix(test$pol2_wt)))


test <- test[order(test$ratio_chd4_brg1),]

# test.melt <- melt(gene_enzyme_enhancer[,c(161,189)])
# ggplot(test.melt, aes(ep400_cluster_merged, (value))) + geom_boxplot() + facet_wrap(~variable, scales="free")

# test.melt <- melt(test[which(test$ep400_cluster_merged=="high" | test$ep400_cluster_merged=="average"),c(189,20,16,42,57,58,59,61,62,63,64,65,66,190,194,25,26,33,35,37,36,10,45,47,51,52)])
# ggplot(test.melt, aes(ratio_chd4_brg1, log(value))) + geom_boxplot() + facet_wrap(~variable, scales="free")

# test.melt <- melt(test[which(test$ep400_cluster_merged=="low" | test$ep400_cluster_merged=="bivalent"),c(189,20,16,42,57,58,59,61,62,63,64,65,66,190,194,25,26,33,35,37,36,10)])
# ggplot(test.melt, aes(ratio_chd4_brg1, log(value))) + geom_boxplot() + facet_wrap(~variable, scales="free")


##----------------------------------------------------##
## plot heatmap for low or bivalent genes (prc2 bound)
##----------------------------------------------------##

test.prc2 <- test[which(test$ep400_cluster_merged=="bivalent" | test$ep400_cluster_merged=="low"),]

#barplot(rep(1, nrow(test.prc2)), col = test.prc2$ep400_cluster_merged_col, border = NA)

test.melt <- melt(summaryBy(chd4 + brg1 ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.5)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3k27me3 ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.5)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(ezh2 ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.9)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(expr_esc ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p4 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(brg1_fc ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p5 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.4)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(chd4_fc ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p6 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.7)(256))) + 
  theme(axis.text.x = element_blank())

g <- arrangeGrob(p1, p2, p3, p4, p5, p6, nrow = 6, ncol=1)
ggsave(g, dpi=300, height=25, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_prc2.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## plot heatmap for average or highly expressed genes (ep400 bound)
##----------------------------------------------------##

test.ep400 <- test[which(test$ep400_cluster_merged=="high" | test$ep400_cluster_merged=="average"),]

# barplot(rep(1, nrow(test.ep400)), col = test.ep400$ep400_cluster_merged_col, border = NA)

test.melt <- melt(summaryBy(chd4 + brg1 ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3k27ac ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_wt ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(ep400 ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p4 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3v3 ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p5 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) +
  theme(axis.text.x = element_blank())

#ggarrange(p2, p3, p4, p5, nrow=4)

test.melt <- melt(summaryBy(h2az ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p6 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.8)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser2p ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p7 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.65)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser5p ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p8 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser7p ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p9 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.5)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(ratio_ser5p_wt ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p10 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log2(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(nelfa + paf1 + leo1 ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log2(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.5)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(expr_esc ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p11 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.6)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(brg1_fc ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p12 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.75)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(chd4_fc ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p13 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) + 
  theme(axis.text.x = element_blank())

# test.melt <- melt(summaryBy(groseq_tssr_rep1_norm ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
# ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.6)(256))) + 
#   theme(axis.text.x = element_blank())

# test.melt <- melt(summaryBy(expr_nelfa_kd ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
# ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log2(value))) + geom_tile() +
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) + 
#   theme(axis.text.x = element_blank())

# ## chd4 KD
# test.melt <- melt(test[,c(161,83,189)])
# ggplot(test.melt, aes(ratio_chd4_brg1, value)) + geom_boxplot() + facet_grid(.~ep400_cluster_merged)

# ## Brg1 KD 
# test.melt <- melt(test[,c(161,71,189)])
# ggplot(test.melt, aes(ratio_chd4_brg1, value)) + geom_boxplot() + facet_grid(.~ep400_cluster_merged)

# pol2_release_rate <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pol2_release_rate_genes")
# colnames(pol2_release_rate) <- c("gene", "bp_per_min", "release_class")
# test.ep400.sub <- merge(test.ep400, pol2_release_rate, by.x="gene", by.y="gene")
# test.melt <- melt(summaryBy(bp_per_min ~ ratio_chd4_brg1, data = test.ep400.sub, FUN = list(median)))
# ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() +
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=2)(256))) + 
#   theme(axis.text.x = element_blank())

# ## check for additional TFs or histone modifications (column 13: cdk9; 15: h3.3S31p)
# test.ep400 <- test[which(test$ep400_cluster_merged=="high" | test$ep400_cluster_merged=="average"),]
# cdk9 <- read.table(pipe("cut -f 4,15 /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/GENES_PROMOTER.BED.RC.CLUSTERS.TPM.additional"))
# colnames(cdk9) <- c("gene", "cdk9")
# test.ep400 <- merge(test.ep400, cdk9, by.x="gene", by.y="gene")
# 
# #test.ep400$cdk9 <- test.ep400$cdk9/test.ep400$pol2_ser5p
# test.melt <- melt(summaryBy(cdk9 ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
# ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.6)(256))) +
#   theme(axis.text.x = element_blank())

g <- arrangeGrob(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, p13, nrow = 13, ncol=1)
ggsave(g, dpi=300, height=25, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_pol2.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## plot containing overview analysis
##----------------------------------------------------##

test.melt <- melt(test[,c(165,25,42,57,58,59,61:66,10)])
p1 <- ggplot(test.melt, aes(ep400_cluster_merged, log(value))) + geom_boxplot() + facet_wrap(~variable, scales = "free") + 
  theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(test[,c(164,25,42,57,58,59,61:66,10)])
p2 <- ggplot(test.melt, aes(promoter_class_merged, log(value))) + geom_boxplot() + facet_wrap(~variable, scales = "free") +
  theme(axis.text.x = element_text(size=10, angle=45))

var="expr_esc"
wilcox.test(test.melt[which(test.melt[,1]=="narrow" & test.melt[,2]==var),]$value, test.melt[which(test.melt[,1]=="bivalent" & test.melt[,2]==var),]$value, alternative = "g")$p.value
wilcox.test(test.melt[which(test.melt[,1]=="medium" & test.melt[,2]==var),]$value, test.melt[which(test.melt[,1]=="narrow" & test.melt[,2]==var),]$value, alternative = "g")$p.value
wilcox.test(test.melt[which(test.melt[,1]=="broad" & test.melt[,2]==var),]$value, test.melt[which(test.melt[,1]=="medium" & test.melt[,2]==var),]$value, alternative = "g")$p.value


test.melt <- melt(gene_enzyme_enhancer[,c(171,25,42,57,58,59,61:66,10)])
p3 <- ggplot(test.melt, aes(directionality_class_merged, log(value))) + geom_boxplot() + facet_wrap(~variable, scales = "free") +
  theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y2"),c(199,25,42,57,58,59,61:66,10)])
p4 <- ggplot(test.melt, aes(groseq_class, log(value))) + geom_boxplot() + facet_wrap(~variable, scales = "free") +
  theme(axis.text.x = element_text(size=10, angle=45))

expr_nelfa_kd <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/expr_nelfa_kd.txt")
df <- merge(test[,c(1:205)], expr_nelfa_kd[,c(1,4)], by.x="gene", by.y="V1")
df <- df[,c(2:4,1,5:206)]
colnames(df)[206] <- "expr_nelfa_kd"

test.melt <- melt(df[,c(173,206)])
p5 <- ggplot(test.melt, aes(ratio_chd4_brg1_class, (value))) + geom_boxplot() + facet_grid(.~variable) + 
  theme(axis.text.x = element_text(size=10, angle=45))

test.melt <- melt(test.ep400[,c("pausing_brg1_wt", "pausing_brg1_ko")])
# test.melt <- melt(test.ep400[,c("pausing_wt", "pausing_brg1_kd")])
p6 <- ggplot(test.melt, aes((value), color=variable)) + stat_ecdf() + xlim(c(0.6,1))
wilcox.test(test.melt[which(test.melt$variable=="pausing_brg1_wt"),]$value, test.melt[which(test.melt$variable=="pausing_brg1_ko"),]$value, alternative="g")$p.value

test.melt <- melt(test.ep400[,c("pausing_mbd3_wt", "pausing_mbd3_ko")])
#test.melt <- melt(test.ep400[,c("pausing_wt", "pausing_chd4_kd")])
p7 <- ggplot(test.melt, aes((value), color=variable)) + stat_ecdf() + xlim(c(0.6,1))
wilcox.test(test.melt[which(test.melt$variable=="pausing_mbd3_wt"),]$value, test.melt[which(test.melt$variable=="pausing_mbd3_ko"),]$value, alternative="l")$p.value

test.melt <- melt(test.ep400[,c("pausing_wt_rep3", "pausing_brg1_kd_rep3", "pausing_chd4_kd_rep3")])
ggplot(test.melt, aes((value), color=variable)) + stat_ecdf() + xlim(c(0.6,1))
wilcox.test(test.melt[which(test.melt$variable=="pausing_wt_rep3"),]$value, test.melt[which(test.melt$variable=="pausing_brg1_kd_rep3"),]$value, alternative="g")$p.value
wilcox.test(test.melt[which(test.melt$variable=="pausing_wt_rep3"),]$value, test.melt[which(test.melt$variable=="pausing_chd4_kd_rep3"),]$value, alternative="l")$p.value

g <- arrangeGrob(p1, p2, p3, p4, p6, p7, nrow = 3, ncol=2)
ggsave(g, dpi=300, height=20, width=15, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_pol2_ecdf.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## plot line plot containing results of pol2 pausing analysis
##----------------------------------------------------##

test.ep400 <- test[which(test$ep400_cluster_merged=="high" | test$ep400_cluster_merged=="average"),]
# test.ep400 <- test.ep400[which(test.ep400$dist_closest_tss > 3000),]
test.ep400$ratio_chd4_brg1 <- log2((test.ep400$chd4+1)/(test.ep400$brg1+1))
# test.ep400$pausing_brg1_wt <- rowMeans(test.ep400[,c(173:174)])
# test.ep400$pausing_brg1_ko <- rowMeans(test.ep400[,c(175:176)])
# test.ep400$pausing_mbd3_wt <- rowMeans(test.ep400[,c(178:179)])
# test.ep400$pausing_mbd3_ko <- rowMeans(test.ep400[,c(181:182)])

test.ep400$pause_class <- cut2(test.ep400$pol2_fc_gb_Brg1_rep1, m=1500)
brg1 <- test.ep400[which(test.ep400$pause_class=="[ 0.4466, 2.4844]"),]
# test.ep400$pause_class <- cut2((test.ep400$h2az / test.ep400$h3v3), m=1500)
test.melt <- melt(test.ep400[,c(220,211:212)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p1 <- ggplot(test.melt, aes(pause_class, value, color=variable)) + 
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) + 
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#a50026", "#313695")) +
  theme(legend.position="none")

test.melt <- melt(test.ep400[,c(220,75)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p2 <- ggplot(test.melt, aes(pause_class, value, color=variable)) + 
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) + 
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#33a02c")) +
  theme(legend.position="none")

test.melt <- melt(test.ep400[,c(220,64)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p3 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#8856a7")) +
  theme(legend.position="none") +
  ylab("Pol2-ser2p")

test.melt <- melt(test.ep400[,c(220,65)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p4 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#8856a7")) +
  theme(legend.position="none") +
  ylab("Pol2-ser5p")

test.melt <- melt(test.ep400[,c(220,16)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p5 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("Brg1")

test.melt <- melt(test.ep400[,c(220,20)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p6 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("Chd4")

test.melt <- melt(test.ep400[,c(220,25)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p7 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("EP400")

test.melt <- melt(test.ep400[,c(220,42)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p8 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("H2A.Z")

test.melt <- melt(test.ep400[,c(220,57)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p9 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("H2A.Zac")

test.melt <- melt(test.ep400[,c(220,58)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p10 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("H3.3")

# test.melt <- melt(summaryBy(brg1_fc ~ pause_class, data = test.ep400, FUN = list(median)))
# ggplot(test.melt, aes(pause_class, variable, fill=(value))) + geom_tile() + 
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
#   theme(axis.text.x = element_blank())

test.ep400 <- test[which(test$ep400_cluster_merged=="high" | test$ep400_cluster_merged=="average"),]
# test.ep400 <- test.ep400[which(test.ep400$dist_closest_tss > 3000),]
test.ep400$ratio_chd4_brg1 <- log2((test.ep400$chd4+1)/(test.ep400$brg1+1))
test.ep400$pause_class <- cut2(test.ep400$pol2_fc_gb_Chd4_rep1, m=1500)

# test.ep400 <- merge(test.ep400, gene_enzyme_divergent[,c(4,231:236)], by.x="gene", by.y="gene")
# test.ep400 <- merge(test.ep400, df[,c(4,31,32)], by.x="gene", by.y="gene")

chd4 <- test.ep400[which(test.ep400$pause_class=="[-0.57685,-0.04763)"),]
test.melt <- melt(test.ep400[,c(220,213:214)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p11 <- ggplot(test.melt, aes(pause_class, value, color=variable)) + 
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) + 
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#a50026", "#313695")) +
  theme(legend.position="none")

test.melt <- melt(test.ep400[,c(220,87)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p12 <- ggplot(test.melt, aes(pause_class, value, color=variable)) + 
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) + 
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#33a02c")) + 
  theme(legend.position="none")

test.melt <- melt(test.ep400[,c(220,64)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p13 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#8856a7")) +
  theme(legend.position="none") +
  ylab("Pol2-ser2p")

test.melt <- melt(test.ep400[,c(220,65)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p14 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#8856a7")) +
  theme(legend.position="none") +
  ylab("Pol2-ser5p")

test.melt <- melt(test.ep400[,c(220,16)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p15 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("Brg1")

test.melt <- melt(test.ep400[,c(220,20)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p16 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("Chd4")

test.melt <- melt(test.ep400[,c(220,25)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p17 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("EP400")

test.melt <- melt(test.ep400[,c(220,42)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p18 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("H2A.Z")

test.melt <- melt(test.ep400[,c(220,57)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p19 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("H2A.Zac")

test.melt <- melt(test.ep400[,c(220,58)])
test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("pause_class", "variable"))
p20 <- ggplot(test.melt, aes(pause_class, value, color=variable)) +
  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  scale_color_manual(values=c("#000000")) +
  theme(legend.position="none") +
  ylab("H3.3")

g <- arrangeGrob(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, p13, p14, p15, p16, p17, p18, p19, p20, nrow = 2, ncol=10)
ggsave(g, dpi=300, height=4, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_pol2_line.pdf", useDingbats=FALSE)

# # chd4 <- chd4[which(chd4$chd4_fc < -0.58),]
# # brg1 <- brg1[which(brg1$brg1_fc > 0.58),]
# rest <- test.ep400[!test.ep400$gene %in% c(as.vector(chd4$gene), as.vector(brg1$gene)),]
# chd4$pause_class <- "chd4"
# brg1$pause_class <- "brg1"
# rest$pause_class <- "rest"
# chd4_brg1 <- rbind(chd4, brg1, rest)
# 
# test.melt <- melt(chd4_brg1[,c(75,87,206)])
# ggplot(test.melt, aes(variable, (value))) + geom_boxplot() + facet_grid(.~pause_class)
# 
# test.melt <- melt(chd4_brg1[,c(200:203,206)])
# ggplot(test.melt, aes(variable, (value))) + geom_boxplot() + facet_grid(.~pause_class)
# 
# test.melt <- melt(chd4_brg1[,c(16,20,25,61,64:67,42,56,57,58,200,202,206)])
# ggplot(test.melt, aes(pause_class, log(value))) + geom_boxplot() + facet_grid(.~variable)

##----------------------------------------------------##
## correlation plots between H2A.Z, H3.3 and Pol2 pausing in ESC (ep400 bound)
##----------------------------------------------------##

groseq_esc <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_divergentTranscription.bed")
colnames(groseq_esc) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss",
                  "groseq_wt_rep1", "groseq_wt_rep2",
                  "cage_wt_rep1", "cage_wt_rep2", "cage_rrp40_rep1", "cage_rrp40_rep2", "cage_rbm7_rep1", "cage_rbm7_rep2",
                  "proseq_wt_rep1", "proseq_wt_rep2", "proseq_rrp40_rep1", "proseq_rrp40_rep2", "proseq_rbm7_rep1", "proseq_rbm7_rep2",
                  "nasseq_wt_rep1", "nasseq_wt_rep2", "nasseq_wt_rep3", "nasseq_kd_rep1", "nasseq_kd_rep2", "nasseq_kd_rep3")
transcript_width <- (groseq_esc$end - groseq_esc$start)
# groseq_esc <- groseq_esc[which(groseq_esc$dist_closest_tss > 3000),]
tssr <- t(apply(groseq_esc[,c(11:12)], 1, function(y) unlist(lapply(y, function(x) sum(as.numeric(unlist(strsplit(as.character(x), "/"))[1:2]))))))
colnames(tssr) <- c("groseq_tssr_rep1", "groseq_tssr_rep2")
gb <- t(apply(groseq_esc[,c(11:12)], 1, function(x) as.numeric(gsub("^.*/", "", x))))
colnames(gb) <- c("groseq_gb_rep1", "groseq_gb_rep2")
groseq_esc <- cbind(groseq_esc[,c(4),drop=F], tssr, gb)

# groseq_esc$groseq_tssr_rep1_norm <- (groseq_esc$groseq_tssr_rep1/350)/((groseq_esc$groseq_tssr_rep1/350)+(groseq_esc$groseq_gb_rep1/transcript_width))
# groseq_esc$groseq_tssr_rep2_norm <- (groseq_esc$groseq_tssr_rep2/350)/((groseq_esc$groseq_tssr_rep2/350)+(groseq_esc$groseq_gb_rep2/transcript_width))
# groseq_esc$groseq_gb_rep1_norm <- (groseq_esc$groseq_gb_rep1/350)/((groseq_esc$groseq_tssr_rep1/350)+(groseq_esc$groseq_gb_rep1/transcript_width))
# groseq_esc$groseq_gb_rep2_norm <- (groseq_esc$groseq_gb_rep2/350)/((groseq_esc$groseq_tssr_rep2/350)+(groseq_esc$groseq_gb_rep2/transcript_width))

groseq_esc$groseq_tssr_rep1_norm <- groseq_esc$groseq_tssr_rep1/(groseq_esc$groseq_tssr_rep1+groseq_esc$groseq_gb_rep1)
groseq_esc$groseq_tssr_rep2_norm <- groseq_esc$groseq_tssr_rep2/(groseq_esc$groseq_tssr_rep2+groseq_esc$groseq_gb_rep2)
groseq_esc$groseq_gb_rep1_norm <- groseq_esc$groseq_gb_rep1/(groseq_esc$groseq_tssr_rep1+groseq_esc$groseq_gb_rep1)
groseq_esc$groseq_gb_rep2_norm <- groseq_esc$groseq_gb_rep2/(groseq_esc$groseq_tssr_rep2+groseq_esc$groseq_gb_rep2)

test <- merge(gene_enzyme_divergent, groseq_esc, by.x="gene", by.y="gene")
test$ratio_chd4_brg1 <- cut2((test$chd4+1)/(test$brg1+1), m=20)
test$ep400_cluster <- cut2(test$groseq_tssr_rep2, m=20)

test.ep400 <- test[which(test$ep400_cluster_merged=="high" | test$ep400_cluster_merged=="average"),]
# test.ep400[,c("groseq_tssr_rep1", "groseq_tssr_rep2", "groseq_gb_rep1", "groseq_gb_rep2")] <- normalize.quantiles(as.matrix(test.ep400[,c("groseq_tssr_rep1", "groseq_tssr_rep2", "groseq_gb_rep1", "groseq_gb_rep2")]))
# test.ep400$groseq_wt_rep1_pausing <- test.ep400$groseq_tssr_rep1/(test.ep400$groseq_tssr_rep1 + test.ep400$groseq_gb_rep1)
# test.ep400$groseq_wt_rep2_pausing <- test.ep400$groseq_tssr_rep2/(test.ep400$groseq_tssr_rep2 + test.ep400$groseq_gb_rep2)
# 
# test.melt <- melt(summaryBy(groseq_wt_rep1_pausing + groseq_wt_rep2_pausing ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
# ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() +
#  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
#  theme(axis.text.x = element_blank())

## Pol2 pausing is not possible below a certain level of gene expression
## Pol2 pausing formula can thus give a false sense of high pausing for many low expressed genes
# par(mfrow=c(2,1))
# plot(groseq_esc$groseq_tssr_rep2_norm, log(groseq_esc$groseq_tssr_rep2))
# plot(test.ep400$groseq_tssr_rep2_norm, log(test.ep400$groseq_tssr_rep2))

test.melt <- summaryBy(h2az + h3v3 + groseq_tssr_rep1 + groseq_tssr_rep2 + pol2_ser5p + pol2_ser2p +
                         groseq_tssr_rep1_norm + groseq_tssr_rep2_norm + groseq_gb_rep1_norm + groseq_gb_rep2_norm + 
                         chd4 + brg1 + atac_wt ~ ep400_cluster, data = test.ep400, FUN = list(median))

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/cor_groseq_h2az_h3v3.pdf")
par(mfrow=c(3,2))
smoothScatter(test.melt$groseq_tssr_rep2_norm.median, test.melt$h2az.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H2A.Z")
lines(lowess(test.melt$groseq_tssr_rep2_norm.median, test.melt$h2az.median),col='black', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm.median, test.melt$h2az.median, method="spearman")

smoothScatter(test.melt$groseq_tssr_rep2_norm.median, test.melt$h3v3.median, colramp=colorRampPalette((brewer.pal(9, "OrRd")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H3.3")
lines(lowess(test.melt$groseq_tssr_rep2_norm.median, test.melt$h3v3.median),col='black', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm.median, test.melt$h3v3.median, method = "spearman")

smoothScatter(test.melt$groseq_tssr_rep2_norm.median, test.melt$chd4.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "Chd")
lines(lowess(test.melt$groseq_tssr_rep2_norm.median, test.melt$chd4.median),col='black', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm.median, test.melt$chd4.median, method="spearman")

smoothScatter(test.melt$groseq_tssr_rep2_norm.median, test.melt$brg1.median, colramp=colorRampPalette((brewer.pal(9, "OrRd")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "Brg1")
lines(lowess(test.melt$groseq_tssr_rep2_norm.median, test.melt$brg1.median),col='black', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm.median, test.melt$brg1.median, method="spearman")

smoothScatter(test.melt$atac_wt.median, test.melt$h2az.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "ATAC-seq", ylab = "H2A.Z")
lines(lowess(test.melt$atac_wt.median, test.melt$h2az.median),col='black', lty="dashed")
cor(test.melt$atac_wt.median, test.melt$h2az.median, method="spearman")

smoothScatter(test.melt$atac_wt.median, test.melt$h3v3.median, colramp=colorRampPalette((brewer.pal(9, "OrRd")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "ATAC-seq", ylab = "H3.3")
lines(lowess(test.melt$atac_wt.median, test.melt$h3v3.median),col='black', lty="dashed")
cor(test.melt$atac_wt.median, test.melt$h3v3.median, method = "spearman")

# smoothScatter(test.melt$groseq_gb_rep2_norm.median, test.melt$h2az.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H2A.Z")
# lines(lowess(test.melt$groseq_gb_rep2_norm.median, test.melt$h2az.median),col='red', lty="dashed")
# cor(test.melt$groseq_gb_rep2_norm.median, test.melt$h2az.median)
# 
# smoothScatter(test.melt$groseq_gb_rep2_norm.median, test.melt$h3v3.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H2A.Z")
# lines(lowess(test.melt$groseq_gb_rep2_norm.median, test.melt$h3v3.median),col='red', lty="dashed")
# cor(test.melt$groseq_gb_rep2_norm.median, test.melt$h3v3.median)

smoothScatter(test.melt$pol2_ser5p.median, test.melt$h2az.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "Pol2-Sep5p", ylab = "H2A.Z")
lines(lowess(test.melt$pol2_ser5p.median, test.melt$h2az.median),col='black', lty="dashed")
cor(test.melt$pol2_ser5p.median, test.melt$h2az.median)

smoothScatter(test.melt$pol2_ser5p.median, test.melt$h3v3.median, colramp=colorRampPalette((brewer.pal(9, "OrRd")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "Pol2-Sep5p", ylab = "H3.3")
lines(lowess(test.melt$pol2_ser5p.median, test.melt$h3v3.median),col='black', lty="dashed")
cor(test.melt$pol2_ser5p.median, test.melt$h3v3.median)

p1 <- ggplot(test.melt, aes(groseq_tssr_rep2_norm.median, h2az.median)) + geom_point(alpha=0.6, color="#784c97") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#784c97", fill="#784c97") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic()

p2 <- ggplot(test.melt, aes(groseq_tssr_rep2_norm.median, h3v3.median)) + geom_point(alpha=0.6, color="#f49d58") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#f49d58", fill="#f49d58") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic() +
  scale_y_continuous(position = "right")

p3 <- ggplot(test.melt, aes(atac_wt.median, h2az.median)) + geom_point(alpha=0.6, color="#784c97") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#784c97", fill="#784c97") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic()

p4 <- ggplot(test.melt, aes(atac_wt.median, h3v3.median)) + geom_point(alpha=0.6, color="#f49d58") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#f49d58", fill="#f49d58") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic() +
  scale_y_continuous(position = "right")

ggarrange(p1, p2, p3, p4, ncol=2, nrow=2)

dev.off()

```

##################################################
#### CHD4 AND BRG1 BINDING FOLD CHANGE AND POL2 PAUSING ANALYSIS (MEF)
##################################################
```{r}
# test <- gene_enzyme_mef[which(gene_enzyme_mef$expr_cluster_mef!="low" & gene_enzyme_mef$chd4_mef > summary(gene_enzyme_mef$chd4_mef)[2] & gene_enzyme_mef$brg1_mef > summary(gene_enzyme_mef$brg1_mef)[2] & gene_enzyme_mef$pol2_wt_mef > summary(gene_enzyme_mef$pol2_wt_mef)[2]),]

#test <- gene_enzyme_mef[which((gene_enzyme_mef$class_mef=="N3" | gene_enzyme_mef$class_mef=="N4" | gene_enzyme_mef$class_mef=="Y1") & !is.na(gene_enzyme_mef$expr_mef)),]
test <- gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad!="other"),]
table(test$class_mef_broad)
test[,c("chd4_mef", "brg1_mef")] <- normalize.quantiles(as.matrix(test[,c("chd4_mef", "brg1_mef")]))

test$ratio_chd4_brg1 <- cut2(log2((test$chd4_mef+1)/(test$brg1_mef+1)), m=20)
# test$ratio_chd4_brg1 <- cut2((test$chd4_mef)/(test$brg1_mef), m=100)
# test[,c("pol2_ser2p_mef", "pol2_ser5p_mef")] <- normalize.quantiles(as.matrix(test[,c("pol2_ser2p_mef", "pol2_ser5p_mef")]))
# test$ratio_ser2p_ser5p <- test$pol2_ser2p_mef/test$pol2_ser5p_mef
test$ratio_ser2p_ser5p <- as.numeric(normalize.quantiles(as.matrix(test$pol2_ser2p_mef))/normalize.quantiles(as.matrix(test$pol2_ser5p_mef)))
test$ratio_ser2p_wt <- as.numeric(normalize.quantiles(as.matrix(test$pol2_ser2p_mef))/normalize.quantiles(as.matrix(test$pol2_wt_mef)))
test$ratio_ser5p_wt <- as.numeric(normalize.quantiles(as.matrix(test$pol2_ser5p_mef))/normalize.quantiles(as.matrix(test$pol2_wt_mef)))
test <- test[order(test$ratio_chd4_brg1),]
# barplot(test$brg1)
# barplot(test$chd4)

## gene class counts
## N1+N2 are inactive (N=8313);
## N3: repressed (N=1851);
## N4+Y1: active (N=10489; N4 + 50 from Y1 are low expressed)
## table(gene_enzyme_mef[which(!is.na(gene_enzyme_mef$expr_mef)),]$class_mef)
table(test[,c("class_mef", "expr_cluster_mef")])

##----------------------------------------------------##
## plot heatmap for average or highly expressed genes (prc2 bound)
##----------------------------------------------------##

#test.prc2 <- test[which(test$expr_cluster_mef=="low" | test$class_mef=="N3" | test$class_mef=="N4"),]
test.prc2 <- test[which(test$expr_cluster_mef=="low" | test$expr_cluster_mef=="repressed"),]

test.melt <- melt(summaryBy(chd4_mef + brg1_mef ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(ezh2_mef ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3k27me3_mef ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(expr_mef ~ ratio_chd4_brg1, data = test.prc2, FUN = list(median)))
p4 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() +
   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.2)(256))) + 
   theme(axis.text.x = element_blank())

g <- arrangeGrob(p1, p2, p3, p4, nrow = 4, ncol=1)
ggsave(g, dpi=300, height=12, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_prc2_mef.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## plot heatmap for average or highly expressed genes (ep400 bound)
##----------------------------------------------------##

#test.ep400 <- test[which(test$expr_cluster_mef!="low" & test$class_mef=="Y1"),]
test.ep400 <- test[which(test$expr_cluster_mef=="average" | test$expr_cluster_mef=="high"),]

test.melt <- melt(summaryBy(chd4_mef + brg1_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.6)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3k27ac_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_wt_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3v3_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p4 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.5)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h2az_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p5 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser2p_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p6 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.4)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser5p_mef ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p7 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.4)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(ratio_ser2p_ser5p ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
p8 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.8)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pausing_wt_rep1_mef ~ ratio_chd4_brg1, data = test.ep400[which(!is.na(test.ep400$pausing_wt_rep1)),], FUN = list(median)))
p9 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=(value))) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.3)(256))) + 
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(expr_mef ~ ratio_chd4_brg1, data = test.ep400[!is.na(test.ep400$expr_mef),], FUN = list(median)))
p10 <- ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() +
   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) + 
   theme(axis.text.x = element_blank())

g <- arrangeGrob(p1, p2, p3, p4, p5, p6, p7, p8, p10, nrow = 9, ncol=1)
ggsave(g, dpi=300, height=25, width=20, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_pol2_mef.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## correlation plots between H2A.Z, H3.3 and Pol2 pausing in MEF (ep400 bound)
##----------------------------------------------------##

groseq_mef <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/mef/allGenes_divergentTranscription.bed")
colnames(groseq_mef) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss",
                  "groseq_wt_rep1_mef", "groseq_wt_rep2_mef", "groseq_wt_rep3_mef")
# groseq_mef <- groseq_mef[which(groseq_mef$dist_closest_tss > 3000),]
tssr <- t(apply(groseq_mef[,c(11:13)], 1, function(y) unlist(lapply(y, function(x) sum(as.numeric(unlist(strsplit(as.character(x), "/"))[1:2]))))))
colnames(tssr) <- c("groseq_tssr_rep1_mef", "groseq_tssr_rep2_mef", "groseq_tssr_rep3_mef")
gb <- t(apply(groseq_mef[,c(11:13)], 1, function(x) as.numeric(gsub("^.*/", "", x))))
colnames(gb) <- c("groseq_gb_rep1_mef", "groseq_gb_rep2_mef", "groseq_gb_rep3_mef")
groseq_mef <- cbind(groseq_mef[,c(4),drop=F], tssr, gb)
groseq_mef$groseq_tssr_rep1_norm_mef <- groseq_mef$groseq_tssr_rep1_mef/(groseq_mef$groseq_tssr_rep1_mef+groseq_mef$groseq_gb_rep1_mef)
groseq_mef$groseq_tssr_rep2_norm_mef <- groseq_mef$groseq_tssr_rep2_mef/(groseq_mef$groseq_tssr_rep2_mef+groseq_mef$groseq_gb_rep2_mef)
groseq_mef$groseq_gb_rep1_norm_mef <- groseq_mef$groseq_gb_rep1_mef/(groseq_mef$groseq_tssr_rep1_mef+groseq_mef$groseq_gb_rep1_mef)
groseq_mef$groseq_gb_rep2_norm_mef <- groseq_mef$groseq_gb_rep2_mef/(groseq_mef$groseq_tssr_rep2_mef+groseq_mef$groseq_gb_rep2_mef)

#test <- merge(gene_enzyme_mef[which((gene_enzyme_mef$class_mef=="N3" | gene_enzyme_mef$class_mef=="N4" | gene_enzyme_mef$class_mef=="Y1") & !is.na(gene_enzyme_mef$expr_mef)),], groseq_mef, by.x="name_mef", by.y="gene")
test <- merge(gene_enzyme_mef[which(gene_enzyme_mef$class_mef_broad!="other"),], groseq_mef, by.x="name_mef", by.y="gene")
test[,c("chd4_mef", "brg1_mef")] <- normalize.quantiles(as.matrix(test[,c("chd4_mef", "brg1_mef")]))
test$ratio_chd4_brg1 <- cut2(log2((test$chd4_mef+1)/(test$brg1_mef+1)), m=20)
test$ep400_cluster <- cut2(test$groseq_tssr_rep2_mef, m=20)

#test.ep400 <- test[which(test$expr_cluster_mef!="low" & test$class_mef=="Y1"),]
test.ep400 <- test[which(test$expr_cluster_mef=="average" | test$expr_cluster_mef=="high"),]
# test.ep400[,c("groseq_tssr_rep1_mef", "groseq_tssr_rep2_mef", "groseq_gb_rep1_mef", "groseq_gb_rep2_mef")] <- normalize.quantiles(as.matrix(test.ep400[,c("groseq_tssr_rep1_mef", "groseq_tssr_rep2_mef", "groseq_gb_rep1_mef", "groseq_gb_rep2_mef")]))
# test.ep400$groseq_wt_rep1_pausing <- test.ep400$groseq_tssr_rep1_mef/(test.ep400$groseq_tssr_rep1_mef + test.ep400$groseq_gb_rep1_mef)
# test.ep400$groseq_wt_rep2_pausing <- test.ep400$groseq_tssr_rep2_mef/(test.ep400$groseq_tssr_rep2_mef + test.ep400$groseq_gb_rep2_mef)
# 
# test.melt <- melt(summaryBy(groseq_wt_rep1_pausing + groseq_wt_rep2_pausing ~ ratio_chd4_brg1, data = test.ep400, FUN = list(median)))
# ggplot(test.melt, aes(ratio_chd4_brg1, variable, fill=log(value))) + geom_tile() +
#  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
#  theme(axis.text.x = element_blank())

test.melt <- summaryBy(h2az_mef + h3v3_mef + groseq_tssr_rep1_mef + groseq_tssr_rep2_mef + pol2_ser5p_mef + pol2_ser2p_mef +
                         groseq_tssr_rep1_norm_mef + groseq_tssr_rep2_norm_mef + groseq_gb_rep1_norm_mef + groseq_gb_rep2_norm_mef + 
                         chd4_mef + brg1_mef + atac_mef ~ ep400_cluster, data = test.ep400, FUN = list(median))

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/cor_groseq_h2az_h3v3_mef.pdf")
par(mfrow=c(3,2))
smoothScatter(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$h2az_mef.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H2A.Z")
lines(lowess(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$h2az_mef.median),col='red', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$h2az_mef.median, method="spearman")

smoothScatter(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$h3v3_mef.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H3.3")
lines(lowess(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$h3v3_mef.median),col='red', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$h3v3_mef.median, method="spearman")

smoothScatter(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$chd4_mef.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "Chd")
lines(lowess(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$chd4_mef.median),col='black', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$chd4_mef.median, method="spearman")

smoothScatter(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$brg1_mef.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "Brg1")
lines(lowess(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$brg1_mef.median),col='black', lty="dashed")
cor(test.melt$groseq_tssr_rep2_norm_mef.median, test.melt$brg1_mef.median, method="spearman")

smoothScatter(test.melt$atac_mef.median, test.melt$chd4_mef.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "Chd")
lines(lowess(test.melt$atac_mef.median, test.melt$chd4_mef.median),col='black', lty="dashed")
cor(test.melt$atac_mef.median, test.melt$chd4_mef.median, method="spearman")

smoothScatter(test.melt$atac_mef.median, test.melt$brg1_mef.median, colramp=colorRampPalette((brewer.pal(9, "BuPu")), bias=1), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "Brg1")
lines(lowess(test.melt$atac_mef.median, test.melt$brg1_mef.median),col='black', lty="dashed")
cor(test.melt$atac_mef.median, test.melt$brg1_mef.median, method="spearman")

# smoothScatter(test.melt$groseq_gb_rep2_norm_mef.median, test.melt$h2az_mef.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H2A.Z")
# lines(lowess(test.melt$groseq_gb_rep2_norm_mef.median, test.melt$h2az_mef.median),col='red', lty="dashed")
# cor(test.melt$groseq_gb_rep2_norm_mef.median, test.melt$h2az_mef.median)
# 
# smoothScatter(test.melt$groseq_gb_rep2_norm_mef.median, test.melt$h3v3_mef.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "GRO-seq", ylab = "H2A.Z")
# lines(lowess(test.melt$groseq_gb_rep2_norm_mef.median, test.melt$h3v3_mef.median),col='red', lty="dashed")
# cor(test.melt$groseq_gb_rep2_norm_mef.median, test.melt$h3v3_mef.median)

smoothScatter(test.melt$pol2_ser5p_mef.median, test.melt$h2az_mef.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Pol2-Sep5p", ylab = "H2A.Z")
lines(lowess(test.melt$pol2_ser5p_mef.median, test.melt$h2az_mef.median),col='red', lty="dashed")
cor(test.melt$pol2_ser5p_mef.median, test.melt$h2az_mef.median, method="spearman")

smoothScatter(test.melt$pol2_ser5p_mef.median, test.melt$h3v3_mef.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Pol2-Sep5p", ylab = "H3.3")
lines(lowess(test.melt$pol2_ser5p_mef.median, test.melt$h3v3_mef.median),col='red', lty="dashed")
cor(test.melt$pol2_ser5p_mef.median, test.melt$h3v3_mef.median, method="spearman")

p1 <- ggplot(test.melt, aes(groseq_tssr_rep2_norm_mef.median, h2az_mef.median)) + geom_point(alpha=0.6, color="#784c97") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#784c97", fill="#784c97") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic()

p2 <- ggplot(test.melt, aes(groseq_tssr_rep2_norm_mef.median, h3v3_mef.median)) + geom_point(alpha=0.6, color="#f49d58") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#f49d58", fill="#f49d58") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic() +
  scale_y_continuous(position = "right")

p3 <- ggplot(test.melt, aes(atac_mef.median, h2az_mef.median)) + geom_point(alpha=0.6, color="#784c97") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#784c97", fill="#784c97") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic()

p4 <- ggplot(test.melt, aes(atac_mef.median, h3v3_mef.median)) + geom_point(alpha=0.6, color="#f49d58") + 
  geom_smooth(method="loess", se=TRUE, fullrange=FALSE, level=0.95, color="#f49d58", fill="#f49d58") +
  #stat_density_2d(aes(fill = stat(level)), geom = "polygon") +
  #scale_fill_viridis_c() +
  theme_classic() +
  scale_y_continuous(position = "right")

ggarrange(p1, p2, p3, p4, ncol=2, nrow=2)
dev.off()

```

##################################################
#### CHD4 AND BRG1 BINDING FOLD CHANGE AND POL2 PAUSING ANALYSIS (ESC vs MEF)
##################################################
```{r}
test <- gene_enzyme_mef[which(abs(gene_enzyme_mef$fc)>1),]
## this condition: gene_enzyme_mef$expr_cluster_esc!="low" | gene_enzyme_mef$expr_cluster_mef!="low" does not matter; total genes reamin 5745

# test <- gene_enzyme_mef[gene_enzyme_mef$name_mef %in% gene_enzyme_enhancer[which(gene_enzyme_enhancer$ep400_cluster_merged=="high"),]$gene,]

test[,c("chd4_mef", "brg1_mef", "chd4", "brg1")] <- normalize.quantiles(as.matrix(test[,c("chd4_mef", "brg1_mef", "chd4", "brg1")]))
test[,c("h3v3", "h3v3_mef")] <- normalize.quantiles(as.matrix(test[,c("h3v3", "h3v3_mef")]))
test[,c("h2az", "h2az_mef")] <- normalize.quantiles(as.matrix(test[,c("h2az", "h2az_mef")]))
test[,c("pol2_ser5p", "pol2_ser5p_mef")] <- normalize.quantiles(as.matrix(test[,c("pol2_ser5p", "pol2_ser5p_mef")]))
test[,c("pol2_ser2p", "pol2_ser2p_mef")] <- normalize.quantiles(as.matrix(test[,c("pol2_ser2p", "pol2_ser2p_mef")]))
test[,c("pol2_wt", "pol2_wt_mef")] <- normalize.quantiles(as.matrix(test[,c("pol2_wt", "pol2_wt_mef")]))
# test[,c("pol2_ser5p", "pol2_ser2p", "pol2_wt", "pol2_ser5p_mef", "pol2_ser2p_mef", "pol2_wt_mef")] <- normalize.quantiles(as.matrix(test[,c("pol2_ser5p", "pol2_ser2p", "pol2_wt", "pol2_ser5p_mef", "pol2_ser2p_mef", "pol2_wt_mef")]))
test[,c("expr_esc", "expr_mef")] <- normalize.quantiles(as.matrix(test[,c("expr_esc", "expr_mef")]))
test[,c("ezh2", "ezh2_mef")] <- normalize.quantiles(as.matrix(test[,c("ezh2", "ezh2_mef")]))
test[,c("h3k27me3", "h3k27me3_mef")] <- normalize.quantiles(as.matrix(test[,c("h3k27me3", "h3k27me3_mef")]))

test$ratio_chd4_brg1_mef <-  (test$chd4_mef/test$brg1_mef)
test$ratio_chd4_brg1 <-  (test$chd4/test$brg1)

test$ratio_class <- cut2(log2((test$ratio_chd4_brg1_mef)/(test$ratio_chd4_brg1)), m = 100)
# test$ratio_class <- cut2(log2(rank(test$ratio_chd4_brg1_mef)/rank(test$ratio_chd4_brg1)), g = 5)
test$ratio_class <- cut2(log2((test$ratio_chd4_brg1_mef)/(test$ratio_chd4_brg1)), c(-0.7343,0.7362))
## very interesting patterns by defining classes using gene expr. fc
# test$ratio_class <- cut2(test$fc, m = 100)
test <- test[order(test$ratio_class),]
table(test$ratio_class)
#test[which()]

# gro-seq pausing
test1 <- groseq_esc
test1[which(log(test1$groseq_tssr_rep2)<4),]$groseq_tssr_rep2_norm <- 0
test2 <- groseq_mef
test2[which(log(test2$groseq_tssr_rep2_mef)<5),]$groseq_tssr_rep2_norm_mef <- 0

test <- merge(test, merge(test1[,c("gene", "groseq_tssr_rep1_norm", "groseq_tssr_rep2_norm")], test2[,c("gene", "groseq_tssr_rep1_norm_mef", "groseq_tssr_rep2_norm_mef")], by.x="gene", by.y="gene"), by.x="name_mef", by.y="gene", all.x=T)
test[,c("groseq_tssr_rep1_norm", "groseq_tssr_rep2_norm", "groseq_tssr_rep1_norm_mef", "groseq_tssr_rep2_norm_mef")] <- normalize.quantiles(as.matrix(test[,c("groseq_tssr_rep1_norm", "groseq_tssr_rep2_norm", "groseq_tssr_rep1_norm_mef", "groseq_tssr_rep2_norm_mef")]))

# pol2 pausing
df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_pol2pausing.bed")
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss", 
                  "pausing_brg1_wt_rep1", "pausing_brg1_wt_rep2",
                  "pausing_brg1_ko_rep1", "pausing_brg1_ko_rep2",
                  "pausing_mbd3_24h_rep1", "pausing_mbd3_24h_rep2", "pausing_mbd3_24h_rep3",
                  "pausing_mbd3_0h_rep1", "pausing_mbd3_0h_rep2", "pausing_mbd3_0h_rep3",
                  "pausing_wt_rep1", "pausing_wt_rep2", "pausing_brg1_kd_rep1", "pausing_brg1_kd_rep2",
                  "pausing_chd4_kd_rep1", "pausing_chd4_kd_rep2", "pausing_ep400_kd_rep1", "pausing_ep400_kd_rep2",
                  "pausing_wt_rep3", "pausing_brg1_kd_rep3", "pausing_chd4_kd_rep3")
test <- merge(test, df[,c("gene","pausing_wt_rep1","pausing_wt_rep2")], by.x="name_mef", by.y="gene", all.x=T)
test[,c("pausing_wt_rep1_mef", "pausing_wt_rep1", "pausing_wt_rep2")] <- normalize.quantiles(as.matrix(test[,c("pausing_wt_rep1_mef", "pausing_wt_rep1", "pausing_wt_rep2")]))

# df <- (merge(test[,c("name_mef", "ratio_class")], gene_enzyme_enhancer[,c("gene", "ep400_cluster_merged", "chd4_fc", "brg1_fc")], by.x="name_mef", by.y="gene"))
# df$ep400_cluster_merged <- as.character(df$ep400_cluster_merged)
# table(df[,c(2,3)])
# 136+29+473 (678); 197+73 (270)
# 108+32+331 (471); 330+107 (437)

# barplot(table((gene_enzyme_enhancer[which(gene_enzyme_enhancer$gene %in% test[which(test$ratio_class=="[-4.952,-0.734)"),]$name_mef),]$ep400_cluster_merged)), main="chd4-driven")
# barplot(table((gene_enzyme_enhancer[which(gene_enzyme_enhancer$gene %in% test[which(test$ratio_class=="[ 0.736, 4.543]"),]$name_mef),]$ep400_cluster_merged)), main="brg1-driven")

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/overlap_mef_esc_specific_genes.pdf")
par(mfrow=c(1,2))
barplot(table(test[test$name_mef %in% gene_mef$V1,]$ratio_class)/table(test$ratio_class), ylim=c(0,0.12), main="MEF-specifc", las=2)
barplot(table(test[test$name_mef %in% gene_esc$V1,]$ratio_class)/table(test$ratio_class), ylim=c(0,0.12), main="ESC-specific", las=2)
dev.off()

# mef48.up <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/mef48.up")
# mef48.down <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/mef48.down")
# plot(Vennerable::Venn(list(chd4=test[which(test$ratio_class=="[-4.952,-0.734)"),]$name_mef, mef48.up=mef48.up$V1, mef48.down=mef48.down$V1)))
# plot(Vennerable::Venn(list(brg1=test[which(test$ratio_class=="[ 0.736, 4.543]"),]$name_mef, mef48.up=mef48.up$V1, mef48.down=mef48.down$V1)))

test.melt <- melt(summaryBy(chd4 + chd4_mef ~ ratio_class, data = test, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  #theme(axis.text.x = element_text(angle = 45, hjust = 1))
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(brg1 + brg1_mef ~ ratio_class, data = test, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h2az + h2az_mef ~ ratio_class, data = test, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(expr_esc + expr_mef ~ ratio_class, data = test, FUN = list(median)))
p4 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser5p + pol2_ser5p_mef ~ ratio_class, data = test, FUN = list(median)))
p5 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_ser2p + pol2_ser2p_mef ~ ratio_class, data = test, FUN = list(median)))
p6 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(pol2_wt + pol2_wt_mef ~ ratio_class, data = test, FUN = list(median)))
p7 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3k27me3 + h3k27me3_mef ~ ratio_class, data = test, FUN = list(median)))
p8 <- ggplot(test.melt, aes(ratio_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.3)(256))) +
  theme(axis.text.x = element_blank())

g1 <- ggarrange(p1, p2, p3, p4, p5, p6, p7, p8, nrow=8)

ggsave(g1, dpi=300, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/analysis_mef_esc_heatmap.pdf", useDingbats=FALSE)

test.melt <- melt(test[,c("chd4_mef","chd4", "ratio_class")])
p1 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("Chd4 signal (log2)") +
  stat_compare_means(comparisons = list(c("chd4_mef", "chd4")))

test.melt <- melt(test[,c("brg1_mef","brg1", "ratio_class")])
p2 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("Brg1 signal (log2)") +
  stat_compare_means(comparisons = list(c("brg1_mef", "brg1")))

test.melt <- melt(test[,c("h3v3_mef","h3v3", "ratio_class")])
p3 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("H3.3 signal (log2)") +
  stat_compare_means(comparisons = list(c("h3v3_mef", "h3v3")))

test.melt <- melt(test[,c("h2az_mef","h2az", "ratio_class")])
p4 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("H2A.Z signal (log2)") +
  stat_compare_means(comparisons = list(c("h2az_mef", "h2az")))

test.melt <- melt(test[,c("expr_mef","expr_esc", "ratio_class")])
p5 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("Gene expr. (log2)") +
  stat_compare_means(comparisons = list(c("expr_mef", "expr_esc")))

# test.melt <- melt(test[,c("pol2_wt_mef","pol2_wt", "ratio_class")])
# ggboxplot(test.melt, x = "variable", y = "value",
#          fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2",
#          facet.by = "ratio_class", legend="none",
#          #add = "boxplot", add.params = list(fill = "white"),
#          palette = c("#784b97", "#ebebeb", "#ffa054")) +
#   xlab("") +
#   ylab("Pol2 Ser5p (log2)") +
#   stat_compare_means(comparisons = list(c("pol2_wt_mef", "pol2_wt")))

# test$pol2_ser5p <- test$pol2_ser5p/test$pol2_wt
# test$pol2_ser5p_mef <- test$pol2_ser5p_mef/test$pol2_wt_mef

test.melt <- melt(test[,c("pol2_ser5p_mef","pol2_ser5p", "ratio_class")])
p6 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("Pol2 Ser5p (log2)") +
  stat_compare_means(comparisons = list(c("pol2_ser5p_mef", "pol2_ser5p")))

test$pol2_ser2p <- test$pol2_ser2p/test$pol2_wt
test$pol2_ser2p_mef <- test$pol2_ser2p_mef/test$pol2_wt_mef

test.melt <- melt(test[,c("pol2_ser2p_mef","pol2_ser2p", "ratio_class")])
p7 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("Pol2 Ser2p/Pol2 (log2)") +
  stat_compare_means(comparisons = list(c("pol2_ser2p_mef", "pol2_ser2p")))

# ggplot(test.melt, aes(value, color=variable)) + stat_ecdf() + facet_grid(.~ratio_class)
# ks.test(test.melt[which(test.melt$ratio_class=="[-4.952,-0.734)" & test.melt$variable=="pol2_ser2p_mef"),]$value,
#             test.melt[which(test.melt$ratio_class=="[-4.952,-0.734)" & test.melt$variable=="pol2_ser2p"),]$value, alternative="g")$p.value
# 
# ks.test(test.melt[which(test.melt$ratio_class=="[ 0.736, 4.543]" & test.melt$variable=="pol2_ser2p_mef"),]$value,
#             test.melt[which(test.melt$ratio_class=="[ 0.736, 4.543]" & test.melt$variable=="pol2_ser2p"),]$value, alternative="l")$p.value

# test.melt <- melt(test[,c("pol2_wt_mef","pol2_wt", "ratio_class")])
# p8 <- ggboxplot(test.melt, x = "variable", y = "value",
#          fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
#          facet.by = "ratio_class", legend="none",
#          #add = "boxplot", add.params = list(fill = "white"),
#          palette = c("#784b97", "#ebebeb", "#ffa054")) + 
#   xlab("") + 
#   ylab("Pol2 (log2)") +
#   stat_compare_means(comparisons = list(c("pol2_wt_mef", "pol2_wt")))

test.melt <- melt(test[,c("groseq_tssr_rep1_norm_mef","groseq_tssr_rep1_norm", "ratio_class")])
ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) + 
  xlab("") + 
  ylab("GRO-seq pausing (log2)") +
  stat_compare_means(comparisons = list(c("groseq_tssr_rep1_norm_mef", "groseq_tssr_rep1_norm")))

# test.melt <- melt(test[,c("pausing_wt_rep1_mef","pausing_wt_rep1", "ratio_class")])
# ggboxplot(test.melt, x = "variable", y = "value",
#          fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
#          facet.by = "ratio_class", legend="none",
#          #add = "boxplot", add.params = list(fill = "white"),
#          palette = c("#784b97", "#ebebeb", "#ffa054")) + 
#   xlab("") + 
#   ylab("Pol2 Ser5p (log2)") +
#   stat_compare_means(comparisons = list(c("pausing_wt_rep1_mef","pausing_wt_rep1")))

test.melt <- melt(test[,c("ezh2_mef","ezh2", "ratio_class")])
p8 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) +
  xlab("") +
  ylab("EZH2 (log2)") +
  stat_compare_means(comparisons = list(c("ezh2_mef", "ezh2")))

test.melt <- melt(test[,c("h3k27me3_mef","h3k27me3", "ratio_class")])
p9 <- ggboxplot(test.melt, x = "variable", y = "value",
         fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="log2", 
         facet.by = "ratio_class", legend="none",
         #add = "boxplot", add.params = list(fill = "white"),
         palette = c("#784b97", "#ebebeb", "#ffa054")) +
  xlab("") +
  ylab("H3K27me3 (log2)") +
  stat_compare_means(comparisons = list(c("h3k27me3_mef", "h3k27me3")))

test <- merge(test,gene_expr[,c("gene", "chd4_fc", "brg1_fc")], by.x="name_mef", by.y="gene")

# test.melt <- melt(test[,c("chd4_fc", "brg1_fc", "ratio_class")])
# ggboxplot(test.melt, x = "variable", y = "value",
#          fill = "ratio_class", notch=1, notchwidth = 0.6, yscale="none", 
#          facet.by = "ratio_class", legend="none",
#          #add = "boxplot", add.params = list(fill = "white"),
#          palette = c("#784b97", "#ebebeb", "#ffa054")) +
#   xlab("") +
#   ylab("Gene expr. (logFC)") +
#   stat_compare_means(comparisons = list(c("chd4_fc", "brg1_fc")))

g2 <- ggarrange(p1, p2, p3, p4, p5, p6, p7, p8, p9, nrow=5, ncol=2)

ggsave(g2, dpi=300, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/analysis_mef_esc_boxplot.pdf", useDingbats=FALSE, width=10, height=15)

# test.melt <- melt(test[,c("expr_mef", "expr_mef48", "expr_ipsc", "expr_esc", "ratio_class")])
# ggplot(test.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~ratio_class) + theme_bw()

# test$ratio_class <- cut2(log2((test$ratio_chd4_brg1_mef)/(test$ratio_chd4_brg1)), m = 100)
# test <- test[order(test$ratio_class),]
# table(test$ratio_class)
# 
# test.melt <- melt(test[,c("expr_mef","expr_esc", "ratio_class")])
# test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("ratio_class", "variable"))
# ggplot(test.melt, aes(ratio_class, value)) +
#   geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
#   geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
#   geom_point(shape = 21, size = 2, fill="white") +
#   facet_grid(.~variable) +
#   theme(legend.position="none") +
#   ylab("Pol2-ser2p")

# test.melt <- melt(test[,c("expr_esc", "ratio_class")])
# test.melt <- summarySE(test.melt, measurevar = "value", groupvars = c("ratio_class", "variable"))
# plot(row.names(test.melt) ~ test.melt$value, data = test.melt)
# sm.spl <- with(test.melt, smooth.spline(value, row.names(test.melt)))
# lines(sm.spl, col = "blue")

write.table(test, "/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/gene_enzyme_mef.bed", quote = F, col.names=T, row.names = F, sep="\t")
```

##################################################
#### GENE EXPRESSION ANALYSIS UPON KD OF DIFFERENT CHROMATIN MODIFIERS
##################################################
```{r}
##----------------------------------------------------##
## Correlation between gene expression upon kd of Chd4, Brg1 and EP400
##----------------------------------------------------##

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/correlation_chd4_ep400.pdf")

#smoothScatter(gene_enzyme$chd4_fc, gene_enzyme$ep400_fc, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (logFC)", ylab = "ep400 (logFC)")
# reg<-lm(chd4_fc ~ ep400_fc, data = gene_enzyme)
# abline(reg, col="red", lty="dashed")
# cor(gene_enzyme$chd4_fc, gene_enzyme$ep400_fc, method="pearson")

# test <- gene_enzyme_enhancer
# test$ep400_cluster <- cut2(test$ep400, m=5)
# test.melt <- (summaryBy(ep400_fc + chd4_fc + brg1_fc + ep400 + pol2_wt + pol2_ser5p + h3k27ac + h3v3 ~ ep400_cluster, data = test, FUN = list(median)))
# corrplot::corrplot(cor(test.melt[,c("ep400_fc.median", "chd4_fc.median", "brg1_fc.median", "ep400.median", "pol2_wt.median", "pol2_ser5p.median", "h3k27ac.median", "h3v3.median"),]), type="upper", method="number", order="hclust", col=c("black", "white"), bg="lightblue")
# corrplot::corrplot(cor(gene_enzyme_enhancer[,c("ep400_fc", "chd4_fc", "brg1_fc", "pol2_wt", "pol2_ser5p", "h3k27ac", "ep400", "h3v3"),], method="spearman"), type="upper", method="number", order="hclust", col=c("black", "white"), bg="lightblue")


test <- gene_enzyme
test$ep400_cluster <- cut2(test$pol2_wt, m=20)
table(test[,c(132,133)])

test <- merge(test, enzyme_tpm[,c(4,57:59)], by.x="gene", by.y="name")
test$atac_fc_brg1 <- log2(test$atac_brg1_kd/test$atac_wt)
test$atac_fc_chd4 <- log2(test$atac_chd4_kd/test$atac_wt)

test.melt <- summaryBy(ep400_fc + chd4_fc + chd6_fc + chd8_fc + brg1_fc + ring1_fc + pol2_wt + ezh2 + chd4 + brg1 ~ ep400_cluster, data = test, FUN = list(median))

smoothScatter(test.melt$chd4_fc.median, test.melt$ep400_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (logFC)", ylab = "ep400 (logFC)")
lines(lowess(test.melt$chd4_fc.median, test.melt$ep400_fc.median),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ ep400_fc.median, data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$chd4_fc.median, test.melt$ep400_fc.median, method = "spearman")

smoothScatter(test.melt$chd4_fc.median, test.melt$brg1_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (logFC)", ylab = "brg1 (logFC)")
lines(lowess(test.melt$chd4_fc.median, test.melt$brg1_fc.median),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ brg1_fc.median, data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$chd4_fc.median, test.melt$brg1_fc.median, method = "spearman")

smoothScatter(test.melt$chd4_fc.median, test.melt$ring1_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 (logFC)", ylab = "ring1 (logFC)")
lines(lowess(test.melt$chd4_fc.median, test.melt$ring1_fc.median),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ ring1_fc.median, data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$chd4_fc.median, test.melt$ring1_fc.median, method = "spearman")

smoothScatter(test.melt$chd4_fc.median, log(test.melt$pol2_wt.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 KD/WT (logFC)", ylab = "pol2 signal (log)")
lines(lowess(test.melt$chd4_fc.median, log(test.melt$pol2_wt.median)),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ log2(test.melt$pol2_wt.median), data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$chd4_fc.median, log(test.melt$pol2_wt.median), method = "spearman")

smoothScatter(test.melt$brg1_fc.median, log(test.melt$pol2_wt.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "brg1  KD/WT (logFC)", ylab = "pol2 signal (log)")
lines(lowess(test.melt$brg1_fc.median, log(test.melt$pol2_wt.median)),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ log(test.melt$pol2_wt.median), data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$brg1_fc.median, log(test.melt$pol2_wt.median), method="spearman")

smoothScatter(test.melt$chd4_fc.median, log(test.melt$ezh2.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "chd4 KD/WT (logFC)", ylab = "ezh2 signal (log)")
lines(lowess(test.melt$chd4_fc.median, log(test.melt$ezh2.median)),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ log2(test.melt$ezh2.median), data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$chd4_fc.median, log(test.melt$ezh2.median), method = "spearman")

smoothScatter(test.melt$brg1_fc.median, log(test.melt$ezh2.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "brg1 KD/WT (logFC)", ylab = "ezh2 signal (log)")
lines(lowess(test.melt$brg1_fc.median, log(test.melt$ezh2.median)),col='red', lty="dashed")
# reg<-lm(brg1_fc.median ~ log2(test.melt$ezh2.median), data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$brg1_fc.median, log(test.melt$ezh2.median), method = "spearman")

smoothScatter(test.melt$brg1_fc.median, log(test.melt$pol2_wt.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "brg1  KD/WT (logFC)", ylab = "pol2 signal (log)")
lines(lowess(test.melt$brg1_fc.median, log(test.melt$pol2_wt.median)),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ log(test.melt$pol2_wt.median), data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(test.melt$brg1_fc.median, log(test.melt$pol2_wt.median), method="spearman")

smoothScatter(log(test.melt$brg1.median), log(test.melt$chd4.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "brg1 signal (log)", ylab = "chd4 signal (log)")
lines(lowess(log(test.melt$brg1.median), log(test.melt$chd4.median)),col='red', lty="dashed")
# reg<-lm(chd4_fc.median ~ log(test.melt$pol2_wt.median), data = test.melt)
# abline(reg, col="red", lty="dashed")
cor(log(test.melt$brg1.median), log(test.melt$chd4.median), method="spearman")

PerformanceAnalytics::chart.Correlation(as.matrix(test.melt[,c(3,6,8,9)]), histogram=FALSE, pch=1, method = "spearman", raster=T)

PerformanceAnalytics::chart.Correlation(as.matrix(log(gene_enzyme[,c(16,20,21,22,25,33,62,63)])), histogram=FALSE, pch=1, method = "spearman", raster=T)

corrplot::corrplot(cor(log(gene_enzyme[,c(16,20,21,22,25,33,62,63)])))

test.melt <- summaryBy(chd4_fc + brg1_fc + pol2_wt + h3k27ac + ezh2 + h3k27me3 ~ ep400_cluster, data = test, FUN = list(median))
par(mfrow=c(2,2))
smoothScatter(test.melt$h3k27ac.median, test.melt$chd4_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$h3k27ac.median, test.melt$chd4_fc.median),col='red', lty="dashed")
smoothScatter(test.melt$h3k27ac.median, test.melt$brg1_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$h3k27ac.median, test.melt$brg1_fc.median),col='red', lty="dashed")

smoothScatter(test.melt$pol2_wt.median, test.melt$chd4_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$pol2_wt.median, test.melt$chd4_fc.median),col='red', lty="dashed")
smoothScatter(test.melt$pol2_wt.median, test.melt$brg1_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$pol2_wt.median, test.melt$brg1_fc.median),col='red', lty="dashed")

smoothScatter(test.melt$h3k27me3.median, test.melt$chd4_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$h3k27me3.median, test.melt$chd4_fc.median),col='red', lty="dashed")
smoothScatter(test.melt$h3k27me3.median, test.melt$brg1_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$h3k27me3.median, test.melt$brg1_fc.median),col='red', lty="dashed")

smoothScatter(test.melt$ezh2.median, test.melt$chd4_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$ezh2.median, test.melt$chd4_fc.median),col='red', lty="dashed")
smoothScatter(test.melt$ezh2.median, test.melt$brg1_fc.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$ezh2.median, test.melt$brg1_fc.median),col='red', lty="dashed")

smoothScatter(test.melt$pol2_wt.median, test.melt$h3k27ac.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$pol2_wt.median, test.melt$h3k27ac.median),col='red', lty="dashed")

smoothScatter(test.melt$ezh2.median, test.melt$h3k27me3.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T)
lines(lowess(test.melt$ezh2.median, test.melt$h3k27me3.median),col='red', lty="dashed")

PerformanceAnalytics::chart.Correlation(as.matrix(test.melt[,c(2:7)]), histogram=FALSE, pch=1, method = "spearman", raster=T)

test.melt <- melt(summaryBy(pol2_wt + ezh2 + chd4 + brg1 ~ ep400_cluster, data = test, FUN = list(median)))
p1 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.8)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(brg1_fc + chd4_fc ~ ep400_cluster, data = test, FUN = list(median)))
p2 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.7)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(atac_fc_brg1 + atac_fc_chd4 ~ ep400_cluster, data = test, FUN = list(median)))
p3 <- ggplot(test.melt, aes(ep400_cluster, variable, fill=(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1)(256))) +
  theme(axis.text.x = element_blank())

ggarrange(p1, p2, p3, nrow=3)

dev.off()

##----------------------------------------------------##
## Changes in spatial profile of various chromatin modifiers upon upon kd of:
## Brg1, Chd4, Chd6, Chd8, Mll2, EP400, Ring1, and H2A.Z
##----------------------------------------------------##

name <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$name))
test <- lapply(profile[,c(8:ncol(profile))], function(x) as.data.frame(matrix(x, nrow=length(x)/8, byrow = T)))
test <- lapply(test, setNames, nm <- c("2up", "1up", "up", "hfr", "nfr", "down", "1down", "2down"))
test <- lapply(test, cbind, class)
test <- lapply(test, cbind, distance)
test <- lapply(test, cbind, name)

for(i in c(9,21,25,29,37,45,57,61)) {
  NAME=gsub("_class", "", colnames(gene_expr[,c(1,i)])[2])
  p <- lapply(names(test)[c(2,3,6,7,8,11,12,15,19,21,22,23,26,28,31,35,43,44,45)], function(x) { 
    test.sub <- merge(test[[x]], gene_expr[,c(1,i,64)], by.x="name", by.y="gene", all=F)
    #test.sub <- merge(test[[11]], gene_expr[,c(1,21,64)], by.x="name", by.y="gene", all=F)
    colnames(test.sub) <- c(colnames(test.sub)[1:11], "geneClass", "promoter_class")
    #test.sub$geneClass <- sprintf("%s_%s", test.sub$geneClass, test.sub$promoter_class)
    plot_profile_by_geneFC(test.sub, x, "promoter")
    #c(table(gene_expr[which(gene_expr$chd4_class=="down" & gene_expr$cluster=="Y2"),]$promoter_class))/table(gene_expr[which(gene_expr$cluster=="Y2"),]$promoter_class)
    #plot_profile(test.sub[,c(2:11)], x)
  })
  test.sub <- merge(test[[2]], gene_expr[,c(1,i,64)], by.x="name", by.y="gene", all=F)
  colnames(test.sub) <- c(colnames(test.sub)[1:11], "geneClass", "promoter_class")
  p[[20]] <- plot_profile_by_geneFC(test.sub, x, "frequency")
  g <- do.call(grid.arrange, p)
  FILE=sprintf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/profile_%s_kd.pdf", NAME)
  ggsave(g, dpi=300, height=35, width=35, filename=FILE, useDingbats=FALSE)
}

# plot_list = list()
# j <- 1
# for(i in seq(5,61,by=4)) {
#   # t <- as.data.frame(table(gene_expr[,c(48,i)]))
#   t <- as.data.frame(table(gene_expr[which(gene_expr[,i]!="neutral"  & (gene_expr[,63]=="Y1" | gene_expr[,63]=="Y2")),c(63,i)]))
#   #t <- as.data.frame(table(gene_expr[which(gene_expr[,i]!="neutral"),c(48,i)]))
#   t <- t[which(t[,2]!="neutral"),]
#   t <- t[which(t[,1]=="Y1" | t[,1]=="Y2"),]
#   t[,1] <- as.factor(as.character(t[,1]))
#   t[,2] <- as.factor(as.character(t[,2]))
#   plot_list[[j]] <- freqPlot(t, "freq", colnames(gene_expr)[i], "dodge")
#   # plot_list[[j]] <- piePlot(t, "freq", colnames(gene_expr)[i], "dodge")
#   j <- j+1
# }
# 
# g <- arrangeGrob(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]], plot_list[[5]],
#                  plot_list[[6]], plot_list[[7]], plot_list[[8]], plot_list[[9]], plot_list[[10]],
#                  plot_list[[11]], plot_list[[12]], plot_list[[13]], plot_list[[14]], plot_list[[15]],  nrow = 5, ncol=3)
# ggsave(g, dpi=300, height=20, width=20, filename="pdf/test.pdf", useDingbats=FALSE)

```


######################################################
#### Gehre et al. 2020 (H3.3 knock out)
#### Subramanian et al. 2013 (H2A.z knock out)
######################################################
```{r}
gehre <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/gehre_2020/gene_de.matrix", header=T)
gehre$gene_class <- "neutral"
gehre[which(gehre$log2FoldChange>0 & gehre$padj<0.05),]$gene_class <- "up"
gehre[which(gehre$log2FoldChange<0 & gehre$padj<0.05),]$gene_class <- "down"
df <- merge(gehre, gene_enzyme_enhancer[,c("gene", "ep400_cluster_merged", "pol2_wt", "cluster", "promoter_class_merged", "chd4_fc", "chd4_class", "brg1_fc", "brg1_class")], by.x="external_gene_name", by.y="gene")

subramanian <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/subramanian_2013/pgen.1003725.s007.txt", header=T, sep="\t")
subramanian <- subramanian[order(subramanian$GeneID, -subramanian$H2A.ZWT_RPKM),]
subramanian <- subramanian[match(unique(subramanian$GeneID), subramanian$GeneID),]
colnames(subramanian)[c(10,11)] <- c("log2FC_H2AZ", "log2FC_H2AZ_AP3")
dim(subramanian)
df <- merge(df, subramanian[,c("GeneID", "log2FC_H2AZ", "log2FC_H2AZ_AP3")], by.x="external_gene_name", by.y="GeneID")

dim(df)
df$ep400_cluster <- cut2(df$pol2_wt, m=25)

df.summary <- summaryBy(log2FoldChange + chd4_fc + brg1_fc + log2FC_H2AZ ~ ep400_cluster, data = df, FUN = list(median))
df.summary <- df.summary[!is.na(df.summary$log2FoldChange.median),]

# ggplot(df.summary, aes(df.summary$log2FoldChange.median, df.summary$brg1_fc.median)) + geom_point(alpha=0.6, color="#784c97") + 
#   geom_smooth(method="lm", se=TRUE, fullrange=FALSE, level=0.95, color="#784c97", fill="#784c97") +
#   theme_classic()

smoothScatter(df.summary$log2FoldChange.median, df.summary$brg1_fc.median, colramp=colorRampPalette(my.cols), 
              pch=19, cex=.8, useRaster=T, xlab = "Gene expr. FC (H3.3)", ylab = "Gene expr. FC (Brg1)")
lines(lowess(df.summary$log2FoldChange.median, df.summary$brg1_fc.median), col='red', lty="dashed")
cor(df.summary$log2FoldChange.median, df.summary$brg1_fc.median, use = "complete.obs")

smoothScatter(df.summary$log2FoldChange.median, df.summary$chd4_fc.median, colramp=colorRampPalette(my.cols), 
              pch=19, cex=.8, useRaster=T, xlab = "Gene expr. FC (H3.3)", ylab = "Gene expr. FC (Chd4)")
lines(lowess(df.summary$log2FoldChange.median, df.summary$chd4_fc.median), col='red', lty="dashed")
cor(df.summary$log2FoldChange.median, df.summary$chd4_fc.median, use = "complete.obs")

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chd4_brg1_h2az_h3v3_expr.pdf")
pheatmap(cor(df.summary[,c(2:5)], use="complete.obs", method = "spearman"), display_numbers = T, color = rev(colorBrewer2palette("RdYlBu", bias=0.7)))
ggplot(melt(df.summary), aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(5, "RdBu"), bias=1.2)(56))) +
  theme(axis.text.x = element_text(size=10, angle=45)) + theme_bw()
dev.off()
```

######################################################
#### PROFILE ANALYSIS OF CHROMATIN REMODELLERS
######################################################
```{r}
class <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$description))
distance <- as.numeric(apply(profile[grep("HFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[7]))
gene <- apply(profile[grep("HFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[4])
test <- lapply(profile[,c(8:ncol(profile))], function(x) as.data.frame(matrix(x, nrow=length(x)/8, byrow = T)))
test <- lapply(test, setNames, nm <- c("2up", "1up", "up", "hfr", "nfr", "down", "1down", "2down"))
test <- lapply(test, cbind, class)
test <- lapply(test, cbind, distance)
test <- lapply(test, cbind, gene)

test <- lapply(test, function(x) { merge(x, gene_enzyme_enhancer[,c(4,10)], by.x="gene", by.y="gene", all=F)[,c(2:9,10,11,1,12)] })
test <- lapply(test, function(x) { 
  cn <- colnames(x);
  x <- cbind(x, apply(x[, c(1:8)], 1, function(y) which(y == max(y))[1]));
  x <- cbind(x, (rowSums(x[,c(5:8)])-rowSums(x[,c(1:4)]))/(rowSums(x[,c(5:8)])+rowSums(x[,c(1:4)])))
  colnames(x) <- c(cn, "max_nucl", "directionality_bias");
  x$directionality_class <- "0"
  x[which(x$directionality_bias > 0.3),]$directionality_class <- "1"
  x[which(x$directionality_bias < -0.3),]$directionality_class <- "-1"
  x$directionality_class <- factor(x$directionality_class, levels=c("-1", "0", "1"), ordered = T)
  x <- cbind(x, rowSums(x[,c(1:8)]))
  colnames(x) <- c(cn, "max_nucl", "directionality_bias", "directionality_class", "total_signal");
  return(x);
})

test <- lapply(test, function(x) cbind(x[,c(1:16)], test$chd4$directionality_class))

#names(test[which(!names(test) %in% profile_order_y1)])
p <- lapply(profile_order_y1, function(x) { plot_profile(test[[x]], x, "promoter_directionality") })
g <- do.call(grid.arrange, p)
ggsave(g, dpi=300, height=30, width=30, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chromatin_remodellers_esc_promoters.pdf", useDingbats=FALSE)

# t.melt <- melt(test[[6]][,c(12,17)])
# ggplot(t.melt, aes(test$chd4$directionality_class, log(value))) + geom_boxplot()
# 
# table(test[[1]][,c(9,17)])

##----------------------------------------------------##

t <- lapply(test[profile_order_y1[1:52]], function(x) { 
  #x <- test[[1]]
  #x <- x[which(x$total_signal>summary(x$total_signal)[2]),]; 
  x <- melt(table(x$max_nucl)); 
  while(nrow(x)<8) { x <- rbind(x, c(nrow(x)+1,0)) }
  x$Var.1 <- as.factor(x$Var.1); 
  return(x); } )
# p <- lapply(seq_along(t), function(i) ggplot(t[[i]], aes(1, value, fill=as.factor(Var1))) + geom_bar(stat="identity", position="stack") + theme(legend.position="none", axis.title=element_blank(), axis.text=element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) + coord_flip() + ggtitle(names(t[i])) +  theme(plot.title = element_text(size = 40, face = "bold")))
# g <- do.call(grid.arrange, p)

t <- as.data.frame(t)[,c(1,seq(2,ncol(as.data.frame(t)),by=2))]
df.melt <- melt(t)
colnames(df.melt) <- c("xlabels", "class", "freq")
total <- as.vector(unlist(ddply(df.melt, .(class), summarise, X2=sum(freq))[2]))
df.melt$total <- rep(total, length(df.melt[,1])/length(total))
df.melt$density <- df.melt$freq/df.melt$total
df.melt$class <- factor(gsub(".value", "", df.melt$class), levels=profile_order_y1, ordered = T)
col <- c("#9970ab", "#c2a5cf", "#e7d4e8", "#bf812d", "#dfc27d","#a6dba0", "#5aae61", "#1b7837")
#col <- c("#9970ab", "#9970ab", "#9970ab", "#bf812d", "#dfc27d","#1b7837", "#1b7837", "#1b7837")
# col <- rev(colorRampPalette(brewer.pal(8, "Spectral"), bias=1)(8))
p <- ggplot(df.melt, aes(class, density, fill=factor(xlabels))) + 
  geom_bar(stat="identity", position = "stack") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  scale_fill_manual(values=col)
ggsave(p, dpi=300, height=20, width=50, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chromatin_remodellers_esc_Y1_promoters_barplot.pdf", useDingbats=FALSE, limitsize = F)

# p_y1 <- compute_cor_matrix_list(test, "pearson", 8)
#col=colorRampPalette(c("dodgerblue4","grey97", "sienna3"))(250);
# p_y1 <- compute_cor_matrix_list(test, "spearman", 8)
col=colorRampPalette(rev(brewer.pal(11, "RdBu")), bias=1)(250);
#col=colorRampPalette(c("dodgerblue4", "dodgerblue4", "grey97", "grey97", "sienna3", "sienna3"), bias=1.5)(250);

myBreaks <- c(seq(min(mat), 0, length.out=ceiling(250/2) + 1), 
              seq(max(mat)/250, max(mat), length.out=floor(250/2)))

# mat <- p_y1[c(2,3,6,7,12,14,15,16,19,21,23,24,25,26,28,29,34,35,36,40,44,45,49),
#             c(2,3,6,7,12,14,15,16,19,21,23,24,25,26,28,29,34,35,36,40,44,45,49)]

# mat <- p_y1[c(1:17,19:33,35:45,47:57),c(1:17,19:33,35:45,47:57)]
mat <- p_y1[which(colnames(p_y1) %in% profile_order_y1[c(1:52,54,55)]), which(colnames(p_y1) %in% profile_order_y1[c(1:52,54,55)])]

# mat <- p_y1
# res <- pvclust(mat, method.hclust = "ward.D", method.dist = "maximum", nboot=1000, iseed=21)
# plot(res)
# pvrect(res, alpha=0.95)

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chromatin_remodellers_esc_Y1_promoters_heatmap.pdf")
pheatmap(mat, color=col, border_color = NA, scale="none", cluster_rows = T,
         cluster_cols = T, legend=T, show_rownames = T, show_colnames = T,
         clustering_distance_rows="maximum", clustering_distance_cols="maximum",
         clustering_method = "mcquitty", breaks=myBreaks)
dev.off()

# ##----------------------------------------------------##
# df <- read.table(pipe("less /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters//esc/GENES_PROMOTER.BED.RC.CLUSTERS.profileTPM | grep -vw N1 | grep -vw N2 | grep -vw N3 | grep -vw Y1"), comment.char = "")
# colnames(df) <- c("chr", "start", "end", "name", "score", "strand", "description", "ash2l", "brg1", "cbx7", "chd1",
#                   "chd2", "chd4", "chd6", "chd8", "chd9", "cxxc1", "ep400", "ezh2", "hdac1", "kdm5b",
#                   "mll2", "mll34", "p300", "rad21", "ring1b", "set1a", "suz12", "h3k27ac", "h3k27me3", "h3k4me1",
#                   "h3k4me3", "h3k9me3", "h3k36me3", "h2az", "h3k4me2", "h3k9ac", "brd4", "lsd1", "cdk8", "ctcf", 
#                   "jarid2", "kdm2a", "med12", "med1", "nipbl", "smc1a", "tbp", "yy1", "h2azAC", "h3v3", "h3Only",
#                   "mbd3", "pol2", "myc", "max", "nelfa", "esrrb", "kdm5c", "cdk9", "cdk7", "paf1", "leo1", "cdc73",
#                   "pol2_ser2p", "pol2_ser5p", "pol2_ser7p")
# 
# class <- as.character(gsub("^.*#", "", df[grep("HFR", df$name),]$description))
# distance <- as.numeric(apply(df[grep("HFR", df$name),], 1, function(x) unlist(strsplit(x[7], "#"))[7]))
# gene <- apply(df[grep("HFR", df$name),], 1, function(x) unlist(strsplit(x[7], "#"))[4])
# test <- lapply(df[,c(8:57)], function(x) as.data.frame(matrix(x, nrow=length(x)/8, byrow = T)))
# test <- lapply(test, setNames, nm <- c("2up", "1up", "up", "hfr", "nfr", "down", "1down", "2down"))
# test <- lapply(test, cbind, class)
# test <- lapply(test, cbind, distance)
# test <- lapply(test, cbind, gene)
# 
# test <- lapply(test, function(x) { merge(x, gene_enzyme_enhancer[,c(4,10)], by.x="gene", by.y="gene", all=F)[,c(2:9,10,11,1,12)] })
# test <- lapply(test, function(x) { 
#   cn <- colnames(x);
#   x <- cbind(x, apply(x[, c(1:8)], 1, function(y) which(y == max(y))[1]));
#   x <- cbind(x, (rowSums(x[,c(5:8)])-rowSums(x[,c(1:4)]))/(rowSums(x[,c(5:8)])+rowSums(x[,c(1:4)])))
#   colnames(x) <- c(cn, "max_nucl", "directionality_bias");
#   x$directionality_class <- "0"
#   x[which(x$directionality_bias > 0.3),]$directionality_class <- "1"
#   x[which(x$directionality_bias < -0.3),]$directionality_class <- "-1"
#   x$directionality_class <- factor(x$directionality_class, levels=c("-1", "0", "1"), ordered = T)
#   return(x);
# })
# 
# t <- lapply(test[profile_order_y2], function(x) { x <- melt(table(x$max_nucl)); if(nrow(x)<8) { x <- rbind(x, c(8,0)) }; x$Var1 <- as.factor(x$Var1); return(x); } )
# # p <- lapply(seq_along(t), function(i) ggplot(t[[i]], aes(1, value, fill=as.factor(Var1))) + geom_bar(stat="identity", position="stack") + theme(legend.position="none", axis.title=element_blank(), axis.text=element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) + coord_flip() + ggtitle(names(t[i])) +  theme(plot.title = element_text(size = 40, face = "bold")))
# # g <- do.call(grid.arrange, p)
# 
# t <- as.data.frame(t)[,c(1,seq(2,100,by=2))]
# df.melt <- melt(t)
# colnames(df.melt) <- c("xlabels", "class", "freq")
# total <- as.vector(unlist(ddply(df.melt, .(class), summarise, X2=sum(freq))[2]))
# df.melt$total <- rep(total, length(df.melt[,1])/length(total))
# df.melt$density <- df.melt$freq/df.melt$total
# df.melt$class <- factor(gsub(".value", "", df.melt$class), levels=profile_order_y2, ordered = T)
# col <- c("#9970ab", "#c2a5cf", "#e7d4e8", "#bf812d", "#dfc27d","#a6dba0", "#5aae61", "#1b7837")
# # col <- rev(colorRampPalette(brewer.pal(8, "Spectral"), bias=1)(8))
# p <- ggplot(df.melt, aes(class, density, fill=factor(xlabels))) + 
#   geom_bar(stat="identity", position = "stack") +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
#   scale_fill_manual(values=col)
# ggsave(p, dpi=300, height=20, width=50, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chromatin_remodellers_esc_Y2_promoters_barplot.pdf", useDingbats=FALSE, limitsize = F)
# 
# # p_y2 <- compute_cor_matrix_list(test, "pearson", 8)
# # col=colorRampPalette(c("dodgerblue4","grey97", "sienna3"), bias=1)(250);
# # p_y2 <- compute_cor_matrix_list(test, "spearman", 8)
# 
# col=colorRampPalette(rev(brewer.pal(11, "RdBu")), bias=1)(250);
# myBreaks <- c(seq(min(mat), 0, length.out=ceiling(250/2) + 1), 
#               seq(max(mat)/250, max(mat), length.out=floor(250/2)))
# 
# mat <- p_y2[c(1,2,4,5,6,7,8,9,10,11,13,14,15,16,17,20,22,24,25,27,28,29,30,31,32,33,34,36,37:44,45,47:50),
#             c(1,2,4,5,6,7,8,9,10,11,13,14,15,16,17,20,22,24,25,27,28,29,30,31,32,33,34,36,37:44,45,47:50)]
# pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/chromatin_remodellers_esc_Y2_promoters_heatmap.pdf")
# pheatmap(mat, color=col, border_color = NA, scale="none", cluster_rows = T, 
#          cluster_cols = T, legend=T, show_rownames = T, show_colnames = T,
#          clustering_distance_rows="maximum", clustering_distance_cols="maximum",
#          clustering_method = "ward.D", breaks=myBreaks)
# dev.off()
# 
# ##----------------------------------------------------##
# ## binding profiles based on ep400 binding at promoters
# profile <- read.table(pipe("less /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters//esc/GENES_PROMOTER.BED.RC.CLUSTERS.profileTPM | grep -vw N1 | grep -vw N2 | grep -vw N3"), comment.char = "")
# colnames(profile) <- c("chr", "start", "end", "name", "score", "strand", "description", "ash2l", "brg1", "cbx7", "chd1",
#                        "chd2", "chd4", "chd6", "chd8", "chd9", "cxxc1", "ep400", "ezh2", "hdac1", "kdm5b",
#                        "mll2", "mll34", "p300", "rad21", "ring1b", "set1a", "suz12", "h3k27ac", "h3k27me3", "h3k4me1",
#                        "h3k4me3", "h3k9me3", "h3k36me3", "h2az", "h3k4me2", "h3k9ac", "brd4", "lsd1", "cdk8", "ctcf", 
#                        "jarid2", "kdm2a", "med12", "med1", "nipbl", "smc1a", "tbp", "yy1", "h2azAC", "h3v3", "h3Only", 
#                        "mbd3", "pol2", "myc", "max", "nelfa", "esrrb", "kdm5c", "cdk9", "cdk7", "paf1", "leo1", "cdc73",
#                        "pol2_ser2p", "pol2_ser5p", "pol2_ser7p")
# 
# class <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$description))
# distance <- as.numeric(apply(profile[grep("HFR", profile$name),], 1, function(x) unlist(strsplit(x[7], "#"))[7]))
# name <- as.character(gsub("^.*#", "", profile[grep("HFR", profile$name),]$name))
# test <- lapply(profile[,c(8:57)], function(x) as.data.frame(matrix(x, nrow=length(x)/8, byrow = T)))
# test <- lapply(test, setNames, nm <- c("2up", "1up", "up", "hfr", "nfr", "down", "1down", "2down"))
# test <- lapply(test, cbind, class)
# test <- lapply(test, cbind, distance)
# test <- lapply(test, cbind, name)
# 
# p <- lapply(names(test[profile_order_y2]), function(x) { 
#   test.sub <- merge(test[[x]], gene_enzyme_enhancer[,c(4,128,129,130,166,10)], by.x="name", by.y="gene", all=F)[,c(2:9,12,13,14,15,16)]
#   test.sub$ep400_cluster <- "average"
#   test.sub[which(test.sub$directionality_class=="-1"),]$ep400_cluster <- "high"
#   test.sub[which(test.sub$directionality_class=="1"),]$ep400_cluster <- "low"
#   # test.sub[which(log(test.sub$expr_esc+1)>3.3),]$ep400_cluster <- "high"
#   # test.sub[which(log(test.sub$expr_esc+1)<1.3),]$ep400_cluster <- "low"
#   colnames(test.sub) <- c(colnames(test.sub)[1:8], "geneClass", "promoter_class", "ep400_cluster")
#   #test.sub$geneClass <- sprintf("%s_%s", test.sub$geneClass, test.sub$promoter_class)
#   plot_profile_by_geneFC(test.sub[,c(1:11)], x, "ep400")
#   #c(table(gene_expr[which(gene_expr$chd4_class=="down" & gene_expr$cluster=="Y2"),]$promoter_class))/table(gene_expr[which(gene_expr$cluster=="Y2"),]$promoter_class)
#   #plot_profile(test.sub[,c(2:11)], x)
# })
# g <- do.call(grid.arrange, p)
# ggsave(g, dpi=300, height=35, width=35, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/profile_ep400_binding.pdf", useDingbats=FALSE)
# 
# p <- lapply(names(test)[c(2,3,6,7,8,11,12,15,19,21,22,23,26,28,31,35,43,44,45)], function(x) { 
#   test.sub <- merge(test[[x]], gene_enzyme[,c(4,123,124,126)], by.x="name", by.y="gene", all=F)[,c(2:9,12,13,14)]
#   colnames(test.sub) <- c(colnames(test.sub)[1:8], "geneClass", "promoter_class", "CpG_length")
#   #test.sub$geneClass <- sprintf("%s_%s", test.sub$geneClass, test.sub$promoter_class)
#   plot_profile_by_geneFC(test.sub, x, "CpG")
#   #c(table(gene_expr[which(gene_expr$chd4_class=="down" & gene_expr$cluster=="Y2"),]$promoter_class))/table(gene_expr[which(gene_expr$cluster=="Y2"),]$promoter_class)
#   #plot_profile(test.sub[,c(2:11)], x)
# })
# g <- do.call(grid.arrange, p)
# FILE=sprintf("pdf/profile_CpG_length.pdf", NAME)
# ggsave(g, dpi=300, height=35, width=35, filename=FILE, useDingbats=FALSE)
# 
# 
# test <- merge(gene_enzyme, df_esc[,c(4,32)], by.x="gene", by.y="gene")
# test.melt <- melt(test[,c(128,130,132)])
# ggplot(test.melt, aes(ep400_cluster, log(value))) + geom_boxplot() + facet_grid(.~cluster)
```

##################################################
#### CHD4/BRG1 DIRECTIONALITY ANALYSIS
##################################################
```{r}
##----------------------------------------------------##
## chd4 directionality enrichment per promoter class
##----------------------------------------------------##

df.melt <- data.frame(density(gene_enzyme_enhancer[which(gene_enzyme_enhancer$promoter_class_merged=="broad"),]$profile_directionality_chd4)[c("x", "y")])
p1 <- ggplot(df.melt, aes(x, y)) + 
  geom_area(data = subset(df.melt, x < -0.3), fill = "#72b796") +
  geom_area(data = subset(df.melt, x >= -0.3 & x <= 0.3), fill = "#c19d6c") +
  geom_area(data = subset(df.melt, x > 0.3), fill = "#b26799")

df.melt <- data.frame(density(gene_enzyme_enhancer[which(gene_enzyme_enhancer$promoter_class_merged=="medium"),]$profile_directionality_chd4)[c("x", "y")])
p2 <- ggplot(df.melt, aes(x, y)) + 
  geom_area(data = subset(df.melt, x < -0.3), fill = "#72b796") +
  geom_area(data = subset(df.melt, x >= -0.3 & x <= 0.3), fill = "#c19d6c") +
  geom_area(data = subset(df.melt, x > 0.3), fill = "#b26799")

df.melt <- data.frame(density(gene_enzyme_enhancer[which(gene_enzyme_enhancer$promoter_class_merged=="narrow"),]$profile_directionality_chd4)[c("x", "y")])
p3 <- ggplot(df.melt, aes(x, y)) + 
  geom_area(data = subset(df.melt, x < -0.3), fill = "#72b796") +
  geom_area(data = subset(df.melt, x >= -0.3 & x <= 0.3), fill = "#c19d6c") +
  geom_area(data = subset(df.melt, x > 0.3), fill = "#b26799")

df.melt <- data.frame(density(gene_enzyme_enhancer[which(gene_enzyme_enhancer$promoter_class_merged=="bivalent"),]$profile_directionality_chd4)[c("x", "y")])
p4 <- ggplot(df.melt, aes(x, y)) + 
  geom_area(data = subset(df.melt, x < -0.3), fill = "#72b796") +
  geom_area(data = subset(df.melt, x >= -0.3 & x <= 0.3), fill = "#c19d6c") +
  geom_area(data = subset(df.melt, x > 0.3), fill = "#b26799")

p5 <- ggplot(gene_enzyme_enhancer, aes(profile_directionality_chd4)) + 
  geom_density(aes(fill=promoter_class_merged, col=promoter_class_merged, alpha=0.05)) +
  #scale_fill_manual(values=c("#cbd5e8", "#7f2704", "#f16913", "#fdd0a2")) +
  #scale_color_manual(values=c("#cbd5e8", "#7f2704", "#f16913", "#fdd0a2"))
  theme(legend.position="none")

df.melt <- melt(table(gene_enzyme_enhancer[,c(164,170)]))
colnames(df.melt) <- c("xlabels", "class", "freq")
total <- as.vector(unlist(ddply(df.melt, .(xlabels), summarise, X2=sum(freq))[2]))
df.melt$total <- rep(total, length(df.melt[,1])/length(total))
df.melt$density <- df.melt$freq/df.melt$total
p6 <- ggplot(df.melt, aes(xlabels, density, fill=factor(class))) +
  geom_bar(stat="identity", position = "stack") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  geom_text(aes(label = sprintf("%0.2f", density)), size = 3, hjust = 0.5, vjust = 2, position = "stack") +
  theme(legend.position="none")

p7 <- ggplot(gene_enzyme_enhancer, aes(profile_directionality_chd4)) + 
  stat_density(aes(x=profile_directionality_chd4, y=..scaled..,fill=directionality_class, col=directionality_class), position="dodge", geom="area", alpha=0.5) +
  scale_fill_manual(values=c("#72b796", "#c19d6c", "#b26799")) +
  scale_color_manual(values=c("#72b796", "#c19d6c", "#b26799")) +
  theme(legend.position="none")
p8 <- ggplot(gene_enzyme_enhancer, aes(log(expr_esc))) + 
  stat_density(aes(x=log(expr_esc), y=..scaled..,fill=directionality_class, col=directionality_class), position="dodge", geom="area", alpha=0.5) +
  scale_fill_manual(values=c("#72b796", "#c19d6c", "#b26799")) +
  scale_color_manual(values=c("#72b796", "#c19d6c", "#b26799")) +
  theme(legend.position="none")
p9 <- ggplot(gene_enzyme_enhancer, aes(1, log(expr_esc), col=directionality_class)) + geom_violin() +
  geom_boxplot(width=0.2) +
  scale_color_manual(values=c("#72b796", "#c19d6c", "#b26799")) +
  #geom_jitter(aes(fill=directionality_class),shape=21,size=0.4,width=0.1,col="black",alpha=0.5)
  facet_grid(.~directionality_class) +
  theme(legend.position="none")
g <- arrangeGrob(p1, p2, p3, p4, p5, p6, p7, p8, p9, nrow=3, ncol=4)
ggsave(g, dpi=300, height=6, width=10, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/directionality_promoter_class.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## relationship between directionality bias and gene expression
##----------------------------------------------------##

test <- dplyr::arrange(gene_enzyme_enhancer, cluster, promoter_class_merged, profile_directionality_chd4, profile_directionality_brg1)
test$directionality_cluster <- cut2(test$profile_directionality_chd4, m=20)
test$ep400_cluster <- cut2(test$ep400, m=20)

test.melt <- melt(summaryBy(expr_esc ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
p1 <- ggplot(test.melt, aes(directionality_cluster, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.5)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())
test.melt <- melt(summaryBy(profile_directionality_chd4 + profile_directionality_brg1 + profile_directionality_ep400 + profile_directionality_ezh2 ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
p2 <- ggplot(test.melt, aes(directionality_cluster, variable, fill=value)) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "PRGn"), bias=1)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(h3k27ac ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
p3 <- ggplot(test.melt, aes(directionality_cluster, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=1.5)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())

# ggplot(test.melt, aes(x=directionality_cluster, y=log(value))) + 
#   geom_point() + facet_grid(.~promoter_class_merged)

# test.melt <- melt(summaryBy(expr_esc ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
# ggplot(test.melt, aes(x=directionality_cluster, y=log(value))) + 
#   geom_point() + facet_grid(.~promoter_class_merged) + 
#   geom_smooth(method=lm, se=FALSE)

test.melt <- melt(summaryBy(h3k27me3 ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
p4 <- ggplot(test.melt, aes(directionality_cluster, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.7)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())

# test.melt <- melt(summaryBy(pol2_pausing ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
# ggplot(test.melt, aes(directionality_cluster, variable, fill=value)) + geom_tile() + 
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.8)(256))) +
#   facet_grid(.~cluster + promoter_class_merged) +
#   theme(axis.text.x = element_blank())

# test.melt <- melt(summaryBy(profile_directionality_chd4 + profile_directionality_brg1 + profile_directionality_ep400 + profile_directionality_ezh2 ~ ep400_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
# ggplot(test.melt, aes(ep400_cluster, variable, fill=value)) + geom_tile() + 
#   scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "PRGn"), bias=1)(256))) +
#   facet_grid(.~cluster + promoter_class_merged) +
#   theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(chd4_fc ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
p5 <- ggplot(test.melt, aes(directionality_cluster, variable, fill=value)) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=0.5)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())
p6 <- ggplot(gene_enzyme_enhancer, aes(directionality_class_merged, chd4_fc)) + geom_boxplot()

test.melt <- melt(summaryBy(brg1_fc ~ directionality_cluster + promoter_class_merged + cluster, data = test, FUN = list(median)))
p7 <- ggplot(test.melt, aes(directionality_cluster, variable, fill=value)) + geom_tile() +
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(9, "RdBu"), bias=1.2)(256))) +
  facet_grid(.~cluster + promoter_class_merged) +
  theme(axis.text.x = element_blank())
p8 <- ggplot(gene_enzyme_enhancer, aes(directionality_class_merged, brg1_fc)) + geom_boxplot()

g <- arrangeGrob(p1, p2, p3, p4, p6, p8, nrow=6, ncol=1)
ggsave(g, dpi=300, height=16, width=6, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/directionality_expression_analysis.pdf", useDingbats=FALSE)

## check influence of closest TSS
pdf("pdf/directionality_dist_to_tss.pdf")
test.sort <- test[order(test$cluster, test$promoter_class_merged, test$directionality_cluster),]
test.sort$dist_closest_tss_class <- ifelse(test.sort$dist_closest_tss < 1000, 0, 1)
par(mfrow=c(5,2))
barplot(test.sort$profile_directionality_chd4, border=NA)
barplot(log(test.sort$expr_esc+1), border=NA)
image(as.matrix(test.sort[which(test.sort$promoter_class_merged=="bivalent"),]$dist_closest_tss_class), col=c("black", "white"))
image(as.matrix(test.sort[which(test.sort$promoter_class_merged=="narrow"),]$dist_closest_tss_class), col=c("black", "white"))
image(as.matrix(test.sort[which(test.sort$promoter_class_merged=="medium"),]$dist_closest_tss_class), col=c("black", "white"))
image(as.matrix(test.sort[which(test.sort$promoter_class_merged=="broad"),]$dist_closest_tss_class), col=c("black", "white"))
image(as.matrix(log(test.sort[which(test.sort$promoter_class_merged=="bivalent"),]$dist_closest_tss)))
image(as.matrix(log(test.sort[which(test.sort$promoter_class_merged=="narrow"),]$dist_closest_tss)))
image(as.matrix(log(test.sort[which(test.sort$promoter_class_merged=="medium"),]$dist_closest_tss)))
image(as.matrix(log(test.sort[which(test.sort$promoter_class_merged=="broad"),]$dist_closest_tss)))
dev.off()

##----------------------------------------------------##
## relationship between directionality bias and pol2 pausing, release
##----------------------------------------------------##

test <- merge(profile.lst$chd4, gene_enzyme_divergent[which(gene_enzyme_divergent$directionality_class_merged!="bivalent_1"),c(4,165,171)], by.x="gene", by.y="gene")
test <- test[,c(2:11,1,12:ncol(test))]
# ggplot(test, aes(ep400_cluster_merged, profile_directionality_chd4)) + geom_boxplot()
# ggplot(test, aes(directionality_class_merged, profile_directionality_ezh2)) + geom_boxplot()

test.melt <- melt(test[,c(1:8,9,13,14)])

ggplot(test.melt, aes(variable, group=1)) +
  stat_smooth(data = test.melt[which(test.melt$directionality_class_merged=="active_1"),], aes(y = log(value), group=1, colour="low"), method=lm, formula = y ~ poly(x,5), level=0.95, lty="dashed") +
  stat_smooth(data = test.melt[which(test.melt$directionality_class_merged=="active_0"),], aes(y = log(value), group=1, colour="average"), method=lm, formula = y ~ poly(x,5), level=0.95) +
  stat_smooth(data = test.melt[which(test.melt$directionality_class_merged=="active_-1"),], aes(y = log(value), group=1, colour="high"), method=lm, formula = y ~ poly(x,5), level=0.95) +
  ## colors are arranged alphabetically, meaning broad, medium and narrow
  scale_colour_manual("promoter", breaks = c("active_1", "active_0", "active_-1"), values = c("#4faf4a","#ef7e1b", "#e31e1e")) +
  theme_bw() +
  ggtitle("y") +
  scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
  theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
  theme(legend.position="none")

ggplot(test.melt, aes(variable, group=1)) +
  stat_smooth(data = test.melt[which(test.melt$ep400_cluster_merged=="low"),], aes(y = log(value), group=1, colour="low"), method=lm, formula = y ~ poly(x,5), level=0.95, lty="dashed") +
  stat_smooth(data = test.melt[which(test.melt$ep400_cluster_merged=="average"),], aes(y = log(value), group=1, colour="average"), method=lm, formula = y ~ poly(x,5), level=0.95) +
  stat_smooth(data = test.melt[which(test.melt$ep400_cluster_merged=="high"),], aes(y = log(value), group=1, colour="high"), method=lm, formula = y ~ poly(x,5), level=0.95) +
  ## colors are arranged alphabetically, meaning broad, medium and narrow
  scale_colour_manual("promoter", breaks = c("active_1", "active_0", "active_-1"), values = c("#4faf4a","#ef7e1b", "#e31e1e")) +
  theme_bw() +
  ggtitle("y") +
  scale_x_discrete(labels=c("-1''", "-1'", "-1", "cNFR", "NFR", "+1", "+1'", "+1''" )) +
  theme(axis.text.x = element_text(size=10, angle=45)) + xlab("") + ylab("Signal (TPKM, log)") +
  theme(legend.position="none")

##----------------------------------------------------##
## relationship between directionality bias, enhancer-promoter interaction and divergent transcription
##----------------------------------------------------##

df.melt <- melt(gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),c(171,152)])
p1 <- ggplot(df.melt, aes(directionality_class_merged, (value))) + geom_boxplot(fill="#bebdbd") + ylab("# enhancers") + ylim(c(0,40))
df.melt <- melt(gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),c(171,153)])
p2 <- ggplot(df.melt, aes(directionality_class_merged, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("Fire score (Dixon et al. 2012; log)")
df.melt <- melt(gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),c(171,154)])
p3 <- ggplot(df.melt, aes(directionality_class_merged, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("Fire score (Fraser et al. 2016; log)")
df.melt <- melt(gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),c(171,56)])
p4 <- ggplot(df.melt, aes(directionality_class_merged, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("YY1")
df.melt <- melt(gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),c(171,10)])
p5 <- ggplot(df.melt, aes(directionality_class_merged, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("Gene expr. (log)")
# df.melt <- melt(gene_enzyme_enhancer[,c(167,163)])
# p6 <- ggplot(df.melt, aes(directionality_class_merged, (value))) + geom_boxplot(fill="#bebdbd") + ylab("Profile directionality (Chd4)")
# df.melt <- melt(test[,c(171,177,178,182,183)])
# p6 <- ggplot(df.melt, aes(directionality_class_merged, (value))) + geom_boxplot(fill="#bebdbd") + ylab("Pol2 pausing")  + facet_grid(.~variable)
df.melt <- melt(gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),c(171,197:198)])
p6 <- ggplot(df.melt, aes(directionality_class_merged, value)) + geom_boxplot() + facet_grid(.~variable) + ylab("Directionality")
# df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y1"),c(159,143)])
# ggplot(df.melt, aes(directionality_class, (value))) + geom_boxplot(fill="#bebdbd") + ylab("# enhancers") + ylim(c(0,40))
# df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y1"),c(159,144)])
# ggplot(df.melt, aes(directionality_class, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("Fire score (Dixon et al. 2012; log)")
# df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y1"),c(159,145)])
# ggplot(df.melt, aes(directionality_class, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("Fire score (Fraser et al. 2016; log)")
# df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y1"),c(159,56)])
# ggplot(df.melt, aes(directionality_class, log(value))) + geom_boxplot(fill="#bebdbd") + ylab("YY1")

g <- arrangeGrob(p1, p2, p3, p4, p5, p6, nrow=3, ncol=2)
ggsave(g, dpi=300, height=6, width=6, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/enhancer_count_analysis.pdf", useDingbats=FALSE)

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/directionality_expression_analysis_corPlot.pdf")
test <- gene_enzyme_enhancer
test$profile_class <- cut2(test$profile_directionality_chd4, m=20)
test.melt <- summaryBy(profile_directionality_chd4 + profile_directionality_brg1 + profile_directionality_ep400 + profile_directionality_ezh2 + ezh2 + ep400 + expr_esc ~ profile_class, data = test, FUN = list(median))

## chd4 and brg1 directionality bias
smoothScatter(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_brg1.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Profile directionality (Chd4)", ylab = "Profile directionality (Brg1)")
lines(lowess(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_brg1.median), col='red', lty="dashed")
cor(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_brg1.median, use = "complete.obs")

smoothScatter(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_ezh2.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Profile directionality (Chd4)", ylab = "Profile directionality (Ezh2)")
lines(lowess(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_ezh2.median), col='red', lty="dashed")
cor(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_ezh2.median, use = "complete.obs")

smoothScatter(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_ep400.median, colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Profile directionality (Chd4)", ylab = "Profile directionality (EP400)")
lines(lowess(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_ep400.median), col='red', lty="dashed")
cor(test.melt$profile_directionality_chd4.median, test.melt$profile_directionality_ep400.median, use = "complete.obs")

## directionality bias (chd4) and expr
smoothScatter(test.melt$profile_directionality_chd4.median, log(test.melt$expr_esc.median+1), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Profile directionality (Chd4)", ylab = "Expr (log)")
cor(test.melt$profile_directionality_chd4.median, log(test.melt$expr_esc.median+1), use =  "complete.obs")
lines(lowess(test.melt$profile_directionality_chd4.median, log(test.melt$expr_esc.median+1)), col='red', lty="dashed")

## directionality bias (chd4) and Ezh2
# test <- gene_enzyme_enhancer[which(gene_enzyme_enhancer$promoter_class_merged=="bivalent"),]
# test$profile_class <- cut2(test$profile_directionality_chd4, m=20)
# test.melt <- summaryBy(profile_directionality_chd4 + profile_directionality_brg1 + profile_directionality_ep400 + profile_directionality_ezh2 + ezh2 + ep400 + expr_esc ~ profile_class, data = test, FUN = list(median))
smoothScatter(test.melt$profile_directionality_chd4.median, log(test.melt$ezh2.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Profile directionality (Chd4)", ylab = "Ezh2 (log)")
lines(lowess(test.melt$profile_directionality_chd4.median, log(test.melt$ezh2.median)), col='red', lty="dashed")
cor(test.melt$profile_directionality_chd4.median, log(test.melt$ezh2.median), use = "complete.obs")

## directionality bias (chd4) and EP400
# test <- gene_enzyme_enhancer[which(gene_enzyme_enhancer$promoter_class_merged!="bivalent"),]
# test$profile_class <- cut2(test$profile_directionality_chd4, m=20)
# test.melt <- summaryBy(profile_directionality_chd4 + profile_directionality_brg1 + profile_directionality_ep400 + profile_directionality_ezh2 + ezh2 + ep400 + expr_esc ~ profile_class, data = test, FUN = list(median))
smoothScatter(test.melt$profile_directionality_chd4.median, log(test.melt$ep400.median), colramp=colorRampPalette(my.cols), pch=19, cex=.8, useRaster=T, xlab = "Profile directionality (Chd4)", ylab = "Ep400 (log)")
cor(test.melt$profile_directionality_chd4.median, log(test.melt$ep400.median), use="complete.obs")
lines(lowess(test.melt$profile_directionality_chd4.median, log(test.melt$ep400.median)), col='red', lty="dashed")
dev.off()

# ggplot(gene_enzyme_enhancer, aes(profile_directionality_chd4, col=promoter_class_merged)) + geom_density(aes(group=promoter_class_merged)) +
#   xlim(c(-1,1))


##----------------------------------------------------##
## check influence of closest TSS
##----------------------------------------------------##

test <- gene_enzyme_divergent[which(gene_enzyme_divergent$cluster=="Y2"),]
test$expr_class <- "low"
test[which(test$expr_esc > summary(test$expr_esc)[2] & test$expr_esc < summary(test$expr_esc)[5]),]$expr_class <- "medium"
test[which(test$expr_esc > summary(test$expr_esc)[5]),]$expr_class <- "high"

## gene expression and check influence of closest TSS
pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/directionality_dist_to_tss.pdf")
test <- test[order(test$profile_directionality_chd4),]
test$directionality_cluster <- cut2(test$profile_directionality_chd4, m=2)
test.melt <- melt(summaryBy(expr_esc ~ directionality_cluster + directionality_class_merged, data = test, FUN = list(median)))
ggplot(test.melt, aes(directionality_cluster, variable, fill=log(value+1))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=0.55)(256))) +
  facet_grid(.~directionality_class_merged) +
  theme(axis.text.x = element_blank())

test.sort <- test[order(test$directionality_class_merged, test$profile_directionality_chd4, test$profile_directionality_brg1),]
test.sort$dist_closest_tss_class <- ifelse(test.sort$dist_closest_tss < 1000, 0, 1)
par(mfrow=c(4,2))
barplot(test.sort$profile_directionality_chd4, border=NA)
barplot(log(test.sort$expr_esc+1), border=NA)
image(as.matrix(test.sort[which(test.sort$directionality_class_merged=="active_-1"),]$dist_closest_tss_class), col=c("black", "yellow"))
image(as.matrix(log(test.sort[which(test.sort$directionality_class_merged=="active_-1"),]$dist_closest_tss)), col=(colorRampPalette(brewer.pal(9, "Blues"), bias=1)(256)))
image(as.matrix(test.sort[which(test.sort$directionality_class_merged=="active_0"),]$dist_closest_tss_class), col=c("black", "yellow"))
image(as.matrix(log(test.sort[which(test.sort$directionality_class_merged=="active_0"),]$dist_closest_tss)), col=(colorRampPalette(brewer.pal(9, "Blues"), bias=1)(256)))
image(as.matrix(test.sort[which(test.sort$directionality_class_merged=="active_1"),]$dist_closest_tss_class), col=c("black", "yellow"))
image(as.matrix(log(test.sort[which(test.sort$directionality_class_merged=="active_1"),]$dist_closest_tss)), col=(colorRampPalette(brewer.pal(9, "Blues"), bias=1)(256)))
dev.off()

##----------------------------------------------------##
## directionality bias relationship with divergent transcription upon exosome kd
##----------------------------------------------------##

## no control - all promoters
# test.sub <- test

## control for similar expression across the three chd4 classes (-1, 0 and 1)
test.sub <- test[which(test$equalExpr=="ExprNorm"),]

## CAGE and PROseq data
df.melt <- melt(test.sub[,c(171,199:204)])
p1 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~directionality_class_merged) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
df.melt <- melt(test.sub[,c(171,205:210)])
p2 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~directionality_class_merged) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
g <- arrangeGrob(p1, p2, nrow = 1, ncol=2)
ggsave(g, dpi=300, height=5, width=10, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/directionality_exosome_kd.pdf", useDingbats=FALSE)

##----------------------------------------------------##
## line plots showing effect on divergent transcription upon changes in Chd4 localization
##----------------------------------------------------##

# test.sub <- test.sub[order(test.sub$profile_directionality_chd4),]
# test.sub$directionality_class <- cut2(test.sub$profile_directionality_chd4, m=200)
# test.sub.melt <- melt(test.sub[,c(170,197:200)])
# test.sub.melt <- summarySE(test.sub.melt, measurevar = "value", groupvars = c("directionality_class", "variable"))
# ggplot(test.sub.melt, aes(directionality_class, value, color=variable)) +
#   geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
#   geom_line(aes(group=factor(variable), color=factor(variable)), alpha=1) +
#   geom_point(shape = 21, size = 2, fill="white")
#   #scale_color_manual(values=c("#a50026", "#313695")) +
#   #theme(legend.position="none")

##----------------------------------------------------##
## directionality bias relationship with divergent transcription upon mbd3 kd
##----------------------------------------------------##

test <- gene_enzyme_divergent[which(gene_enzyme_divergent$cluster=="Y2"),]
test.melt <- melt(test[,c(171,196)])
ggplot(test.melt, aes(directionality_class_merged, value)) + geom_boxplot() +
  stat_summary(fun.y=median, colour="red", geom="text", 
               vjust=-0.7, aes( label=round(..y.., digits=6)))
test.melt <- melt(test[,c(171,211:216)])
test.melt$condition <- unlist(lapply(test.melt$variable, function(x) unlist(strsplit(as.character(x), "_"))[2]))
test.melt$replicate <- unlist(lapply(test.melt$variable, function(x) unlist(strsplit(as.character(x), "_"))[3]))
ggplot(test.melt, aes(condition, value)) + geom_boxplot() + facet_grid(.~replicate+directionality_class_merged)


df <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/promoters/esc/allGenes_divergentTranscription.bed")
colnames(df) <- c("chr", "start", "end", "name", "score", "strand", "gene", "tssr", "tssr_signal", "dist_to_tss", 
                  "groseq_esc_rep1","groseq_esc_rep2",
                  "cage_esc_wt_rep1","cage_esc_wt_rep2","cage_esc_rrp40_kd_rep1","cage_esc_rrp40_kd_rep2",
                  "cage_esc_rbm7_kd_rep1","cage_esc_rbm7_kd_rep2","proseq_esc_wt_rep1","proseq_esc_wt_rep2",
                  "proseq_esc_rrp40_kd_rep1","proseq_esc_rrp40_kd_rep2","proseq_esc_rbm7_kd_rep1","proseq_esc_rbm7_kd_rep2",
                  "nasseq_esc_mbd3_wt_rep1","nasseq_esc_mbd3_wt_rep2","nasseq_esc_mbd3_wt_rep3","nasseq_esc_mbd3_kd_rep1",
                  "nasseq_esc_mbd3_kd_rep2","nasseq_esc_mbd3_kd_rep3")

mat <- t(apply(df[,c(25:30)], 1, function(x) unlist(lapply(as.character(x), function(y) as.numeric(unlist(strsplit(y,"/"))[1])))))
res.reverse <- matrix2deseq2(mat, c("wt,wt,wt,ko,ko,ko"))
df$nasseq_esc_fc_reverse <- res.reverse$log2FoldChange

mat <- t(apply(df[,c(25:30)], 1, function(x) unlist(lapply(as.character(x), function(y) as.numeric(unlist(strsplit(y,"/"))[2])))))
res.forward <- matrix2deseq2(mat, c("wt,wt,wt,ko,ko,ko"))
df$nasseq_esc_fc_forward <- res.forward$log2FoldChange

mat <- t(apply(df[,c(25:30)], 1, function(x) unlist(lapply(as.character(x), function(y) as.numeric(unlist(strsplit(y,"/"))[3])))))
res.gb <- matrix2deseq2(mat, c("wt,wt,wt,ko,ko,ko"))
df$nasseq_esc_fc_gb <- res.gb$log2FoldChange

test <- gene_enzyme_divergent[which(gene_enzyme_divergent$equalExpr=="ExprNorm"),]
test <- merge(test, df[,c(4,25:33)], by.x="gene", by.y="name")
test <- test[,c(2:4,1,5:ncol(test))]

test.melt <- melt((cbind(test$directionality_class, as.data.frame(t(apply(test[,c(241:243)], 1, function(x) unlist(lapply(as.character(x), function(y) as.numeric(unlist(strsplit(y,"/"))[1])))))))))
colnames(test.melt) <- c("directionality_class", "variable", "value")
ggplot(test.melt, aes(directionality_class, log(value))) + geom_boxplot() + facet_grid(.~variable)

test.melt <- melt(test[,c(170,247:249)])
ggplot(test.melt, aes(directionality_class, value)) + geom_boxplot() + facet_grid(.~variable)

test.melt <- melt(test[,c(170,10,56,61,64,65,60,153:155,176,66,68:70)])
ggplot(test.melt, aes(directionality_class, log(value))) + geom_boxplot() + facet_grid(.~variable)

test.melt <- melt(test[,c(170,183,186)])
ggplot(test.melt, aes(variable, value)) + geom_boxplot() + facet_grid(.~directionality_class)

test.melt <- melt(test[,c(170,64,65)])
# test.melt <- melt(test[,c(170,87)])
# test.melt <- melt(gene_enzyme_divergent[,c(170,87)])
# test.melt <- melt(test[,c(170,16)])
ggplot(test.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~directionality_class) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

##################################################
#### CHD4/BRG1 DIRECTIONALITY RELATIONSHIP WITH GENE EXPRESSION CHANGES (MEF/ESC)
##################################################
```{r}
##----------------------------------------------------##
## relationship between directionality bias and gene expression changes (mef/esc)
##----------------------------------------------------##

# pdf/directionality_esc_mef.pdf
df <- esc_mef[which(esc_mef$gene_specificity!="none"),]
# pdf/directionality_esc_mef1.pdf
# df <- esc_mef[which(esc_mef$promoter_class!="none"),]
# df$directionality_class <- "0_0"
# df[which(df$groseq_class_esc=="bidirectional" & df$groseq_class_mef=="unidirectional"),]$directionality_class <- "-1_1"
# df[which(df$groseq_class_esc=="unidirectional" & df$groseq_class_mef=="bidirectional"),]$directionality_class <- "1_-1"
# df$directionality_class <- factor(df$directionality_class, levels=c("-1_1", "0_0", "1_-1"), ordered=T)
table(df$directionality_class)
table(df[,c(16,24)])
table(df[,c(16,35)])

#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@#
## sanity check
# df.melt <- melt(df[,c(10,12,16)])
# df.melt <- melt(df[,c(23,34,16)])
# ggplot(df.melt, aes(variable, value)) + geom_boxplot() + facet_grid(.~directionality_class)
# par(mfrow=c(1,2))
# barplot(table(df[which(df$groseq_wt_down_rep2_esc > summary(esc_mef$groseq_wt_down_rep2_esc)[3]),]$directionality_class)/table(df$directionality_class))
# barplot(table(df[which(df$groseq_wt_down_rep3_mef > summary(esc_mef$groseq_wt_down_rep3_mef)[3]),]$directionality_class)/table(df$directionality_class))
# 
# df <- df[order(df$diff),]
# barplot(ifelse(df$gene_specificity=="none", 0, 1), border=NA)
# barplot(ifelse(df$groseq_class_esc=="bidirectional", 0, 1), border=NA)
# barplot(df$directionality_esc)
# barplot(df$directionality_mef)
# barplot(df$fc)
# barplot(df$groseq_wt_down_rep2_esc)
# barplot(df$groseq_wt_up_rep2_esc)
# barplot(df$groseq_wt_down_rep3_mef)
# barplot(df$groseq_wt_up_rep3_mef)
# table(esc_mef[which(esc_mef$directionality_esc_class=="-1"),]$groseq_class_esc)
# table(esc_mef[which(esc_mef$directionality_mef_class=="1"),]$groseq_class_mef)
# table(esc_mef[which(esc_mef$directionality_esc_class=="1"),]$groseq_class_esc)
# table(esc_mef[which(esc_mef$directionality_mef_class=="-1"),]$groseq_class_mef)

#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@#

df.melt <- melt(df[,c(18:19,16)])
df.melt$variable <- factor(df.melt$variable, levels=c(unique(grep("up", df.melt$variable, value = T)), unique(grep("down", df.melt$variable, value = T))), ordered = T)
p1 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~directionality_class) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
df.melt <- melt(df[,c(20:21,16)])
df.melt$variable <- factor(df.melt$variable, levels=c(unique(grep("up", df.melt$variable, value = T)), unique(grep("down", df.melt$variable, value = T))), ordered = T)
p2 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~directionality_class) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

df.melt <- melt(df[,c(26:27,16)])
df.melt$variable <- factor(df.melt$variable, levels=c(unique(grep("up", df.melt$variable, value = T)), unique(grep("down", df.melt$variable, value = T))), ordered = T)
p3 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~ directionality_class) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
#df.melt <- melt(df[,c(28:29,16)])
df.melt <- melt(df[,c(30:31,16)])
df.melt$variable <- factor(df.melt$variable, levels=c(unique(grep("up", df.melt$variable, value = T)), unique(grep("down", df.melt$variable, value = T))), ordered = T)
p4 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~ directionality_class) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

df.melt <- melt(df[,c(7,8,16)])
p5 <- ggplot(df.melt, aes(variable, log(value))) + geom_boxplot() + facet_grid(.~directionality_class)
# ggplot(df.melt, aes(log(value), color=variable)) + stat_ecdf() + facet_grid(.~directionality_class)
wilcox.test(df[which(df$directionality_class=="-1_1"),]$expr_mef, df[which(df$directionality_class=="-1_1"),]$expr_esc, alternative = "l")$p.value
wilcox.test(df[which(df$directionality_class=="1_-1"),]$expr_mef, df[which(df$directionality_class=="1_-1"),]$expr_esc, alternative = "g")$p.value
wilcox.test(df[which(df$directionality_class=="0_0"),]$expr_mef, df[which(df$directionality_class=="0_0"),]$expr_esc)$p.value

df$diff_class <- cut2(df$diff, m=30)
test.melt <- melt(summaryBy(expr_mef + expr_esc ~ diff_class, data = df, FUN = list(median)))
p6 <- ggplot(test.melt, aes(diff_class, variable, fill=log(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.1)(256))) +
  theme(axis.text.x = element_blank())

test.melt <- melt(summaryBy(directionality_esc + directionality_mef ~ diff_class, data = df, FUN = list(median)))
p7 <- ggplot(test.melt, aes(diff_class, variable, fill=(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "PRGn"), bias=1.1)(256))) +
  theme(axis.text.x = element_blank())


test.melt <- melt(summaryBy(groseq_directionality_rep2_esc + groseq_directionality_rep3_mef ~ diff_class, data = df, FUN = list(mean), na.rm = T))
p8 <- ggplot(test.melt, aes(diff_class, variable, fill=(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "PRGn"), bias=1.1)(256))) +
  theme(axis.text.x = element_blank())

df.melt <- melt(df[,c(23,34,16)])
p9 <- ggplot(df.melt, aes(variable, value)) + geom_boxplot() + facet_grid(.~directionality_class)

test.melt <- melt(summaryBy(fc ~ diff_class, data = df, FUN = list(median)))
p10 <- ggplot(test.melt, aes(diff_class, variable, fill=(value))) + geom_tile() + 
  scale_fill_gradientn(colours = rev(colorRampPalette(brewer.pal(11, "RdBu"), bias=1.1)(256))) +
  theme(axis.text.x = element_blank())
g <- arrangeGrob(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, nrow = 5, ncol=2)
ggsave(g, dpi=300, height=20, width=10, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/directionality_esc_mef.pdf", useDingbats=FALSE)

#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@#
## sanity check; what happes to genes from -1_1, 0_0 and 1_-1 classes upon Chd4 KD

df <- merge(gene_expr, esc_mef[which(esc_mef$gene_specificity!="none"),c(4,7,8,16)], by.x="gene", by.y="gene")
ggplot(df, aes(directionality_class, chd4_fc)) + geom_boxplot()
wilcox.test(df[which(df$directionality_class=="-1_1"),]$chd4_fc, df[which(df$directionality_class=="1_-1"),]$chd4_fc)$p.value
```

##################################################
#### EP400, CHD4 AND PCG PROTEIN COMPLEX ANALYSIS
##################################################
```{r}
## paf1 complex genes (source: https://www.nature.com/articles/ncb3424)
paf1_complex <- c("Ctr9", "Rtf1", "Leo1", "Paf1", "Cdc73", "Wdr61")
test <- gene_enzyme[gene_enzyme$gene %in% c(paf1_complex),]
test$gene <- factor(test$gene, levels=test$gene, ordered = T)
test.melt <- melt(test[,c(4,75,87)])
p1 <- ggplot(test.melt, aes(gene, value, fill=variable)) + geom_bar(stat="identity", position = "dodge") + theme(axis.text.x = element_text(size=10, angle=45)) + scale_fill_manual(values=c("#9970ab", "#5aae61")) + xlab("EP400 Complex") + ylab("log FC (expr.)")
# p-values 
# test[,c(4,72:75,84:87)]

## ep400 complex genes (source: http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006668)
ep400_complex <- c("Trrap", "Ep400", "Kat5", "Epc1", "Epc2", "Ing3", "Dmap1", "Ruvbl1", "Ruvbl2", "Actl6a", "Actl6b",
                   "Morf4l1", "Morf4l2", "Meaf6", "Mrgbp", "Yeats4", "Brd8", "Vps72", "Mbtd1")
mad_complex <- c("Mxd1", "Mxd3", "Mxd4", "Mnt", "Sin3a", "Mga", "E2f6", "Tfdp1", "Pcgf6", "Ring1", "Rnf2", "Cbx3", 
                 "L3mbtl2", "Rybp", "Yaf2")
max_complex <- c("Max", "Mycl", "Mycn")
st_complex <- c("Ppp2r1a", "Ppp2r1b", "Ppp2ca", "Ppp2cb")
# lt_complex <- c("Rb1", "Vps39")

test <- gene_enzyme[gene_enzyme$gene %in% c(ep400_complex, max_complex, st_complex),]
test$gene <- factor(test$gene, levels=test$gene, ordered = T)
test.melt <- melt(test[,c(4,87,103)])
p2 <- ggplot(test.melt, aes(gene, value, fill=variable)) + geom_bar(stat="identity", position = "dodge") + theme(axis.text.x = element_text(size=10, angle=45)) + scale_fill_manual(values=c("#9970ab", "#5aae61")) + xlab("EP400 Complex") + ylab("log FC (expr.)")

#nurd complex (source: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4157524/)
#Rbbp4/7 alias RBAP48/46
nurd_complex <- c("Hdac1", "Hdac2", "Chd3", "Chd4", "Rbbp4", "Rbbp7", "Mta1", "Mta2", "Mta3", "Mbd2", "Mbd3", 
                  "Gatad2a", "Gatad2b")

test <- gene_enzyme[gene_enzyme$gene %in% c(nurd_complex),]
test$gene <- factor(test$gene, levels=test$gene, ordered = T)
test.melt <- melt(test[,c(4,87,103)])
p3 <- ggplot(test.melt, aes(gene, value, fill=variable)) + geom_bar(stat="identity", position = "dodge") + theme(axis.text.x = element_text(size=10, angle=45)) + scale_fill_manual(values=c("#9970ab", "#5aae61"))  + xlab("NuRD Complex") + ylab("log FC (expr.)")

## Prc2 complex (source: http://www.jbc.org/content/early/2017/09/14/jbc.R117.800367.full.pdf)
prc2_complex <- c("Suz12", "Ezh1", "Ezh2", "Eed", "Rbbp4", "Rbbp7", "Aebp2", "Jarid2", "Rybp")
## Prc1 complex (source: http://www.jbc.org/content/291/1/182.full)
## PcG complex (source: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5310723/)
prc1_complex <- c("Ring1", "Bmi1", "Pcgf1", "Pcgf2", "Pcgf3", "Pcgf4", "Pcgf5", "Pcgf6")

test <- gene_enzyme[gene_enzyme$gene %in% c(prc2_complex, prc1_complex),]
test$gene <- factor(test$gene, levels=test$gene, ordered = T)
test.melt <- melt(test[,c(4,87,103)])
p4 <- ggplot(test.melt, aes(gene, value, fill=variable)) + geom_bar(stat="identity", position = "dodge") + theme(axis.text.x = element_text(size=10, angle=45)) + scale_fill_manual(values=c("#9970ab", "#5aae61")) + xlab("PcG Complex") + ylab("log FC (expr.)")

## Swi/Snf complex
swi_snf_complex <- c("Ss18", "Baf57", "Snf5", "Brd7", "Brd9", "Pbrm1", "Arid1a", "Arid1b", "Arid2", "Baf155", 
                     "Baf45a", "Baf45d", "Baf60a", "Baf60b", "Baf53a", "Smarca4", "Actb")

test <- gene_enzyme[gene_enzyme$gene %in% c(swi_snf_complex),]
test$gene <- factor(test$gene, levels=test$gene, ordered = T)
test.melt <- melt(test[,c(4,87,103)])
p5 <- ggplot(test.melt, aes(gene, value, fill=variable)) + geom_bar(stat="identity", position = "dodge") + theme(axis.text.x = element_text(size=10, angle=45)) + scale_fill_manual(values=c("#9970ab", "#5aae61")) + xlab("SWI/SNF Complex") + ylab("log FC (expr.)")

g <- arrangeGrob(p1, p2, p3, p4, p5, nrow = 3, ncol=2)
ggsave(g, dpi=300, height=10, width=10, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/protein_complex_analysis.pdf", useDingbats=FALSE)


test <- gene_enzyme[gene_enzyme$gene %in% c("Myc", "Mycn"),]
test$gene <- factor(test$gene, levels=test$gene, ordered = T)
test.melt <- melt(test[,c(4,75,87,103)])
ggplot(test.melt, aes(gene, value, fill=variable)) + geom_bar(stat="identity", position = "dodge") + theme(axis.text.x = element_text(size=10, angle=45)) 

scale_fill_manual(values=c("#9970ab", "#5aae61")) + xlab("SWI/SNF Complex") + ylab("log FC (expr.)")
```

##################################################
#### GVIZ PLOTS
##################################################
```{r}
browser_files <- read.table(pipe("grep -v '#' /home/pundhir/project/chip-seq-analysis/analysis_me1_me3/FILES_FOR_GENOME_BROWSER"))

## find interesting example genes for the four gene classes (bivalent, low, average and high)
cbind(unique(names(table(gene_enzyme_enhancer$ep400_cluster_merged))),
      unlist(lapply(unique(names(table(gene_enzyme_enhancer$ep400_cluster_merged))), function(x) median(gene_enzyme_enhancer[which(gene_enzyme_enhancer$ep400_cluster_merged==x),]$chd4))),
      unlist(lapply(unique(names(table(gene_enzyme_enhancer$ep400_cluster_merged))), function(x) median(gene_enzyme_enhancer[which(gene_enzyme_enhancer$ep400_cluster_merged==x),]$brg1))),
      unlist(lapply(unique(names(table(gene_enzyme_enhancer$ep400_cluster_merged))), function(x) median(gene_enzyme_enhancer[which(gene_enzyme_enhancer$ep400_cluster_merged==x),]$ep400))))

test <- gene_enzyme_enhancer[which(round(gene_enzyme_enhancer$chd4,0)==1 & round(gene_enzyme_enhancer$brg1,0)==1 & 
                                     (gene_enzyme_enhancer$chd4_qVal<0.05 | gene_enzyme_enhancer$brg1_qVal<0.05) &
                                     ((gene_enzyme_enhancer$chd4_fc > 0 & gene_enzyme_enhancer$brg1_fc < 0)  | 
                                        (gene_enzyme_enhancer$chd4_fc < 0 & gene_enzyme_enhancer$brg1_fc > 0))),
                             c("chr", "start", "end", "gene", "promoter_class_merged", "ep400_cluster_merged", 
                               "chd4_fc", "brg1_fc", "ep400_fc", "gene_coor", 
                               "expr_mef", "expr_mef48", "expr_ipsc", "expr_esc", 
                               "chd4_pVal", "chd4_qVal", "brg1_pVal", "brg1_qVal")]

# set.seed(101)
# ddply(test, .(ep400_cluster_merged), function(x) x[sample(nrow(x),1),])
test <- test[order(test$ep400_cluster_merged),]
dim(test)
table(test$ep400_cluster_merged)
barplot(table(test$promoter_class_merged)/table(gene_enzyme_enhancer$promoter_class_merged))

write.table(test, "/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/goAna/genes_antagonistic_expr.txt", sep="\t", quote=F, row.names = F, col.names = F)

# pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gviz.pdf")
# #COOR <- apply(test, 1, function(x) sprintf("%s:%s-%s", x[1], x[2], x[3]))[116]
# lapply(test$gene_coor, function(coor) {
#   plot_gviz_track(GEN = "mm9", COOR = coor,
#                 BAMFILE = browser_files[c(1:3,23,4:7),]$V2,
#                 type="a", cex=0.5, FLANK_WIN = 10000)
# })
# dev.off()

table(test$ep400_cluster_merged)
# bivalent (page 1-40): Parm1 (25)*, Slit2 (29)
# low (page 260-306): Rhobtb3 (294), Cd47 (267), Egln1 (270), Fam107b (273)*
# average (page 166-259): 169, 184, 194*, 199, 206, 212, 219, 229, 232
# high (page 41-165): 93*, 116, 124, 134,

#Rfc1|Rfc3|Ccnh|Dna2|Orc4|Rnaseh2a|^Atm$|Pold3|Atr
gene_enzyme_enhancer[c(grep("Wnt3a|Tcf3|^Trp53$", gene_enzyme_enhancer$gene)),c("gene", "chd4_fc", "chd4_qVal", "brg1_fc", "brg1_qVal", "ep400_cluster_merged", "expr_mef", "expr_esc")]

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gviz_selected.pdf")
# Ets1/Cpe/Wnt4/Ppp3cc (bivalent)
plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("Ets1", test$gene)],
                BAMFILE = c(as.character(browser_files[c(1:3,23,4:7,13:16,9:11),]$V2),
                            "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                            "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw"
                            ),
                type="a", cex=0.5, FLANK_WIN = 10000, maxlimit = "12,10,60,35,200,25,25,15,40,40,10,10,15,15,15")

plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("^Ets1$", test$gene)],
              BAMFILE = c(
                #as.character(browser_files[c(1:2),]$V2), 
                "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_h3k27me3_log2.bw",
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_h3k27me3_subtract.bw",
                "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_h3k27me3_log2.bw",
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_h3k27me3_subtract.bw"
                ),
              type="a", cex=0.5, FLANK_WIN = 10000)

# # Parm1 (bivalent)
# plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("Parm1", test$gene)],
#                 BAMFILE = browser_files[c(1:3,23,4:7,13:16,9:11),]$V2,
#                 type="a", cex=0.5, FLANK_WIN = 10000, maxlimit = "12,10,60,35,200,25,25,15,40,40,10,10,15,15,15")

# Fam107b/Zfhx3 (low)
plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("Fam107b", test$gene)],
                BAMFILE = c(as.character(browser_files[c(1:3,23,4:7,13:16,9:11),]$V2),
                            "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                            "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw"
                            ),
                type="a", cex=0.5, FLANK_WIN = 10000, maxlimit = "12,10,60,35,200,25,25,15,40,40,10,10,15,15,15")

plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("^Fam107b$", test$gene)],
              BAMFILE = c(
                #as.character(browser_files[c(1:2),]$V2), 
                "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_h3k27me3_log2.bw",
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_h3k27me3_subtract.bw",
                "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_h3k27me3_log2.bw",
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_h3k27me3_subtract.bw"
                ),
              type="a", cex=0.5, FLANK_WIN = 10000)

# Ercc6l (average)
plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("Ercc6l", test$gene)],
              BAMFILE = c(as.character(browser_files[c(1:3,23,4:7,27:30,9:11),]$V2),
                          "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                          "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw"
                          ),
              type="a", cex=0.5, FLANK_WIN = 5000, maxlimit = "12,10,60,35,200,25,25,15,12,12,10,10,15,15,15")
# gene_enzyme_enhancer[grep("^Ercc6l$", gene_enzyme_enhancer$gene),c(177:189)]

plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("^Ercc6l$", test$gene)],
              BAMFILE = c(
                #as.character(browser_files[c(1:2),]$V2),
                "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_log2.bw",
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_subtract.bw",
                "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_log2.bw"
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_subtract.bw"
                ),
              type="a", cex=0.5, FLANK_WIN = 500)

# # Mettl13 (high)
# plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("Mettl13", test$gene)],
#               BAMFILE = browser_files[c(1:3,23,4:7,27:30,9:11),]$V2,
#               type="a", cex=0.5, FLANK_WIN = 10000, maxlimit = "12,10,60,35,200,25,25,15,40,40,10,10,15,15,15")

# Cdk1 (high)
plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("^Cdk1$", test$gene)],
              BAMFILE = c(as.character(browser_files[c(1:3,23,4:7,27:30,9:11),]$V2),
                          "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                          "/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw"
                          ),
              type="a", cex=0.5, FLANK_WIN = 5000, maxlimit = "15,15,60,120,300,100,25,15,30,30,30,30,24,24,24")

plot_gviz_track(GEN = "mm9", COOR = test$gene_coor[grep("^Cdk1$", test$gene)],
              BAMFILE = c(
                #as.character(browser_files[c(1:2),]$V2),
                #"/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/chd4_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_log2.bw",
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/test_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/atac-seq/mouse/plath/KD/brg1_atac_subtract.bw",
                #"/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_log2.bw"
                "/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/test_subtract.bw"
                ),
              type="h", cex=0.5, FLANK_WIN = 500)

df.melt <- melt(test[c(grep("Ets1|Fam107b|Ercc6l|^Cdk1$", test$gene)),c(4,7,8)])
df.melt$gene <- factor(df.melt$gene, levels=c("Ets1", "Fam107b", "Cdk1", "Ercc6l"), ordered = T)
df.melt$variable <- factor(df.melt$variable, levels=rev(c("chd4_fc", "brg1_fc")), ordered = T)
ggplot(df.melt, aes(variable, value)) + geom_bar(stat="identity") + 
  facet_wrap(.~gene, scales="free") + theme_bw() + coord_flip()

plot_gviz_track(GEN = "mm9", COOR = gene_enzyme_enhancer[grep("Chd4", gene_enzyme_enhancer$gene),c("gene_coor")], #chr6:125044280-125049651
                BAMFILE = browser_files[c(1:3,23,4:7),]$V2,
                type="a", cex=0.5, FLANK_WIN = 100, maxlimit = "12,10,35,35,200,25,25,15")


pol2_chd4_ko <- CoverageBamFile("/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/pol2S5P_esc_mbd3_0h_Rep3.bam")
pol2_chd4_wt <- CoverageBamFile("/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/mbd3_kd/pol2S5P_esc_mbd3_24h_Rep3.bam")
pol2_brg1_ko <- CoverageBamFile("/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/pol2_brg1_ko_Rep1.bam")
pol2_brg1_wt <- CoverageBamFile("/home/pundhir/project/chip-seq-analysis/data/tf_marks/mouse/chromatin_remodellers/brg1_ko/pol2_brg1_wt_Rep1.bam")

ercc6l <- cbind(cov.interval(pol2_chd4_ko, pol2_chd4_wt, bin_width=10, chr="chr10", start=68798882, end=68816186, do='substraction'),
                cov.interval(pol2_brg1_ko, pol2_brg1_wt, bin_width=10, chr="chr10", start=68798882, end=68816186, do='substraction'))
ercc6l <- as.data.frame(cbind(seq(1, nrow(ercc6l)), ercc6l))
colnames(ercc6l) <- c("pos", "ratio_chd4", "ratio_brg1")
ercc6l$pos <- as.factor(ercc6l$pos)

#spline.d <- as.data.frame(spline(ercc6l$pos, ercc6l$ratio))
p1 <- ggplot(melt(ercc6l), aes(pos, value)) + 
  #geom_point() + 
  stat_smooth(aes(x = as.numeric(pos), y = value), method = "lm", formula = y ~ poly(x, 25), se = FALSE) +
  #geom_line(data = spline.d, aes(x = x, y = y)) +
  #geom_smooth(method = "loess") +
  facet_wrap(.~variable, scales = "free") + theme_bw()

cdk1 <- cbind(cov.interval(pol2_chd4_ko, pol2_chd4_wt,bin_width=10,chr="chrX", start=99336554,end=99352930,do='substraction'),
              cov.interval(pol2_brg1_ko, pol2_brg1_wt,bin_width=10,chr="chrX", start=99336554,end=99352930,do='substraction'))
cdk1 <- as.data.frame(cbind(seq(1, nrow(cdk1)), cdk1))
colnames(cdk1) <- c("pos", "ratio_chd4", "ratio_brg1")
cdk1$pos <- as.factor(cdk1$pos)

#spline.d <- as.data.frame(spline(cdk1$pos, cdk1$ratio))
p2 <-ggplot(melt(cdk1), aes(pos, value)) + 
  #geom_point() + 
  stat_smooth(aes(x = as.numeric(pos), y = value), method = "lm", formula = y ~ poly(x, 25), se = FALSE) +
  #geom_line(data = spline.d, aes(x = x, y = y)) +
  #geom_smooth(method = "loess") +
  facet_wrap(.~variable, scales = "free") + theme_bw()
ggarrange(p1, p2, nrow=2, ncol=1)
dev.off()

##---------------------------------------------------------------##
## pathway (WNT and Cellular senesence)
library("pathview")
pluripotency <- c("Fgfr3", "Fzd3", "Fzd8", "Pik3cb", "Pik3r1", "Wnt4", "Wnt6", "Wnt7b", "Zfhx3")
wnt <- c("Fzd3", "Fzd8", "Nfatc1", "Nkd1", "Nkd2", "Ppp3cc", "Prickle2", "Ror2", "Rspo3", "Wnt4", "Wnt6", "Wnt7b") #mmu04310
cell_senesence <- c("Capn2", "Ets1", "Hipk2", "Mapkapk2", "Nfatc1", "Ppp3cc", "Tgfb2", "Tgfbr2") #mmu04218
cell_cycle <- c("Anapc1", "Atm", "Atr", "Bub3", "Ccnb1", "Ccnb2", "Ccne2", "Ccnh", "Cdc25a", "Cdc27", "Cdk1", "Chek1", "Cul1", "Mcm2", "Mcm3", "Mcm5", "Mcm6", "Mcm7", "Orc1", "Orc2", "Orc3", "Orc4", "Orc5", "Orc6", "Skp1a", "Smad2", "Smc3", "Ercc6l")
dna_replication <- c("Dna2", "Fen1", "Mcm2", "Mcm3", "Mcm5", "Mcm6", "Mcm7", "Pold3", "Rfc1", "Rfc3", "Rfc4", "Rnaseh1", "Rnaseh2a", "Ssbp1")
nanog_oct3_tcf3 <- read.table("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/analysis/martello_2013/nanog_oct3_tcf3_target.txt")$V1

pathview(gene.data  = cell_senesence,
                     pathway.id = "mmu04218",
                     species    = "mmu", same.layer = F, gene.idtype = gene.idtype.list[1], out.suffix="LPM", kegg.native = F)

##---------------------------------------------------------------##
## heatmap with gene names
mat <- gene_enzyme_enhancer
mat$promoter_class_merged <- factor(mat$promoter_class_merged, levels=c("bivalent", "narrow", "medium", "broad"), ordered = T)
mat <- mat[order(mat$promoter_class_merged, -mat$pol2_wt),]
mat$ep400_cluster <- as.numeric(cut2(mat$pol2_wt, m=25))
barplot(mat$ep400_cluster)

mat$funcClass <- "none"
mat[which(mat$gene %in% pluripotency),]$funcClass <- "pluripotency"
mat[which(mat$gene %in% wnt),]$funcClass <- "wnt"
mat[which(mat$gene %in% cell_senesence),]$funcClass <- "cell_senesence"
mat[which(mat$gene %in% cell_cycle),]$funcClass <- "cell_cycle"
mat[which(mat$gene %in% dna_replication),]$funcClass <- "dna_replication"
mat$promoter_class_merged <- factor(mat$promoter_class_merged, levels=c("bivalent", "narrow", "medium", "broad"), ordered = T)
table(mat$ep400_cluster)

labels_gp = gpar(fontsize = rep(1, length(mat[which(mat$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$gene)))
labels_at <- which(mat$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication))
labels_genes <- sprintf("%s_%s", mat[which(mat$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$gene,
                        mat[which(mat$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$funcClass)

# labels_gp = gpar(fontsize = rep(1, length(mat[which(mat$gene %in% test[which(test$gene %in% nanog_oct3_tcf3),]$gene),]$gene)))
# labels_at <- which(mat$gene %in% test[which(test$gene %in% nanog_oct3_tcf3),]$gene)
# labels_genes <- mat[which(mat$gene %in% test[which(test$gene %in% nanog_oct3_tcf3),]$gene),]$gene

la <- rowAnnotation(cluster_de = as.factor(mat$promoter_class_merged))

ra <- rowAnnotation(na_col = NA,
                        foo = anno_mark(at = labels_at, labels = labels_genes)
                    )
pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/heatmap_ep400_cluster.pdf", height=20)
print(Heatmap(as.matrix(t(scale(t(mat[,c("expr_esc", "expr_mef")])))), cluster_rows=F, cluster_columns=F,
              col=colorRampPalette(c("dodgerblue4","grey97", "sienna3"), bias=1.5)(250), 
              use_raster=T, show_row_names = F, raster_device="tiff", raster_quality = 10, left_annotation = la, right_annotation = ra))

lapply(c("bivalent", "medium", "broad"), function(x) {
    mat.sub <- mat[which(mat$promoter_class_merged==x),]
    mat.sub$ep400_cluster <- as.numeric(as.factor(-mat.sub$ep400_cluster))
    labels_gp = gpar(fontsize = rep(1, length(mat.sub[which(mat.sub$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$gene)))
    labels_at <- mat.sub[which(mat.sub$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$ep400_cluster
    labels_genes <- sprintf("%s_%s", mat.sub[which(mat.sub$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$gene,
                        mat.sub[which(mat.sub$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),]$funcClass)
  
    ra <- rowAnnotation(na_col = NA, foo = anno_mark(at = labels_at, labels = labels_genes))
    t <- mat.sub[match(unique(mat.sub[which(mat.sub$promoter_class_merged==x),]$ep400_cluster),mat.sub[which(mat.sub$promoter_class_merged==x),]$ep400_cluster),]
    
    print(Heatmap(as.matrix(t[,c("ep400_cluster")]), cluster_rows=F, cluster_columns=F,
              col=colorRampPalette(c("dodgerblue4","grey97", "sienna3"), bias=1.5)(250), 
              use_raster=T, show_row_names = F, raster_device="tiff", raster_quality = 10, right_annotation = ra))

})
dev.off()

## expression profile during reprogramming
df.melt <- melt(mat[which(mat$gene %in% c(pluripotency, wnt, cell_senesence, cell_cycle, dna_replication)),
    c("expr_mef", "expr_mef48", "expr_ipsc", "expr_esc", "funcClass")])
df.melt$variableNumeric <- as.numeric(df.melt$variable)
df.melt$value <- df.melt$value
df.melt <- summarySE(df.melt[!is.na(df.melt$value),], measurevar = "value", groupvars = c("funcClass", "variable"))

ggplot(df.melt, aes(variable, value, color=funcClass)) +  geom_errorbar(aes(ymin=value-se, ymax=value+se), width=.1, alpha=1) +
  geom_line(aes(group=factor(funcClass), color=factor(funcClass)), alpha=1) +
  geom_point(shape = 21, size = 2, fill="white") +
  #scale_color_manual(values=c("#313695", "#000000", "#a50026")) +
  theme_bw() +  theme(legend.position="top") +
  geom_smooth(method = "lm", formula = y ~ poly(x, 4), se = F) +
  facet_wrap(~funcClass, scales="free")


##---------------------------------------------------------------##
## other genes
## Rbbp4
CHR <- "chr4"
GEN <- "mm9"
START=128956073
END=129040885

itrack <- IdeogramTrack(genome = GEN, chromosome = CHR)
gtrack <- GenomeAxisTrack()
txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene
txTr <- GeneRegionTrack(txdb, chromosome = CHR, start = START, end = END)

dtrack_atac <- apply(as.data.frame(browser_files[grep("atac_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_ep400 <- apply(as.data.frame(browser_files[grep("ep400_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_chd4 <- apply(as.data.frame(browser_files[grep("chd4_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_brg1 <- apply(as.data.frame(browser_files[grep("brg1_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac <- apply(as.data.frame(browser_files[grep("h3k27ac_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27me3 <- apply(as.data.frame(browser_files[grep("h3k27me3_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_mdb3_wt <- apply(as.data.frame(browser_files[grep("h3k27ac_mbd3_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_mdb3_ko <- apply(as.data.frame(browser_files[grep("h3k27ac_mbd3_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_brg1_wt <- apply(as.data.frame(browser_files[grep("h3k27ac_brg1_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_brg1_ko <- apply(as.data.frame(browser_files[grep("h3k27ac_brg1_ko", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_wt <- apply(as.data.frame(browser_files[grep("pol2_esc_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_chd4_kd <- apply(as.data.frame(browser_files[grep("pol2_esc_chd4_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_brg1_kd <- apply(as.data.frame(browser_files[grep("pol2_esc_brg1_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_ep400_kd <- apply(as.data.frame(browser_files[grep("pol2_esc_ep400_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/rbbp4_track.pdf")
my_list <- c(itrack, txTr, dtrack_ep400, dtrack_h3k27ac, dtrack_h3k27me3, dtrack_h3k27ac_mdb3_wt, dtrack_h3k27ac_mdb3_ko, dtrack_h3k27ac_brg1_wt, dtrack_h3k27ac_brg1_ko, dtrack_pol2_esc_wt, dtrack_pol2_esc_chd4_kd, dtrack_pol2_esc_brg1_kd, dtrack_pol2_esc_ep400_kd)
plotTracks(my_list, from=START, to=END)
dev.off()

## Ezh2
CHR <- "chr6"
GEN <- "mm9"
START=47462486
END=47562816

itrack <- IdeogramTrack(genome = GEN, chromosome = CHR)
gtrack <- GenomeAxisTrack()
txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene
txTr <- GeneRegionTrack(txdb, chromosome = CHR, start = START, end = END)

dtrack_atac <- apply(as.data.frame(browser_files[grep("atac_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_ep400 <- apply(as.data.frame(browser_files[grep("ep400_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_chd4 <- apply(as.data.frame(browser_files[grep("chd4_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_brg1 <- apply(as.data.frame(browser_files[grep("brg1_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac <- apply(as.data.frame(browser_files[grep("h3k27ac_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27me3 <- apply(as.data.frame(browser_files[grep("h3k27me3_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_mdb3_wt <- apply(as.data.frame(browser_files[grep("h3k27ac_mbd3_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_mdb3_ko <- apply(as.data.frame(browser_files[grep("h3k27ac_mbd3_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_brg1_wt <- apply(as.data.frame(browser_files[grep("h3k27ac_brg1_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_h3k27ac_brg1_ko <- apply(as.data.frame(browser_files[grep("h3k27ac_brg1_ko", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_wt <- apply(as.data.frame(browser_files[grep("pol2_esc_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_chd4_kd <- apply(as.data.frame(browser_files[grep("pol2_esc_chd4_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_brg1_kd <- apply(as.data.frame(browser_files[grep("pol2_esc_brg1_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
dtrack_pol2_esc_ep400_kd <- apply(as.data.frame(browser_files[grep("pol2_esc_ep400_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/ezh2_track.pdf")
my_list <- c(itrack, gtrack, txTr, dtrack_ep400, dtrack_h3k27ac, dtrack_h3k27me3, dtrack_h3k27ac_mdb3_wt, dtrack_h3k27ac_mdb3_ko, dtrack_h3k27ac_brg1_wt, dtrack_h3k27ac_brg1_ko)
plotTracks(my_list, from=START, to=END)
dev.off()

#gene_enzyme[which(gene_enzyme$gene=="Rbbp4"),c(7:10,65,81,97,101,123,124,125)]
test <- gene_enzyme[which(gene_enzyme$gene=="Chd4" | gene_enzyme$gene=="Ep400" | gene_enzyme$gene=="Smarca4" | gene_enzyme$gene=="Rbbp4" | gene_enzyme$gene=="Suz12" | gene_enzyme$gene=="Ezh2"),c(4,7:10)]
test.melt <- melt(test)
test.melt$variable <- factor(test.melt$variable, levels=c("expr_esc", "expr_ipsc", "expr_mef48", "expr_mef"), ordered = T)
p <- ggplot(test.melt, aes(variable, log(value), group=gene, color=gene)) + geom_line() + geom_point() + xlab("") + ylab("Expression (log)")
ggsave(p, dpi=300, width=4, height=2, filename="/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gene_expr_differentiation.pdf", useDingbats=FALSE)

df_esc[which(df_esc$gene=="Gata6"),c(4,41:44)]

## -------------------------------------------------- ##
## gviz for the promoters
## -------------------------------------------------- ##

plot_genome_browser_example <- function(CHR, START, END) {
  CHR <- CHR  
  GEN <- "mm9"  
  START=as.numeric(START)-5000  
  END=as.numeric(END)+5000  
  
  txTr <- GeneRegionTrack(txdb, chromosome = CHR, start = START, end = END)
  itrack <- IdeogramTrack(genome = GEN, chromosome = CHR)  
  
  dtrack_atac <- apply(as.data.frame(browser_files[grep("atac", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a", 100))
  dtrack_ep400 <- apply(as.data.frame(browser_files[grep("ep400", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a", 20))
  dtrack_chd4 <- apply(as.data.frame(browser_files[grep("chd4", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a", 15))
  dtrack_brg1 <- apply(as.data.frame(browser_files[grep("brg1_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a", 15))
  dtrack_pol2 <- apply(as.data.frame(browser_files[grep("pol2", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a", 220))
  dtrack_ezh2 <- apply(as.data.frame(browser_files[grep("ezh2", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
  dtrack_h3k27ac <- apply(as.data.frame(browser_files[grep("h3k27ac_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a", 70))
  dtrack_h3k27me3 <- apply(as.data.frame(browser_files[grep("h3k27me3_esc", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
  dtrack_h3k27ac_mdb3_wt <- apply(as.data.frame(browser_files[grep("h3k27ac_mbd3_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
  dtrack_h3k27ac_mdb3_ko <- apply(as.data.frame(browser_files[grep("h3k27ac_mbd3_kd", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
  dtrack_h3k27ac_brg1_wt <- apply(as.data.frame(browser_files[grep("h3k27ac_brg1_wt", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
  dtrack_h3k27ac_brg1_ko <- apply(as.data.frame(browser_files[grep("h3k27ac_brg1_ko", browser_files$V1),]), 1, function(x) create_data_track(x, GEN, CHR, "a"))
  
  my_list <- c(itrack, gtrack, txTr, dtrack_atac, dtrack_ep400, dtrack_chd4, dtrack_brg1, dtrack_pol2, dtrack_ezh2, dtrack_h3k27ac, dtrack_h3k27me3)
  plotTracks(my_list, from=START, to=END)
}

gtrack <- GenomeAxisTrack()
txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene
# txTr <- GeneRegionTrack(txdb)

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gviz.pdf")
gbPlots <- gene_enzyme_enhancer[order(gene_enzyme_enhancer$cluster, gene_enzyme_enhancer$profile_directionality_chd4),]
# head(gbPlots[,c(1:3,10,36,155)])
dev.off()

pdf("/home/pundhir/project/chip-seq-analysis/analysis_me1_me3/pdf/gviz_directionality_bias.pdf")
apply(gbPlots[9203,c(1:3)], 1, function(x) plot_genome_browser_example(x[1], x[2], x[3]))
apply(gbPlots[9214,c(1:3)], 1, function(x) plot_genome_browser_example(x[1], x[2], x[3]))
apply(gbPlots[9155,c(1:3)], 1, function(x) plot_genome_browser_example(x[1], x[2], x[3]))
apply(gbPlots[1979,c(1:3)], 1, function(x) plot_genome_browser_example(x[1], x[2], x[3]))
apply(gbPlots[2240,c(1:3)], 1, function(x) plot_genome_browser_example(x[1], x[2], x[3]))
dev.off()
```

##################################################
#### TRACKVIEWER PLOTS
##################################################
```{r}
## load libraries
library(rlist)
library(GenomicRanges)
library(GenomicFeatures)
library(Gviz)
library(rtracklayer)
library(trackViewer)
library(TxDb.Mmusculus.UCSC.mm9.knownGene)
library(org.Mm.eg.db)

## define genomic region of interest
gr <- GRanges("chr4", IRanges(128956073, 129040885), strand="*")

## load transcripts model
trs <- geneModelFromTxdb(TxDb.Mmusculus.UCSC.mm9.knownGene, org.Mm.eg.db, gr=gr)

## load tracks
track_h3k4me3 <- importBam("/home/pundhir/project/chip-seq-analysis/analysis_kdm5c/test/h3k4me3_scgfp_Rep1.bam", ranges = gr, pairs=FALSE)
track_h3k4me3$dat <- coverageGR(track_h3k4me3$dat)

track_h3k27me3 <- importScore("/home/pundhir/project/chip-seq-analysis/analysis_kdm5c/test/h3k27me3_gmp_p30_Rep1.bw", format="BigWig", ranges = gr, ignore.strand = T)
track_h3k27me3$dat <- coverageGR(track_h3k27me3$dat)

## make tracks and their style list
optSty <- optimizeStyle(trackList(track_h3k4me3, track_h3k27me3, trs))
trackList <- optSty$tracks
viewerStyle <- optSty$style

## optimize tracks
setTrackXscaleParam(trackList$track_h3k4me3, "draw", TRUE)
setTrackStyleParam(trackList$track_h3k27me3, "ylim", c(0, 100))

setTrackStyleParam(trackList$track_h3k4me3, "color", c("green", "black"))
setTrackStyleParam(trackList$track_h3k27me3, "color", c("blue", "black"))
for(i in 3:length(trackList)) setTrackStyleParam(trackList[[i]], "color", "black")

## visualize
dev.off(); viewTracks(trackList, gr=gr, viewerStyle=viewerStyle, newpage=FALSE)
```

##################################################
#### GVIZ PLOTS
##################################################
```{r}
CHR <- "chr4"
GEN <- "mm9"
START=128956073
END=129040885

itrack <- IdeogramTrack(genome = GEN, chromosome = CHR)
gtrack <- GenomeAxisTrack()
txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene
txTr <- GeneRegionTrack(txdb, chromosome = CHR, start = START, end = END)

track_h3k4me3 <- DataTrack(range = "/home/pundhir/project/chip-seq-analysis/analysis_kdm5c/test/h3k4me3_scgfp_Rep1.bam", genome = "mm9", chromosome = "chr4",
                        type = "histogram", ylim = c(0,100), name = "Coverage", window = -1, col.histogram="green",  baseline=0, span=0.1)

track_h3k27me3 <- DataTrack(range = "/home/pundhir/project/chip-seq-analysis/analysis_kdm5c/test/h3k27me3_gmp_p30_Rep1.bw", genome = "mm9", chromosome = "chr4",
                         type = "histogram", name = "Coverage", window = -1, col.histogram="blue", baseline=0, span=0.1)

my_list <- c(itrack, gtrack, txTr, track_h3k27me3, track_h3k4me3)
plotTracks(my_list, from=START, to=END, smooth=0)
```

######################################################
#### MISC ANALYSIS
######################################################
```{r}
test <- gene_enzyme_enhancer[which(gene_enzyme_enhancer$cluster=="Y2" & gene_enzyme_enhancer$dist_closest_tss>5000),]
ggplot(test, aes(directionality_class, log(dist_closest_tss))) + geom_boxplot()
ggplot(test, aes(directionality_class, groseq_directionality_rep2)) + geom_boxplot()
ggplot(test, aes(directionality_class, log(expr_esc))) + geom_boxplot()
ggplot(test, aes(groseq_class, profile_directionality_chd4)) + geom_boxplot()
ggplot(test, aes(groseq_class, log(expr_esc))) + geom_boxplot()
test$ep400_cluster <- factor(test$ep400_cluster, levels=c("high", "average", "low"), ordered = T)
ggplot(test, aes(ep400_cluster, profile_directionality_chd4)) + geom_boxplot()
ggplot(test, aes(ep400_cluster, groseq_directionality_rep2)) + geom_boxplot()

df <- read.table("t")
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "dist_tss_down", "cluster", "cage_directionality_rep1")
test <- merge(test, df[,c(4,9)], by.x="gene", by.y="gene")
test <- test[,c(2:4,1,5:ncol(test))]
test$groseq_class <- "0"
test[which(test$groseq_directionality_rep2 < -0.35),]$groseq_class <- "-1"
test[which(test$groseq_directionality_rep2 > 0.35),]$groseq_class <- "1"
table(test$groseq_class)

df.melt <- melt(test[,c(170,133,132,200)])
ggplot(df.melt, aes(directionality_class, (value))) + geom_boxplot() + facet_grid(.~cluster + promoter_class)

## -------------------------------------------------- ##
## exploratory analysis
## -------------------------------------------------- ##

df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$dist_closest_tss > 5000),c(170,133,132,198)])
ggplot(df.melt, aes(directionality_class, (value))) + geom_boxplot() + facet_grid(.~cluster + promoter_class)

df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$dist_closest_tss > 5000),c(170,133,132,10)])
ggplot(df.melt, aes(directionality_class, log(value))) + geom_boxplot() + facet_grid(.~cluster + promoter_class)

df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$dist_closest_tss > 5000),c(170,176)])
ggplot(df.melt, aes(directionality_class, log(value))) + geom_boxplot()

df.melt <- melt(gene_enzyme_enhancer[which(gene_enzyme_enhancer$dist_closest_tss > 5000),c(170,166:169,197:198)])
ggplot(df.melt, aes(directionality_class, (value))) + geom_boxplot() + facet_grid(.~variable)

df.melt <- melt(gene_enzyme_enhancer[,c(164,197:198)])
ggplot(df.melt, aes(promoter_class_merged, value)) + geom_boxplot() + facet_grid(.~variable)

df.melt <- melt(gene_enzyme_enhancer[,c(199,10)])
ggplot(df.melt, aes(groseq_class, log(value))) + geom_boxplot()

# groseq_directionality_bias <- read.table(pipe("grep -w Y2 t"))
# colnames(groseq_directionality_bias) <- c("chr", "start", "end", "gene", "score", "strand", "distance_tss_down", "class", 
#                                           "groseq_bias_rep1", "groseq_bias_rep2")
# t <- merge(gene_enzyme_enhancer, groseq_directionality_bias[,c(4,9,10)], by.x="gene", by.y="gene")
# cor(t$profile_directionality_chd4, t$groseq_bias_rep2)
# t.melt <- melt(t[,c(197,198,170)])
# ggplot(t.melt, aes(directionality_class, value)) + geom_boxplot()


test <- gene_enzyme_enhancer
mll2_bivalent_genes <- read.table("mll2_bivalent.genes")
test$mll2_bivalent_genes <- "no"
test[which(test$gene %in% mll2_bivalent_genes$V1),]$mll2_bivalent_genes <- "yes"
#test.melt <- melt(test[which(test$cluster=="Y1"),c(10,15:67,194)])
test.melt <- melt(test[which(test$cluster=="Y1"),c(163:166,194)])
ggplot(test.melt, aes(mll2_bivalent_genes, (value))) + geom_boxplot() + facet_wrap(~variable)

df <- read.table("promoters/esc/allGenes_pol2pausingRaw.bed")
colnames(df) <- c("chr", "start", "end", "gene", "score", "strand", "id", "TSSR", "pol2_TSSR", "dist_closest_tss", 
                  "pausing_brg1_wt_rep1", "pausing_brg1_wt_rep2",
                  "pausing_brg1_ko_rep1", "pausing_brg1_ko_rep2",
                  "pausing_mbd3_24h_rep1", "pausing_mbd3_24h_rep2", "pausing_mbd3_24h_rep3",
                  "pausing_mbd3_0h_rep1", "pausing_mbd3_0h_rep2", "pausing_mbd3_0h_rep3",
                  "pausing_wt_rep1", "pausing_brg1_kd_rep1", "pausing_chd4_kd_rep1", "pausing_ep400_kd_rep1")
df$gene_coor <- sprintf("%s:%d-%d", df$chr, df$start, df$end)
df$gene_length <- df$end - df$start
#length(apply(t(as.data.frame(strsplit(as.character(df[,c(11)]), "/"))), 1, function(x) as.numeric(x[1])/(as.numeric(x[1])+as.numeric(x[2]))))

# df$pausing_brg1_change_rep1 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(11)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(14)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_brg1_change_rep2 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(12)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(15)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_brg1_change_rep3 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(13)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(16)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_mbd3_change_rep1 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(17)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(21)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_mbd3_change_rep2 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(18)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(22)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_mbd3_change_rep3 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(19)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(23)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_mbd3_change_rep4 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(20)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(24)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")
# 
# df$pausing_ep400_change_rep1 <- p.adjust(apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(25)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(26)]), "/")))), 1, function(x) fisher.test(round(matrix(c(as.numeric(x[1])*100, as.numeric(x[2])*100, as.numeric(x[3])*100, as.numeric(x[4])*100), nrow=2)))$p.value), method="BH")

# df$pausing_brg1_change_rep1 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(11)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(14)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_brg1_change_rep2 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(12)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(15)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_brg1_change_rep3 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(13)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(16)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_mbd3_change_rep1 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(17)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(21)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_mbd3_change_rep2 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(18)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(22)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_mbd3_change_rep3 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(19)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(23)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_mbd3_change_rep4 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(20)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(24)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))
# 
# df$pausing_ep400_change_rep1 <- apply(cbind(t(as.data.frame(strsplit(as.character(df[,c(25)]), "/"))), t(as.data.frame(strsplit(as.character(df[,c(26)]), "/")))), 1, function(x) log2((as.numeric(x[4])*100)/(as.numeric(x[2])*100)))

mat <- t(apply(df[,c(11:24)], 1, function(x) as.numeric(gsub("^.*/", "", x))))
#res.brg1 <- matrix2deseq2(mat[,c(1:2,3:4)], c("wt,wt,ko,ko"))
#res.chd4 <- matrix2deseq2(mat[,c(6:7,9:10)], c("wt,wt,ko,ko"))
df$fc_gb_Brg1 <- res.brg1$log2FoldChange
df$fc_gb_Chd4 <- res.chd4$log2FoldChange
mat <- t(apply(df[,c(11:24)], 1, function(x) as.numeric(gsub("/.*$", "", x))))
#res.brg1 <- matrix2deseq2(mat[,c(1:2,3:4)], c("wt,wt,ko,ko"))
#res.chd4 <- matrix2deseq2(mat[,c(6:7,9:10)], c("wt,wt,ko,ko"))
df$fc_tssr_Brg1 <- res.brg1$log2FoldChange
df$fc_tssr_Chd4 <- res.chd4$log2FoldChange

df1 <- merge(gene_enzyme_enhancer[,c(1:169)], df[,c(4,27:30)],  by.x="gene", by.y="gene", all.x=T)
df1 <- df1[,c(2:4,1,5:173)]

df1$pause_class <- cut2(df1$fc_gb_Brg1, m=500)
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$fc_gb_Brg1, m=500)
df1.melt <- melt(df1.ep400[,c(174,71)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)

df1$pause_class <- cut2(df1$fc_gb_Chd4, m=500)
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$fc_gb_Chd4, m=500)
df1.melt <- melt(df1.ep400[,c(174,83)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)


df1 <- gene_enzyme_enhancer
df1$pause_class <- cut2(df1$pausing_brg1_ko_rep2/df1$pausing_brg1_wt_rep2, m=500)
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$pausing_brg1_ko_rep2/df1.ep400$pausing_brg1_wt_rep2, m=500)
df1.melt <- melt(df1.ep400[,c(191,71)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)

df1 <- gene_enzyme_enhancer
df1$pause_class <- cut2(df1$pausing_mbd3_0h_rep3/df1$pausing_mbd3_24h_rep3, m=500)
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$pausing_mbd3_0h_rep3/df1.ep400$pausing_mbd3_24h_rep3, m=500)
df1.melt <- melt(df1.ep400[,c(191,83)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)


df1 <- merge(gene_enzyme_enhancer[,c(1:169)], df[,c(4,29:36)], by.x="gene", by.y="gene", all.x=T)
df1 <- df1[,c(2:4,1,5:177)]

df1$pause_class <- cut2(df1$pausing_brg1_change_rep1, m=500)
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$pausing_brg1_change_rep1, m=500)
df1.melt <- melt(df1.ep400[,c(178,71)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)

df1$pause_class <- cut2(df1$pausing_mbd3_change_rep3, m=500)
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$pausing_mbd3_change_rep3, m=500)
df1.melt <- melt(df1.ep400[,c(178,83)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)

df1 <- test
df1.ep400 <- df1[which(df1$ep400_cluster_merged=="high" | df1$ep400_cluster_merged=="average"),]
df1.ep400$pause_class <- cut2(df1.ep400$pausing_brg1_ko, m=500)
df1.melt <- melt(df1.ep400[,c(197,71)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)

df1.ep400$pause_class <- cut2(df1.ep400$pausing_mbd3_ko, m=500)
df1.melt <- melt(df1.ep400[,c(197,83)])
ggplot(df1.melt, aes(pause_class, value)) + geom_boxplot() + 
  theme(axis.text.x = element_text(size=10, angle=45)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "red", size=1)

## -------------------------------------------------- ##
## overlap analysis with prc2 ko genes
## -------------------------------------------------- ##

gene_overlap <- read.table(pipe("grep '_Y1' /home/pundhir/project/chip-seq-analysis/data/gene/esc/GENE_OVERLAP"))
gene_overlap$V2 <- as.factor(as.character(gsub(pattern = "_.*", "", gene_overlap$V2)))
gene_overlap$V2 <- factor(gene_overlap$V2, levels=enzyme_order, ordered = T)
res <- sapply(unique(gene_overlap$V2), function(x) getPval(testGeneOverlap(newGeneOverlap(gene_overlap[which(gene_overlap$V2==x),]$V1, gene_overlap[which(gene_overlap$V2=="ring1" | gene_overlap$V2=="ezh2" | gene_overlap$V2=="suz12"),]$V1, genome.size=14473))))
# res <- sapply(unique(gene_overlap$V2), function(x) getPval(testGeneOverlap(newGeneOverlap(gene_overlap[which(gene_overlap$V2==x),]$V1, gene_overlap[which(gene_overlap$V2=="ep400"),]$V1, genome.size=9598))))
res <- data.frame(unique(gene_overlap$V2), res)
colnames(res) <- c("enzyme", "pvalue")
ggplot(res[c(2,3,4,5,6,8),], aes(enzyme, -log10(pvalue))) + geom_bar(stat="identity") + theme_bw() +
  geom_hline(yintercept = -log10(0.00000000001), linetype="dashed", color="red", 1)
total <- as.data.frame(table(gene_overlap$V2))
colnames(total) <- c("enzyme", "total")
p2 <- ggplot(total[c(2,3,4,5,6,8,12,13),], aes(enzyme, total)) + geom_bar(stat="identity") + theme_bw()

enz <- c("chd2", "chd9", "chd1", "chd6", "chd4", "brg1")
res <- sapply(c(1:1000), function(y) { set.seed(y); sapply(enz, function(x) length(getIntersection(testGeneOverlap(newGeneOverlap(sample(gene_overlap[which(gene_overlap$V2==x),]$V1, 12), gene_overlap[which(gene_overlap$V2=="ring1" | gene_overlap$V2=="ezh2" | gene_overlap$V2=="suz12"),]$V1, genome.size=14473))))) })
res <- data.frame(res, enz)
colnames(res) <- c(c(1:1000), "enzyme")
res.melt <- melt(res)
p3 <- ggplot(res.melt, aes(enzyme, value)) + geom_boxplot() + theme_bw()
apply(res, 1, function(x) x[1001])
apply(res, 1, function(x) wilcox.test(as.numeric(x[1:1000]), as.numeric(res[which(res$enzyme=="chd1"),c(1:1000)]), alternative = "g")$p.value)

enz <- c("chd1", "chd6", "chd4", "brg1")
res <- sapply(c(1:1000), function(y) { set.seed(y); sapply(enz, function(x) getPval(testGeneOverlap(newGeneOverlap(sample(gene_overlap[which(gene_overlap$V2==x),]$V1, 46), gene_overlap[which(gene_overlap$V2=="ring1" | gene_overlap$V2=="ezh2" | gene_overlap$V2=="suz12"),]$V1, genome.size=14473)))) })
res <- data.frame(res, enz)
colnames(res) <- c(c(1:1000), "enzyme")
res.melt <- melt(res)
pVal <- data.frame(apply(res, 1, function(x) length(which(as.numeric(x)<0.01))), res$enzyme)
colnames(pVal) <- c("significant", "enzyme")
ggplot(pVal, aes(enzyme, significant)) + geom_point(stat="identity")
g <- arrangeGrob(p1, p2, p3, p4, nrow = 2, ncol=2)
ggsave(g, dpi=300, height=20, width=20, filename="pdf/genes_overlap_prc2.pdf", useDingbats=FALSE)

test <- read.table("/home/pundhir/project/chip-seq-analysis/data/gene/me1_me3/OVERLAP.ALL")
colnames(test) <- c("brg1", "chd1", "chd4", "chd6", "chd8", "chd9", "ep400", "brd4", "prc2", "mll34", "chd2")
test <- test[order(-test$prc2),]
mat <- (as.matrix(test))
pheatmap(mat, color=col, border_color = NA, scale="none", cluster_rows = T,
         cluster_cols = T, legend=T, show_rownames = T, show_colnames = T,
         clustering_distance_rows="maximum", clustering_distance_cols="maximum")

test <- read.table(pipe("less /home/pundhir/project/chip-seq-analysis/data/gene/me1_me3/GENE.MATRIX | cut -f 1,2,5,8,11,14,17,20,23,26,32,34"))
colnames(test) <- c("gene", "brg1", "chd1", "chd4", "chd6", "chd8", "chd9", "ep400", "brd4", "prc2", "mll34", "chd2")
mat <- as.matrix(test[,c(2:12)])
mat[is.na(mat)] <- 0
mat[is.infinite(mat)] <- 0
corrplot(cor(mat), order = "hclust")


## -------------------------------------------------- ##
## Correlation heatmaps using ngsPlot matrices
## -------------------------------------------------- ##

test <- list()
i=1;
SAMPLE=c("ash2l", "brg1", "cbx7", "chd1", "chd2", "chd4", "chd6", "chd8", "chd9", "ctcf", "cxxc1", "ep400", "ezh2", "hdac1", "kdm5b", "mll2", "mll34", "p300", "rad21", "ring1b", "set1a", "suz12", "h3k27ac", "h3k27me3", "h3k4me1", "h3k4me3", "h3k9me3", "h2az", "h3k4me2", "h3k9ac", "brd4", "lsd1", 
         "cdk8", "jarid2", "kdm2a", "med12", "med1", "nipbl", "smc1a", "tbp", "yy1")
for(NAMES in SAMPLE) { 
  load(sprintf("ngsPlot/esc/promoters/%s/heatmap.heatmap.RData", NAMES))
  test[[i]] <- as.data.frame(enrichList[[5]])
  i <- i+1
}
names(test) <- SAMPLE
cor_y2 <- compute_cor_matrix_list(test, "pearson", 101)

col=colorRampPalette(c("dodgerblue4","grey97", "sienna3"))(250);
mat <- cor_y2[c(5,25,9,4,29,31,2,6,8,7,28,26,30,15,23,11,1,18,12,16,14,21,17,32),c(5,25,9,4,29,31,2,6,8,7,28,26,30,15,23,11,1,18,12,16,14,21,17,32)]
pheatmap(mat, color=col, border_color = NA, scale="none", cluster_rows = T,
         cluster_cols = T, legend=T, show_rownames = T, show_colnames = T)

test <- list()
i=1;
SAMPLE=c("ash2l", "brg1", "cbx7", "chd1", "chd2", "chd4", "chd6", "chd8", "chd9", "ctcf", "cxxc1", "ep400", "ezh2", "hdac1", "kdm5b", "mll2", "mll34", "p300", "rad21", "ring1b", "set1a", "suz12", "h3k27ac", "h3k27me3", "h3k4me1", "h3k4me3", "h3k9me3", "h2az", "h3k4me2", "h3k9ac", "brd4", "lsd1",
         "cdk8", "jarid2", "kdm2a", "med12", "med1", "nipbl", "smc1a", "tbp", "yy1")
for(NAMES in SAMPLE) { 
  load(sprintf("ngsPlot/esc/promoters/%s/heatmap.heatmap.RData", NAMES))
  test[[i]] <- as.data.frame(enrichList[[4]])
  i <- i+1
}
names(test) <- SAMPLE
#cor_y1 <- compute_cor_matrix_list(test, "pearson", 101)

col=colorRampPalette(c("dodgerblue4","grey97", "sienna3"))(250);
mat <- cor_y1[c(2,3,6,7,13,14,15,16,17,18,20,21,22,24,25,26,27,29,30,33:41),c(2,3,6,7,13,14,15,16,17,18,20,21,22,24,25,26,27,29,30,33:41)]
pheatmap(mat, color=col, border_color = NA, scale="none", cluster_rows = T,
         cluster_cols = T, legend=T, show_rownames = T, show_colnames = T)
```

