---
title: "R Notebook for Chd4 and Brg1 project (author: sachin pundhir)"
output:
  pdf_document: default
  html_notebook: default
  html_document:
    df_print: paged
    fig_caption: yes
editor_options:
  chunk_output_type: console
---

```{r global options, include = F}
knitr::opts_chunk$set(echo=F, include = T, results="hide", warning=FALSE, message=FALSE, tidy = TRUE, tidy.opts = list(comment = FALSE),
                      fig.width=12, fig.height=8, out.width = "100%", out.height = "70%")
## edit a R function
## trace("makevp.eqsc",edit=TRUE) 
```

```{r cachecontrol, cache = TRUE}
knitr::opts_chunk$set(cache.rebuild = TRUE)
```

######################################################
#### LOAD LIBRARIES
######################################################
```{r setup}
library(DESeq2)
library(preprocessCore)
library(reshape)
library(ggplot2)
library(doBy)
library(ggpubr)
library(gridExtra)
library(Hmisc)
library(plyr)
library(pheatmap)
library(Gviz)
library(rlist)
library(GenomicRanges)
library(GenomicFeatures)
library(TxDb.Mmusculus.UCSC.mm9.knownGene)
library(rtracklayer)
library(trackViewer)
library(plyr)
library(ComplexHeatmap)
theme_set(theme_classic2())
source("~/software/myScripts/R-scripts/FUNCTIONS.R")
```

###########################################
## methods
<!-- TCGA cohorts correlation analyses -->

<!-- The phenotype dataset (with survival outcomes) and the RNAseq dataset of each cancer cohort (the gene-level transcription estimates, as in log2(x+1) transformed RSEM normalized count) were downloaded from UCSC Xena (https://xenabrowser.net/) where genes are mapped onto the human genome coordinates using UCSC Xena HUGO probeMap in the RNAseq dataset. Spearman correlation coefficient was calculated between TREM2 and each of the genes in each cancer cohort, accompanied with sample size and p values. The individual scatterplot of each of the top genes with TREM2 (with fitted linear lines) was generated with Spearman correlation in the plot. -->
<!-- The Kaplan-Meier curve for overall survival (OS) and relapse free survival (RFS) was generated for TREM2 dichotomized by 75% quantile of TREM2 expression into high/low TREM2 expression group. The log-rank test was applied to test the survival difference between high/low TREM2 expression. -->
###########################################

###########################################
## load TCGA data
## https://shixiangwang.github.io/home/en/post/ucscxenatools-201908/
###########################################
```{r}
#gene_expr <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_gene_expected_count.gz", header=T, check.names=FALSE)

ensembl_protein_coding <- read.table(pipe("grep protein_coding /scratch/genomes/annotations/BED/hg19_ensembl_gene.bed"))
gene_expr_PC <- gene_expr[(gsub("\\..*", "", gene_expr$sample) %in% ensembl_protein_coding$V7),]

tissue_barCode <- read.table(pipe("less ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/code_table/tissueSourceSite.tsv | cut -f 1,3"), header=T, sep="\t")

mutation_info <- read.table(pipe("zcat ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_mutation_phenotype.txt.gz | cut -f 1,7,8,9,12,13"), header=T, check.names=FALSE, sep="\t", stringsAsFactors=FALSE, quote="\"")

survival_info <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TCGA_survival_data.gz", header=T, check.names=FALSE, sep="\t", stringsAsFactors=FALSE, quote="\"")

#sample_info <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGTEX_phenotype.txt.gz", header=T, sep="\t", stringsAsFactors=FALSE)
sample_info <- read.table(pipe("less ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/clinical_data/clinical.tsv | cut -f 1,2,3,4,9,11,12,15,16,20,25,26,27,28,29,41,47,48,68,80,96,108,110,111,112,116,119,120,125,128,152,154 | cut -f 2,4,6,7,8,24,28,30,31"), header=T, sep="\t", stringsAsFactors=FALSE, check.names=FALSE, quote="\"")
colnames(sample_info)[1] <- "sample"
sample_info <- unique(sample_info)

smarca4_info <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/41594_2017_7_MOESM6_ESM.txt", sep="\t", header=T, stringsAsFactors=FALSE)
smarca4_info[which(smarca4_info$Study=="Testicular Germ Cell Cancer"),]$Study <- "Testicular Germ Cell Tumors"

chd4_info <- read.table(pipe("zcat ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_mutation_phenotype.txt.gz | cut -f 1,7,8,13 | grep -w CHD4"), header=F, check.names=FALSE, sep="\t", stringsAsFactors=FALSE, quote="\"")
colnames(chd4_info) <- c("sample", "gene", "effect", "tissueSource")
chd4_info[which(chd4_info$effect!="Missense_Mutation"),]$effect <- "other"

t <- mutation_info[which(grepl("Uterine Corpus Endometrial Carcinoma|Colon adenocarcinoma|Stomach adenocarcinoma", mutation_info$tissueSource) &
                      mutation_info$gene!="CHD4"),]
t$gene <- "Not_CHD4"
t$effect <- "No_mutation"
t <- unique(t[,c(1,2,3,6)])
t <- t[which(!(t$sample %in% chd4_info$sample)),]
#length(grep(paste(unlist(lapply(strsplit(chd4_info[which(chd4_info$tissueSource=="Uterine Corpus Endometrial Carcinoma"),]$sample, "-"), function(x) sprintf("TCGA-%s-.*$", x[2]))), collapse="|"), t$sample))
chd4_info <- rbind(chd4_info, t)
table(chd4_info$effect)

gtex_info <- read.table(pipe("cat ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt | cut -f 1,6,7"), sep="\t", header=T)
```

###########################################
## Analysis for SMARCA4 (confounding factors)
###########################################
```{r}
GENE="SMARCA4"
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene==GENE),c("sample", "tissueSource")])$tissueSource))

gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
if(GENE=="CHD4") {
  names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")
} else{
  names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")
}

LIST <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
LIST <- LIST[stringr::str_to_title(LIST) %in% stringr::str_to_title(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study))]

## test for confounding factors (phenotype)
confounding_smarca4 <- lapply(LIST, function(TISSUE) {
  unlist(lapply(colnames(sample_info)[2:ncol(sample_info)], function(FACTOR) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 # &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    
    info_missense <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_missense), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_missense$class <- "missense"
    info_no_mutation <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_no_mutation), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_no_mutation$class <- "no_mutation"
    info_test <- rbind(info_missense, info_no_mutation)
    info_test$class <- as.factor(info_test$class)
    cat(TISSUE, length(ID_no_mutation), length(ID_missense), length(info_test$class)); cat("\n");

    if(length(unique(info_test[,FACTOR]))>1) {
      if(FACTOR=="age_at_index") {
        summary(glm(as.numeric(as.character(info_test[,FACTOR])) ~ info_test$class, data=info_test))$coefficients[2,4]
      } else {
        chisq.test(table(info_test[,c(3,2)]))$p.value
      }
    } else {
      1
    }
  }))
})
names(confounding_smarca4) <- LIST
confounding_smarca4 <- as.data.frame(confounding_smarca4)
row.names(confounding_smarca4) <- colnames(sample_info)[2:ncol(sample_info)]
confounding_smarca4$factor <- colnames(sample_info)[2:ncol(sample_info)]
apply(confounding_smarca4[,c(1:ncol(confounding_smarca4)-1)], 2, function(x) length(which(x<0.01)))

## test for confounding factors (genotype)
confounding_smarca4_genotypes <- lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 # &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    
    info_missense <- unique(mutation_info[which(grepl(paste(ID_missense, collapse="|"), mutation_info$sample) &
                                           grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    info_no_mutation <- unique(mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) &
                                        grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    unlist(lapply(attributes(Vennerable::Venn(list(
      A=paste(as.data.frame(table(info_no_mutation$gene))[which(as.data.frame(table(info_no_mutation$gene))$Freq >= 
                                                            length(ID_no_mutation)),]$Var1, collapse=","),
      B=paste(as.data.frame(table(info_missense$gene))[which(as.data.frame(table(info_missense$gene))$Freq >= 
                                                         length(ID_missense)),]$Var1, collapse=","))))$IntersectionSets,
           function(x) ifelse(x=="", "NA", x)))
})
names(confounding_smarca4_genotypes) <- LIST
confounding_smarca4_genotypes <- as.data.frame(confounding_smarca4_genotypes)

## test for survival difference
survival_pVal <- lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    cat(TISSUE); cat("\n")
    wilcox.test(survival_info[grep(paste(ID_missense, collapse="|"), survival_info$sample),]$OS.time, survival_info[grep(paste(ID_no_mutation, collapse="|"), survival_info$sample),]$OS.time)$p.value
})
names(survival_pVal) <- LIST
survival_info <- t(as.data.frame(survival_pVal))
```

###########################################
## Analysis for SMARCA4
###########################################
```{r}
TISSUE_STUDY <- unlist(lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    cat(TISSUE, length(ID_no_mutation), length(ID_missense)); cat("\n");
    if(length(ID_missense)>0 & length(ID_no_mutation)>0) { return(TISSUE); }
}))

test_smarca4 <- lapply(TISSUE_STUDY, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    #SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID_missense, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMeanCustom <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMeanCustom),]
    mat$expr_class <- cut2(mat$baseMeanCustom, g=9)
    #mat$expr_class <- cut2(c(1:nrow(mat)), g=5)

    ## DESeq2 analysis
    test <- mat[,grep("mutation", colnames(mat))]
    colnames(test) <- gsub("\\..*", "", colnames(test))
    colnames(test) <- gsub("no_mutation", "wt", colnames(test))
    colnames(test) <- gsub("mutation", "ko", colnames(test))
    test <- matrix2deseq2(countTable = test, treatment = colnames(test))
    row.names(test) <- test$gene
    res <- merge(test, mat[,c("baseMeanCustom", "expr_class")], by="row.names")
    colnames(res) <- gsub("\\..*", "", colnames(res))
    return(res)
})
names(test_smarca4) <- TISSUE_STUDY

p_smarca4 <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  # plot(log2(test$baseMeanCustom), test$log2FoldChange)
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  # test$expr_class <- cut2(1:nrow(test), g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  #cat(ncol(SAMPLE_MUTATION), ncol(SAMPLE_NO_MUTATION))

  test.melt <- melt(test)
  cat(TISSUE); cat("\n")
  #test.melt$value <- log2(test.melt$value)
  # ggline(test.melt, x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
  #   theme_bw() +
  #   ggtitle(TISSUE) +
  #   scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
  #   ylab("Gene expr.") + xlab("") +
  #   theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
    #scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test_GENE[[TISSUE]]$expr_class), function(x) mean(test_GENE[[TISSUE]][which(test_GENE[[TISSUE]]$expr_class==x),]$baseMean)))))

  ggboxplot(test.melt, x = "expr_class", y = "value", fill = "variable", yscale = "log2",
            #notch=1, notchwidth = 0.6, yscale="log2",
            palette = c("#e31a1c", "#1f78b4"), outlier.shape=NA) +
    xlab("") +
    ylab("Gene exor. (log2)") + ggtitle(TISSUE)
  })

pVal_smarca4 <- unlist(lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  wilcox.test(test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="ko_RAW"),]$value,
              test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="wt_RAW"),]$value, alternative="g")$p.value
}))

p_smarca4_fc <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  test$expr_class <- cut2(test$baseMeanCustom, g=9)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  test.melt <-summaryBy(value ~ variable + expr_class, data = test.melt, FUN = list(median))
  log2(test.melt[which(test.melt$variable=="ko_RAW"),]$value.median/test.melt[which(test.melt$variable=="wt_RAW"),]$value.median)
})
names(p_smarca4_fc) <- names(test_smarca4)
p_smarca4_fc <- as.data.frame(p_smarca4_fc)
p_smarca4_fc$class <- as.factor(row.names(p_smarca4_fc))
p_smarca4_fc <- melt(p_smarca4_fc)

p_smarca4_deseq2 <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  ggline(melt(test[,c("log2FoldChange", "expr_class")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
    theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c")) +
    ylab("Gene expr. (Log2FC)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
})

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4.pdf", width=20)
ggarrange(plotlist = p_smarca4, nrow=2, ncol=8)
ggarrange(plotlist = p_smarca4_deseq2, nrow=2, ncol=8)
dev.off()

# expr_smarca4 <- lapply(TISSUE_STUDY, function(TISSUE) {
#   cat(TISSUE); cat("\n")
#   sample <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]$sample)[unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]$sample) %in% colnames(gene_expr)]
#   rowMeans(gene_expr[grep("ENSG00000127616", gene_expr$sample), sample])
# })
# names(expr_smarca4) <- TISSUE_STUDY
# expr_smarca4 <- as.data.frame(t(as.data.frame(expr_smarca4)))
# expr_smarca4$sample <- row.names(expr)
```

###########################################
## Analysis for CHD4 
###########################################
```{r}
GENE="CHD4"
# TISSUE_STAT <- as.data.frame(table(mutation_info[which(mutation_info$gene==GENE),]$tissueSource))
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene==GENE),c("sample", "tissueSource")])$tissueSource))

gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
if(GENE=="CHD4") {
  names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")
} else{
  names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")
}

LIST <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)

## test for confounding factors (phenotype)
confounding_chd4 <- lapply(LIST, function(TISSUE) {
  unlist(lapply(colnames(sample_info)[2:ncol(sample_info)], function(FACTOR) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 # &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="CHD4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    info_missense <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_missense), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_missense$class <- "missense"
    info_no_mutation <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_no_mutation), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_no_mutation$class <- "no_mutation"
    info_test <- rbind(info_missense, info_no_mutation)
    info_test$class <- as.factor(info_test$class)
    cat(TISSUE, length(ID_no_mutation), length(ID_missense), length(info_test$class)); cat("\n");
    
    if(length(unique(info_test[,FACTOR]))>1) {
      if(FACTOR=="age_at_index") {
        summary(glm(as.numeric(as.character(info_test[,FACTOR])) ~ info_test$class, data=info_test))$coefficients[2,4]
        # wilcox.test(as.numeric(as.character(info_test[which(info_test$class=="missense"),FACTOR])), as.numeric(as.character(info_test[which(info_test$class=="no_mutation"),FACTOR])))$p.value
      } else {
        chisq.test(table(info_test[,c(3,2)]))$p.value
      }
    } else {
      1
    }
    # info_missense <- mutation_info[grep(paste(ID_missense, collapse="|"), mutation_info$sample),c("sample", "gene")]
    # info_missense$class <- "missense"
    # info_no_mutation <- mutation_info[grep(paste(ID_no_mutation, collapse="|"), mutation_info$sample),c("sample", "gene")]
    # info_no_mutation$class <- "no_mutation"
    # info_test <- rbind(info_missense, info_no_mutation)
    # info_test$class <- as.factor(info_test$class)
  }))
})
names(confounding_chd4) <- LIST
confounding_chd4 <- as.data.frame(confounding_chd4)
row.names(confounding_chd4) <- colnames(sample_info)[2:ncol(sample_info)]
confounding_chd4$factor <- colnames(sample_info)[2:ncol(sample_info)]
apply(confounding_chd4[,c(1:ncol(confounding_chd4)-1)], 2, function(x) length(which(x<0.01)))

## test for confounding factors (genotype)
confounding_chd4_genotypes <- lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                #  &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    
    info_missense <- unique(mutation_info[which(grepl(paste(ID_missense, collapse="|"), mutation_info$sample) &
                                           grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    info_no_mutation <- unique(mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) &
                                        grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    unlist(lapply(attributes(Vennerable::Venn(list(
      A=paste(as.data.frame(table(info_no_mutation$gene))[which(as.data.frame(table(info_no_mutation$gene))$Freq >= 
                                                            length(ID_no_mutation)),]$Var1, collapse=","),
      B=paste(as.data.frame(table(info_missense$gene))[which(as.data.frame(table(info_missense$gene))$Freq >= 
                                                         length(ID_missense)),]$Var1, collapse=","))))$IntersectionSets,
           function(x) ifelse(x=="", "NA", x)))
})
names(confounding_chd4_genotypes) <- LIST
confounding_chd4_genotypes <- as.data.frame(confounding_chd4_genotypes)
```

###########################################
## Analysis for CHD4
###########################################
```{r}
GENE="CHD4"
TISSUE_STAT <- as.data.frame(table(mutation_info[which(mutation_info$gene==GENE),]$tissueSource))

gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
if(GENE=="CHD4") {
  names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")
} else{
  names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")
}

LIST <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)

TISSUE_STUDY <- unlist(lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="CHD4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    cat(TISSUE, length(ID_no_mutation), length(ID_missense)); cat("\n");
    if(length(ID_missense)>0 & length(ID_no_mutation)>0) { return(TISSUE); }
}))

test_chd4 <- lapply(TISSUE_STUDY, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    #SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID_missense, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMeanCustom <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMeanCustom),]
    mat$expr_class <- cut2(mat$baseMeanCustom, g=9)
    #mat$expr_class <- cut2(c(1:nrow(mat)), g=5)

    ## DESeq2 analysis
    test <- mat[,grep("mutation", colnames(mat))]
    colnames(test) <- gsub("\\..*", "", colnames(test))
    colnames(test) <- gsub("no_mutation", "wt", colnames(test))
    colnames(test) <- gsub("mutation", "ko", colnames(test))
    test <- matrix2deseq2(countTable = test, treatment = colnames(test))
    row.names(test) <- test$gene
    res <- merge(test, mat[,c("baseMeanCustom", "expr_class")], by="row.names")
    colnames(res) <- gsub("\\..*", "", colnames(res))
    return(res)
})
names(test_chd4) <- TISSUE_STUDY

p_chd4 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[(gsub("\\..*", "", test$gene) %in% ensembl_protein_coding$V7),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  #cat(ncol(SAMPLE_MUTATION), ncol(SAMPLE_NO_MUTATION))

  test.melt <- melt(test)
  #test.melt$value <- log2(test.melt$value)
  # ggline(test.melt, x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
  #   theme_bw() +
  #   ggtitle(TISSUE) +
  #   scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
  #   ylab("Gene expr.") + xlab("") +
  #   theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
    #scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test_GENE[[TISSUE]]$expr_class), function(x) mean(test_GENE[[TISSUE]][which(test_GENE[[TISSUE]]$expr_class==x),]$baseMean)))))

  ggboxplot(test.melt, x = "expr_class", y = "value", fill = "variable", yscale = "log2",
            #notch=1, notchwidth = 0.6, yscale="log2",
            palette = c("#e31a1c", "#1f78b4"), outlier.shape=NA) +
    xlab("") +
    ylab("Gene exor. (log2)") + ggtitle(TISSUE)
  })

pVal_chd4 <- unlist(lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[(gsub("\\..*", "", test$gene) %in% ensembl_protein_coding$V7),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  wilcox.test(test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="ko_RAW"),]$value,
              test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="wt_RAW"),]$value, alternative="l")$p.value
}))

p_chd4_fc <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[(gsub("\\..*", "", test$gene) %in% ensembl_protein_coding$V7),]
  test$expr_class <- cut2(test$baseMeanCustom, g=9)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  test.melt <-summaryBy(value ~ variable + expr_class, data = test.melt, FUN = list(median))
  log2(test.melt[which(test.melt$variable=="ko_RAW"),]$value.median/test.melt[which(test.melt$variable=="wt_RAW"),]$value.median)
})
names(p_chd4_fc) <- names(test_chd4)
p_chd4_fc <- as.data.frame(p_chd4_fc)
p_chd4_fc$class <- as.factor(row.names(p_chd4_fc))
p_chd4_fc <- melt(p_chd4_fc)

p_chd4_deseq2 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  ggline(melt(test[,c("log2FoldChange", "expr_class")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
    theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c")) +
    ylab("Gene expr. (Log2FC)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
})

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/CHD4.pdf", width=20)
ggarrange(plotlist = p_chd4, nrow=2, ncol=8)
ggarrange(plotlist = p_chd4_deseq2, nrow=2, ncol=8)
dev.off()
```

########################################
## summary plots for SMARCA4 and CHD4
########################################
```{r}
t <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  # test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # ggplot(test_smarca4[[TISSUE]][,c("expr_class", "baseMeanCustom")], aes(expr_class, baseMeanCustom)) + geom_boxplot()
  # boxplot(test_smarca4[[TISSUE]]$baseMeanCustom)
  # table(test_smarca4[[TISSUE]]$expr_class)
  # test <- test[which(test$baseMeanCustom > 3.97e-05 & test$baseMeanCustom < 7.04e+00),]
  test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean=T)
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  summaryBy(log2FoldChange ~ expr_class, data = test, FUN = list(median))$log2FoldChange.median
})
t <- as.data.frame(t)
colnames(t) <- names(test_smarca4)
t$expr_class <- as.factor(row.names(t))

t1 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # test <- test[which(test$baseMeanCustom > 3.97e-05 & test$baseMeanCustom < 7.04e+00),]
  test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean = T)
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  summaryBy(log2FoldChange ~ expr_class, data = test, FUN = list(median))$log2FoldChange.median
})
t1 <- as.data.frame(t1)
colnames(t1) <- names(test_chd4)
t1$expr_class <- as.factor(row.names(t1))

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4_CHD4.pdf", width=15, height = 5)
ggplot(test[,c("expr_class", "baseMeanCustom")], aes(expr_class, baseMeanCustom)) + geom_boxplot() + theme_bw()

ggarrange(ggplot(p_smarca4_fc, aes(as.numeric(class), value)) + geom_point(aes(color=variable, size = 1)) +
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 2), alpha=0.1) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            ylab("FC in median gene expr. (Log2)"),
          ggplot(p_chd4_fc, aes(as.numeric(class), value)) + geom_point(aes(color=variable, size = 1)) +
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 2), alpha=0.1) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            ylab("FC in median gene expr. (Log2)"), 
          nrow=1, ncol=2)

ggarrange(ggplot(melt(t), aes(as.numeric(expr_class) , value)) + geom_point(aes(color=variable, size = 1)) + 
            geom_smooth(method = "glm", se = T, formula = y ~ poly(x, 1), span=0.2, alpha=0.2) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))), 
          ggplot(melt(t1), aes(as.numeric(expr_class) , value)) + geom_point(aes(color=variable, size = 1)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 1), alpha=0.2) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))), nrow=1, ncol=2)

lm(as.numeric(melt(t)$expr_class) ~ melt(t)$value)$coefficients[2]
summary(lm(as.numeric(melt(t)$expr_class) ~ melt(t)$value))$coefficients[2, 4]

lm(as.numeric(melt(t1)$expr_class) ~ melt(t1)$value)$coefficients[2]
summary(lm(as.numeric(melt(t1)$expr_class) ~ melt(t1)$value))$coefficients[2, 4]

t.melt <- melt(t)
t.melt$gene <- "Smarca4"
t1.melt <- melt(t1)
t1.melt$gene <- "Chd4"
df <- rbind(t.melt, t1.melt)
ggarrange(ggline(df, x = "expr_class", y = "value", add=c("mean"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)"),
          ggplot(df, aes(as.numeric(expr_class), value, color=gene)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 2), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c("#784b97", "#ffa054")), nrow=1, ncol=2)

TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="SMARCA4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")

df_smarca4 <- lapply(names(test_smarca4)[-5], function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  # test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean=T)

  # test$baseMean <- log2(test$baseMean)
  # test$baseMean <- log2(rowMedians(as.matrix(test[,grep("RAW", colnames(test))])))
  ID_BASELINE <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
  test <- merge(test[,c("gene", "baseMean", "log2FoldChange")], gene_expr_PC[,c("sample", grep(paste(ID_BASELINE, collapse="|"), colnames(gene_expr_PC), value=TRUE))], by.x="gene", by.y="sample")
  test$baseMean <- rowMedians(as.matrix(test[,grep("GTEX", colnames(test))]))
  
  # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
  test <- test[order(test$baseMean),]
  test$expr_class <- cut2(seq(1, nrow(test)), g=50, levels.mean=T)
  # test$expr_class <- as.numeric(as.character(test$expr_class))
  # test <- test[which(test$expr_class > summary(test$expr_class)[2] & test$expr_class < summary(test$expr_class)[5]),]
  # test$expr_class <- as.factor(test$expr_class)
  test$expr_class <- mapvalues(test$expr_class, from = levels(test$expr_class), to = seq(1, length(unique(test$expr_class))))
  
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  test <- melt(test[,c("expr_class", "log2FoldChange")])[,c(1,3)]
  test$variable <- TISSUE
  return(test)
})
# df_smarca4 <- as.data.frame(df_smarca4)
# df_smarca4 <- as.data.frame(cbind(df_smarca4$expr_class, df_smarca4[,grep("value", colnames(df_smarca4))]))
# colnames(df_smarca4) <- c("expr_class", names(test_smarca4))
# df_smarca4 <- melt(df_smarca4)
df_smarca4 <- do.call(rbind, df_smarca4)
df_smarca4$gene <- "Smarca4"

df_chd4 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  # test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean=T)

  #test$baseMean <- log2(test$baseMean)
  test$baseMean <- log2(rowMedians(as.matrix(test[,grep("RAW", colnames(test))])))
  # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
  test <- test[order(test$baseMean),]
  test$expr_class <- cut2(seq(1, nrow(test)), g=50, levels.mean=T)
  # test$expr_class <- as.numeric(as.character(test$expr_class))
  # test <- test[which(test$expr_class > summary(test$expr_class)[2] & test$expr_class < summary(test$expr_class)[5]),]
  # test$expr_class <- as.factor(test$expr_class)
  test$expr_class <- mapvalues(test$expr_class, from = levels(test$expr_class), to = seq(1, length(unique(test$expr_class))))
  
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  test <- melt(test[,c("expr_class", "log2FoldChange")])[,c(1,3)]
  test$variable <- TISSUE
  return(test)
})
# df_chd4 <- as.data.frame(df_chd4)
# df_chd4 <- as.data.frame(cbind(df_chd4$expr_class, df_chd4[,grep("value", colnames(df_chd4))]))
# colnames(df_chd4) <- c("expr_class", names(test_chd4))
# df_chd4 <- melt(df_chd4)
df_chd4 <- do.call(rbind, df_chd4)
df_chd4$gene <- "Chd4"

df <- rbind(df_smarca4, df_chd4)
# df <- rbind(df_smarca4[which(df_smarca4$value<0),], df_chd4[which(df_chd4$value<0),])
ggarrange(ggline(df, x = "expr_class", y = "value", add=c("median"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_smarca4, x = "expr_class", y = "value", add=c("median"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_chd4, x = "expr_class", y = "value", add=c("median"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"), nrow=1, ncol=3)

ggarrange(ggplot(df, aes(as.numeric(expr_class), value, color=gene)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 10), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c("#784b97", "#ffa054")),
          ggplot(df_smarca4, aes(as.numeric(expr_class), value, color=variable)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 10), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            theme(legend.position="none"),
          ggplot(df_chd4, aes(as.numeric(expr_class), value, color=variable)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 10), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            theme(legend.position="none"), nrow=1, ncol=3)

matrix2pheatmap(t(confounding_smarca4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
matrix2pheatmap(t(confounding_chd4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
dev.off()
```

###########################################
## Analysis for SMARCA4 and CHD4 (regression-based analysis)
###########################################
```{r}
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="SMARCA4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")

df_smarca4_sampleStat <- lapply(names(test_smarca4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="SMARCA4" &
                                                mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    ID_frame_shift <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                            mutation_info$gene=="SMARCA4" &
                                            (mutation_info$effect=="Frame_Shift_Del" | mutation_info$effect=="Frame_Shift_Ins" |
                                            mutation_info$effect=="Splice_Site")
                                          ),]$sample)
    ID_frame_shift <- ID_frame_shift[which(!ID_frame_shift %in% ID_missense)]
    return(c(length(ID_missense), length(ID_no_mutation), length(ID_frame_shift)));
})
names(df_smarca4_sampleStat) <- names(test_smarca4)
df_smarca4_sampleStat <- as.data.frame(df_smarca4_sampleStat)
df_smarca4_sampleStat$mutation <- c("missense", "no_mutation", "frame_shift")
freqPlot(melt(df_smarca4_sampleStat)[,c(2,1,3)]) + theme_bw()

df_smarca4_regression <- lapply(names(test_smarca4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="SMARCA4" &
                                                ((mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)))
                                              ),]$sample)
    # t <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
    #                                                mutation_info$gene=="SMARCA4" &
    #                                                ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
    #                                                mutation_info$effect=="Silent")
    #                                              ),]$sample)
    # ID_missense <- t[which(!t %in% ID_missense)]
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
    # set.seed(1); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
    
    mat$baseMean <- log2(rowMedians(as.matrix(mat)))
    # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
    mat <- mat[order(mat$baseMean),]
    mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
    mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
    
    # t <- t(apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
    #   scale(x)
    # }))

    mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
      t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
      colnames(t) <- c("expr", "class")
      t$class <- relevel(t$class, ref = "no_mutation")
      if(sd(t$expr) > 0) {
        # t$expr <- scale(t$expr)
        summary(glm(expr ~ class, data=t))$coefficients[2,1]
      } else { 0; }
    })

    cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
    mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
    mat$variable <- TISSUE
    return(mat)
})

df_smarca4_regression <- do.call(rbind, df_smarca4_regression)
df_smarca4_regression$gene <- "Smarca4"

TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="CHD4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")

df_chd4_sampleStat <- lapply(names(test_chd4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="CHD4" &
                                                mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    ID_frame_shift <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                            mutation_info$gene=="CHD4" &
                                            (mutation_info$effect=="Frame_Shift_Del" | mutation_info$effect=="Frame_Shift_Ins" |
                                            mutation_info$effect=="Splice_Site")
                                          ),]$sample)
    ID_frame_shift <- ID_frame_shift[which(!ID_frame_shift %in% ID_missense)]
    return(c(length(ID_missense), length(ID_no_mutation), length(ID_frame_shift)));
})
names(df_chd4_sampleStat) <- names(test_chd4)
df_chd4_sampleStat <- as.data.frame(df_chd4_sampleStat)
df_chd4_sampleStat$mutation <- c("missense", "no_mutation", "frame_shift")
freqPlot(melt(df_chd4_sampleStat)[,c(2,1,3)]) + theme_bw()

df_chd4_regression <- lapply(names(test_chd4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="CHD4" &
                                                ((mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)))
                                              ),]$sample)
    # t <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
    #                                                mutation_info$gene=="CHD4" &
    #                                                ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
    #                                                mutation_info$effect=="Silent")
    #                                              ),]$sample)
    # ID_missense <- t[which(!t %in% ID_missense)]
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
    # set.seed(1); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
    
    mat$baseMean <- log2(rowMedians(as.matrix(mat)))
    # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
    mat <- mat[order(mat$baseMean),]
    mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
    mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))

    
    mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
      t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
      colnames(t) <- c("expr", "class")
      t$class <- relevel(t$class, ref = "no_mutation")
      if(sd(t$expr) > 0) {
        # t$expr <- scale(t$expr)
        summary(glm(expr ~ class, data=t))$coefficients[2,1]
      } else { 0; }
    }) 
    
    cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
    mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
    mat$variable <- TISSUE
    return(mat)
})
df_chd4_regression <- do.call(rbind, df_chd4_regression)
df_chd4_regression$gene <- "Chd4"

df <- rbind(df_smarca4_regression, df_chd4_regression)

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4_CHD4.pdf", width=15, height = 5)
# df <- rbind(df_smarca4_regression[which(df_smarca4_regression$value<0),], df_chd4_regression[which(df_chd4_regression$value<0),])
ggarrange(ggline(df, x = "expr_class", y = "value", add=c("mean_se"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_smarca4_regression, x = "expr_class", y = "value", add=c("mean_se"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_chd4_regression, x = "expr_class", y = "value", add=c("mean_se"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"), nrow=1, ncol=3)
matrix2pheatmap(t(confounding_smarca4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
matrix2pheatmap(t(confounding_chd4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
dev.off()
```

###########################################
## Analysis for SMARCA4 and CHD4 (regression + sampling-based analysis)
###########################################
```{r}
# df_smarca4_sampling <- lapply(seq(1:100), function(SEED) {
#   do.call(rbind, lapply(names(test_smarca4), function(TISSUE) {
#       ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
#                                                   mutation_info$gene=="SMARCA4" &
#                                                   mutation_info$effect=="Missense_Mutation" &
#                                                   grepl("damaging", mutation_info$PolyPhen)
#                                                 ),]$sample)
#       ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
#                                                    ),]$sample)
#       ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
#   
#       SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#       SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#   
#       mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#       row.names(mat) <- gene_expr$sample
#       cat(SEED, TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#       colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#       mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#       set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
#       
#       mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#       mat <- mat[order(mat$baseMean),]
#       mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#       mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
#   
#       mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#         t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#         colnames(t) <- c("expr", "class")
#         t$class <- relevel(t$class, ref = "no_mutation")
#         if(sd(t$expr) > 0) {
#           # t$expr <- scale(t$expr)
#           summary(glm(expr ~ class, data=t))$coefficients[2,1]
#         } else { 0; }
#       })
#   
#       cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
#       mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
#       mat$variable <- TISSUE
#       mat$seed <- as.factor(SEED)
#       return(mat)
#   }))
# })
# df_smarca4_sampling <- lapply(seq(1:100), function(SEED) { df_smarca4_sampling[[SEED]]$seed <- as.factor(SEED); return(df_smarca4_sampling[[SEED]]); })
# write.table(do.call(rbind, df_smarca4_sampling), "~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4_sampling.txt", sep="\t", quote=F, row.names = F, col.names =T)
# ggline(do.call(rbind, df_smarca4_sampling), x = "expr_class", y = "value", add=c("median"), color = "seed") + 
#             geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_smarca4 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_smarca4_regression[which(as.numeric(df_smarca4_regression$expr_class) > 0),]$value, 
          df_smarca4_sampling[[SEED]][which(as.numeric(df_smarca4_sampling[[SEED]]$expr_class) > 0),]$value, alternative = "l")$p.value; })), method="BH")
length(which(pVal_smarca4 >= 0.05))/100

# df_chd4_sampling <- lapply(seq(1:100), function(SEED) {
#   do.call(rbind, lapply(names(test_chd4), function(TISSUE) {
#       ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
#                                                   mutation_info$gene=="CHD4" &
#                                                   mutation_info$effect=="Missense_Mutation" &
#                                                   grepl("damaging", mutation_info$PolyPhen)
#                                                 ),]$sample)
#       ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
#                                                    ),]$sample)
#       ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
#   
#       SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#       SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#   
#       mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#       row.names(mat) <- gene_expr$sample
#       cat(SEED, TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#       colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#       mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#       set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
#       
#       mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#       mat <- mat[order(mat$baseMean),]
#       mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#       mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
#   
#       mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#         t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#         colnames(t) <- c("expr", "class")
#         t$class <- relevel(t$class, ref = "no_mutation")
#         if(sd(t$expr) > 0) {
#           # t$expr <- scale(t$expr)
#           summary(glm(expr ~ class, data=t))$coefficients[2,1]
#         } else { 0; }
#       })
#   
#       cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
#       mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
#       mat$variable <- TISSUE
#       mat$seed <- as.factor(SEED)
#       return(mat)
#   }))
# })
# df_chd4_sampling <- lapply(seq(1:100), function(SEED) { df_chd4_sampling[[SEED]]$seed <- as.factor(SEED); return(df_chd4_sampling[[SEED]]); })
# write.table(do.call(rbind, df_chd4_sampling), "~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/CHD4_sampling.txt", sep="\t", quote=F, row.names = F, col.names =T)
# ggline(do.call(rbind, df_chd4_sampling), x = "expr_class", y = "value", add=c("median"), color = "seed") +
#             geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_chd4 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_chd4_regression[which(as.numeric(df_chd4_regression$expr_class) > 0),]$value, 
          df_chd4_sampling[[SEED]][which(as.numeric(df_chd4_sampling[[SEED]]$expr_class) > 0),]$value, alternative = "g")$p.value; })), method="BH")
length(which(pVal_chd4 >= 0.05))/100
```

###########################################
## Analysis for SMARCA4 and CHD4 (regression-based analysis - irrespective to TISSUE)
###########################################
```{r}
## SMARCA4
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="SMARCA4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")

ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4)) &
                                            mutation_info$gene=="SMARCA4" &
                                            ((mutation_info$effect=="Missense_Mutation" &
                                            grepl("damaging", mutation_info$PolyPhen)))
                                          ),]$sample)
ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4))
                                          ),]$sample)
ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))

mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
row.names(mat) <- gene_expr$sample
cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]

mat$baseMean <- log2(rowMedians(as.matrix(mat)))
mat <- mat[order(mat$baseMean),]
mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))

mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
  t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
  colnames(t) <- c("expr", "class")
  t$class <- relevel(t$class, ref = "no_mutation")
  if(sd(t$expr) > 0) {
    # t$expr <- scale(t$expr)
    summary(glm(expr ~ class, data=t))$coefficients[2,1]
  } else { 0; }
})

cat(length(table(mat$expr_class))); cat("\n");
df_smarca4_regression_v2 <- mat
df_smarca4_regression_v2$gene <- "Smarca4"

## CHD4
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="CHD4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")

ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4)) &
                                            mutation_info$gene=="CHD4" &
                                            ((mutation_info$effect=="Missense_Mutation" &
                                            grepl("damaging", mutation_info$PolyPhen)))
                                          ),]$sample)
ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4))
                                          ),]$sample)
ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))

mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
row.names(mat) <- gene_expr$sample
cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]

mat$baseMean <- log2(rowMedians(as.matrix(mat)))
mat <- mat[order(mat$baseMean),]
mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))

mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
  t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
  colnames(t) <- c("expr", "class")
  t$class <- relevel(t$class, ref = "no_mutation")
  if(sd(t$expr) > 0) {
    # t$expr <- scale(t$expr)
    summary(glm(expr ~ class, data=t))$coefficients[2,1]
  } else { 0; }
})

cat(length(table(mat$expr_class))); cat("\n");
df_chd4_regression_v2 <- mat
df_chd4_regression_v2$gene <- "chd4"

df_smarca4_regression_v2$expr_class <- cut2(seq(1, nrow(df_smarca4_regression_v2)), g=50, levels.mean=T)

df_chd4_regression_v2$expr_class <- cut2(seq(1, nrow(df_chd4_regression_v2)), g=50, levels.mean=T)

df <- rbind(df_smarca4_regression_v2[,c("expr_class", "coefficients", "gene")], df_chd4_regression_v2[,c("expr_class", "coefficients", "gene")])

# df <- rbind(df_smarca4_regression[which(df_smarca4_regression$value<0),], df_chd4_regression[which(df_chd4_regression$value<0),])
ggarrange(ggline(melt(df), x = "expr_class", y = "value", add=c("mean_se"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(melt(df_smarca4_regression_v2[,c("expr_class", "coefficients")]), x = "expr_class", y = "value", add=c("mean_se"), 
                 color = "variable", palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(melt(df_chd4_regression_v2[,c("expr_class", "coefficients")]), x = "expr_class", y = "value", add=c("mean_se"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"), nrow=1, ncol=3)

## SMARCA4 sampling
# df_smarca4_sampling_v2 <- lapply(seq(1:100), function(SEED) {
#   ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4)) &
#                                               mutation_info$gene=="SMARCA4" &
#                                               ((mutation_info$effect=="Missense_Mutation" &
#                                               grepl("damaging", mutation_info$PolyPhen)))
#                                             ),]$sample)
#   ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4))
#                                             ),]$sample)
#   ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
#   
#   SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
#   SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))
#   
#   mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#   row.names(mat) <- gene_expr$sample
#   cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#   colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#   mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#   set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
# 
#   mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#   mat <- mat[order(mat$baseMean),]
#   mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#   mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
# 
#   mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#     t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#     colnames(t) <- c("expr", "class")
#     t$class <- relevel(t$class, ref = "no_mutation")
#     if(sd(t$expr) > 0) {
#       # t$expr <- scale(t$expr)
#       summary(glm(expr ~ class, data=t))$coefficients[2,1]
#     } else { 0; }
#   })
# 
#   cat("\t"); cat(length(table(mat$expr_class))); cat("\n");
#   mat <- mat[,c("baseMean", "expr_class", "coefficients")]
#   mat$seed <- as.factor(SEED)
#   return(mat)
# })
# df_smarca4_sampling_v2 <- lapply(seq(1:100), function(SEED) { df_smarca4_sampling_v2[[SEED]]$seed <- as.factor(SEED); return(df_smarca4_sampling_v2[[SEED]]); })
ggline(do.call(rbind, df_smarca4_sampling_v2), x = "expr_class", y = "coefficients", add=c("median"), color = "seed") +
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_smarca4_v2 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_smarca4_regression_v2[which(as.numeric(df_smarca4_regression_v2$expr_class) > 0),]$coefficients, 
          df_smarca4_sampling_v2[[SEED]][which(as.numeric(df_smarca4_sampling_v2[[SEED]]$expr_class) > 0),]$coefficients, alternative = "l")$p.value; })), method="BH")
length(which(pVal_smarca4_v2 >= 0.05))/100

## CHD4 sampling
# df_chd4_sampling_v2 <- lapply(seq(1:100), function(SEED) {
#   ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4)) &
#                                               mutation_info$gene=="CHD4" &
#                                               ((mutation_info$effect=="Missense_Mutation" &
#                                               grepl("damaging", mutation_info$PolyPhen)))
#                                             ),]$sample)
#   ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4))
#                                             ),]$sample)
#   ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
# 
#   SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
#   SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))
# 
#   mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#   row.names(mat) <- gene_expr$sample
#   cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#   colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#   mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#   set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
# 
#   mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#   mat <- mat[order(mat$baseMean),]
#   mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#   mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
# 
#   mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#     t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#     colnames(t) <- c("expr", "class")
#     t$class <- relevel(t$class, ref = "no_mutation")
#     if(sd(t$expr) > 0) {
#       # t$expr <- scale(t$expr)
#       summary(glm(expr ~ class, data=t))$coefficients[2,1]
#     } else { 0; }
#   })
# 
#   cat("\t"); cat(length(table(mat$expr_class))); cat("\n");
#   mat <- mat[,c("baseMean", "expr_class", "coefficients")]
#   mat$seed <- as.factor(SEED)
#   return(mat)
# })
# df_chd4_sampling_v2 <- lapply(seq(1:100), function(SEED) { df_chd4_sampling_v2[[SEED]]$seed <- as.factor(SEED); return(df_chd4_sampling_v2[[SEED]]); })
ggline(do.call(rbind, df_chd4_sampling_v2), x = "expr_class", y = "coefficients", add=c("median"), color = "seed") +
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_chd4_v2 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_chd4_regression_v2[which(as.numeric(df_chd4_regression_v2$expr_class) > 0),]$coefficients, 
          df_chd4_sampling_v2[[SEED]][which(as.numeric(df_chd4_sampling_v2[[SEED]]$expr_class) > 0),]$coefficients, alternative = "g")$p.value; })), method="BH")
length(which(pVal_chd4_v2 >= 0.05))/100
```

########################################
## refined analysis on individual mutations
########################################
```{r}
# Stomach adenocarcinoma (G782S); Lung adenocarcinoma (G784V); Skin Cutaneous Melanoma (K785R, E861K, E882K); Bladder Urothelial Carcinoma (T786N, E882D); Lung squamous cell carcinoma (E861K); Kidney renal papillary cell carcinoma (E882D); Breast Invasive Carcinoma (E882K); Brain Lower Grade Glioma (H884N); Kidney renal clear cell carcinoma (H884R)
MUTATION <- c("T786N", "E882K", "E882D", "G784V", "E861K", "H884R")
TISSUE <- c("Bladder Urothelial Carcinoma", "Breast invasive carcinoma", "Kidney renal papillary cell carcinoma", "Lung adenocarcinoma", "Lung squamous cell carcinoma", "Kidney renal clear cell carcinoma")
p1 <- lapply(seq(1,6), function(i) {
  #grepl(MUTATION[i], mutation_info$Amino_Acid_Change) &
  ID <- unique(mutation_info[which(grepl(MUTATION[i], mutation_info$Amino_Acid_Change) & 
                                     mutation_info$tissueSource==TISSUE[i] &
                                     mutation_info$effect=="Missense_Mutation" & 
                                     mutation_info$gene=="SMARCA4"),]$sample)
  SAMPLE_INTEREST_MUTATION <- as.data.frame(gene_expr[,unique(mutation_info[which(grepl(MUTATION[i], mutation_info$Amino_Acid_Change) &
                                                                                    mutation_info$tissueSource==TISSUE[i] &
                                                                                    mutation_info$effect=="Missense_Mutation" & 
                                                                                    mutation_info$gene=="SMARCA4"),]$sample)])
  colnames(SAMPLE_INTEREST_MUTATION) <- rep("interest_mutation", ncol(SAMPLE_INTEREST_MUTATION))
  
  
  SAMPLE_OTHER_MUTATION <- gene_expr[,unique(mutation_info[which(!grepl(MUTATION[i], mutation_info$Amino_Acid_Change) &
                                                                   mutation_info$tissueSource==TISSUE[i] &
                                                                   mutation_info$effect=="Missense_Mutation" & 
                                                                   mutation_info$gene=="SMARCA4"),]$sample)]
  #SAMPLE_OTHER_MUTATION <- as.data.frame(SAMPLE_OTHER_MUTATION[,grep(paste(unlist(lapply(strsplit(ID, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_OTHER_MUTATION), value=TRUE)])
  colnames(SAMPLE_OTHER_MUTATION) <- rep("other_mutation", ncol(SAMPLE_OTHER_MUTATION))
  
  SAMPLE_NO_MUTATION <- gene_expr[,unique(mutation_info[which(mutation_info$tissueSource==TISSUE[i] &
                                                                mutation_info$effect=="Missense_Mutation" &
                                                                mutation_info$gene!="SMARCA4"),]$sample)]
  SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
  colnames(SAMPLE_NO_MUTATION) <- rep("no_mutation", ncol(SAMPLE_NO_MUTATION))
  
  SAMPLE_NORMAL <- gene_expr[,grep(paste(unlist(lapply(strsplit(ID, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(gene_expr), value=TRUE)]
  colnames(SAMPLE_NORMAL) <- rep("normal", ncol(SAMPLE_NORMAL))
  # test <- cbind(SAMPLE_INTEREST_MUTATION, SAMPLE_OTHER_MUTATION, SAMPLE_NO_MUTATION, SAMPLE_NORMAL)
  test <- cbind(SAMPLE_INTEREST_MUTATION, SAMPLE_NO_MUTATION)
  table(colnames(test))
  # test <- as.data.frame(normalize.quantiles(as.matrix(test)))
  # colnames(test) <- colnames(cbind(SAMPLE_INTEREST_MUTATION, SAMPLE_OTHER_MUTATION, SAMPLE_NO_MUTATION))
  test$expr_class <- cut2(rowMeans(test), g=3)
  table(test$expr_class)
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  ggplot(test.melt, aes(expr_class, (value), fill=variable)) + geom_boxplot(notch = F) + theme_bw() + ggtitle(TISSUE[i]) +
    scale_fill_manual(values=c("#e31a1c", "#1f78b4")) + ylab("Gene expression")
})

MUTATION <- c("T786N", "E882K", "E882D", "G784V", "E861K", "H884R")
TISSUE <- c("Bladder Urothelial Carcinoma", "Breast invasive carcinoma", "Kidney renal papillary cell carcinoma", "Lung adenocarcinoma", "Lung squamous cell carcinoma", "Kidney renal clear cell carcinoma")
p1 <- lapply(seq(1,6), function(i) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE[i], ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMean <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMean),]
    mat$expr_class <- cut2(mat$baseMean, g=3)
    mat <- mat[,which(!grepl("baseMean", colnames(mat)))]
    colnames(mat) <- gsub("\\..*", "", colnames(mat))
    table(mat$expr_class)
    ggplot(melt(mat), aes(expr_class, log(value), fill=variable)) + geom_boxplot(notch = F) + theme_bw() + ggtitle(TISSUE[i]) +
      scale_fill_manual(values=c("#e31a1c", "#1f78b4")) + ylab("Gene expression")
})
```

