---
title: "R Notebook for Chd4 and Brg1 project (author: sachin pundhir)"
output:
  pdf_document: default
  html_notebook: default
  html_document:
    df_print: paged
    fig_caption: yes
editor_options:
  chunk_output_type: console
---

```{r global options, include = F}
knitr::opts_chunk$set(echo=F, include = T, results="hide", warning=FALSE, message=FALSE, tidy = TRUE, tidy.opts = list(comment = FALSE),
                      fig.width=12, fig.height=8, out.width = "100%", out.height = "70%")
## edit a R function
## trace("makevp.eqsc",edit=TRUE) 
```

```{r cachecontrol, cache = TRUE}
knitr::opts_chunk$set(cache.rebuild = TRUE)
```

######################################################
#### LOAD LIBRARIES
######################################################
```{r setup}
library(DESeq2)
library(preprocessCore)
library(reshape)
library(ggplot2)
library(doBy)
library(ggpubr)
library(gridExtra)
library(Hmisc)
library(plyr)
library(pheatmap)
library(Gviz)
library(rlist)
library(GenomicRanges)
library(GenomicFeatures)
library(TxDb.Mmusculus.UCSC.mm9.knownGene)
library(rtracklayer)
library(trackViewer)
library(plyr)
library(ComplexHeatmap)
theme_set(theme_classic2())
source("~/software/myScripts/R-scripts/FUNCTIONS.R")
```

###########################################
## methods
<!-- TCGA cohorts correlation analyses -->

<!-- The phenotype dataset (with survival outcomes) and the RNAseq dataset of each cancer cohort (the gene-level transcription estimates, as in log2(x+1) transformed RSEM normalized count) were downloaded from UCSC Xena (https://xenabrowser.net/) where genes are mapped onto the human genome coordinates using UCSC Xena HUGO probeMap in the RNAseq dataset. Spearman correlation coefficient was calculated between TREM2 and each of the genes in each cancer cohort, accompanied with sample size and p values. The individual scatterplot of each of the top genes with TREM2 (with fitted linear lines) was generated with Spearman correlation in the plot. -->
<!-- The Kaplan-Meier curve for overall survival (OS) and relapse free survival (RFS) was generated for TREM2 dichotomized by 75% quantile of TREM2 expression into high/low TREM2 expression group. The log-rank test was applied to test the survival difference between high/low TREM2 expression. -->
###########################################

###########################################
## load TCGA data
## https://shixiangwang.github.io/home/en/post/ucscxenatools-201908/
###########################################
```{r}
#gene_expr <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_gene_expected_count.gz", header=T, check.names=FALSE)

ensembl_protein_coding <- read.table(pipe("grep protein_coding /scratch/genomes/annotations/BED/hg19_ensembl_gene.bed"))
gene_expr_PC <- gene_expr[(gsub("\\..*", "", gene_expr$sample) %in% ensembl_protein_coding$V7),]

tissue_barCode <- read.table(pipe("less ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/code_table/tissueSourceSite.tsv | cut -f 1,3"), header=T, sep="\t")

mutation_info <- read.table(pipe("zcat ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_mutation_phenotype.txt.gz | cut -f 1,7,8,9,12,13"), header=T, check.names=FALSE, sep="\t", stringsAsFactors=FALSE, quote="\"")

survival_info <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TCGA_survival_data.gz", header=T, check.names=FALSE, sep="\t", stringsAsFactors=FALSE, quote="\"")

#sample_info <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGTEX_phenotype.txt.gz", header=T, sep="\t", stringsAsFactors=FALSE)
sample_info <- read.table(pipe("less ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/clinical_data/clinical.tsv | cut -f 1,2,3,4,9,11,12,15,16,20,25,26,27,28,29,41,47,48,68,80,96,108,110,111,112,116,119,120,125,128,152,154 | cut -f 2,4,6,7,8,24,28,30,31"), header=T, sep="\t", stringsAsFactors=FALSE, check.names=FALSE, quote="\"")
colnames(sample_info)[1] <- "sample"
sample_info <- unique(sample_info)

smarca4_info <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/41594_2017_7_MOESM6_ESM.txt", sep="\t", header=T, stringsAsFactors=FALSE)
smarca4_info[which(smarca4_info$Study=="Testicular Germ Cell Cancer"),]$Study <- "Testicular Germ Cell Tumors"

chd4_info <- read.table(pipe("zcat ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_mutation_phenotype.txt.gz | cut -f 1,7,8,13 | grep -w CHD4"), header=F, check.names=FALSE, sep="\t", stringsAsFactors=FALSE, quote="\"")
colnames(chd4_info) <- c("sample", "gene", "effect", "tissueSource")
chd4_info[which(chd4_info$effect!="Missense_Mutation"),]$effect <- "other"

t <- mutation_info[which(grepl("Uterine Corpus Endometrial Carcinoma|Colon adenocarcinoma|Stomach adenocarcinoma", mutation_info$tissueSource) &
                      mutation_info$gene!="CHD4"),]
t$gene <- "Not_CHD4"
t$effect <- "No_mutation"
t <- unique(t[,c(1,2,3,6)])
t <- t[which(!(t$sample %in% chd4_info$sample)),]
#length(grep(paste(unlist(lapply(strsplit(chd4_info[which(chd4_info$tissueSource=="Uterine Corpus Endometrial Carcinoma"),]$sample, "-"), function(x) sprintf("TCGA-%s-.*$", x[2]))), collapse="|"), t$sample))
chd4_info <- rbind(chd4_info, t)
table(chd4_info$effect)

gtex_info <- read.table(pipe("cat ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt | cut -f 1,6,7"), sep="\t", header=T)
```

###########################################
## Analysis for SMARCA4 (confounding factors)
###########################################
```{r}
GENE="SMARCA4"
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene==GENE),c("sample", "tissueSource")])$tissueSource))

gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
if(GENE=="CHD4") {
  names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")
} else{
  names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")
}

LIST <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
LIST <- LIST[stringr::str_to_title(LIST) %in% stringr::str_to_title(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study))]

## test for confounding factors (phenotype)
confounding_smarca4 <- lapply(LIST, function(TISSUE) {
  unlist(lapply(colnames(sample_info)[2:ncol(sample_info)], function(FACTOR) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 # &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    
    info_missense <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_missense), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_missense$class <- "missense"
    info_no_mutation <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_no_mutation), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_no_mutation$class <- "no_mutation"
    info_test <- rbind(info_missense, info_no_mutation)
    info_test$class <- as.factor(info_test$class)
    cat(TISSUE, length(ID_no_mutation), length(ID_missense), length(info_test$class)); cat("\n");

    if(length(unique(info_test[,FACTOR]))>1) {
      if(FACTOR=="age_at_index") {
        summary(glm(as.numeric(as.character(info_test[,FACTOR])) ~ info_test$class, data=info_test))$coefficients[2,4]
      } else {
        chisq.test(table(info_test[,c(3,2)]))$p.value
      }
    } else {
      1
    }
  }))
})
names(confounding_smarca4) <- LIST
confounding_smarca4 <- as.data.frame(confounding_smarca4)
row.names(confounding_smarca4) <- colnames(sample_info)[2:ncol(sample_info)]
confounding_smarca4$factor <- colnames(sample_info)[2:ncol(sample_info)]
apply(confounding_smarca4[,c(1:ncol(confounding_smarca4)-1)], 2, function(x) length(which(x<0.01)))

## test for confounding factors (genotype)
confounding_smarca4_genotypes <- lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 # &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    
    info_missense <- unique(mutation_info[which(grepl(paste(ID_missense, collapse="|"), mutation_info$sample) &
                                           grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    info_no_mutation <- unique(mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) &
                                        grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    unlist(lapply(attributes(Vennerable::Venn(list(
      A=paste(as.data.frame(table(info_no_mutation$gene))[which(as.data.frame(table(info_no_mutation$gene))$Freq >= 
                                                            length(ID_no_mutation)),]$Var1, collapse=","),
      B=paste(as.data.frame(table(info_missense$gene))[which(as.data.frame(table(info_missense$gene))$Freq >= 
                                                         length(ID_missense)),]$Var1, collapse=","))))$IntersectionSets,
           function(x) ifelse(x=="", "NA", x)))
})
names(confounding_smarca4_genotypes) <- LIST
confounding_smarca4_genotypes <- as.data.frame(confounding_smarca4_genotypes)

## test for survival difference
survival_pVal <- lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    cat(TISSUE); cat("\n")
    wilcox.test(survival_info[grep(paste(ID_missense, collapse="|"), survival_info$sample),]$OS.time, survival_info[grep(paste(ID_no_mutation, collapse="|"), survival_info$sample),]$OS.time)$p.value
})
names(survival_pVal) <- LIST
survival_info <- t(as.data.frame(survival_pVal))
```

###########################################
## Analysis for SMARCA4
###########################################
```{r}
TISSUE_STUDY <- unlist(lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    cat(TISSUE, length(ID_no_mutation), length(ID_missense)); cat("\n");
    if(length(ID_missense)>0 & length(ID_no_mutation)>0) { return(TISSUE); }
}))

test_smarca4 <- lapply(TISSUE_STUDY, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    #SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID_missense, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMeanCustom <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMeanCustom),]
    mat$expr_class <- cut2(mat$baseMeanCustom, g=9)
    #mat$expr_class <- cut2(c(1:nrow(mat)), g=5)

    ## DESeq2 analysis
    test <- mat[,grep("mutation", colnames(mat))]
    colnames(test) <- gsub("\\..*", "", colnames(test))
    colnames(test) <- gsub("no_mutation", "wt", colnames(test))
    colnames(test) <- gsub("mutation", "ko", colnames(test))
    test <- matrix2deseq2(countTable = test, treatment = colnames(test))
    row.names(test) <- test$gene
    res <- merge(test, mat[,c("baseMeanCustom", "expr_class")], by="row.names")
    colnames(res) <- gsub("\\..*", "", colnames(res))
    return(res)
})
names(test_smarca4) <- TISSUE_STUDY

p_smarca4 <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  # plot(log2(test$baseMeanCustom), test$log2FoldChange)
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  # test$expr_class <- cut2(1:nrow(test), g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  #cat(ncol(SAMPLE_MUTATION), ncol(SAMPLE_NO_MUTATION))

  test.melt <- melt(test)
  cat(TISSUE); cat("\n")
  #test.melt$value <- log2(test.melt$value)
  # ggline(test.melt, x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
  #   theme_bw() +
  #   ggtitle(TISSUE) +
  #   scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
  #   ylab("Gene expr.") + xlab("") +
  #   theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
    #scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test_GENE[[TISSUE]]$expr_class), function(x) mean(test_GENE[[TISSUE]][which(test_GENE[[TISSUE]]$expr_class==x),]$baseMean)))))

  ggboxplot(test.melt, x = "expr_class", y = "value", fill = "variable", yscale = "log2",
            #notch=1, notchwidth = 0.6, yscale="log2",
            palette = c("#e31a1c", "#1f78b4"), outlier.shape=NA) +
    xlab("") +
    ylab("Gene exor. (log2)") + ggtitle(TISSUE)
  })

pVal_smarca4 <- unlist(lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  wilcox.test(test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="ko_RAW"),]$value,
              test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="wt_RAW"),]$value, alternative="g")$p.value
}))

p_smarca4_fc <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  test$expr_class <- cut2(test$baseMeanCustom, g=9)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  test.melt <-summaryBy(value ~ variable + expr_class, data = test.melt, FUN = list(median))
  log2(test.melt[which(test.melt$variable=="ko_RAW"),]$value.median/test.melt[which(test.melt$variable=="wt_RAW"),]$value.median)
})
names(p_smarca4_fc) <- names(test_smarca4)
p_smarca4_fc <- as.data.frame(p_smarca4_fc)
p_smarca4_fc$class <- as.factor(row.names(p_smarca4_fc))
p_smarca4_fc <- melt(p_smarca4_fc)

p_smarca4_deseq2 <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[order(test$baseMeanCustom),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  ggline(melt(test[,c("log2FoldChange", "expr_class")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
    theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c")) +
    ylab("Gene expr. (Log2FC)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
})

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4.pdf", width=20)
ggarrange(plotlist = p_smarca4, nrow=2, ncol=8)
ggarrange(plotlist = p_smarca4_deseq2, nrow=2, ncol=8)
dev.off()

# expr_smarca4 <- lapply(TISSUE_STUDY, function(TISSUE) {
#   cat(TISSUE); cat("\n")
#   sample <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]$sample)[unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]$sample) %in% colnames(gene_expr)]
#   rowMeans(gene_expr[grep("ENSG00000127616", gene_expr$sample), sample])
# })
# names(expr_smarca4) <- TISSUE_STUDY
# expr_smarca4 <- as.data.frame(t(as.data.frame(expr_smarca4)))
# expr_smarca4$sample <- row.names(expr)
```

###########################################
## Analysis for CHD4 
###########################################
```{r}
GENE="CHD4"
# TISSUE_STAT <- as.data.frame(table(mutation_info[which(mutation_info$gene==GENE),]$tissueSource))
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene==GENE),c("sample", "tissueSource")])$tissueSource))

gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
if(GENE=="CHD4") {
  names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")
} else{
  names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")
}

LIST <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)

## test for confounding factors (phenotype)
confounding_chd4 <- lapply(LIST, function(TISSUE) {
  unlist(lapply(colnames(sample_info)[2:ncol(sample_info)], function(FACTOR) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 # &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="CHD4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    info_missense <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_missense), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_missense$class <- "missense"
    info_no_mutation <- unique(sample_info[grep(paste(gsub("\\-[0-9]+$", "", ID_no_mutation), collapse = "|"), sample_info$sample),c("sample", FACTOR)])
    info_no_mutation$class <- "no_mutation"
    info_test <- rbind(info_missense, info_no_mutation)
    info_test$class <- as.factor(info_test$class)
    cat(TISSUE, length(ID_no_mutation), length(ID_missense), length(info_test$class)); cat("\n");
    
    if(length(unique(info_test[,FACTOR]))>1) {
      if(FACTOR=="age_at_index") {
        summary(glm(as.numeric(as.character(info_test[,FACTOR])) ~ info_test$class, data=info_test))$coefficients[2,4]
        # wilcox.test(as.numeric(as.character(info_test[which(info_test$class=="missense"),FACTOR])), as.numeric(as.character(info_test[which(info_test$class=="no_mutation"),FACTOR])))$p.value
      } else {
        chisq.test(table(info_test[,c(3,2)]))$p.value
      }
    } else {
      1
    }
    # info_missense <- mutation_info[grep(paste(ID_missense, collapse="|"), mutation_info$sample),c("sample", "gene")]
    # info_missense$class <- "missense"
    # info_no_mutation <- mutation_info[grep(paste(ID_no_mutation, collapse="|"), mutation_info$sample),c("sample", "gene")]
    # info_no_mutation$class <- "no_mutation"
    # info_test <- rbind(info_missense, info_no_mutation)
    # info_test$class <- as.factor(info_test$class)
  }))
})
names(confounding_chd4) <- LIST
confounding_chd4 <- as.data.frame(confounding_chd4)
row.names(confounding_chd4) <- colnames(sample_info)[2:ncol(sample_info)]
confounding_chd4$factor <- colnames(sample_info)[2:ncol(sample_info)]
apply(confounding_chd4[,c(1:ncol(confounding_chd4)-1)], 2, function(x) length(which(x<0.01)))

## test for confounding factors (genotype)
confounding_chd4_genotypes <- lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                #  &
                                                   # mutation_info$gene==GENE &
                                                   # ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   # mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="SMARCA4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    
    info_missense <- unique(mutation_info[which(grepl(paste(ID_missense, collapse="|"), mutation_info$sample) &
                                           grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    info_no_mutation <- unique(mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) &
                                        grepl("damaging", mutation_info$PolyPhen)),c("sample", "gene")])
    unlist(lapply(attributes(Vennerable::Venn(list(
      A=paste(as.data.frame(table(info_no_mutation$gene))[which(as.data.frame(table(info_no_mutation$gene))$Freq >= 
                                                            length(ID_no_mutation)),]$Var1, collapse=","),
      B=paste(as.data.frame(table(info_missense$gene))[which(as.data.frame(table(info_missense$gene))$Freq >= 
                                                         length(ID_missense)),]$Var1, collapse=","))))$IntersectionSets,
           function(x) ifelse(x=="", "NA", x)))
})
names(confounding_chd4_genotypes) <- LIST
confounding_chd4_genotypes <- as.data.frame(confounding_chd4_genotypes)
```

###########################################
## Analysis for CHD4
###########################################
```{r}
GENE="CHD4"
TISSUE_STAT <- as.data.frame(table(mutation_info[which(mutation_info$gene==GENE),]$tissueSource))

gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
if(GENE=="CHD4") {
  names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")
} else{
  names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")
}

LIST <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)

TISSUE_STUDY <- unlist(lapply(LIST, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)

    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    
    ## TEST
    #mutation_info[which(grepl(paste(ID_no_mutation, collapse="|"), mutation_info$sample) & mutation_info$gene=="CHD4" & stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE)),]
    cat(TISSUE, length(ID_no_mutation), length(ID_missense)); cat("\n");
    if(length(ID_missense)>0 & length(ID_no_mutation)>0) { return(TISSUE); }
}))

test_chd4 <- lapply(TISSUE_STUDY, function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    #SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID_missense, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMeanCustom <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMeanCustom),]
    mat$expr_class <- cut2(mat$baseMeanCustom, g=9)
    #mat$expr_class <- cut2(c(1:nrow(mat)), g=5)

    ## DESeq2 analysis
    test <- mat[,grep("mutation", colnames(mat))]
    colnames(test) <- gsub("\\..*", "", colnames(test))
    colnames(test) <- gsub("no_mutation", "wt", colnames(test))
    colnames(test) <- gsub("mutation", "ko", colnames(test))
    test <- matrix2deseq2(countTable = test, treatment = colnames(test))
    row.names(test) <- test$gene
    res <- merge(test, mat[,c("baseMeanCustom", "expr_class")], by="row.names")
    colnames(res) <- gsub("\\..*", "", colnames(res))
    return(res)
})
names(test_chd4) <- TISSUE_STUDY

p_chd4 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[(gsub("\\..*", "", test$gene) %in% ensembl_protein_coding$V7),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  #cat(ncol(SAMPLE_MUTATION), ncol(SAMPLE_NO_MUTATION))

  test.melt <- melt(test)
  #test.melt$value <- log2(test.melt$value)
  # ggline(test.melt, x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
  #   theme_bw() +
  #   ggtitle(TISSUE) +
  #   scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
  #   ylab("Gene expr.") + xlab("") +
  #   theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
    #scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test_GENE[[TISSUE]]$expr_class), function(x) mean(test_GENE[[TISSUE]][which(test_GENE[[TISSUE]]$expr_class==x),]$baseMean)))))

  ggboxplot(test.melt, x = "expr_class", y = "value", fill = "variable", yscale = "log2",
            #notch=1, notchwidth = 0.6, yscale="log2",
            palette = c("#e31a1c", "#1f78b4"), outlier.shape=NA) +
    xlab("") +
    ylab("Gene exor. (log2)") + ggtitle(TISSUE)
  })

pVal_chd4 <- unlist(lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[(gsub("\\..*", "", test$gene) %in% ensembl_protein_coding$V7),]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  wilcox.test(test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="ko_RAW"),]$value,
              test.melt[which(test.melt$expr_class==names(table(test.melt$expr_class)[length(unique(test.melt$expr_class))]) &
                                test.melt$variable=="wt_RAW"),]$value, alternative="l")$p.value
}))

p_chd4_fc <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[(gsub("\\..*", "", test$gene) %in% ensembl_protein_coding$V7),]
  test$expr_class <- cut2(test$baseMeanCustom, g=9)
  test <- test[,grep("RAW|expr_class", colnames(test))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  test.melt <-summaryBy(value ~ variable + expr_class, data = test.melt, FUN = list(median))
  log2(test.melt[which(test.melt$variable=="ko_RAW"),]$value.median/test.melt[which(test.melt$variable=="wt_RAW"),]$value.median)
})
names(p_chd4_fc) <- names(test_chd4)
p_chd4_fc <- as.data.frame(p_chd4_fc)
p_chd4_fc$class <- as.factor(row.names(p_chd4_fc))
p_chd4_fc <- melt(p_chd4_fc)

p_chd4_deseq2 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  test$expr_class <- cut2(test$baseMeanCustom, g=3)
  ggline(melt(test[,c("log2FoldChange", "expr_class")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
    theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c")) +
    ylab("Gene expr. (Log2FC)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
})

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/CHD4.pdf", width=20)
ggarrange(plotlist = p_chd4, nrow=2, ncol=8)
ggarrange(plotlist = p_chd4_deseq2, nrow=2, ncol=8)
dev.off()
```

########################################
## summary plots for SMARCA4 and CHD4
########################################
```{r}
t <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  # test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # ggplot(test_smarca4[[TISSUE]][,c("expr_class", "baseMeanCustom")], aes(expr_class, baseMeanCustom)) + geom_boxplot()
  # boxplot(test_smarca4[[TISSUE]]$baseMeanCustom)
  # table(test_smarca4[[TISSUE]]$expr_class)
  # test <- test[which(test$baseMeanCustom > 3.97e-05 & test$baseMeanCustom < 7.04e+00),]
  test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean=T)
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  summaryBy(log2FoldChange ~ expr_class, data = test, FUN = list(median))$log2FoldChange.median
})
t <- as.data.frame(t)
colnames(t) <- names(test_smarca4)
t$expr_class <- as.factor(row.names(t))

t1 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  # test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # test <- test[which(test$baseMeanCustom > 3.97e-05 & test$baseMeanCustom < 7.04e+00),]
  test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean = T)
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  summaryBy(log2FoldChange ~ expr_class, data = test, FUN = list(median))$log2FoldChange.median
})
t1 <- as.data.frame(t1)
colnames(t1) <- names(test_chd4)
t1$expr_class <- as.factor(row.names(t1))

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4_CHD4.pdf", width=15, height = 5)
ggplot(test[,c("expr_class", "baseMeanCustom")], aes(expr_class, baseMeanCustom)) + geom_boxplot() + theme_bw()

ggarrange(ggplot(p_smarca4_fc, aes(as.numeric(class), value)) + geom_point(aes(color=variable, size = 1)) +
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 2), alpha=0.1) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            ylab("FC in median gene expr. (Log2)"),
          ggplot(p_chd4_fc, aes(as.numeric(class), value)) + geom_point(aes(color=variable, size = 1)) +
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 2), alpha=0.1) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            ylab("FC in median gene expr. (Log2)"), 
          nrow=1, ncol=2)

ggarrange(ggplot(melt(t), aes(as.numeric(expr_class) , value)) + geom_point(aes(color=variable, size = 1)) + 
            geom_smooth(method = "glm", se = T, formula = y ~ poly(x, 1), span=0.2, alpha=0.2) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))), 
          ggplot(melt(t1), aes(as.numeric(expr_class) , value)) + geom_point(aes(color=variable, size = 1)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 1), alpha=0.2) + theme_bw() +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))), nrow=1, ncol=2)

lm(as.numeric(melt(t)$expr_class) ~ melt(t)$value)$coefficients[2]
summary(lm(as.numeric(melt(t)$expr_class) ~ melt(t)$value))$coefficients[2, 4]

lm(as.numeric(melt(t1)$expr_class) ~ melt(t1)$value)$coefficients[2]
summary(lm(as.numeric(melt(t1)$expr_class) ~ melt(t1)$value))$coefficients[2, 4]

t.melt <- melt(t)
t.melt$gene <- "Smarca4"
t1.melt <- melt(t1)
t1.melt$gene <- "Chd4"
df <- rbind(t.melt, t1.melt)
ggarrange(ggline(df, x = "expr_class", y = "value", add=c("mean"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)"),
          ggplot(df, aes(as.numeric(expr_class), value, color=gene)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 2), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c("#784b97", "#ffa054")), nrow=1, ncol=2)

TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="SMARCA4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")

df_smarca4 <- lapply(names(test_smarca4)[-5], function(TISSUE) {
  test <- test_smarca4[[TISSUE]]
  test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  # test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean=T)

  # test$baseMean <- log2(test$baseMean)
  # test$baseMean <- log2(rowMedians(as.matrix(test[,grep("RAW", colnames(test))])))
  ID_BASELINE <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
  test <- merge(test[,c("gene", "baseMean", "log2FoldChange")], gene_expr_PC[,c("sample", grep(paste(ID_BASELINE, collapse="|"), colnames(gene_expr_PC), value=TRUE))], by.x="gene", by.y="sample")
  test$baseMean <- rowMedians(as.matrix(test[,grep("GTEX", colnames(test))]))
  
  # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
  test <- test[order(test$baseMean),]
  test$expr_class <- cut2(seq(1, nrow(test)), g=50, levels.mean=T)
  # test$expr_class <- as.numeric(as.character(test$expr_class))
  # test <- test[which(test$expr_class > summary(test$expr_class)[2] & test$expr_class < summary(test$expr_class)[5]),]
  # test$expr_class <- as.factor(test$expr_class)
  test$expr_class <- mapvalues(test$expr_class, from = levels(test$expr_class), to = seq(1, length(unique(test$expr_class))))
  
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  test <- melt(test[,c("expr_class", "log2FoldChange")])[,c(1,3)]
  test$variable <- TISSUE
  return(test)
})
# df_smarca4 <- as.data.frame(df_smarca4)
# df_smarca4 <- as.data.frame(cbind(df_smarca4$expr_class, df_smarca4[,grep("value", colnames(df_smarca4))]))
# colnames(df_smarca4) <- c("expr_class", names(test_smarca4))
# df_smarca4 <- melt(df_smarca4)
df_smarca4 <- do.call(rbind, df_smarca4)
df_smarca4$gene <- "Smarca4"

df_chd4 <- lapply(names(test_chd4), function(TISSUE) {
  test <- test_chd4[[TISSUE]]
  test <- test[which(test$gene %in% gene_expr_PC$sample),]
  # test <- test[which(test$baseMeanCustom > summary(test$baseMeanCustom)[2] & test$baseMeanCustom < summary(test$baseMeanCustom)[5]),]
  # test$expr_class <- cut2(test$baseMeanCustom, g=50, levels.mean=T)

  #test$baseMean <- log2(test$baseMean)
  test$baseMean <- log2(rowMedians(as.matrix(test[,grep("RAW", colnames(test))])))
  # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
  test <- test[order(test$baseMean),]
  test$expr_class <- cut2(seq(1, nrow(test)), g=50, levels.mean=T)
  # test$expr_class <- as.numeric(as.character(test$expr_class))
  # test <- test[which(test$expr_class > summary(test$expr_class)[2] & test$expr_class < summary(test$expr_class)[5]),]
  # test$expr_class <- as.factor(test$expr_class)
  test$expr_class <- mapvalues(test$expr_class, from = levels(test$expr_class), to = seq(1, length(unique(test$expr_class))))
  
  cat(TISSUE); cat("\t"); cat(length(table(test$expr_class))); cat("\n"); 
  test <- melt(test[,c("expr_class", "log2FoldChange")])[,c(1,3)]
  test$variable <- TISSUE
  return(test)
})
# df_chd4 <- as.data.frame(df_chd4)
# df_chd4 <- as.data.frame(cbind(df_chd4$expr_class, df_chd4[,grep("value", colnames(df_chd4))]))
# colnames(df_chd4) <- c("expr_class", names(test_chd4))
# df_chd4 <- melt(df_chd4)
df_chd4 <- do.call(rbind, df_chd4)
df_chd4$gene <- "Chd4"

df <- rbind(df_smarca4, df_chd4)
# df <- rbind(df_smarca4[which(df_smarca4$value<0),], df_chd4[which(df_chd4$value<0),])
ggarrange(ggline(df, x = "expr_class", y = "value", add=c("median"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_smarca4, x = "expr_class", y = "value", add=c("median"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_chd4, x = "expr_class", y = "value", add=c("median"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"), nrow=1, ncol=3)

ggarrange(ggplot(df, aes(as.numeric(expr_class), value, color=gene)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 10), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c("#784b97", "#ffa054")),
          ggplot(df_smarca4, aes(as.numeric(expr_class), value, color=variable)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 10), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            theme(legend.position="none"),
          ggplot(df_chd4, aes(as.numeric(expr_class), value, color=variable)) + 
            geom_smooth(method = "lm", se = T, formula = y ~ poly(x, 10), span=1, alpha=0.2) + theme_bw() +
            geom_hline(yintercept = 0, lty=2) + ylab("Median FC in gene expr. (log2)") +
            scale_color_manual(values=c(colorBrewer2palette(name="Paired", count = 11))) +
            theme(legend.position="none"), nrow=1, ncol=3)

matrix2pheatmap(t(confounding_smarca4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
matrix2pheatmap(t(confounding_chd4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
dev.off()
```

###########################################
## Analysis for SMARCA4 and CHD4 (regression-based analysis)
###########################################
```{r}
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="SMARCA4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")

df_smarca4_sampleStat <- lapply(names(test_smarca4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="SMARCA4" &
                                                mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    ID_frame_shift <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                            mutation_info$gene=="SMARCA4" &
                                            (mutation_info$effect=="Frame_Shift_Del" | mutation_info$effect=="Frame_Shift_Ins" |
                                            mutation_info$effect=="Splice_Site")
                                          ),]$sample)
    ID_frame_shift <- ID_frame_shift[which(!ID_frame_shift %in% ID_missense)]
    return(c(length(ID_missense), length(ID_no_mutation), length(ID_frame_shift)));
})
names(df_smarca4_sampleStat) <- names(test_smarca4)
df_smarca4_sampleStat <- as.data.frame(df_smarca4_sampleStat)
df_smarca4_sampleStat$mutation <- c("missense", "no_mutation", "frame_shift")
freqPlot(melt(df_smarca4_sampleStat)[,c(2,1,3)]) + theme_bw()

df_smarca4_regression <- lapply(names(test_smarca4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="SMARCA4" &
                                                ((mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)))
                                              ),]$sample)
    # t <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
    #                                                mutation_info$gene=="SMARCA4" &
    #                                                ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
    #                                                mutation_info$effect=="Silent")
    #                                              ),]$sample)
    # ID_missense <- t[which(!t %in% ID_missense)]
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
    # set.seed(1); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
    
    mat$baseMean <- log2(rowMedians(as.matrix(mat)))
    # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
    mat <- mat[order(mat$baseMean),]
    mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
    mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
    
    # t <- t(apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
    #   scale(x)
    # }))

    mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
      t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
      colnames(t) <- c("expr", "class")
      t$class <- relevel(t$class, ref = "no_mutation")
      if(sd(t$expr) > 0) {
        # t$expr <- scale(t$expr)
        summary(glm(expr ~ class, data=t))$coefficients[2,1]
      } else { 0; }
    })

    cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
    mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
    mat$variable <- TISSUE
    return(mat)
})

df_smarca4_regression <- do.call(rbind, df_smarca4_regression)
df_smarca4_regression$gene <- "Smarca4"

TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="CHD4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")

df_chd4_sampleStat <- lapply(names(test_chd4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="CHD4" &
                                                mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
    ID_frame_shift <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                            mutation_info$gene=="CHD4" &
                                            (mutation_info$effect=="Frame_Shift_Del" | mutation_info$effect=="Frame_Shift_Ins" |
                                            mutation_info$effect=="Splice_Site")
                                          ),]$sample)
    ID_frame_shift <- ID_frame_shift[which(!ID_frame_shift %in% ID_missense)]
    return(c(length(ID_missense), length(ID_no_mutation), length(ID_frame_shift)));
})
names(df_chd4_sampleStat) <- names(test_chd4)
df_chd4_sampleStat <- as.data.frame(df_chd4_sampleStat)
df_chd4_sampleStat$mutation <- c("missense", "no_mutation", "frame_shift")
freqPlot(melt(df_chd4_sampleStat)[,c(2,1,3)]) + theme_bw()

df_chd4_regression <- lapply(names(test_chd4), function(TISSUE) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene=="CHD4" &
                                                ((mutation_info$effect=="Missense_Mutation" &
                                                grepl("damaging", mutation_info$PolyPhen)))
                                              ),]$sample)
    # t <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
    #                                                mutation_info$gene=="CHD4" &
    #                                                ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
    #                                                mutation_info$effect=="Silent")
    #                                              ),]$sample)
    # ID_missense <- t[which(!t %in% ID_missense)]
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
                                                 ),]$sample)
    # ID_no_mutation <- unique(as.character(gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID[gtex_info[which(gtex_info$SMTS==names(gtex[which(gtex==TISSUE)])),]$SAMPID %in% colnames(gene_expr)]))
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    # if(length(ID_no_mutation) > length(ID_missense)) {
    #   set.seed(6)
    #   SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    # }
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
    # set.seed(1); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
    
    mat$baseMean <- log2(rowMedians(as.matrix(mat)))
    # test <- test[which(test$baseMean> summary(test$baseMean)[2] & test$baseMean < summary(test$baseMean)[5]),]
    mat <- mat[order(mat$baseMean),]
    mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
    mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))

    
    mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
      t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
      colnames(t) <- c("expr", "class")
      t$class <- relevel(t$class, ref = "no_mutation")
      if(sd(t$expr) > 0) {
        # t$expr <- scale(t$expr)
        summary(glm(expr ~ class, data=t))$coefficients[2,1]
      } else { 0; }
    }) 
    
    cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
    mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
    mat$variable <- TISSUE
    return(mat)
})
df_chd4_regression <- do.call(rbind, df_chd4_regression)
df_chd4_regression$gene <- "Chd4"

df <- rbind(df_smarca4_regression, df_chd4_regression)

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4_CHD4.pdf", width=15, height = 5)
# df <- rbind(df_smarca4_regression[which(df_smarca4_regression$value<0),], df_chd4_regression[which(df_chd4_regression$value<0),])
ggarrange(ggline(df, x = "expr_class", y = "value", add=c("mean_se"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_smarca4_regression, x = "expr_class", y = "value", add=c("mean_se"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(df_chd4_regression, x = "expr_class", y = "value", add=c("mean_se"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"), nrow=1, ncol=3)
matrix2pheatmap(t(confounding_smarca4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
matrix2pheatmap(t(confounding_chd4[,c(1:11)]), bias = 3, displayN = T, col="Reds")[[4]]
dev.off()
```

###########################################
## Analysis for SMARCA4 and CHD4 (regression + sampling-based analysis)
###########################################
```{r}
# df_smarca4_sampling <- lapply(seq(1:100), function(SEED) {
#   do.call(rbind, lapply(names(test_smarca4), function(TISSUE) {
#       ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
#                                                   mutation_info$gene=="SMARCA4" &
#                                                   mutation_info$effect=="Missense_Mutation" &
#                                                   grepl("damaging", mutation_info$PolyPhen)
#                                                 ),]$sample)
#       ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
#                                                    ),]$sample)
#       ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
#   
#       SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#       SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#   
#       mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#       row.names(mat) <- gene_expr$sample
#       cat(SEED, TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#       colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#       mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#       set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
#       
#       mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#       mat <- mat[order(mat$baseMean),]
#       mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#       mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
#   
#       mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#         t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#         colnames(t) <- c("expr", "class")
#         t$class <- relevel(t$class, ref = "no_mutation")
#         if(sd(t$expr) > 0) {
#           # t$expr <- scale(t$expr)
#           summary(glm(expr ~ class, data=t))$coefficients[2,1]
#         } else { 0; }
#       })
#   
#       cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
#       mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
#       mat$variable <- TISSUE
#       mat$seed <- as.factor(SEED)
#       return(mat)
#   }))
# })
# df_smarca4_sampling <- lapply(seq(1:100), function(SEED) { df_smarca4_sampling[[SEED]]$seed <- as.factor(SEED); return(df_smarca4_sampling[[SEED]]); })
# write.table(do.call(rbind, df_smarca4_sampling), "~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4_sampling.txt", sep="\t", quote=F, row.names = F, col.names =T)
# ggline(do.call(rbind, df_smarca4_sampling), x = "expr_class", y = "value", add=c("median"), color = "seed") + 
#             geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_smarca4 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_smarca4_regression[which(as.numeric(df_smarca4_regression$expr_class) > 0),]$value, 
          df_smarca4_sampling[[SEED]][which(as.numeric(df_smarca4_sampling[[SEED]]$expr_class) > 0),]$value, alternative = "l")$p.value; })), method="BH")
length(which(pVal_smarca4 >= 0.05))/100

# df_chd4_sampling <- lapply(seq(1:100), function(SEED) {
#   do.call(rbind, lapply(names(test_chd4), function(TISSUE) {
#       ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
#                                                   mutation_info$gene=="CHD4" &
#                                                   mutation_info$effect=="Missense_Mutation" &
#                                                   grepl("damaging", mutation_info$PolyPhen)
#                                                 ),]$sample)
#       ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) 
#                                                    ),]$sample)
#       ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
#   
#       SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#       SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
#   
#       mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#       row.names(mat) <- gene_expr$sample
#       cat(SEED, TISSUE, ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#       colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#       mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#       set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
#       
#       mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#       mat <- mat[order(mat$baseMean),]
#       mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#       mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
#   
#       mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#         t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#         colnames(t) <- c("expr", "class")
#         t$class <- relevel(t$class, ref = "no_mutation")
#         if(sd(t$expr) > 0) {
#           # t$expr <- scale(t$expr)
#           summary(glm(expr ~ class, data=t))$coefficients[2,1]
#         } else { 0; }
#       })
#   
#       cat(TISSUE); cat("\t"); cat(length(table(mat$expr_class))); cat("\n"); 
#       mat <- melt(mat[,c("expr_class", "coefficients")])[,c(1,3)]
#       mat$variable <- TISSUE
#       mat$seed <- as.factor(SEED)
#       return(mat)
#   }))
# })
# df_chd4_sampling <- lapply(seq(1:100), function(SEED) { df_chd4_sampling[[SEED]]$seed <- as.factor(SEED); return(df_chd4_sampling[[SEED]]); })
# write.table(do.call(rbind, df_chd4_sampling), "~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/CHD4_sampling.txt", sep="\t", quote=F, row.names = F, col.names =T)
# ggline(do.call(rbind, df_chd4_sampling), x = "expr_class", y = "value", add=c("median"), color = "seed") +
#             geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_chd4 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_chd4_regression[which(as.numeric(df_chd4_regression$expr_class) > 0),]$value, 
          df_chd4_sampling[[SEED]][which(as.numeric(df_chd4_sampling[[SEED]]$expr_class) > 0),]$value, alternative = "g")$p.value; })), method="BH")
length(which(pVal_chd4 >= 0.05))/100
```

###########################################
## Analysis for SMARCA4 and CHD4 (regression-based analysis - irrespective to TISSUE)
###########################################
```{r}
## SMARCA4
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="SMARCA4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Brain", "Breast", "Cervix Uteri", "Colon", "Esophagus", "NA", "Kidney", "Kidney", "Liver", "Lung", "Lung", "Skin", "Stomach", "Uterus")

ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4)) &
                                            mutation_info$gene=="SMARCA4" &
                                            ((mutation_info$effect=="Missense_Mutation" &
                                            grepl("damaging", mutation_info$PolyPhen)))
                                          ),]$sample)
ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4))
                                          ),]$sample)
ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))

mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
row.names(mat) <- gene_expr$sample
cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]

mat$baseMean <- log2(rowMedians(as.matrix(mat)))
mat <- mat[order(mat$baseMean),]
mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))

mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
  t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
  colnames(t) <- c("expr", "class")
  t$class <- relevel(t$class, ref = "no_mutation")
  if(sd(t$expr) > 0) {
    # t$expr <- scale(t$expr)
    summary(glm(expr ~ class, data=t))$coefficients[2,1]
  } else { 0; }
})

cat(length(table(mat$expr_class))); cat("\n");
df_smarca4_regression_v2 <- mat
df_smarca4_regression_v2$gene <- "Smarca4"

## CHD4
TISSUE_STAT <- as.data.frame(table(unique(mutation_info[which(mutation_info$gene=="CHD4"),c("sample", "tissueSource")])$tissueSource))
gtex <- as.character(TISSUE_STAT[which(TISSUE_STAT$Freq>=10),]$Var1)
names(gtex) <- c("Bladder", "Breast", "Cervix Uteri", "Colon", "NA", "Lung", "Lung", "Skin", "Stomach", "Uterus", "Uterus")

ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4)) &
                                            mutation_info$gene=="CHD4" &
                                            ((mutation_info$effect=="Missense_Mutation" &
                                            grepl("damaging", mutation_info$PolyPhen)))
                                          ),]$sample)
ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4))
                                          ),]$sample)
ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))

mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
row.names(mat) <- gene_expr$sample
cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]

mat$baseMean <- log2(rowMedians(as.matrix(mat)))
mat <- mat[order(mat$baseMean),]
mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))

mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
  t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
  colnames(t) <- c("expr", "class")
  t$class <- relevel(t$class, ref = "no_mutation")
  if(sd(t$expr) > 0) {
    # t$expr <- scale(t$expr)
    summary(glm(expr ~ class, data=t))$coefficients[2,1]
  } else { 0; }
})

cat(length(table(mat$expr_class))); cat("\n");
df_chd4_regression_v2 <- mat
df_chd4_regression_v2$gene <- "chd4"

df_smarca4_regression_v2$expr_class <- cut2(seq(1, nrow(df_smarca4_regression_v2)), g=50, levels.mean=T)

df_chd4_regression_v2$expr_class <- cut2(seq(1, nrow(df_chd4_regression_v2)), g=50, levels.mean=T)

df <- rbind(df_smarca4_regression_v2[,c("expr_class", "coefficients", "gene")], df_chd4_regression_v2[,c("expr_class", "coefficients", "gene")])

# df <- rbind(df_smarca4_regression[which(df_smarca4_regression$value<0),], df_chd4_regression[which(df_chd4_regression$value<0),])
ggarrange(ggline(melt(df), x = "expr_class", y = "value", add=c("mean_se"), color = "gene", palette = c("#784b97", "#ffa054")) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(melt(df_smarca4_regression_v2[,c("expr_class", "coefficients")]), x = "expr_class", y = "value", add=c("mean_se"), 
                 color = "variable", palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"),
          ggline(melt(df_chd4_regression_v2[,c("expr_class", "coefficients")]), x = "expr_class", y = "value", add=c("mean_se"), color = "variable", 
                 palette = c(colorBrewer2palette(name="Paired", count = 11))) + 
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)"), nrow=1, ncol=3)

## SMARCA4 sampling
# df_smarca4_sampling_v2 <- lapply(seq(1:100), function(SEED) {
#   ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4)) &
#                                               mutation_info$gene=="SMARCA4" &
#                                               ((mutation_info$effect=="Missense_Mutation" &
#                                               grepl("damaging", mutation_info$PolyPhen)))
#                                             ),]$sample)
#   ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_smarca4))
#                                             ),]$sample)
#   ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
#   
#   SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
#   SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))
#   
#   mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#   row.names(mat) <- gene_expr$sample
#   cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#   colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#   mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#   set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
# 
#   mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#   mat <- mat[order(mat$baseMean),]
#   mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#   mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
# 
#   mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#     t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#     colnames(t) <- c("expr", "class")
#     t$class <- relevel(t$class, ref = "no_mutation")
#     if(sd(t$expr) > 0) {
#       # t$expr <- scale(t$expr)
#       summary(glm(expr ~ class, data=t))$coefficients[2,1]
#     } else { 0; }
#   })
# 
#   cat("\t"); cat(length(table(mat$expr_class))); cat("\n");
#   mat <- mat[,c("baseMean", "expr_class", "coefficients")]
#   mat$seed <- as.factor(SEED)
#   return(mat)
# })
# df_smarca4_sampling_v2 <- lapply(seq(1:100), function(SEED) { df_smarca4_sampling_v2[[SEED]]$seed <- as.factor(SEED); return(df_smarca4_sampling_v2[[SEED]]); })
ggline(do.call(rbind, df_smarca4_sampling_v2), x = "expr_class", y = "coefficients", add=c("median"), color = "seed") +
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_smarca4_v2 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_smarca4_regression_v2[which(as.numeric(df_smarca4_regression_v2$expr_class) > 0),]$coefficients, 
          df_smarca4_sampling_v2[[SEED]][which(as.numeric(df_smarca4_sampling_v2[[SEED]]$expr_class) > 0),]$coefficients, alternative = "l")$p.value; })), method="BH")
length(which(pVal_smarca4_v2 >= 0.05))/100

## CHD4 sampling
# df_chd4_sampling_v2 <- lapply(seq(1:100), function(SEED) {
#   ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4)) &
#                                               mutation_info$gene=="CHD4" &
#                                               ((mutation_info$effect=="Missense_Mutation" &
#                                               grepl("damaging", mutation_info$PolyPhen)))
#                                             ),]$sample)
#   ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource) %in% stringr::str_to_title(names(test_chd4))
#                                             ),]$sample)
#   ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]
# 
#   SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_missense]*log(2)), 0))
#   SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,colnames(gene_expr) %in% ID_no_mutation]*log(2)), 0))
# 
#   mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#   row.names(mat) <- gene_expr$sample
#   cat(ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
#   colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
#   mat <- mat[which(row.names(mat) %in% gene_expr_PC$sample),]
#   set.seed(SEED); colnames(mat) <- rep("no_mutation", ncol(mat)); colnames(mat)[sample(seq(1:ncol(mat)), ncol(SAMPLE_MUTATION))] <- "mutation"
# 
#   mat$baseMean <- log2(rowMedians(as.matrix(mat)))
#   mat <- mat[order(mat$baseMean),]
#   mat$expr_class <- cut2(seq(1, nrow(mat)), g=50, levels.mean=T)
#   mat$expr_class <- mapvalues(mat$expr_class, from = levels(mat$expr_class), to = seq(1, length(unique(mat$expr_class))))
# 
#   mat$coefficients <- apply(mat[,grep("mutation", colnames(mat))], 1, function(x) {
#     t <- cbind((as.data.frame(x)), as.data.frame(gsub("\\..*", "", colnames(mat[1,grep("mutation", colnames(mat))]))))
#     colnames(t) <- c("expr", "class")
#     t$class <- relevel(t$class, ref = "no_mutation")
#     if(sd(t$expr) > 0) {
#       # t$expr <- scale(t$expr)
#       summary(glm(expr ~ class, data=t))$coefficients[2,1]
#     } else { 0; }
#   })
# 
#   cat("\t"); cat(length(table(mat$expr_class))); cat("\n");
#   mat <- mat[,c("baseMean", "expr_class", "coefficients")]
#   mat$seed <- as.factor(SEED)
#   return(mat)
# })
# df_chd4_sampling_v2 <- lapply(seq(1:100), function(SEED) { df_chd4_sampling_v2[[SEED]]$seed <- as.factor(SEED); return(df_chd4_sampling_v2[[SEED]]); })
ggline(do.call(rbind, df_chd4_sampling_v2), x = "expr_class", y = "coefficients", add=c("median"), color = "seed") +
            geom_hline(yintercept = 0, lty=2) + ylab("Mean FC in gene expr. (log2)")
pVal_chd4_v2 <- p.adjust(unlist(lapply(seq(1:100), function(SEED) { 
  ks.test(df_chd4_regression_v2[which(as.numeric(df_chd4_regression_v2$expr_class) > 0),]$coefficients, 
          df_chd4_sampling_v2[[SEED]][which(as.numeric(df_chd4_sampling_v2[[SEED]]$expr_class) > 0),]$coefficients, alternative = "g")$p.value; })), method="BH")
length(which(pVal_chd4_v2 >= 0.05))/100
```

########################################
## refined analysis on individual mutations
########################################
```{r}
# Stomach adenocarcinoma (G782S); Lung adenocarcinoma (G784V); Skin Cutaneous Melanoma (K785R, E861K, E882K); Bladder Urothelial Carcinoma (T786N, E882D); Lung squamous cell carcinoma (E861K); Kidney renal papillary cell carcinoma (E882D); Breast Invasive Carcinoma (E882K); Brain Lower Grade Glioma (H884N); Kidney renal clear cell carcinoma (H884R)
MUTATION <- c("T786N", "E882K", "E882D", "G784V", "E861K", "H884R")
TISSUE <- c("Bladder Urothelial Carcinoma", "Breast invasive carcinoma", "Kidney renal papillary cell carcinoma", "Lung adenocarcinoma", "Lung squamous cell carcinoma", "Kidney renal clear cell carcinoma")
p1 <- lapply(seq(1,6), function(i) {
  #grepl(MUTATION[i], mutation_info$Amino_Acid_Change) &
  ID <- unique(mutation_info[which(grepl(MUTATION[i], mutation_info$Amino_Acid_Change) & 
                                     mutation_info$tissueSource==TISSUE[i] &
                                     mutation_info$effect=="Missense_Mutation" & 
                                     mutation_info$gene=="SMARCA4"),]$sample)
  SAMPLE_INTEREST_MUTATION <- as.data.frame(gene_expr[,unique(mutation_info[which(grepl(MUTATION[i], mutation_info$Amino_Acid_Change) &
                                                                                    mutation_info$tissueSource==TISSUE[i] &
                                                                                    mutation_info$effect=="Missense_Mutation" & 
                                                                                    mutation_info$gene=="SMARCA4"),]$sample)])
  colnames(SAMPLE_INTEREST_MUTATION) <- rep("interest_mutation", ncol(SAMPLE_INTEREST_MUTATION))
  
  
  SAMPLE_OTHER_MUTATION <- gene_expr[,unique(mutation_info[which(!grepl(MUTATION[i], mutation_info$Amino_Acid_Change) &
                                                                   mutation_info$tissueSource==TISSUE[i] &
                                                                   mutation_info$effect=="Missense_Mutation" & 
                                                                   mutation_info$gene=="SMARCA4"),]$sample)]
  #SAMPLE_OTHER_MUTATION <- as.data.frame(SAMPLE_OTHER_MUTATION[,grep(paste(unlist(lapply(strsplit(ID, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_OTHER_MUTATION), value=TRUE)])
  colnames(SAMPLE_OTHER_MUTATION) <- rep("other_mutation", ncol(SAMPLE_OTHER_MUTATION))
  
  SAMPLE_NO_MUTATION <- gene_expr[,unique(mutation_info[which(mutation_info$tissueSource==TISSUE[i] &
                                                                mutation_info$effect=="Missense_Mutation" &
                                                                mutation_info$gene!="SMARCA4"),]$sample)]
  SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
  colnames(SAMPLE_NO_MUTATION) <- rep("no_mutation", ncol(SAMPLE_NO_MUTATION))
  
  SAMPLE_NORMAL <- gene_expr[,grep(paste(unlist(lapply(strsplit(ID, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(gene_expr), value=TRUE)]
  colnames(SAMPLE_NORMAL) <- rep("normal", ncol(SAMPLE_NORMAL))
  # test <- cbind(SAMPLE_INTEREST_MUTATION, SAMPLE_OTHER_MUTATION, SAMPLE_NO_MUTATION, SAMPLE_NORMAL)
  test <- cbind(SAMPLE_INTEREST_MUTATION, SAMPLE_NO_MUTATION)
  table(colnames(test))
  # test <- as.data.frame(normalize.quantiles(as.matrix(test)))
  # colnames(test) <- colnames(cbind(SAMPLE_INTEREST_MUTATION, SAMPLE_OTHER_MUTATION, SAMPLE_NO_MUTATION))
  test$expr_class <- cut2(rowMeans(test), g=3)
  table(test$expr_class)
  colnames(test) <- gsub("\\..*", "", colnames(test))
  test.melt <- melt(test)
  ggplot(test.melt, aes(expr_class, (value), fill=variable)) + geom_boxplot(notch = F) + theme_bw() + ggtitle(TISSUE[i]) +
    scale_fill_manual(values=c("#e31a1c", "#1f78b4")) + ylab("Gene expression")
})

MUTATION <- c("T786N", "E882K", "E882D", "G784V", "E861K", "H884R")
TISSUE <- c("Bladder Urothelial Carcinoma", "Breast invasive carcinoma", "Kidney renal papillary cell carcinoma", "Lung adenocarcinoma", "Lung squamous cell carcinoma", "Kidney renal clear cell carcinoma")
p1 <- lapply(seq(1,6), function(i) {
    ID_missense <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                mutation_info$gene==GENE &
                                                mutation_info$effect=="Missense_Mutation" & 
                                                grepl("damaging", mutation_info$PolyPhen)
                                              ),]$sample)
    ID_no_mutation <- unique(mutation_info[which(stringr::str_to_title(mutation_info$tissueSource)==stringr::str_to_title(TISSUE) &
                                                   mutation_info$gene==GENE &
                                                   ((mutation_info$effect=="Missense_Mutation" & grepl("benign", mutation_info$PolyPhen)) |
                                                   mutation_info$effect=="Silent")
                                                 ),]$sample)
    ID_no_mutation <- ID_no_mutation[which(!ID_no_mutation %in% ID_missense)]

    SAMPLE_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    SAMPLE_NO_MUTATION <- as.data.frame(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    cat(TISSUE[i], ncol(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION)); cat("\n")
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMean <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMean),]
    mat$expr_class <- cut2(mat$baseMean, g=3)
    mat <- mat[,which(!grepl("baseMean", colnames(mat)))]
    colnames(mat) <- gsub("\\..*", "", colnames(mat))
    table(mat$expr_class)
    ggplot(melt(mat), aes(expr_class, log(value), fill=variable)) + geom_boxplot(notch = F) + theme_bw() + ggtitle(TISSUE[i]) +
      scale_fill_manual(values=c("#e31a1c", "#1f78b4")) + ylab("Gene expression")
})
```














########################################
## OLD (v2)
########################################
```{r}
# Stomach adenocarcinoma (G782S); Lung adenocarcinoma (G784V); Skin Cutaneous Melanoma (K785R, E861K, E882K); Bladder Urothelial Carcinoma (T786N, E882D); Lung squamous cell carcinoma (E861K); Kidney renal papillary cell carcinoma (E882D); Breast Invasive Carcinoma (E882K); Brain Lower Grade Glioma (H884N); Kidney renal clear cell carcinoma (H884N); Head and Neck squamous cell carcinoma (H884N); Liver hepatocellular carcinoma (G911S)
INDEX <- which(unlist(lapply(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study), function(TISSUE) {
  x <- grep(paste(unlist(lapply(tissue_barCode[which(stringr::str_to_title(tissue_barCode$Study.Name)==stringr::str_to_title(TISSUE)),]$TSS.Code, function(x) sprintf("TCGA-%s.*11$", x))), collapse = "|"), colnames(gene_expr), value=T)
  IDs <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status=="Missense"),]$Sample;
  COUNT_MISSENSE <- length(grep(paste(unlist(lapply(strsplit(IDs, "-"), function(x) sprintf("TCGA-%s.*11$", x[2]))), collapse="|"), x))
  IDs <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status=="No_mutation"),]$Sample;
  COUNT_NOMUTATION <- length(grep(paste(unlist(lapply(strsplit(IDs, "-"), function(x) sprintf("TCGA-%s.*11$", x[2]))), collapse="|"), x))
  if(COUNT_MISSENSE > 0 & COUNT_NOMUTATION > 0) { return(1); } else { return(0); }
  }))>0)

test_smarca4 <- lapply(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)[INDEX], function(TISSUE) {
  lapply(c("Missense", "No mutation"), function(MUTATION) {
    IDs <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status==MUTATION),]$Sample;
    SAMPLE_NORMAL <- round(exp(gene_expr[,grep(paste(unlist(lapply(strsplit(IDs, "-"), function(x) 
      sprintf("TCGA-%s.*11$", x[2]))), collapse="|"), colnames(gene_expr), value=TRUE)]*log(2)), 0)
    SAMPLE_MUTATION <- round(exp(gene_expr[,c(grep(paste(IDs, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    if(MUTATION=="No mutation") {
      set.seed(1)
      SAMPLE_MUTATION <- SAMPLE_MUTATION[,sample(colnames(SAMPLE_MUTATION), ncol(SAMPLE_NORMAL))]
    }
    mat <- cbind(SAMPLE_NORMAL, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    # cat("%s %s", ncol(SAMPLE_NORMAL), ncol(SAMPLE_MUTATION))
    mat <- matrix2deseq2(mat, c(rep("wt", ncol(SAMPLE_NORMAL)),  rep("ko", ncol(SAMPLE_MUTATION))))
    # mat <- mat[which(apply(mat, 1, function(x) ifelse(max(x)>4, 1, 0))!=0),]
    # mat <- matrix2edgeR(mat, c(rep("wt", length(grep("TCGA.*11$", colnames(mat)))),  rep("ko", length(grep("TCGA.*01$|TCGA.*06$", colnames(mat))))))
    # colnames(mat)[2] <- "baseMean"; colnames(mat)[3] <- "log2FoldChange"; colnames(mat)[7] <- "padj"
    mat$fc_class <- "neutral"
    mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
    mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
    mat$expr_class <- cut2(rowMeans(mat[,c(grep("TCGA.*11$", colnames(mat), value=T))]), m=1000)
    return(mat)
    })
  })
names(test_smarca4) <- unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)[INDEX]

p1 <- lapply(names(test_smarca4), function(TISSUE) {
  test1 <- test_smarca4[[TISSUE]];
  names(test1) <- c("Missense", "No_mutation");
  
  test1 <- as.data.frame(cbind(test1[["Missense"]]$log2FoldChange, test1[["No_mutation"]]$log2FoldChange, test1[["No_mutation"]]$baseMean));
  # test1 <- as.data.frame(cbind(test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["No_mutation"]][which(test1[["No_mutation"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$baseMean));
  # test1 <- as.data.frame(cbind(test1[["Missense"]]$log2FoldChange,
  #                              test1[["No_mutation"]]$log2FoldChange,
  #                              rowMeans(cbind(test1[["Missense"]][,c(grep("RAW", colnames(test1[["Missense"]]), value=T))],
  #                                             test1[["No_mutation"]][,c(grep("RAW", colnames(test1[["No_mutation"]]), value=T))]))))
  colnames(test1) <- c("Missense", "No_mutation", "baseMean")
  #test1$expr_class <- cut2(test1$baseMean, g=50)
  test1 <- test1[order(test1$baseMean),]
  test1$expr_class <- cut2(c(1:nrow(test1)), g=15)
  ggline(melt(test1[,c("expr_class", "Missense", "No_mutation")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) + theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
    ylab("FC in Gene expr. (Cancer vs Normal; Log2)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
    scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test1$expr_class), function(x) mean(test1[which(test1$expr_class==x),]$baseMean)))))
})

p2 <- lapply(names(test_smarca4), function(TISSUE) {
  test1 <- test_smarca4[[TISSUE]];
  names(test1) <- c("Missense", "No_mutation");
  test1 <- as.data.frame(cbind(test1[["Missense"]]$log2FoldChange, test1[["No_mutation"]]$log2FoldChange, test1[["No_mutation"]]$baseMean));
  # test1 <- as.data.frame(cbind(test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["No_mutation"]][which(test1[["No_mutation"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$baseMean));
  colnames(test1) <- c("Missense", "No_mutation", "baseMean")
  #test1$expr_class <- cut2(test1$baseMean, g=50)
  test1 <- test1[order(test1$baseMean),]
  test1$expr_class <- cut2(c(1:nrow(test1)), g=15)
  test1.melt <- melt(summaryBy(Missense + No_mutation  ~ expr_class, data = test1, FUN = list(median)))
  ggplot(test1.melt, aes(expr_class, value, color = variable)) +
    geom_point() +
    stat_smooth(aes(group = variable), method = "lm", formula = y ~ poly(x, 2), alpha=0.2, se=T) + theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
    ylab("FC in Gene expr. (Cancer vs Normal; Log2)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
    scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test1$expr_class), function(x) mean(test1[which(test1$expr_class==x),]$baseMean)))))
})

SAMPLE_COUNTS <- lapply(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)[INDEX], function(TISSUE) {
  unlist(lapply(c("Missense", "No mutation"), function(MUTATION) {
    IDs <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status==MUTATION),]$Sample;
    SAMPLE_NORMAL <- round(exp(gene_expr[,grep(paste(unlist(lapply(strsplit(IDs, "-"), function(x) 
      sprintf("TCGA-%s.*11$", x[2]))), collapse="|"), colnames(gene_expr), value=TRUE)]*log(2)), 0)
    SAMPLE_MUTATION <- round(exp(gene_expr[,c(grep(paste(IDs, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    #if(MUTATION=="No mutation") {
    #  set.seed(1)
    #  SAMPLE_MUTATION <- SAMPLE_MUTATION[,sample(colnames(SAMPLE_MUTATION), ncol(SAMPLE_NORMAL))]
    #}
    c(ncol(SAMPLE_NORMAL), ncol(SAMPLE_MUTATION))
    }))
  })
names(SAMPLE_COUNTS) <- unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)[INDEX]
SAMPLE_COUNTS <- t(as.data.frame(SAMPLE_COUNTS))
colnames(SAMPLE_COUNTS) <- c("missense_normal", "missense_tumor", "no_mutation_normal", "no_mutation_tumor")
p3 <- freqPlot(melt(SAMPLE_COUNTS), "density") + ggtitle("sample counts")

## compare only cancer patients having missense mutation with those having no mutation in SMARCA4
# test_smarca4_v2 <- lapply(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)[INDEX], function(TISSUE) {
#     ID_missense <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status=="Missense"),]$Sample;
#     ID_no_mutation <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status=="No mutation"),]$Sample;
#     
#     SAMPLE_MUTATION <- round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
#     SAMPLE_NO_MUTATION <- round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
#     set.seed(1)
#     SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
#     mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
#     row.names(mat) <- gene_expr$sample
#     #cat(ncol(SAMPLE_MUTATION), ncol(SAMPLE_NO_MUTATION))
#     mat <- matrix2deseq2(mat, c(rep("wt", ncol(SAMPLE_NO_MUTATION)),  rep("ko", ncol(SAMPLE_MUTATION))))
#     # mat <- mat[which(apply(mat, 1, function(x) ifelse(max(x)>4, 1, 0))!=0),]
#     # mat <- matrix2edgeR(mat, c(rep("wt", length(grep("TCGA.*11$", colnames(mat)))),  rep("ko", length(grep("TCGA.*01$|TCGA.*06$", colnames(mat))))))
#     # colnames(mat)[2] <- "baseMean"; colnames(mat)[3] <- "log2FoldChange"; colnames(mat)[7] <- "padj"
#     mat$fc_class <- "neutral"
#     mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
#     mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
#     mat$expr_class <- cut2(rowMeans(SAMPLE_NO_MUTATION), m=1000)
#     return(mat)
#   })
# names(test_smarca4_v2) <- unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)[INDEX]

# p4 <- lapply(names(test_smarca4_v2), function(TISSUE) {
#   test1 <- test_smarca4_v2[[TISSUE]];
#   test1 <- test1[order(test1$baseMean),]
#   test1$expr_class <- cut2(c(1:nrow(test1)), g=15)
#   ggline(melt(test1[,c("expr_class", "log2FoldChange")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) + theme_bw() +
#     ggtitle(TISSUE) +
#     #scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
#     ylab("FC in Gene expr. (Missense vs No mutation; Log2)") + xlab("") +
#     theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
#     scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test1$expr_class), function(x) mean(test1[which(test1$expr_class==x),]$baseMean)))))
# })

test_smarca4 <- lapply(unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study), function(TISSUE) {
    ID_missense <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status=="Missense"),]$Sample;
    ID_no_mutation <- smarca4_info[which(smarca4_info$Study==TISSUE & smarca4_info$Smarca4.Status=="No mutation"),]$Sample;
    
    SAMPLE_MUTATION <- round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    SAMPLE_NO_MUTATION <- round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,grep(paste(unlist(lapply(strsplit(ID_missense, "-"), function(x) sprintf("TCGA-%s", x[2]))), collapse="|"), colnames(SAMPLE_NO_MUTATION), value=TRUE)]
    #set.seed(6)
    #SAMPLE_NO_MUTATION <- SAMPLE_NO_MUTATION[,sample(colnames(SAMPLE_NO_MUTATION), ncol(SAMPLE_MUTATION))]
    mat <- cbind(SAMPLE_NO_MUTATION, SAMPLE_MUTATION)
    row.names(mat) <- gene_expr$sample
    colnames(mat) <- c(rep("no_mutation", ncol(SAMPLE_NO_MUTATION)),  rep("mutation", ncol(SAMPLE_MUTATION)))
    mat$baseMean <- rowMeans(gene_expr[,grep("TCGA.*11$", colnames(gene_expr))])
    mat <- mat[order(mat$baseMean),]
    #mat$expr_class <- cut2(mat$baseMean, g=15)
    mat$expr_class <- cut2(c(1:nrow(mat)), g=5)
    return(mat)
})
names(test_smarca4) <- unique(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)

p4_test <- lapply(names(test_smarca4), function(TISSUE) {
  test <- test_smarca4[[TISSUE]][,which(!grepl("baseMean", colnames(test_smarca4[[TISSUE]])))]
  colnames(test) <- gsub("\\..*", "", colnames(test))
  #cat(ncol(SAMPLE_MUTATION), ncol(SAMPLE_NO_MUTATION))
  test.melt <- melt(test)
  #test.melt$value <- log2(test.melt$value)
  ggline(test.melt, x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
    theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
    ylab("Gene expr.") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
    scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test_smarca4[[TISSUE]]$expr_class), function(x) mean(test_smarca4[[TISSUE]][which(test_smarca4[[TISSUE]]$expr_class==x),]$baseMean)))))
  })
pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4.1.pdf", width=20)
ggarrange(plotlist = p4_test, nrow=2, ncol=8)
dev.off()

p5 <- lapply(names(test_smarca4), function(x) {
  freqPlot(melt(cbind(table(test_smarca4[[x]][[1]]$fc_class), table(test_smarca4[[x]][[2]]$fc_class)))[,c(2,1,3)], "density") +
    ggtitle(x)
})

mutation_stat <- read.table(pipe("zless ~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/TcgaTargetGtex_mutation_phenotype.txt.gz | grep -w SMARCA4 | grep -w Missense_Mutation | cut -f 8,12,13 | sed -E 's/\\(.*\\)//g' | cut -f 2,3"), sep="\t", stringsAsFactors=FALSE)
p6 <- freqPlot(melt(table(mutation_stat[stringr::str_to_title(mutation_stat$V2) %in% stringr::str_to_title(names(test_smarca4)),c(2,1)])), "density")

p7 <- ggarrange(plotlist = lapply(names(test_smarca4), function(x) {
  test1 <- test_smarca4[[TISSUE]];
  names(test1) <- c("Missense", "No_mutation");
  ggplot(melt(test1[["Missense"]][,c("fc_class", grep("RAW", colnames(test1[["Missense"]]), value=T))]), aes(variable, log(value))) + 
    geom_boxplot() + facet_grid(.~fc_class) + theme_bw() +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) + ggtitle(TISSUE)
}), nrow=2, ncol=4)

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/SMARCA4.pdf", width=20)
ggarrange(plotlist = p1, nrow=2, ncol=4)
ggarrange(plotlist = p2, nrow=2, ncol=4)
p3
ggarrange(plotlist = p4, nrow=2, ncol=4)
ggarrange(plotlist = p5, nrow=2, ncol=4)
p6
p7
dev.off()

t <- read.table("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/t", sep="\t")
freqPlot(melt(table(t[,c(6,7)])), "density")
```

```{r}
# Uterine Corpus Endometrial Carcinoma (R1162W (benign), R572Q); Colon adenocarcinoma (R572Q); Stomach adenocarcinoma (R572Q)
head(grep("TCGA.*01$", colnames(gene_expr[,(!(colnames(gene_expr) %in% unique(mutation_info$sample)))]), value=T))

test_chd4 <- lapply(c("Uterine Corpus Endometrial Carcinoma", "Colon adenocarcinoma", "Stomach adenocarcinoma"), function(TISSUE) {
  lapply(c("Missense_Mutation", "No_mutation"), function(MUTATION) {
  IDs <- chd4_info[which(chd4_info$tissueSource==TISSUE & chd4_info$effect==MUTATION),]$sample;
  mat <- cbind(
    round(exp(gene_expr[,grep(paste(unlist(lapply(strsplit(IDs, "-"), function(x) sprintf("TCGA-%s.*11$", x[2]))), collapse="|"), colnames(gene_expr), value=TRUE)]*log(2)), 0),
    round(exp(gene_expr[,c(grep(paste(IDs, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    )

    # ID_missense <- chd4_info[which(chd4_info$tissueSource==TISSUE & chd4_info$effect=="Missense_Mutation"),]$sample;
    # ID_no_mutation <- chd4_info[which(chd4_info$tissueSource==TISSUE & chd4_info$effect=="No_mutation"),]$sample;
    # mat <- cbind(
    #   round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0),
    #   round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    #   )
    
    row.names(mat) <- gene_expr$sample
    mat <- matrix2deseq2(mat, c(rep("wt", length(grep("TCGA.*11$", colnames(mat)))),  rep("ko", length(grep("TCGA.*01$|TCGA.*06$", colnames(mat))))))
    # mat <- mat[which(apply(mat, 1, function(x) ifelse(max(x)>4, 1, 0))!=0),]
    # mat <- matrix2edgeR(mat, c(rep("wt", length(grep("TCGA.*11$", colnames(mat)))),  rep("ko", length(grep("TCGA.*01$|TCGA.*06$", colnames(mat))))))
    # colnames(mat)[2] <- "baseMean"; colnames(mat)[3] <- "log2FoldChange"; colnames(mat)[7] <- "padj"
    # mat <- matrix2deseq2(mat, c(rep("wt", ncol(round(exp(gene_expr[,c(grep(paste(ID_no_mutation, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0))), rep("ko", ncol(round(exp(gene_expr[,c(grep(paste(ID_missense, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)))))
    mat$fc_class <- "neutral"
    mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
    mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
    mat$expr_class <- cut2(rowMeans(mat[,c(grep("TCGA.*11$", colnames(mat), value=T))]), m=1000)
    return(mat)
    })
})
names(test_chd4) <- c("Uterine Corpus Endometrial Carcinoma", "Colon adenocarcinoma", "Stomach adenocarcinoma")

p1 <- lapply(names(test_chd4), function(TISSUE) {
  test1 <- test_chd4[[TISSUE]];
  names(test1) <- c("Missense", "No_mutation");
  test1 <- as.data.frame(cbind(test1[["Missense"]]$log2FoldChange, test1[["No_mutation"]]$log2FoldChange, test1[["Missense"]]$baseMean));
  # test1 <- as.data.frame(cbind(test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["No_mutation"]][which(test1[["No_mutation"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$baseMean));
  colnames(test1) <- c("Missense", "No_mutation", "baseMean")
  #test1$expr_class <- cut2(test1$baseMean, g=50)
  test1 <- test1[order(test1$baseMean),]
  test1$expr_class <- cut2(c(1:nrow(test1)), g=15)
  ggline(melt(test1[,c("expr_class", "Missense", "No_mutation")]), x = "expr_class", y = "value", color = "variable", add = c("mean_se")) +
    theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
    ylab("FC in Gene expr. (Cancer vs Normal; Log2)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
    scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test1$expr_class), function(x) mean(test1[which(test1$expr_class==x),]$baseMean)))))
})

SAMPLE_COUNTS  <- lapply(c("Uterine Corpus Endometrial Carcinoma", "Colon adenocarcinoma", "Stomach adenocarcinoma"), function(TISSUE) {
  unlist(lapply(c("Missense_Mutation", "No_mutation"), function(MUTATION) {
    IDs <- chd4_info[which(chd4_info$tissueSource==TISSUE & chd4_info$effect==MUTATION),]$sample;
    SAMPLE_NORMAL <- round(exp(gene_expr[,grep(paste(unlist(lapply(strsplit(IDs, "-"), function(x) sprintf("TCGA-%s.*11$", x[2]))), collapse="|"), colnames(gene_expr), value=TRUE)]*log(2)), 0)
    SAMPLE_MUTATION <- round(exp(gene_expr[,c(grep(paste(IDs, collapse="|"), colnames(gene_expr), value=TRUE))]*log(2)), 0)
    #if(MUTATION=="No mutation") {
    #  set.seed(1)
    #  SAMPLE_MUTATION <- SAMPLE_MUTATION[,sample(colnames(SAMPLE_MUTATION), ncol(SAMPLE_NORMAL))]
    #}
    c(ncol(SAMPLE_NORMAL), ncol(SAMPLE_MUTATION))
    }))
  })
names(SAMPLE_COUNTS) <- c("Uterine Corpus Endometrial Carcinoma", "Colon adenocarcinoma", "Stomach adenocarcinoma")
SAMPLE_COUNTS <- t(as.data.frame(SAMPLE_COUNTS))
colnames(SAMPLE_COUNTS) <- c("missense_normal", "missense_tumor", "no_mutation_normal", "no_mutation_tumor")
p2 <- freqPlot(melt(SAMPLE_COUNTS), "density") + ggtitle("sample counts")

p3 <- lapply(names(test_chd4), function(TISSUE) {
  test1 <- test_chd4[[TISSUE]];
  names(test1) <- c("Missense", "No_mutation");
  test1 <- as.data.frame(cbind(test1[["Missense"]]$log2FoldChange, test1[["No_mutation"]]$log2FoldChange, test1[["Missense"]]$baseMean));
  # test1 <- as.data.frame(cbind(test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["No_mutation"]][which(test1[["No_mutation"]]$fc_class!="neutral"),]$log2FoldChange,
  #                              test1[["Missense"]][which(test1[["Missense"]]$fc_class!="neutral"),]$baseMean));
  colnames(test1) <- c("Missense", "No_mutation", "baseMean")
  #test1$expr_class <- cut2(test1$baseMean, g=50)
  test1 <- test1[order(test1$baseMean),]
  test1$expr_class <- cut2(c(1:nrow(test1)), g=15)
  test1.melt <- melt(summaryBy(Missense + No_mutation  ~ expr_class, data = test1, FUN = list(median)))
  ggplot(test1.melt, aes(expr_class, value, color = variable)) +
    geom_point() +
    stat_smooth(aes(group = variable), method = "lm", formula = y ~ poly(x, 2), alpha=0.2, se=T) + theme_bw() +
    ggtitle(TISSUE) +
    scale_color_manual(values=c("#e31a1c", "#1f78b4")) +
    ylab("FC in Gene expr. (Cancer vs Normal; Log2)") + xlab("") +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
    scale_x_discrete(labels=sprintf("%0.2f", unlist(lapply(unique(test1$expr_class), function(x) mean(test1[which(test1$expr_class==x),]$baseMean)))))
})

pdf("~/bric_share/bp_nextgen/09_ALL_PUBLIC/cbio-hub/CHD4.pdf", width=20)
ggarrange(plotlist = p1, nrow=1, ncol=3)
p2
ggarrange(plotlist = p3, nrow=1, ncol=3)
dev.off()
```



########################################
## OLD (V1)
########################################
```{r}
## missense ##

gene_expr_interest <- cbind(gene_expr$sample, 
                            gene_expr[,which(colnames(gene_expr) %in% 
                                               as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Sample))],
                            gene_expr[,c(grep("TCGA.*11$", colnames(gene_expr), value=T))])
colnames(gene_expr_interest)[1] <- "gene"

#tissue_barCode[c(which(tissue_barCode$Study.Name %in% as.character(unique(smarca4_info$Study)[1]))),]$TSS.Code
gene_expr_fc_missense <- lapply(unique(as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study)), function(x) {
  test <- gene_expr_interest[,(colnames(gene_expr_interest[,c("gene", grep(paste(tissue_barCode[c(which(stringr::str_to_title(tissue_barCode$Study.Name) %in% stringr::str_to_title(x))),]$TSS.Code, collapse="|"), colnames(gene_expr_interest), value=T))]))]
  mat <- cbind(round(exp(test[,colnames(test) %in% grep("TCGA.*11", colnames(test), value=T)]*log(2)), 0),
             round(exp(test[,colnames(test) %in% grep("TCGA.*01|TCGA.*06", colnames(test), value=T)]*log(2)), 0))
  row.names(mat) <- test$gene
  mat <- matrix2deseq2(mat, c(rep("wt", length(grep("TCGA.*11", colnames(test)))),  rep("ko", length(grep("TCGA.*01|TCGA.*06", colnames(test))))))
  mat$fc_class <- "neutral"
  mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
  mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
  mat$expr_class <- "low"
  mat[which(mat$baseMean > summary(mat$baseMean)[2] & mat$baseMean < summary(mat$baseMean)[5]),]$expr_class <- "average"
  mat[which(mat$baseMean > summary(mat$baseMean)[5]),]$expr_class <- "high"
  return(mat);
         })

names(gene_expr_fc_missense) <- unique(as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="Missense"),]$Study))
names(gene_expr_fc_missense)

freqPlot(melt(table(gene_expr_fc_missense[[7]][,c("expr_class", "fc_class")])), "density")

## not sequenced ##
gene_expr_interest <- cbind(gene_expr$sample, 
                            gene_expr[,which(colnames(gene_expr) %in% 
                                               as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="Not Sequenced"),]$Sample))],
                            gene_expr[,c(grep("TCGA.*11$", colnames(gene_expr), value=T))])
colnames(gene_expr_interest)[1] <- "gene"

#tissue_barCode[c(which(tissue_barCode$Study.Name %in% as.character(unique(smarca4_info$V2)[1]))),]$TSS.Code
gene_expr_fc_notSeq <- lapply(unique(as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="Not Sequenced"),]$Study)), function(x) {
  test <- gene_expr_interest[,(colnames(gene_expr_interest[,c("gene", grep(paste(tissue_barCode[c(which(stringr::str_to_title(tissue_barCode$Study.Name) %in% stringr::str_to_title(x))),]$TSS.Code, collapse="|"), colnames(gene_expr_interest), value=T))]))]
  mat <- cbind(round(exp(test[,colnames(test) %in% grep("TCGA.*11", colnames(test), value=T)]*log(2)), 0),
             round(exp(test[,colnames(test) %in% grep("TCGA.*01|TCGA.*06", colnames(test), value=T)]*log(2)), 0))
  row.names(mat) <- test$gene
  mat <- matrix2deseq2(mat, c(rep("wt", length(grep("TCGA.*11", colnames(test)))),  rep("ko", length(grep("TCGA.*01|TCGA.*06", colnames(test))))))
  mat$fc_class <- "neutral"
  mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
  mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
  mat$expr_class <- "low"
  mat[which(mat$baseMean > summary(mat$baseMean)[2] & mat$baseMean < summary(mat$baseMean)[5]),]$expr_class <- "average"
  mat[which(mat$baseMean > summary(mat$baseMean)[5]),]$expr_class <- "high"
  return(mat);
         })

names(gene_expr_fc_notSeq) <- unique(as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="Not Sequenced"),]$Study))
names(gene_expr_fc_notSeq)

freqPlot(melt(table(gene_expr_fc_notSeq[[14]][,c("expr_class", "fc_class")])), "density")

## No_mutation ##
gene_expr_interest <- cbind(gene_expr$sample, 
                            gene_expr[,which(colnames(gene_expr) %in% 
                                               as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="No_mutation"),]$Sample))],
                            gene_expr[,c(grep("TCGA.*11$", colnames(gene_expr), value=T))])
colnames(gene_expr_interest)[1] <- "gene"

#tissue_barCode[c(which(tissue_barCode$Study.Name %in% as.character(unique(smarca4_info$V2)[1]))),]$TSS.Code
gene_expr_fc_noMutation <- lapply(unique(as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="No_mutation"),]$Study)), function(x) {
  test <- gene_expr_interest[,(colnames(gene_expr_interest[,c("gene", grep(paste(tissue_barCode[c(which(stringr::str_to_title(tissue_barCode$Study.Name) %in% stringr::str_to_title(x))),]$TSS.Code, collapse="|"), colnames(gene_expr_interest), value=T))]))]
  mat <- cbind(round(exp(test[,colnames(test) %in% grep("TCGA.*11", colnames(test), value=T)]*log(2)), 0),
             round(exp(test[,colnames(test) %in% grep("TCGA.*01|TCGA.*06", colnames(test), value=T)]*log(2)), 0))
  row.names(mat) <- test$gene
  mat <- matrix2deseq2(mat, c(rep("wt", length(grep("TCGA.*11", colnames(test)))),  rep("ko", length(grep("TCGA.*01|TCGA.*06", colnames(test))))))
  mat$fc_class <- "neutral"
  mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
  mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
  mat$expr_class <- "low"
  mat[which(mat$baseMean > summary(mat$baseMean)[2] & mat$baseMean < summary(mat$baseMean)[5]),]$expr_class <- "average"
  mat[which(mat$baseMean > summary(mat$baseMean)[5]),]$expr_class <- "high"
  return(mat);
         })

names(gene_expr_fc_noMutation) <- unique(as.character(smarca4_info[which(smarca4_info$Smarca4.Status=="No_mutation"),]$Study))
names(gene_expr_fc_noMutation)

freqPlot(melt(table(gene_expr_fc_noMutation[[14]][,c("expr_class", "fc_class")])), "density")


# test <- gene_expr_list[[1]]
# mat <- cbind(round(exp(test[,colnames(test) %in% grep("TCGA.*11", colnames(test), value=T)]*log(2)), 0),
#              round(exp(test[,colnames(test) %in% grep("TCGA.*01|TCGA.*06", colnames(test), value=T)]*log(2)), 0))
# row.names(mat) <- test$gene
# mat <- matrix2deseq2(mat, c(rep("wt", length(grep("TCGA.*11", colnames(test)))),  rep("ko", length(grep("TCGA.*01|TCGA.*06", colnames(test))))))
# mat$fc_class <- "neutral"
# mat[which(mat$log2FoldChange > 0 & mat$padj < 0.05),]$fc_class <- "up"
# mat[which(mat$log2FoldChange < 0 & mat$padj < 0.05),]$fc_class <- "down"
# mat$expr_class <- "low"
# mat[which(mat$baseMean > summary(mat$baseMean)[2] & mat$baseMean < summary(mat$baseMean)[5]),]$expr_class <- "average"
# mat[which(mat$baseMean > summary(mat$baseMean)[5]),]$expr_class <- "high"
# freqPlot(melt(table(mat[,c("expr_class", "fc_class")])), "density")
# 
# test$cancer <- rowMeans(test[,colnames(test) %in% grep("TCGA.*01|TCGA.*06", colnames(test), value=T)])
# test$normal <- rowMeans(test[,colnames(test) %in% grep("TCGA.*11", colnames(test), value=T)])
# #test <- test[which(test$normal>0 & test$cancer>0),]
# test$fc_class <- "neutral"
# test[which(test$cancer > test$normal),]$fc_class <- "up"
# test[which(test$cancer < test$normal),]$fc_class <- "down"
# test$expr_class <- "low"
# test[which(test$normal > summary(test$normal)[2] & test$normal <= summary(test$normal)[5]),]$expr_class <- "average"
# test[which(test$normal > summary(test$normal)[5]),]$expr_class <- "high"
# freqPlot(melt(table(mat[,c("expr_class", "fc_class")])), "density")
# barplot(test[order(test$normal),]$fc)
```


```{r}
p1 <- lapply(gene_expr_fc_missense, function(test) {
  test <- test[which(test$padj<0.05),]
  test$expr_class <- cut2(rowMeans(test[,c(grep("TCGA.*11$", colnames(test), value=T))]), m=50)
  ggline(test, x = "expr_class", y = "log2FoldChange", add = c("mean_se"))
})
ggarrange(plotlist = p1, nrow=4, ncol=4)

p2 <- lapply(gene_expr_fc_notSeq, function(test) {
  test <- test[which(test$padj<0.05),]
  test$expr_class <- cut2(rowMeans(test[,c(grep("TCGA.*11$", colnames(test), value=T))]), m=50)
  ggline(test, x = "expr_class", y = "log2FoldChange", add = c("mean_se"))
})
ggarrange(plotlist = p2, nrow=4, ncol=4)

ggplot(test, aes(expr_class, log2FoldChange)) + geom_boxplot() + ylim(c(-2,2))
cor(test$baseMean, test$log2FoldChange)
freqPlot(melt(table(test[,c("expr_class", "fc_class")])), "density")
```

#### revision (Pol II ChIP-seq data analysis - brg1)
```{r}
## Pol II ser5p
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser5p.deepToolsMatrix.gz"
# file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/test/pol2_ser2p_choudhary.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

df_dirk <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer_dirk.bed")
table(df_dirk$V18)
df$class <- "neutral"
df[df$V4 %in% df_dirk[which(df_dirk$V18=="down"),]$V4,]$class <- "down"
df[df$V4 %in% df_dirk[which(df_dirk$V18=="up"),]$V4,]$class <- "up"
table(df$class)

mat <- (df[,-c(1:6)])
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))]
  })

samples <- unlist(strsplit("BRM014,BRM014,BRM014,DMSO,DMSO,DMSO", ","))
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1-mat2)
reformatMat$class <- mat$class
# reformatMat$class <- "neutral"
# reformatMat[c(params$group_boundaries[1]):params$group_boundaries[2],]$class <- "down"
# reformatMat[c(params$group_boundaries[3]):params$group_boundaries[4],]$class <- "up"
# # reformatMat[which(reformatMat$class=="neutral"),]$class <- cut2(as.numeric(row.names(reformatMat[which(reformatMat$class=="neutral"),])), g=3)
# table(reformatMat$class)
# reformatMat.melt <- melt(reformatMat)
# reformatMat.melt$variable <- factor(reformatMat.melt$variable, levels=seq(1,ncol(mat1)), ordered=T)
# ggline(reformatMat.melt, x="variable", y="value", add=c("mean_se"), color="class", palette = "jco")

pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser5p_heatmap.pdf")
# matrix2pheatmap(reformatMat[order(reformatMat$class),c(-201)], scale = "none", bias = 1, 
#                 col = c("#053061", "#2166ac", "#4393c3", "white", "white", "#d6604d", "#b2182b", "#67001f"), 
#                 breaks=c(-15, -10, -5, -2, 0, 2, 5, 10, 15))
Heatmap(reformatMat[,c(-ncol(reformatMat))], col = circlize::colorRamp2(c(-500, -200, 0, 200, 500), c("#2166ac", "#4393c3", "white", "#d6604d", "#b2182b")),
        cluster_rows=F, cluster_columns=F)
dev.off()

## Pol II ser2p
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser2p_genebody.deepToolsMatrix.gz"
# file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/test/pol2_ser2p_choudhary.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

df_dirk <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer_dirk.bed")
table(df_dirk$V18)
df$class <- "neutral"
df[df$V4 %in% df_dirk[which(df_dirk$V18=="down"),]$V4,]$class <- "down"
df[df$V4 %in% df_dirk[which(df_dirk$V18=="up"),]$V4,]$class <- "up"
table(df$class)

mat <- (df[,-c(1:6)])
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))]
  })

samples <- unlist(strsplit("BRM014,BRM014,BRM014,DMSO,DMSO,DMSO", ","))
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1-mat2)
reformatMat$class <- mat$class
# reformatMat$class <- "neutral"
# reformatMat[c(params$group_boundaries[1]):params$group_boundaries[2],]$class <- "down"
# reformatMat[c(params$group_boundaries[3]):params$group_boundaries[4],]$class <- "up"
# # reformatMat[which(reformatMat$class=="neutral"),]$class <- cut2(as.numeric(row.names(reformatMat[which(reformatMat$class=="neutral"),])), g=3)
# table(reformatMat$class)
# reformatMat.melt <- melt(reformatMat)
# reformatMat.melt$variable <- factor(reformatMat.melt$variable, levels=seq(1,ncol(mat1)), ordered=T)
# ggline(reformatMat.melt, x="variable", y="value", add=c("mean_se"), color="class", palette = "jco")

pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser2p_heatmap.pdf")
# matrix2pheatmap(reformatMat[order(reformatMat$class),c(-ncol(reformatMat))], scale = "none", bias = 1, 
#                 col = c("#053061", "#2166ac", "#4393c3", "white", "white", "#d6604d", "#b2182b", "#67001f"), 
#                 breaks=c(-20, -10, -5, -1, 0, 1, 5, 10, 20))
#                 # breaks=c(-500, -250, -50, -5, 0, 5, 50, 250, 500))
Heatmap(reformatMat[,c(-ncol(reformatMat))], col = circlize::colorRamp2(c(-20, -10, 0, 10, 20), c("#2166ac", "#4393c3", "white", "#d6604d", "#b2182b")),
        cluster_rows=F, cluster_columns=F)
dev.off()

barplot(as.matrix(reformatMat[grep("^Cdk1$", df$V4), c(-201)])) ## Cdk1
barplot(as.matrix(reformatMat[grep("^Ercc6l$", df$V4), c(-201)])) ## Ercc6l

## Pol II ser2p (gene regulation)
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser2p_dirk.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

mat <- (df[,-c(1:6)])
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))]
  })

samples <- unlist(strsplit("CHD4i,CHD4i,CHD4i,H2O,H2O,H2O", ","))
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1-mat2)

reformatMat$class <- "neutral"
reformatMat[c(1:608),]$class <- "down"
reformatMat[c(7935:9065),]$class <- "up"
df.melt <- melt(reformatMat[,c(100:201)])
ggline(df.melt, x = "variable", y = "value", add = "mean", color="class")
```

#### revision (Gviz data tracks - brg1)
```{r}
pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/GVIZ_brg1.pdf")
BWFILE <- read.table(pipe("cat ~/project/chip-seq-analysis/analysis_chd4_brg1/BW.CONFIG | grep 'DMSO\\|BRM014' | grep normBySIF_COV"))
## chr10:68798882-68816186 (full gene)
plot_gviz_track(GEN = "mm9", COOR = "chr10:68805900-68816010",
                BAMFILE = c(grep("ser5p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=20000)
plot_gviz_track(GEN = "mm9", COOR = "chr10:68805900-68816010",
                BAMFILE = c(grep("ser2p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=7000)
## chrX:99336554-99352930 (full gene)
plot_gviz_track(GEN = "mm9", COOR = "chrX:99340554-99352930",
                BAMFILE = c(grep("ser5p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=8000)
plot_gviz_track(GEN = "mm9", COOR = "chrX:99340554-99352930",
                BAMFILE = c(grep("ser2p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=3000)
dev.off()
```

#### revision (Histone variants ratio plots - brg1)
```{r}
## h3.3
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/brg1_h3v3_class.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

brg1_signal <- read.table(pipe("less ~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.bed | cut -f 1-6,16 | cut -f 4,7"), header=T)
colnames(brg1_signal) <- c("gene", "brg1")
df <- merge(df, brg1_signal, by.x="V4", by.y="gene")

mat <- (df[,-c(1:6)])
mat <- mat[order(mat$brg1),]
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  cbind(mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))], mat$brg1)
  })

samples <- params$sample_labels
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1[,-ncol(mat1)]-mat2[,-ncol(mat2)])
reformatMat$class <- mat$class

# pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_ser5p_heatmap.pdf")
# matrix2pheatmap(reformatMat[,c(-201)], scale = "none", bias = 1,
#                 col = c("#053061", "#2166ac", "#4393c3", "white", "white", "#d6604d", "#b2182b", "#67001f"),
#                 breaks=c(-15, -10, -5, -2, 0, 2, 5, 10, 15))
# dev.off()

brm014 <- as.data.frame(rowMeans(mat1[,-ncol(mat1)]))
colnames(brm014) <- c("value")
brm014$class <- cut2(c(1:nrow(brm014)), g = 100)
brm014$brg1 <- mat$brg1
brm014$group <- "brm014"
dmso <- as.data.frame(rowMeans(mat2[,-ncol(mat2)]))
colnames(dmso) <- c("value")
dmso$class <- cut2(c(1:nrow(dmso)), g = 100)
dmso$brg1 <- mat$brg1
dmso$group <- "dmso"
matJ <- rbind(brm014, dmso)
# matJ$class <- cut2(matJ$brg1, g=50)
ggerrorplot(matJ, x = "class", y = "value", color="group")

matJ <- as.data.frame(rowMeans(reformatMat))
colnames(matJ) <- c("value")
matJ$brg1 <- mat$brg1
# matJ$class <- cut2(c(1:nrow(matJ)), g = 500)
matJ$class <- cut2(matJ$brg1, g = 300)
matJ$classS <- as.numeric(gsub("]", "", gsub(")", "", gsub("^.*,", "", matJ$class))))

pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/brg1_h3v3_class.pdf")
ggerrorplot(matJ, x = "classS", y = "value") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
  geom_smooth(method = "glm", se = T, formula = y ~ poly(x, 3), span=0.2, alpha=0.2)
matJ$classS <- cut(matJ$brg1, breaks = c(0, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2, 7))
ggerrorplot(matJ, x = "classS", y = "value") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
ggline(matJ, x = "classS", y = "value", add = "mean_se")
dev.off()
```

#### revision (Pol II ChIP-seq data analysis - chd4)
```{r}
## Pol II ser5p
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_Ser5p.deepToolsMatrix.gz"
# file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/test/pol2_ser2p_choudhary.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

mat <- (df[,-c(1:6)])
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))]
  })

samples <- unlist(strsplit("1_shRNA,1_shRNA,1_shRNA,2_SCR,2_SCR,2_SCR", ","))
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1-mat2)
reformatMat$class <- mat$class

pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_ser5p_heatmap.pdf")
# matrix2pheatmap(reformatMat[,c(-201)], scale = "none", bias = 1, 
#                 col = c("#053061", "#2166ac", "#4393c3", "white", "white", "#d6604d", "#b2182b", "#67001f"), 
#                 breaks=c(-15, -10, -5, -2, 0, 2, 5, 10, 15))
Heatmap(reformatMat[,c(-ncol(reformatMat))], col = circlize::colorRamp2(c(-2000, -1000, 0, 1000, 2000), c("#2166ac", "#4393c3", "white", "#d6604d", "#b2182b")),
        cluster_rows=F, cluster_columns=F)
dev.off()

## Pol II ser2p
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_Ser2p_genebody.deepToolsMatrix.gz"
# file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/test/pol2_ser2p_choudhary.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

mat <- (df[,-c(1:6)])
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))]
  })

samples <- unlist(strsplit("1_shRNA,1_shRNA,1_shRNA,2_SCR,2_SCR,2_SCR", ","))
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

# reformatMat <- as.data.frame(log(mat1+1)/log(mat2+1))
# reformatMat <- na.omit(reformatMat[!is.infinite(rowSums(reformatMat)),])
# Heatmap(reformatMat[,c(-ncol(reformatMat))], cluster_rows=F, cluster_columns=F, col=rev(colorBrewer2palette(name = "RdBu", count = 50, bias=1)))

reformatMat <- as.data.frame(mat1-mat2)

pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_ser2p_heatmap.pdf")
# matrix2pheatmap(reformatMat[,c(-ncol(reformatMat))], scale = "none", bias = 1,
#                 col = c("#053061", "#2166ac", "#4393c3", "white", "white", "#d6604d", "#b2182b", "#67001f"), 
#                 breaks=c(-3, -2, -1, -0.2, 0, 0.2, 1, 2, 3))
#                 # breaks=c(-500, -250, -50, -5, 0, 5, 50, 250, 500))

Heatmap(reformatMat[,c(-ncol(reformatMat))], col = circlize::colorRamp2(c(-20, -10, 0, 300, 400), c("#2166ac", "white", "white", "white", "#b2182b")),
        cluster_rows=F, cluster_columns=F)
dev.off()

barplot(as.matrix(reformatMat[grep("^Cdk1$", df$V4), c(-201)])) ## Cdk1
barplot(as.matrix(reformatMat[grep("^Ercc6l$", df$V4), c(-201)])) ## Ercc6l

## Pol II ser2p (gene regulation)
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_ser2p_class.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

mat <- (df[,-c(1:6)])
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))]
  })

samples <- unlist(strsplit("CHD4i,CHD4i,CHD4i,H2O,H2O,H2O", ","))
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1-mat2)
reformatMat <- as.data.frame(log2(mat1/mat2))

reformatMat$class <- "neutral"
reformatMat[c(1:608),]$class <- "down"
reformatMat[c(7935:9065),]$class <- "up"
df.melt <- melt(reformatMat[,c(100:201)])
ggline(df.melt, x = "variable", y = "value", add = "mean", color="class")
```

#### revision (Gviz data tracks - chd4)
```{r}
pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/GVIZ_chd4.pdf")
BWFILE <- read.table(pipe("cat ~/project/chip-seq-analysis/analysis_chd4_brg1/BW.CONFIG | grep 'Chd4_shR\\|Scr_shR'"))
plot_gviz_track(GEN = "mm9", COOR = "chr10:68805900-68816010",
                BAMFILE = c(grep("Ser5p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=8000)
plot_gviz_track(GEN = "mm9", COOR = "chr10:68805900-68816010",
                BAMFILE = c(grep("Ser2p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=3000)
plot_gviz_track(GEN = "mm9", COOR = "chrX:99340554-99352930",
                BAMFILE = c(grep("Ser5p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=2500)
plot_gviz_track(GEN = "mm9", COOR = "chrX:99340554-99352930",
                BAMFILE = c(grep("Ser2p", BWFILE$V1, value = T)),
                type="a", cex=0.5, FLANK_WIN = 1000, maxlimit=1500)
dev.off()
```

#### revision (Histone variants ratio plots - chd4)
```{r}
## h2a.z
file <- "~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/chd4_h2az_class.deepToolsMatrix.gz"
df <- read.table(pipe(sprintf("zgrep -v '@' %s", file)))

chd4_signal <- read.table(pipe("less ~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.bed | cut -f 1-6,20 | cut -f 4,7"), header=T)
colnames(chd4_signal) <- c("gene", "chd4")
df <- merge(df, chd4_signal, by.x="V4", by.y="gene")

mat <- (df[,-c(1:6)])
mat <- mat[order(mat$chd4),]
library("jsonlite")
h <- scan(file, n=1, sep="\n", what=character())
params <- fromJSON(txt=gsub("@", "", h))

len <- params$sample_boundaries[2]-params$sample_boundaries[1]
mat.lst <- lapply(seq(1,length(params$sample_boundaries)-1), function(x) {
  cbind(mat[,c((params$sample_boundaries[x]+1):(params$sample_boundaries[x]+len))], mat$chd4)
  })

samples <- params$sample_labels
index <- lapply(unique(samples), function(x) grep(x, samples))

mat1 <- mat.lst[[1]]
colnames(mat1) <- seq(1,ncol(mat1))
mat2 <- mat.lst[[2]]
colnames(mat2) <- seq(1,ncol(mat2))

reformatMat <- as.data.frame(mat1[,-ncol(mat1)]-mat2[,-ncol(mat2)])
reformatMat$class <- mat$class

chd4_shRNA <- as.data.frame(rowMeans(mat1[,-ncol(mat1)]))
colnames(chd4_shRNA) <- c("value")
chd4_shRNA$class <- cut2(c(1:nrow(chd4_shRNA)), g = 100)
chd4_shRNA$chd4 <- mat$chd4
chd4_shRNA$group <- "chd4_shRNA"
scr_shRNA <- as.data.frame(rowMeans(mat2[,-ncol(mat2)]))
colnames(scr_shRNA) <- c("value")
scr_shRNA$class <- cut2(c(1:nrow(scr_shRNA)), g = 100)
scr_shRNA$chd4 <- mat$chd4
scr_shRNA$group <- "scr_shRNA"
matJ <- rbind(chd4_shRNA, scr_shRNA)
# matJ$class <- cut2(matJ$chd4, g=50)
ggerrorplot(matJ, x = "class", y = "value", color="group")

matJ <- as.data.frame(rowMeans(reformatMat))
colnames(matJ) <- c("value")
matJ$chd4 <- mat$chd4
# matJ$class <- cut2(c(1:nrow(matJ)), g = 500)
matJ$class <- cut2(matJ$chd4, g = 50)
matJ$classS <- as.numeric(gsub("]", "", gsub(")", "", gsub("^.*,", "", matJ$class))))

# pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/chd4_h2az_class.pdf")
ggerrorplot(matJ, x = "classS", y = "value") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
  geom_smooth(method = "glm", se = T, formula = y ~ poly(x, 3), span=0.2, alpha=0.2)
matJ$classS <- cut(matJ$chd4, breaks = c(0, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2, 7))
ggerrorplot(matJ, x = "classS", y = "value") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
ggline(matJ, x = "classS", y = "value", add = "mean_se")
# dev.off()
```

## histone variants edgeR analysis
```{r}
## Chd4
chd4_h2az <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/megadepth/chd4_h2az.signal")
colnames(chd4_h2az) <- c("chr", "start", "end", "shRNA1", "shRNA3", "SCR1", "SCR3")
df <- read.table(pipe("less ~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/REGIONS_INTEREST.BED | cut -f 1-3,4"), header=F)
colnames(df) <- c("chr", "start", "end", "gene")
chd4_h2az$gene <- df$gene

chd4_signal <- read.table(pipe("less ~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.bed | cut -f 1-6,20,29 | cut -f 4,7,8"), header=T)
chd4_h2az <- merge(chd4_h2az, chd4_signal, by.x="gene", by.y="gene")

# chd4_h2az$class_chd4 <- cut(chd4_h2az$chd4, breaks = c(0, as.vector(quantile(chd4_h2az$chd4))[2:5]))
# chd4_h2az$class_mll2 <- cut(chd4_h2az$mll2, breaks = c(0, as.vector(quantile(chd4_h2az$mll2))[2:5]))
chd4_h2az$class_chd4 <- cut(chd4_h2az$chd4, breaks = c(0, 0.6, 0.8, 1, 1.2, 1.4, 1.6, 100))
chd4_h2az$class_mll2 <- cut(chd4_h2az$mll2, breaks = c(0, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5, 100))
table(chd4_h2az$class_chd4)
chd4_h2az$shRNA <- rowMeans(chd4_h2az[,c("shRNA1", "shRNA3")])
chd4_h2az$SCR <- rowMeans(chd4_h2az[,c("SCR1", "SCR3")])
chd4_h2az$value <- chd4_h2az$shRNA - chd4_h2az$SCR

df.melt <- melt((chd4_h2az[,c("shRNA", "SCR", "class_chd4")]))
df.melt$variable <- factor(df.melt$variable, levels=c("SCR", "shRNA"), ordered=T)
p1 <- ggline(df.melt, x = "class_chd4", y = "value", color="variable", add = "mean_se", error.plot = "errorbar") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
  scale_color_manual(values=c("#b2182b", "#2166ac"))
p2 <- ggerrorplot(df.melt, x = "variable", y = "value", color="variable", facet.by = "class_chd4") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
  stat_compare_means(comparisons = list( c("SCR", "shRNA") ),
                     label.y=c(-70, -80, -90, -100)) +
  scale_color_manual(values=c("#b2182b", "#2166ac"))
# df.melt <- melt((chd4_h2az[,c("shRNA", "SCR", "class_mll2")]))
# df.melt$variable <- factor(df.melt$variable, levels=c("SCR", "shRNA"), ordered=T)
# p2 <- ggerrorplot(df.melt, x = "class_mll2", y = "value", color="variable") + 
#   theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
#   stat_compare_means(comparisons = list( c("(0,1.09]", "(1.09,1.37]"), c("(1.09,1.37]", "(1.37,1.69]"), c("(1.37,1.69]", "(1.69,4.82]") ),
#                      label.y=c(-70, -80, -90, -100)) +
#   scale_color_manual(values=c("#b2182b", "#2166ac"))

## Brg1
brg1_h3v3 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/megadepth/brg1_h3v3.signal")
colnames(brg1_h3v3) <- c("chr", "start", "end", "BRM014_1", "BRM014_2", "DMSO2", "DMSO3")
df <- read.table(pipe("less ~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/REGIONS_INTEREST.BED | cut -f 1-3,4"), header=F)
colnames(df) <- c("chr", "start", "end", "gene")
brg1_h3v3$gene <- df$gene

brg1_signal <- read.table(pipe("less ~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.bed | cut -f 1-6,16,29 | cut -f 4,7,8"), header=T)
brg1_h3v3 <- merge(brg1_h3v3, brg1_signal, by.x="gene", by.y="gene")

# brg1_h3v3$class_brg1 <- cut(brg1_h3v3$brg1, breaks = c(0, as.vector(quantile(brg1_h3v3$brg1))[2:5]))
# brg1_h3v3$class_mll2 <- cut(brg1_h3v3$mll2, breaks = c(0, as.vector(quantile(brg1_h3v3$mll2))[2:5]))
brg1_h3v3$class_brg1 <- cut(brg1_h3v3$brg1, breaks = c(0, 0.6, 0.8, 1, 1.2, 1.4, 1.6, 100))
brg1_h3v3$class_mll2 <- cut(brg1_h3v3$mll2, breaks = c(0, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5, 100))
# table(brg1_h3v3$class_brg1)
brg1_h3v3$BRM014 <- rowMeans(brg1_h3v3[,c("BRM014_1", "BRM014_2")])
brg1_h3v3$DMSO <- rowMeans(brg1_h3v3[,c("DMSO2", "DMSO3")])
brg1_h3v3$value <- brg1_h3v3$BRM014 - brg1_h3v3$DMSO

df.melt <- melt((brg1_h3v3[,c("BRM014", "DMSO", "class_brg1")]))
df.melt$variable <- factor(df.melt$variable, levels=c("DMSO", "BRM014"), ordered=T)
p3 <- ggline(df.melt, x = "class_brg1", y = "value", color="variable", add = "mean_se", error.plot = "errorbar") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
  scale_color_manual(values=c("#b2182b", "#2166ac"))
p4 <- ggerrorplot(df.melt, x = "variable", y = "value", color="variable", facet.by = "class_brg1") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
  stat_compare_means(comparisons = list( c("DMSO", "BRM014") ),
                     label.y=c(-70, -80, -90, -100)) +
  scale_color_manual(values=c("#b2182b", "#2166ac"))

pdf("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/histone_variants_diff_per_remodeler_signal.pdf")
ggarrange(p1, p2, p3, p4, nrow=2, ncol=2, labels = c("A) H2A.Z (CHD4)", "B) H2A.Z (CHD4)", "C) H3.3 (BRG1)", "D) H3.3 (BRG1)"))
dev.off()

# #####################
# ## EdgeR
# #####################
# df <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/gene_enzyme_enhancer.bed.histoneVariants", header=T)
# 
# #####################
# ## Brg1
# #####################
# 
# SPIS <- read.table(pipe("cat ~/bric_share/bp_nextgen/00_ALL_CHIP-SEQ_RAW/CHD4_BRG1/histone_variants/MAPSTAT.xls | cut -f 1,16 | grep '^BRM014\\|^DMSO'"), header=F)
# 
# ## Brg1 (H2A.Z)
# mat <- df[,grep("^BRM014_H2A.Z|^DMSO_H2A.Z", colnames(df), value=T)]
# row.names(mat) <- df$name
# mat <- mat[,c(1:5)]
# res.brg1_h2az <- matrix2edgeR(mat, "2,2,1,1,1", coefficient = 2, SPIS[grep("H2A-Z", SPIS$V1)[1:5],]$V2)
# ggplot(res.brg1_h2az, aes(logCPM, logFC, color=class)) + geom_point()
# 
# ## Brg1 (H3.3)
# mat <- df[,grep("^BRM014_H3.3|^DMSO_H3.3", colnames(df), value=T)]
# row.names(mat) <- df$name
# mat <- mat[,c(1:5)]
# res.brg1_h3v3 <- matrix2edgeR(mat, "2,2,1,1,1", coefficient = 2, SPIS[grep("H3-3", SPIS$V1)[1:5],]$V2)
# ggplot(res.brg1_h3v3, aes(logCPM, logFC, color=class)) + geom_point()
# 
# #####################
# ## Chd4
# #####################
# 
# SPIS <- read.table(pipe("cat ~/bric_share/bp_nextgen/00_ALL_CHIP-SEQ_RAW/CHD4_BRG1/histone_variants/MAPSTAT.xls | cut -f 1,15 | grep '^mESC'"), header=F)
# 
# ## Chd4 (H2A.Z)
# mat <- df[,grep("^mESC_H2A.Z", colnames(df), value=T)]
# row.names(mat) <- df$name
# res.chd4_h2az <- matrix2edgeR(mat, "2,2,2,1,1,1", coefficient = 2, SPIS[grep("H2A-Z", SPIS$V1),]$V2)
# res.chd4_h2az$shRNA <- rowMeans(res.chd4_h2az[,grep("^mESC.*Chd4_shR.*mm9$", colnames(res.chd4_h2az), value=T)])
# res.chd4_h2az$SCR <- rowMeans(res.chd4_h2az[,grep("^mESC.*Scr_shR.*mm9$", colnames(res.chd4_h2az), value=T)])
# res.chd4_h2az$value <- res.chd4_h2az$shRNA - res.chd4_h2az$SCR
# # ggplot(res.chd4_h2az, aes(logCPM, logFC, color=class)) + geom_point()
# 
# res.chd4_h2az <- merge(res.chd4_h2az, chd4_signal, by.x="gene", by.y="gene")
# res.chd4_h2az$class_signal <- cut2(res.chd4_h2az$chd4, g = 3)
# res.chd4_h2az$class_signal <- cut(res.chd4_h2az$chd4, breaks = c(0, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2, 7))
# ggerrorplot(res.chd4_h2az, x = "class_signal", y = "value") +
#   theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))
# 
# ## Chd4 (H3.3)
# mat <- df[,grep("^mESC_H3.3", colnames(df), value=T)]
# row.names(mat) <- df$name
# res.chd4_h3v3 <- matrix2edgeR(mat, "2,2,2,1,1,1", coefficient = 2, SPIS[grep("H3.3", SPIS$V1),]$V2)
# ggplot(res.chd4_h3v3, aes(logCPM, logFC, color=class)) + geom_point()

```

## histone variants analysis (binding ratio)
```{r}
df <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.bed", header=T)
df_h3v3_brg1 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/brg1_h3v3_replicates_multiBigwigSummary.bedSignal", header=T)
df_h3v3_brg1$DMSO_H3.3_mESC_rep1_mm9 <- NULL
df_h2az_brg1 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/brg1_h2az_replicates_multiBigwigSummary.bedSignal", header=T)
df_h3v3_chd4 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/chd4_h3v3_replicates_multiBigwigSummary.bedSignal", header=T)
df_h2az_chd4 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/histone_variants/chd4_h2az_replicates_multiBigwigSummary.bedSignal", header=T)


test <- merge(df[,c("gene", "chd4", "brg1", "h2az", "h3v3")], 
              df_h3v3_brg1[,c("name", grep("BRM014|DMSO", colnames(df_h3v3_brg1), value=T))], 
              by.x="gene", by.y="name")
test <- merge(test, df_h2az_brg1[,c("name", grep("BRM014|DMSO", colnames(df_h2az_brg1), value=T))], by.x="gene", by.y="name")
test <- merge(test, df_h3v3_chd4[,c("name", grep("CHD4i|H2O", colnames(df_h3v3_chd4), value=T))], by.x="gene", by.y="name")
test <- merge(test, df_h2az_chd4[,c("name", grep("CHD4i|H2O", colnames(df_h2az_chd4), value=T))], by.x="gene", by.y="name")

pheatmap(cor(test[,-1], method="spearman"))

## brg1
test[,grep("BRM014|DMSO", colnames(test), value=T)] <- as.data.frame(normalize.quantiles(as.matrix(test[,grep("BRM014|DMSO", colnames(test), value=T)])))

test$ratio_brm014_r1 <- log2(test[,grep("BRM014_H2A.Z.*r1", colnames(test), value=T)]/test[,grep("BRM014_H3.3.*r1", colnames(test), value=T)])
test$ratio_brm014_r2 <- log2(test[,grep("BRM014_H2A.Z.*r2", colnames(test), value=T)]/test[,grep("BRM014_H3.3.*r2", colnames(test), value=T)])
test$ratio_dmso_r1 <- log2(test[,grep("DMSO_H2A.Z.*r1", colnames(test), value=T)]/test[,grep("DMSO_H3.3.*r1", colnames(test), value=T)])
test$ratio_dmso_r2 <- log2(test[,grep("DMSO_H2A.Z.*r2", colnames(test), value=T)]/test[,grep("DMSO_H3.3.*r2", colnames(test), value=T)])
test$ratio_dmso_r3 <- log2(test[,grep("DMSO_H2A.Z.*r3", colnames(test), value=T)]/test[,grep("DMSO_H3.3.*r3", colnames(test), value=T)])

p1 <- ggplot(melt(test[,grep("ratio_brm014|ratio_dmso", colnames(test))]), aes(value, color=variable)) + stat_ecdf() + xlim(c(-2,10)) +
  scale_color_manual(values=c("#08519c", "#3182bd", "#a50f15", "#de2d26", "#fb6a4a")) + ylab("density") + 
  xlab("H2A-Z vs H3.3 signal (log2)")

## chd4
test[,grep("CHD4i|H2O", colnames(test), value=T)] <- as.data.frame(normalize.quantiles(as.matrix(test[,grep("CHD4i|H2O", colnames(test), value=T)])))

test$ratio_chd4i_r1 <- log2(test[,grep("CHD4i_H2A.Z.*r1", colnames(test), value=T)]/test[,grep("CHD4i_H3.3.*r1", colnames(test), value=T)])
test$ratio_chd4i_r2 <- log2(test[,grep("CHD4i_H2A.Z.*r2", colnames(test), value=T)]/test[,grep("CHD4i_H3.3.*r2", colnames(test), value=T)])
test$ratio_chd4i_r3 <- log2(test[,grep("CHD4i_H2A.Z.*r3", colnames(test), value=T)]/test[,grep("CHD4i_H3.3.*r3", colnames(test), value=T)])
test$ratio_h2o_r1 <- log2(test[,grep("H2O_H2A.Z.*r1", colnames(test), value=T)]/test[,grep("H2O_H3.3.*r1", colnames(test), value=T)])
test$ratio_h2o_r2 <- log2(test[,grep("H2O_H2A.Z.*r2", colnames(test), value=T)]/test[,grep("H2O_H3.3.*r2", colnames(test), value=T)])
test$ratio_h2o_r3 <- log2(test[,grep("H2O_H2A.Z.*r3", colnames(test), value=T)]/test[,grep("H2O_H3.3.*r3", colnames(test), value=T)])

p2 <- ggplot(melt(test[,grep("ratio_chd4i|ratio_h2o", colnames(test))]), aes(value, color=variable)) + stat_ecdf() + xlim(c(-2,10)) +
  scale_color_manual(values=c("#08519c", "#3182bd", "#6baed6", "#a50f15", "#de2d26", "#fb6a4a")) + ylab("density") + 
  xlab("H2A-Z vs H3.3 signal (log2)")
```

## Pol II analysis (binding ratio)
```{r}
df <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.bed", header=T)
df_ser2p_brg1 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser2p_replicates_multiBigwigSummary.bedSignal", header=T)
df_ser5p_brg1 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/brg1/pol2_ser5p_replicates_multiBigwigSummary.bedSignal", header=T)
df_ser2p_chd4 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_ser2p_replicates_multiBigwigSummary.bedSignal", header=T)
df_ser5p_chd4 <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/deepPlot/chd4/pol2_ser5p_replicates_multiBigwigSummary.bedSignal", header=T)


test <- merge(df[,c("gene", "chd4", "brg1", "pol2_ser2p", "pol2_ser5p")], 
              df_ser2p_brg1[,c("name", grep("BRM014|DMSO", colnames(df_ser2p_brg1), value=T))], 
              by.x="gene", by.y="name")
test <- merge(test, df_ser5p_brg1[,c("name", grep("BRM014|DMSO", colnames(df_ser5p_brg1), value=T))], by.x="gene", by.y="name")
test <- merge(test, df_ser2p_chd4[,c("name", grep("CHD4i|H2O", colnames(df_ser2p_chd4), value=T))], by.x="gene", by.y="name")
test <- merge(test, df_ser5p_chd4[,c("name", grep("CHD4i|H2O", colnames(df_ser5p_chd4), value=T))], by.x="gene", by.y="name")

pheatmap(cor(test[,-1], method="spearman"))

## brg1
test[,grep("BRM014|DMSO", colnames(test), value=T)] <- as.data.frame(normalize.quantiles(as.matrix(test[,grep("BRM014|DMSO", colnames(test), value=T)])))

test$ratio_brm014_r1 <- log2(test[,grep("ser2p_BRM014_rep1", colnames(test), value=T)]/test[,grep("ser5p_BRM014_rep1", colnames(test), value=T)])
test$ratio_brm014_r2 <- log2(test[,grep("ser2p_BRM014_rep2", colnames(test), value=T)]/test[,grep("ser5p_BRM014_rep2", colnames(test), value=T)])
test$ratio_brm014_r3 <- log2(test[,grep("ser2p_BRM014_rep4", colnames(test), value=T)]/test[,grep("ser5p_BRM014_rep3", colnames(test), value=T)])
test$ratio_dmso_r1 <- log2(test[,grep("ser2p_DMSO_rep1", colnames(test), value=T)]/test[,grep("ser5p_DMSO_rep1", colnames(test), value=T)])
test$ratio_dmso_r2 <- log2(test[,grep("ser2p_DMSO_rep2", colnames(test), value=T)]/test[,grep("ser5p_DMSO_rep2", colnames(test), value=T)])
test$ratio_dmso_r3 <- log2(test[,grep("ser2p_DMSO_rep3", colnames(test), value=T)]/test[,grep("ser5p_DMSO_rep3", colnames(test), value=T)])

ggplot(melt(test[,grep("ratio_brm014|ratio_dmso", colnames(test))]), aes(value, color=variable)) + stat_ecdf() + 
  scale_color_manual(values=c("#08519c", "#3182bd", "#6baed6", "#a50f15", "#de2d26", "#fb6a4a")) + ylab("density") + 
  xlab("ser2p vs ser5p signal (log2)")

## chd4
test[,grep("CHD4i|H2O", colnames(test), value=T)] <- as.data.frame(normalize.quantiles(as.matrix(test[,grep("CHD4i|H2O", colnames(test), value=T)])))

test$ratio_chd4i_r1 <- log2(test[,grep("ser2p_CHD4i_rep1", colnames(test), value=T)]/test[,grep("ser5p_CHD4i_rep1", colnames(test), value=T)])
test$ratio_chd4i_r2 <- log2(test[,grep("ser2p_CHD4i_rep2", colnames(test), value=T)]/test[,grep("ser5p_CHD4i_rep2", colnames(test), value=T)])
test$ratio_chd4i_r3 <- log2(test[,grep("ser2p_CHD4i_rep3", colnames(test), value=T)]/test[,grep("ser5p_CHD4i_rep3", colnames(test), value=T)])
test$ratio_h2o_r1 <- log2(test[,grep("ser2p_H2O_rep1", colnames(test), value=T)]/test[,grep("ser5p_H2O_rep1", colnames(test), value=T)])
test$ratio_h2o_r2 <- log2(test[,grep("ser2p_H2O_rep2", colnames(test), value=T)]/test[,grep("ser5p_H2O_rep2", colnames(test), value=T)])
test$ratio_h2o_r3 <- log2(test[,grep("ser2p_H2O_rep3", colnames(test), value=T)]/test[,grep("ser5p_H2O_rep3", colnames(test), value=T)])

ggplot(melt(test[,grep("ratio_chd4i|ratio_h2o", colnames(test))]), aes(value, color=variable)) + stat_ecdf() + 
  scale_color_manual(values=c("#08519c", "#3182bd", "#6baed6", "#a50f15", "#de2d26", "#fb6a4a")) + ylab("density") + 
  xlab("ser2p vs ser5p signal (log2)")

## identify new candidates for qPCR (Pol-II Ser2p and Ser5p ChIP)
mat <- df_ser5p_chd4[,grep("PolII", colnames(df_ser5p_chd4))]
row.names(mat) <- df_ser5p_chd4$name
res_ser5p <- matrix2deseq2(round(mat,0), "ko,ko,ko,wt,wt,wt")
res_ser5p <- res_ser5p[order(-res_ser5p$log2FoldChange),]
head(res_ser5p[which(!is.na(res_ser5p$padj)),])

mat <- df_ser2p_chd4[,grep("PolII", colnames(df_ser2p_chd4))]
row.names(mat) <- df_ser2p_chd4$name
res_ser2p <- matrix2deseq2(round(mat,0), "ko,ko,ko,wt,wt,wt")
res_ser2p <- res_ser2p[order(-res_ser2p$log2FoldChange),]
head(res_ser2p[which(!is.na(res_ser2p$padj)),])

ggplot(res_ser5p[which(!is.na(res_ser5p$padj)),], aes(log2FoldChange, -log10(pvalue))) + geom_point()
ggplot(res_ser2p[which(!is.na(res_ser2p$padj)),], aes(log2FoldChange, -log10(pvalue))) + geom_point()

df.melt <- melt(df_ser5p_chd4[,grep("description6|PolII", colnames(df_ser5p_chd4))])
df.melt$variable <- factor(df.melt$variable, levels=c(grep("H2O", levels(df.melt$variable), value=T), 
                                                      grep("CHD4i", levels(df.melt$variable), value=T)), ordered = T)
ggplot(df.melt, aes(variable, log2(value))) + geom_boxplot() + facet_grid(.~description6)
```


## spike-in analysis
```{r}
## brg1
df_spike_in_brg1 <- read.table(pipe("cut -f 1,8,9 ~/bric_share/bp_nextgen/00_ALL_CHIP-SEQ_RAW/CHD4_BRG1/brg1/MAPSTAT.xls"), header=T, sep="\t")
df_spike_in_brg1$dm6_mapped..proper_pairs. <- as.numeric(gsub("\\s+.*", "", df_spike_in_brg1$dm6_mapped..proper_pairs.))
df_spike_in_brg1$mm9_mapped..proper_pairs. <- as.numeric(gsub("\\s+.*", "", df_spike_in_brg1$mm9_mapped..proper_pairs.))
df_spike_in_brg1$class <- gsub("_rep.*", "", df_spike_in_brg1$id)

df_spike_in_brg1$dm6_vs_mm9 <- df_spike_in_brg1$dm6_mapped..proper_pairs./df_spike_in_brg1$mm9_mapped..proper_pairs.
df_spike_in_brg1$scaling_factor <- 1000000/df_spike_in_brg1$dm6_mapped..proper_pairs.

ggplot(df_spike_in_brg1, aes(class, (dm6_vs_mm9*100))) + geom_boxplot()

## chd4
df_spike_in_chd4 <- read.table(pipe("cut -f 1,8,9 ~/bric_share/bp_nextgen/00_ALL_CHIP-SEQ_RAW/CHD4_BRG1/chd4/MAPSTAT.xls"), header=T, sep="\t")
df_spike_in_chd4 <- df_spike_in_chd4[grep("PolII", df_spike_in_chd4$id),]
df_spike_in_chd4$dm6_mapped..proper_pairs. <- as.numeric(gsub("\\s+.*", "", df_spike_in_chd4$dm6_mapped..proper_pairs.))
df_spike_in_chd4$mm9_mapped..proper_pairs. <- as.numeric(gsub("\\s+.*", "", df_spike_in_chd4$mm9_mapped..proper_pairs.))
df_spike_in_chd4$class <- gsub("_rep.*", "", df_spike_in_chd4$id)

df_spike_in_chd4$dm6_vs_mm9 <- df_spike_in_chd4$dm6_mapped..proper_pairs./df_spike_in_chd4$mm9_mapped..proper_pairs.
df_spike_in_chd4$scaling_factor <- 1000000/df_spike_in_chd4$dm6_mapped..proper_pairs.

ggplot(df_spike_in_chd4, aes(class, (dm6_vs_mm9*100))) + geom_boxplot()

## histone variants
df_spike_in_histone_variants <- read.table(pipe("cut -f 1,8,9 ~/bric_share/bp_nextgen/00_ALL_CHIP-SEQ_RAW/CHD4_BRG1/histone_variants/MAPSTAT.xls"), header=T, sep="\t")
df_spike_in_histone_variants <- df_spike_in_histone_variants[grep("H2A-Z|H3-3", df_spike_in_histone_variants$id),]
df_spike_in_histone_variants$dm6_mapped..proper_pairs. <- as.numeric(gsub("\\s+.*", "", df_spike_in_histone_variants$dm6_mapped..proper_pairs.))
df_spike_in_histone_variants$mm9_mapped..proper_pairs. <- as.numeric(gsub("\\s+.*", "", df_spike_in_histone_variants$mm9_mapped..proper_pairs.))
df_spike_in_histone_variants$class <- gsub("_r.*", "", df_spike_in_histone_variants$id)

df_spike_in_histone_variants$dm6_vs_mm9 <- df_spike_in_histone_variants$dm6_mapped..proper_pairs./df_spike_in_histone_variants$mm9_mapped..proper_pairs.
df_spike_in_histone_variants$scaling_factor <- 1000000/df_spike_in_histone_variants$dm6_mapped..proper_pairs.

par(mfrow=c(2,2))
plot(df_spike_in_brg1$dm6_vs_mm9, df_spike_in_brg1$scaling_factor)
plot(df_spike_in_chd4$dm6_vs_mm9, df_spike_in_chd4$scaling_factor)
plot(df_spike_in_histone_variants$dm6_vs_mm9, df_spike_in_histone_variants$scaling_factor)

ggplot(df_spike_in_histone_variants[df_spike_in_histone_variants$id %in% grep("BRM014_H3-3|DMSO_H3-3", df_spike_in_histone_variants$id, value=TRUE),], aes(class, (dm6_vs_mm9*100))) + geom_boxplot()
ggplot(df_spike_in_histone_variants[df_spike_in_histone_variants$id %in% grep("BRM014_H2A-Z|DMSO_H2A-Z", df_spike_in_histone_variants$id, value=TRUE),], aes(class, (dm6_vs_mm9*100))) + geom_boxplot()

ggplot(df_spike_in_histone_variants[df_spike_in_histone_variants$id %in% grep("CHD4i_H3-3|H2O_H3-3", df_spike_in_histone_variants$id, value=TRUE),], aes(class, (dm6_vs_mm9*100))) + geom_boxplot()
ggplot(df_spike_in_histone_variants[df_spike_in_histone_variants$id %in% grep("CHD4i_H2A-Z|H2O_H2A-Z", df_spike_in_histone_variants$id, value=TRUE),], aes(class, (dm6_vs_mm9*100))) + geom_boxplot()
```

## ChIPseqSpikeInFree analysis
```{r}
library("ChIPseqSpikeInFree")
metaFile <- "~/project/chip-seq-analysis/analysis_chd4_brg1/BAM.CONFIG"
bams <- sprintf("/localhome/bric/xfd783/project/misc/copy_and_analyze/chd4_brg1/chd4/%s", read.table(metaFile, header = T)$ID)
ChIPseqSpikeInFree(bamFiles = bams, chromFile = "mm9", metaFile = metaFile, prefix = "/localhome/bric/xfd783/project/chip-seq-analysis/analysis_chd4_brg1/peaks/test_ser2p")

SF_ser2p <- (calcNormFactors(read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/peaks/test_ser2p_rawCounts.txt", header=T)[,c(2:7)])*1000000)/colSums(test_tss[,grep("ser2p_CHD4i|ser2p_H2O", colnames(test_tss), value=T)])
SF_ser5p <- calcNormFactors(read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/peaks/test_ser5p_rawCounts.txt", header=T)[,c(2:7)])

##---------------##
dat <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/peaks/test_ser2p_SF.txt", sep="\t",header=TRUE,fill=TRUE,stringsAsFactors = FALSE, quote="",check.names=F)
SF <- dat$SF
countData <- test_genebody[,grep("ser2p_CHD4i|ser2p_H2O", colnames(test_genebody), value=T)]
condition <- factor(c("ko","ko","ko","wt", "wt", "wt"))
dds <- DESeqDataSetFromMatrix(countData, DataFrame(condition), ~ condition)
dds <- estimateSizeFactors(dds)
coldata<- colData(dds)
coldata$sizeFactor <- coldata$sizeFactor *SF
colData(dds) <- coldata
dds <- estimateDispersions(dds)
dds <- nbinomWaldTest(dds)
ggplot(as.data.frame(results(dds)), aes(log2FoldChange, -log10(padj))) + geom_point()

##---------------##
test <- read.table("~/project/chip-seq-analysis/analysis_chd4_brg1/gene_enzyme_enhancer.signalValues", header=T)
test_tss <- as.data.frame(lapply(grep("PolII", colnames(test), value=T), function(x) as.numeric(unlist(lapply(test[,x], function(y) unlist(strsplit(y, "/"))[1])))))
colnames(test_tss) <- grep("PolII", colnames(test), value=T)
test_genebody <- as.data.frame(lapply(grep("PolII", colnames(test), value=T), function(x) as.numeric(unlist(lapply(test[,x], function(y) unlist(strsplit(y, "/"))[2])))))
colnames(test_genebody) <- grep("PolII", colnames(test), value=T)
df <- as.data.frame(normalize.quantiles(as.matrix(test_tss[,grep("CHD4i|H2O", colnames(test_tss), value=T)])))
colnames(df) <- colnames(test_tss[,grep("CHD4i|H2O", colnames(test_tss), value=T)])
df$class <- test$gene
df$chd4i_rep1 <- log2(df$PolII_ser2p_CHD4i_rep1_mm9/df$PolII_ser5p_CHD4i_rep1_mm9)
df$chd4i_rep2 <- log2(df$PolII_ser2p_CHD4i_rep2_mm9/df$PolII_ser5p_CHD4i_rep2_mm9)
df$chd4i_rep3 <- log2(df$PolII_ser2p_CHD4i_rep3_mm9/df$PolII_ser5p_CHD4i_rep3_mm9)
df$h2o_rep1 <- log2(df$PolII_ser2p_H2O_rep1_mm9/df$PolII_ser5p_H2O_rep1_mm9)
df$h2o_rep2 <- log2(df$PolII_ser2p_H2O_rep2_mm9/df$PolII_ser5p_H2O_rep2_mm9)
df$h2o_rep3 <- log2(df$PolII_ser2p_H2O_rep3_mm9/df$PolII_ser5p_H2O_rep3_mm9)

chd4.melt <- melt(df[,grep("chd4i|h2o|class", colnames(df), value=T)])
chd4.melt$class <- factor(chd4.melt$class, levels=c("bivalent", "low", "average", "high"), ordered=T)
chd4.melt$variable <- gsub("_.*", "", chd4.melt$variable)

df <- as.data.frame(normalize.quantiles(as.matrix(test_tss[,grep("BRM014|DMSO", colnames(test_tss), value=T)])))
colnames(df) <- colnames(test_tss[,grep("BRM014|DMSO", colnames(test_tss), value=T)])
df$class <- test$gene
df$brm014_rep1 <- log2(df$PolII_ser2p_BRM014_rep1_mm9/df$PolII_ser5p_BRM014_rep1_mm9)
df$brm014_rep2 <- log2(df$PolII_ser2p_BRM014_rep2_mm9/df$PolII_ser5p_BRM014_rep2_mm9)
df$brm014_rep3 <- log2(df$PolII_ser2p_BRM014_rep4_mm9/df$PolII_ser5p_BRM014_rep3_mm9)
df$dmso_rep1 <- log2(df$PolII_ser2p_DMSO_rep1_mm9/df$PolII_ser5p_DMSO_rep1_mm9)
df$dmso_rep2 <- log2(df$PolII_ser2p_DMSO_rep2_mm9/df$PolII_ser5p_DMSO_rep2_mm9)
df$dmso_rep3 <- log2(df$PolII_ser2p_DMSO_rep3_mm9/df$PolII_ser5p_DMSO_rep3_mm9)

brg1.melt <- melt(df[,grep("brm014|dmso|class", colnames(df), value=T)])
brg1.melt$class <- factor(brg1.melt$class, levels=c("bivalent", "low", "average", "high"), ordered=T)
brg1.melt$variable <- gsub("_.*", "", brg1.melt$variable)

ks.test(chd4.melt[which(chd4.melt$class=="high" & chd4.melt$variable=="chd4i"),]$value, chd4.melt[which(chd4.melt$class=="high" & chd4.melt$variable=="h2o"),]$value, alternative = "l")$p.value
ks.test(brg1.melt[which(brg1.melt$class=="high" & brg1.melt$variable=="brm014"),]$value, brg1.melt[which(chd4.melt$class=="high" & brg1.melt$variable=="dmso"),]$value, alternative = "g")$p.value
ggarrange(ggplot(chd4.melt, aes(value, color=variable)) + stat_ecdf() + facet_wrap(.~class, scales="free", nrow = 1),
          ggplot(brg1.melt, aes(value, color=variable)) + stat_ecdf() + facet_wrap(.~class, scales="free", nrow = 1), nrow=2)
```

